<html><head><meta content="text/html; charset=UTF-8" http-equiv="content-type"><style type="text/css">@import url('https://themes.googleusercontent.com/fonts/css?kit=fpjTOVmNbO4Lz34iLyptLTi9jKYd1gJzj5O2gWsEpXol-nTHck7FFkZplK5meosG');ol.lst-kix_lswckwbst11a-3.start{counter-reset:lst-ctn-kix_lswckwbst11a-3 0}ol.lst-kix_drcsydhb4z7b-3.start{counter-reset:lst-ctn-kix_drcsydhb4z7b-3 0}ul.lst-kix_list_1-0{list-style-type:none}.lst-kix_ocekgj1d9cfy-8>li:before{content:"" counter(lst-ctn-kix_ocekgj1d9cfy-8,lower-roman) ". "}ol.lst-kix_ocekgj1d9cfy-5.start{counter-reset:lst-ctn-kix_ocekgj1d9cfy-5 0}.lst-kix_ocekgj1d9cfy-2>li:before{content:"" counter(lst-ctn-kix_ocekgj1d9cfy-2,lower-roman) ". "}ol.lst-kix_lswckwbst11a-8{list-style-type:none}ol.lst-kix_lswckwbst11a-6{list-style-type:none}.lst-kix_ocekgj1d9cfy-1>li:before{content:"" counter(lst-ctn-kix_ocekgj1d9cfy-1,lower-latin) ". "}ol.lst-kix_lswckwbst11a-7{list-style-type:none}ol.lst-kix_lswckwbst11a-4{list-style-type:none}ol.lst-kix_lswckwbst11a-5{list-style-type:none}ol.lst-kix_lswckwbst11a-2{list-style-type:none}.lst-kix_ocekgj1d9cfy-0>li:before{content:"" counter(lst-ctn-kix_ocekgj1d9cfy-0,decimal) ". "}ol.lst-kix_lswckwbst11a-3{list-style-type:none}ul.lst-kix_list_1-3{list-style-type:none}ol.lst-kix_lswckwbst11a-0{list-style-type:none}ul.lst-kix_list_1-4{list-style-type:none}ol.lst-kix_lswckwbst11a-1{list-style-type:none}ul.lst-kix_list_1-1{list-style-type:none}ul.lst-kix_list_1-2{list-style-type:none}ul.lst-kix_list_1-5{list-style-type:none}ul.lst-kix_list_1-6{list-style-type:none}.lst-kix_ocekgj1d9cfy-1>li{counter-increment:lst-ctn-kix_ocekgj1d9cfy-1}ol.lst-kix_tgkyxxtz27uw-0.start{counter-reset:lst-ctn-kix_tgkyxxtz27uw-0 0}.lst-kix_drcsydhb4z7b-4>li{counter-increment:lst-ctn-kix_drcsydhb4z7b-4}ol.lst-kix_list_3-7.start{counter-reset:lst-ctn-kix_list_3-7 0}ol.lst-kix_tgkyxxtz27uw-6.start{counter-reset:lst-ctn-kix_tgkyxxtz27uw-6 0}.lst-kix_list_5-0>li:before{content:"\002022  "}.lst-kix_list_5-3>li:before{content:"\002013  "}.lst-kix_list_5-2>li:before{content:"\002022  "}.lst-kix_list_5-1>li:before{content:"\002013  "}ul.lst-kix_s3kte6nvgtp0-0{list-style-type:none}ol.lst-kix_drcsydhb4z7b-8.start{counter-reset:lst-ctn-kix_drcsydhb4z7b-8 0}ul.lst-kix_s3kte6nvgtp0-1{list-style-type:none}.lst-kix_list_5-7>li:before{content:" "}ul.lst-kix_list_8-4{list-style-type:none}ul.lst-kix_s3kte6nvgtp0-2{list-style-type:none}ul.lst-kix_list_8-5{list-style-type:none}ul.lst-kix_s3kte6nvgtp0-3{list-style-type:none}.lst-kix_list_5-6>li:before{content:"\002022  "}.lst-kix_list_5-8>li:before{content:" "}ul.lst-kix_list_8-2{list-style-type:none}ul.lst-kix_s3kte6nvgtp0-4{list-style-type:none}ol.lst-kix_tgkyxxtz27uw-5.start{counter-reset:lst-ctn-kix_tgkyxxtz27uw-5 0}ul.lst-kix_list_8-3{list-style-type:none}ul.lst-kix_s3kte6nvgtp0-5{list-style-type:none}ul.lst-kix_list_8-8{list-style-type:none}ul.lst-kix_s3kte6nvgtp0-6{list-style-type:none}ul.lst-kix_s3kte6nvgtp0-7{list-style-type:none}ul.lst-kix_list_8-6{list-style-type:none}ul.lst-kix_s3kte6nvgtp0-8{list-style-type:none}.lst-kix_tgkyxxtz27uw-1>li{counter-increment:lst-ctn-kix_tgkyxxtz27uw-1}ul.lst-kix_list_8-7{list-style-type:none}.lst-kix_list_5-4>li:before{content:"\002022  "}.lst-kix_drcsydhb4z7b-0>li:before{content:"" counter(lst-ctn-kix_drcsydhb4z7b-0,decimal) ". "}.lst-kix_list_5-5>li:before{content:"\002013  "}ul.lst-kix_list_8-0{list-style-type:none}ul.lst-kix_list_8-1{list-style-type:none}ol.lst-kix_lswckwbst11a-8.start{counter-reset:lst-ctn-kix_lswckwbst11a-8 0}.lst-kix_ocekgj1d9cfy-8>li{counter-increment:lst-ctn-kix_ocekgj1d9cfy-8}.lst-kix_list_6-1>li:before{content:"o  "}.lst-kix_list_6-3>li:before{content:"\0025cf  "}.lst-kix_drcsydhb4z7b-2>li:before{content:"" counter(lst-ctn-kix_drcsydhb4z7b-2,lower-roman) ". "}.lst-kix_drcsydhb4z7b-4>li:before{content:"" counter(lst-ctn-kix_drcsydhb4z7b-4,lower-latin) ". "}.lst-kix_drcsydhb4z7b-2>li{counter-increment:lst-ctn-kix_drcsydhb4z7b-2}.lst-kix_list_6-0>li:before{content:"\0025cf  "}.lst-kix_list_6-4>li:before{content:"o  "}.lst-kix_drcsydhb4z7b-1>li:before{content:"" counter(lst-ctn-kix_drcsydhb4z7b-1,lower-latin) ". "}.lst-kix_drcsydhb4z7b-5>li:before{content:"" counter(lst-ctn-kix_drcsydhb4z7b-5,lower-roman) ". "}.lst-kix_list_6-2>li:before{content:"\0025aa  "}.lst-kix_drcsydhb4z7b-3>li:before{content:"" counter(lst-ctn-kix_drcsydhb4z7b-3,decimal) ". "}.lst-kix_ocekgj1d9cfy-3>li:before{content:"" counter(lst-ctn-kix_ocekgj1d9cfy-3,decimal) ". "}ol.lst-kix_ocekgj1d9cfy-4.start{counter-reset:lst-ctn-kix_ocekgj1d9cfy-4 0}.lst-kix_list_2-8>li{counter-increment:lst-ctn-kix_list_2-8}.lst-kix_list_6-8>li:before{content:"\0025aa  "}.lst-kix_ocekgj1d9cfy-4>li:before{content:"" counter(lst-ctn-kix_ocekgj1d9cfy-4,lower-latin) ". "}.lst-kix_drcsydhb4z7b-8>li{counter-increment:lst-ctn-kix_drcsydhb4z7b-8}.lst-kix_lswckwbst11a-7>li{counter-increment:lst-ctn-kix_lswckwbst11a-7}.lst-kix_list_6-5>li:before{content:"\0025aa  "}.lst-kix_list_6-7>li:before{content:"o  "}.lst-kix_drcsydhb4z7b-6>li:before{content:"" counter(lst-ctn-kix_drcsydhb4z7b-6,decimal) ". "}.lst-kix_drcsydhb4z7b-8>li:before{content:"" counter(lst-ctn-kix_drcsydhb4z7b-8,lower-roman) ". "}.lst-kix_ocekgj1d9cfy-5>li:before{content:"" counter(lst-ctn-kix_ocekgj1d9cfy-5,lower-roman) ". "}.lst-kix_ocekgj1d9cfy-7>li:before{content:"" counter(lst-ctn-kix_ocekgj1d9cfy-7,lower-latin) ". "}.lst-kix_list_6-6>li:before{content:"\0025cf  "}.lst-kix_drcsydhb4z7b-7>li:before{content:"" counter(lst-ctn-kix_drcsydhb4z7b-7,lower-latin) ". "}.lst-kix_ocekgj1d9cfy-6>li:before{content:"" counter(lst-ctn-kix_ocekgj1d9cfy-6,decimal) ". "}.lst-kix_list_2-7>li:before{content:" "}.lst-kix_list_2-7>li{counter-increment:lst-ctn-kix_list_2-7}.lst-kix_list_7-4>li:before{content:"o  "}.lst-kix_list_7-6>li:before{content:"\0025cf  "}.lst-kix_list_2-5>li:before{content:"   "}.lst-kix_tgkyxxtz27uw-5>li{counter-increment:lst-ctn-kix_tgkyxxtz27uw-5}.lst-kix_list_7-2>li:before{content:"\0025aa  "}ol.lst-kix_ocekgj1d9cfy-0{list-style-type:none}ol.lst-kix_ocekgj1d9cfy-1{list-style-type:none}.lst-kix_tgkyxxtz27uw-3>li:before{content:"" counter(lst-ctn-kix_tgkyxxtz27uw-3,decimal) ". "}ol.lst-kix_ocekgj1d9cfy-6{list-style-type:none}ul.lst-kix_list_3-1{list-style-type:none}ol.lst-kix_ocekgj1d9cfy-7{list-style-type:none}ul.lst-kix_list_3-2{list-style-type:none}.lst-kix_list_5-7>li{counter-increment:lst-ctn-kix_list_5-7}.lst-kix_ocekgj1d9cfy-5>li{counter-increment:lst-ctn-kix_ocekgj1d9cfy-5}ol.lst-kix_ocekgj1d9cfy-8{list-style-type:none}.lst-kix_list_7-8>li:before{content:"\0025aa  "}ul.lst-kix_list_3-0{list-style-type:none}ol.lst-kix_ocekgj1d9cfy-2{list-style-type:none}ol.lst-kix_list_1-7{list-style-type:none}ul.lst-kix_list_3-5{list-style-type:none}ol.lst-kix_ocekgj1d9cfy-3{list-style-type:none}ol.lst-kix_list_1-8{list-style-type:none}ul.lst-kix_list_3-6{list-style-type:none}.lst-kix_tgkyxxtz27uw-4>li{counter-increment:lst-ctn-kix_tgkyxxtz27uw-4}ol.lst-kix_ocekgj1d9cfy-4{list-style-type:none}ul.lst-kix_list_3-3{list-style-type:none}.lst-kix_tgkyxxtz27uw-5>li:before{content:"" counter(lst-ctn-kix_tgkyxxtz27uw-5,lower-roman) ". "}ol.lst-kix_ocekgj1d9cfy-5{list-style-type:none}ul.lst-kix_list_3-4{list-style-type:none}.lst-kix_tgkyxxtz27uw-6>li{counter-increment:lst-ctn-kix_tgkyxxtz27uw-6}ol.lst-kix_drcsydhb4z7b-2.start{counter-reset:lst-ctn-kix_drcsydhb4z7b-2 0}.lst-kix_ocekgj1d9cfy-7>li{counter-increment:lst-ctn-kix_ocekgj1d9cfy-7}.lst-kix_lswckwbst11a-0>li{counter-increment:lst-ctn-kix_lswckwbst11a-0}.lst-kix_tgkyxxtz27uw-7>li:before{content:"" counter(lst-ctn-kix_tgkyxxtz27uw-7,lower-latin) ". "}.lst-kix_list_4-1>li:before{content:"o  "}.lst-kix_list_4-3>li:before{content:"\0025cf  "}.lst-kix_list_4-5>li:before{content:"\0025aa  "}ol.lst-kix_list_5-7.start{counter-reset:lst-ctn-kix_list_5-7 0}.lst-kix_list_1-8>li{counter-increment:lst-ctn-kix_list_1-8}.lst-kix_s3kte6nvgtp0-5>li:before{content:"\0025a0  "}ol.lst-kix_tgkyxxtz27uw-1.start{counter-reset:lst-ctn-kix_tgkyxxtz27uw-1 0}ol.lst-kix_ocekgj1d9cfy-0.start{counter-reset:lst-ctn-kix_ocekgj1d9cfy-0 0}.lst-kix_s3kte6nvgtp0-7>li:before{content:"\0025cb  "}ol.lst-kix_list_2-8.start{counter-reset:lst-ctn-kix_list_2-8 0}ol.lst-kix_drcsydhb4z7b-5.start{counter-reset:lst-ctn-kix_drcsydhb4z7b-5 0}.lst-kix_lswckwbst11a-5>li{counter-increment:lst-ctn-kix_lswckwbst11a-5}ul.lst-kix_list_2-2{list-style-type:none}ul.lst-kix_list_2-3{list-style-type:none}ul.lst-kix_list_2-0{list-style-type:none}ul.lst-kix_list_2-1{list-style-type:none}ul.lst-kix_list_2-6{list-style-type:none}.lst-kix_list_1-1>li:before{content:"   "}ul.lst-kix_list_2-4{list-style-type:none}ul.lst-kix_list_2-5{list-style-type:none}.lst-kix_list_1-3>li:before{content:"   "}ol.lst-kix_tgkyxxtz27uw-2.start{counter-reset:lst-ctn-kix_tgkyxxtz27uw-2 0}ol.lst-kix_ocekgj1d9cfy-1.start{counter-reset:lst-ctn-kix_ocekgj1d9cfy-1 0}ol.lst-kix_drcsydhb4z7b-4.start{counter-reset:lst-ctn-kix_drcsydhb4z7b-4 0}.lst-kix_list_1-7>li:before{content:" "}ol.lst-kix_list_5-8.start{counter-reset:lst-ctn-kix_list_5-8 0}ol.lst-kix_list_2-7.start{counter-reset:lst-ctn-kix_list_2-7 0}.lst-kix_list_1-5>li:before{content:"   "}.lst-kix_tgkyxxtz27uw-1>li:before{content:"" counter(lst-ctn-kix_tgkyxxtz27uw-1,lower-latin) ". "}.lst-kix_list_2-1>li:before{content:"   "}.lst-kix_ocekgj1d9cfy-6>li{counter-increment:lst-ctn-kix_ocekgj1d9cfy-6}.lst-kix_ocekgj1d9cfy-0>li{counter-increment:lst-ctn-kix_ocekgj1d9cfy-0}.lst-kix_list_2-3>li:before{content:"   "}.lst-kix_s3kte6nvgtp0-1>li:before{content:"\0025cb  "}.lst-kix_s3kte6nvgtp0-3>li:before{content:"\0025cf  "}ol.lst-kix_ocekgj1d9cfy-2.start{counter-reset:lst-ctn-kix_ocekgj1d9cfy-2 0}ol.lst-kix_drcsydhb4z7b-0{list-style-type:none}ol.lst-kix_drcsydhb4z7b-2{list-style-type:none}ol.lst-kix_drcsydhb4z7b-1{list-style-type:none}.lst-kix_list_3-0>li:before{content:"\002022  "}ol.lst-kix_drcsydhb4z7b-8{list-style-type:none}ol.lst-kix_lswckwbst11a-0.start{counter-reset:lst-ctn-kix_lswckwbst11a-0 0}ol.lst-kix_drcsydhb4z7b-7{list-style-type:none}.lst-kix_list_3-1>li:before{content:"\002013  "}.lst-kix_list_3-2>li:before{content:"\002022  "}.lst-kix_lswckwbst11a-6>li:before{content:"" counter(lst-ctn-kix_lswckwbst11a-6,decimal) ". "}.lst-kix_lswckwbst11a-7>li:before{content:"" counter(lst-ctn-kix_lswckwbst11a-7,lower-latin) ". "}ul.lst-kix_list_5-5{list-style-type:none}ul.lst-kix_list_5-6{list-style-type:none}ol.lst-kix_drcsydhb4z7b-4{list-style-type:none}.lst-kix_list_8-1>li:before{content:"o  "}ol.lst-kix_list_1-8.start{counter-reset:lst-ctn-kix_list_1-8 0}ol.lst-kix_drcsydhb4z7b-0.start{counter-reset:lst-ctn-kix_drcsydhb4z7b-0 0}ol.lst-kix_drcsydhb4z7b-3{list-style-type:none}.lst-kix_lswckwbst11a-8>li{counter-increment:lst-ctn-kix_lswckwbst11a-8}ol.lst-kix_drcsydhb4z7b-6{list-style-type:none}.lst-kix_list_8-2>li:before{content:"\0025aa  "}ol.lst-kix_drcsydhb4z7b-5{list-style-type:none}.lst-kix_list_3-5>li:before{content:"\002013  "}ul.lst-kix_list_5-0{list-style-type:none}.lst-kix_list_3-4>li:before{content:"\002022  "}ul.lst-kix_list_5-3{list-style-type:none}.lst-kix_list_3-3>li:before{content:"\002013  "}ul.lst-kix_list_5-4{list-style-type:none}.lst-kix_lswckwbst11a-8>li:before{content:"" counter(lst-ctn-kix_lswckwbst11a-8,lower-roman) ". "}ul.lst-kix_list_5-1{list-style-type:none}.lst-kix_drcsydhb4z7b-3>li{counter-increment:lst-ctn-kix_drcsydhb4z7b-3}.lst-kix_list_8-0>li:before{content:"\0025cf  "}ol.lst-kix_list_3-7{list-style-type:none}ul.lst-kix_list_5-2{list-style-type:none}ol.lst-kix_list_3-8{list-style-type:none}.lst-kix_list_8-7>li:before{content:"o  "}.lst-kix_lswckwbst11a-6>li{counter-increment:lst-ctn-kix_lswckwbst11a-6}.lst-kix_list_3-8>li:before{content:" "}.lst-kix_list_8-5>li:before{content:"\0025aa  "}.lst-kix_list_8-6>li:before{content:"\0025cf  "}ol.lst-kix_ocekgj1d9cfy-8.start{counter-reset:lst-ctn-kix_ocekgj1d9cfy-8 0}.lst-kix_list_8-3>li:before{content:"\0025cf  "}.lst-kix_list_3-6>li:before{content:"\002022  "}.lst-kix_list_3-7>li:before{content:" "}.lst-kix_tgkyxxtz27uw-0>li{counter-increment:lst-ctn-kix_tgkyxxtz27uw-0}.lst-kix_list_8-4>li:before{content:"o  "}ol.lst-kix_lswckwbst11a-6.start{counter-reset:lst-ctn-kix_lswckwbst11a-6 0}.lst-kix_drcsydhb4z7b-1>li{counter-increment:lst-ctn-kix_drcsydhb4z7b-1}ol.lst-kix_tgkyxxtz27uw-3.start{counter-reset:lst-ctn-kix_tgkyxxtz27uw-3 0}ol.lst-kix_drcsydhb4z7b-6.start{counter-reset:lst-ctn-kix_drcsydhb4z7b-6 0}.lst-kix_list_8-8>li:before{content:"\0025aa  "}.lst-kix_list_4-8>li:before{content:"\0025aa  "}.lst-kix_list_4-7>li:before{content:"o  "}ol.lst-kix_lswckwbst11a-1.start{counter-reset:lst-ctn-kix_lswckwbst11a-1 0}ul.lst-kix_list_4-8{list-style-type:none}ul.lst-kix_list_4-6{list-style-type:none}ul.lst-kix_list_4-7{list-style-type:none}ol.lst-kix_tgkyxxtz27uw-4.start{counter-reset:lst-ctn-kix_tgkyxxtz27uw-4 0}.lst-kix_ocekgj1d9cfy-2>li{counter-increment:lst-ctn-kix_ocekgj1d9cfy-2}ul.lst-kix_list_4-0{list-style-type:none}ol.lst-kix_ocekgj1d9cfy-3.start{counter-reset:lst-ctn-kix_ocekgj1d9cfy-3 0}ul.lst-kix_list_4-1{list-style-type:none}ul.lst-kix_list_4-4{list-style-type:none}ul.lst-kix_list_4-5{list-style-type:none}ol.lst-kix_list_2-7{list-style-type:none}.lst-kix_tgkyxxtz27uw-7>li{counter-increment:lst-ctn-kix_tgkyxxtz27uw-7}ul.lst-kix_list_4-2{list-style-type:none}ol.lst-kix_list_2-8{list-style-type:none}ul.lst-kix_list_4-3{list-style-type:none}.lst-kix_drcsydhb4z7b-5>li{counter-increment:lst-ctn-kix_drcsydhb4z7b-5}ol.lst-kix_tgkyxxtz27uw-0{list-style-type:none}ol.lst-kix_drcsydhb4z7b-7.start{counter-reset:lst-ctn-kix_drcsydhb4z7b-7 0}ol.lst-kix_tgkyxxtz27uw-1{list-style-type:none}.lst-kix_lswckwbst11a-3>li:before{content:"" counter(lst-ctn-kix_lswckwbst11a-3,decimal) ". "}ol.lst-kix_tgkyxxtz27uw-2{list-style-type:none}ol.lst-kix_tgkyxxtz27uw-3{list-style-type:none}ol.lst-kix_tgkyxxtz27uw-4{list-style-type:none}ol.lst-kix_tgkyxxtz27uw-5{list-style-type:none}.lst-kix_lswckwbst11a-4>li:before{content:"" counter(lst-ctn-kix_lswckwbst11a-4,lower-latin) ". "}.lst-kix_lswckwbst11a-5>li:before{content:"" counter(lst-ctn-kix_lswckwbst11a-5,lower-roman) ". "}ol.lst-kix_tgkyxxtz27uw-6{list-style-type:none}ol.lst-kix_tgkyxxtz27uw-7{list-style-type:none}ol.lst-kix_tgkyxxtz27uw-8{list-style-type:none}.lst-kix_lswckwbst11a-1>li{counter-increment:lst-ctn-kix_lswckwbst11a-1}ol.lst-kix_lswckwbst11a-7.start{counter-reset:lst-ctn-kix_lswckwbst11a-7 0}.lst-kix_lswckwbst11a-4>li{counter-increment:lst-ctn-kix_lswckwbst11a-4}ol.lst-kix_drcsydhb4z7b-1.start{counter-reset:lst-ctn-kix_drcsydhb4z7b-1 0}.lst-kix_lswckwbst11a-2>li:before{content:"" counter(lst-ctn-kix_lswckwbst11a-2,lower-roman) ". "}.lst-kix_list_7-0>li:before{content:"\0025cf  "}.lst-kix_lswckwbst11a-0>li:before{content:"" counter(lst-ctn-kix_lswckwbst11a-0,decimal) ". "}.lst-kix_lswckwbst11a-1>li:before{content:"" counter(lst-ctn-kix_lswckwbst11a-1,lower-latin) ". "}.lst-kix_list_2-6>li:before{content:"   "}.lst-kix_list_3-7>li{counter-increment:lst-ctn-kix_list_3-7}.lst-kix_list_2-4>li:before{content:"   "}.lst-kix_list_2-8>li:before{content:" "}.lst-kix_list_7-1>li:before{content:"o  "}.lst-kix_list_7-5>li:before{content:"\0025aa  "}.lst-kix_list_7-3>li:before{content:"\0025cf  "}ul.lst-kix_list_7-5{list-style-type:none}ul.lst-kix_list_7-6{list-style-type:none}ul.lst-kix_list_7-3{list-style-type:none}ul.lst-kix_list_7-4{list-style-type:none}.lst-kix_tgkyxxtz27uw-2>li:before{content:"" counter(lst-ctn-kix_tgkyxxtz27uw-2,lower-roman) ". "}ul.lst-kix_list_7-7{list-style-type:none}ul.lst-kix_list_7-8{list-style-type:none}ol.lst-kix_list_5-7{list-style-type:none}ol.lst-kix_ocekgj1d9cfy-6.start{counter-reset:lst-ctn-kix_ocekgj1d9cfy-6 0}.lst-kix_tgkyxxtz27uw-4>li:before{content:"" counter(lst-ctn-kix_tgkyxxtz27uw-4,lower-latin) ". "}ol.lst-kix_list_5-8{list-style-type:none}.lst-kix_lswckwbst11a-2>li{counter-increment:lst-ctn-kix_lswckwbst11a-2}ul.lst-kix_list_7-1{list-style-type:none}ol.lst-kix_tgkyxxtz27uw-7.start{counter-reset:lst-ctn-kix_tgkyxxtz27uw-7 0}ul.lst-kix_list_7-2{list-style-type:none}.lst-kix_list_1-7>li{counter-increment:lst-ctn-kix_list_1-7}ol.lst-kix_list_3-8.start{counter-reset:lst-ctn-kix_list_3-8 0}ul.lst-kix_list_7-0{list-style-type:none}.lst-kix_list_7-7>li:before{content:"o  "}.lst-kix_ocekgj1d9cfy-4>li{counter-increment:lst-ctn-kix_ocekgj1d9cfy-4}.lst-kix_tgkyxxtz27uw-8>li:before{content:"" counter(lst-ctn-kix_tgkyxxtz27uw-8,lower-roman) ". "}.lst-kix_lswckwbst11a-3>li{counter-increment:lst-ctn-kix_lswckwbst11a-3}.lst-kix_tgkyxxtz27uw-6>li:before{content:"" counter(lst-ctn-kix_tgkyxxtz27uw-6,decimal) ". "}.lst-kix_list_5-8>li{counter-increment:lst-ctn-kix_list_5-8}.lst-kix_list_4-0>li:before{content:"\0025cf  "}.lst-kix_list_3-8>li{counter-increment:lst-ctn-kix_list_3-8}ol.lst-kix_lswckwbst11a-2.start{counter-reset:lst-ctn-kix_lswckwbst11a-2 0}.lst-kix_tgkyxxtz27uw-3>li{counter-increment:lst-ctn-kix_tgkyxxtz27uw-3}.lst-kix_drcsydhb4z7b-7>li{counter-increment:lst-ctn-kix_drcsydhb4z7b-7}ol.lst-kix_list_1-7.start{counter-reset:lst-ctn-kix_list_1-7 0}.lst-kix_list_4-4>li:before{content:"o  "}.lst-kix_list_4-2>li:before{content:"\0025aa  "}.lst-kix_list_4-6>li:before{content:"\0025cf  "}.lst-kix_s3kte6nvgtp0-4>li:before{content:"\0025cb  "}.lst-kix_drcsydhb4z7b-6>li{counter-increment:lst-ctn-kix_drcsydhb4z7b-6}ol.lst-kix_tgkyxxtz27uw-8.start{counter-reset:lst-ctn-kix_tgkyxxtz27uw-8 0}.lst-kix_s3kte6nvgtp0-0>li:before{content:"\0025cf  "}.lst-kix_s3kte6nvgtp0-8>li:before{content:"\0025a0  "}.lst-kix_s3kte6nvgtp0-6>li:before{content:"\0025cf  "}.lst-kix_tgkyxxtz27uw-8>li{counter-increment:lst-ctn-kix_tgkyxxtz27uw-8}ul.lst-kix_list_6-6{list-style-type:none}ul.lst-kix_list_6-7{list-style-type:none}ul.lst-kix_list_6-4{list-style-type:none}ul.lst-kix_list_6-5{list-style-type:none}ul.lst-kix_list_6-8{list-style-type:none}ol.lst-kix_lswckwbst11a-5.start{counter-reset:lst-ctn-kix_lswckwbst11a-5 0}.lst-kix_list_1-0>li:before{content:"   "}.lst-kix_drcsydhb4z7b-0>li{counter-increment:lst-ctn-kix_drcsydhb4z7b-0}ul.lst-kix_list_6-2{list-style-type:none}ul.lst-kix_list_6-3{list-style-type:none}.lst-kix_list_1-2>li:before{content:"   "}ul.lst-kix_list_6-0{list-style-type:none}ul.lst-kix_list_6-1{list-style-type:none}.lst-kix_list_1-4>li:before{content:"   "}.lst-kix_list_1-6>li:before{content:"   "}.lst-kix_tgkyxxtz27uw-0>li:before{content:"" counter(lst-ctn-kix_tgkyxxtz27uw-0,decimal) ". "}.lst-kix_list_2-0>li:before{content:"   "}.lst-kix_tgkyxxtz27uw-2>li{counter-increment:lst-ctn-kix_tgkyxxtz27uw-2}.lst-kix_list_1-8>li:before{content:" "}.lst-kix_list_2-2>li:before{content:"   "}.lst-kix_ocekgj1d9cfy-3>li{counter-increment:lst-ctn-kix_ocekgj1d9cfy-3}ol.lst-kix_lswckwbst11a-4.start{counter-reset:lst-ctn-kix_lswckwbst11a-4 0}.lst-kix_s3kte6nvgtp0-2>li:before{content:"\0025a0  "}ol.lst-kix_ocekgj1d9cfy-7.start{counter-reset:lst-ctn-kix_ocekgj1d9cfy-7 0}ol{margin:0;padding:0}table td,table th{padding:0}.c6{-webkit-text-decoration-skip:none;color:#000000;font-weight:700;text-decoration:underline;vertical-align:baseline;text-decoration-skip-ink:none;font-size:11pt;font-family:"Arial";font-style:normal}.c19{margin-left:36pt;padding-top:0pt;padding-left:0pt;padding-bottom:0pt;line-height:2.0;orphans:2;widows:2;text-align:left}.c7{color:#000000;font-weight:400;text-decoration:none;vertical-align:baseline;font-size:18pt;font-family:"Arial";font-style:normal}.c0{color:#000000;font-weight:400;text-decoration:none;vertical-align:baseline;font-size:11pt;font-family:"Arial";font-style:normal}.c3{color:#000000;font-weight:700;text-decoration:none;vertical-align:baseline;font-size:11pt;font-family:"Arial";font-style:normal}.c8{-webkit-text-decoration-skip:none;color:#000000;text-decoration:underline;vertical-align:baseline;text-decoration-skip-ink:none;font-style:normal}.c32{padding-top:0pt;padding-bottom:10pt;line-height:1.0;orphans:2;widows:2;text-align:left}.c23{padding-top:9pt;padding-bottom:9pt;line-height:1.0;orphans:2;widows:2;text-align:left}.c13{-webkit-text-decoration-skip:none;color:#1155cc;text-decoration:underline;vertical-align:baseline;text-decoration-skip-ink:none;font-style:normal}.c36{padding-top:0pt;padding-bottom:0pt;line-height:1.0;orphans:2;widows:2;text-align:right}.c1{padding-top:0pt;padding-bottom:0pt;line-height:1.0;orphans:2;widows:2;text-align:left}.c30{padding-top:0pt;padding-bottom:0pt;line-height:1.0;orphans:2;widows:2;text-align:center}.c15{-webkit-text-decoration-skip:none;color:#1155cc;text-decoration:underline;text-decoration-skip-ink:none}.c33{font-weight:400;font-size:12pt;font-family:"Cambria";font-style:normal}.c20{padding-top:0pt;padding-bottom:0pt;line-height:1.15;text-align:left}.c29{font-weight:400;font-size:10pt;font-family:"Cambria";font-style:normal}.c25{color:#808080;font-weight:400;font-size:10pt;font-family:"Cambria"}.c18{color:#000000;text-decoration:none;vertical-align:super;font-style:normal}.c2{font-size:11pt;font-family:"Arial";font-weight:400}.c21{-webkit-text-decoration-skip:none;text-decoration:underline;text-decoration-skip-ink:none}.c14{color:#000000;text-decoration:none;vertical-align:baseline}.c12{font-size:11pt;font-family:"Arial";font-weight:700}.c10{text-decoration:none;vertical-align:baseline;font-style:normal}.c22{margin-left:0pt;padding-left:0pt}.c5{padding:0;margin:0}.c31{text-decoration:none;font-style:normal}.c11{color:inherit;text-decoration:inherit}.c28{text-decoration:none;vertical-align:baseline}.c34{max-width:540pt;padding:18pt 36pt 18pt 36pt}.c26{margin-left:36pt;padding-left:0pt}.c9{font-style:italic}.c24{vertical-align:super}.c27{margin-left:36pt}.c17{color:#4f81bd}.c35{color:#000000}.c16{background-color:#ffffff}.c4{height:12pt}.title{padding-top:24pt;color:#335b8a;font-weight:700;font-size:18pt;padding-bottom:12pt;font-family:"Calibri";line-height:1.0;page-break-after:avoid;orphans:2;widows:2;text-align:center}.subtitle{padding-top:12pt;color:#335b8a;font-weight:700;font-size:15pt;padding-bottom:12pt;font-family:"Calibri";line-height:1.0;page-break-after:avoid;orphans:2;widows:2;text-align:center}li{color:#000000;font-size:12pt;font-family:"Cambria"}p{margin:0;color:#000000;font-size:12pt;font-family:"Cambria"}h1{padding-top:24pt;color:#335b8a;font-weight:700;font-size:16pt;padding-bottom:0pt;font-family:"Calibri";line-height:1.0;page-break-after:avoid;orphans:2;widows:2;text-align:left}h2{padding-top:10pt;color:#4f81bd;font-weight:700;font-size:16pt;padding-bottom:0pt;font-family:"Calibri";line-height:1.0;page-break-after:avoid;orphans:2;widows:2;text-align:left}h3{padding-top:10pt;color:#4f81bd;font-weight:700;font-size:14pt;padding-bottom:0pt;font-family:"Calibri";line-height:1.0;page-break-after:avoid;orphans:2;widows:2;text-align:left}h4{padding-top:10pt;color:#4f81bd;font-weight:700;font-size:12pt;padding-bottom:0pt;font-family:"Calibri";line-height:1.0;page-break-after:avoid;orphans:2;widows:2;text-align:left}h5{padding-top:10pt;color:#4f81bd;font-size:12pt;padding-bottom:0pt;font-family:"Calibri";line-height:1.0;page-break-after:avoid;font-style:italic;orphans:2;widows:2;text-align:left}h6{padding-top:10pt;color:#4f81bd;font-size:12pt;padding-bottom:0pt;font-family:"Calibri";line-height:1.0;page-break-after:avoid;orphans:2;widows:2;text-align:left}</style></head><body class="c16 c34"><div><p class="c1 c4"><span class="c10 c25"></span></p><p class="c1 c4"><span class="c14 c33"></span></p></div><p class="c1"><span class="c7">Genoppi</span></p><p class="c20 c4"><span class="c0"></span></p><p class="c1"><span class="c6">Getting started</span></p><p class="c23"><span class="c0">Genoppi is a user-friendly application </span><span class="c2">integrating</span><span class="c0">&nbsp;quantitative proteomics and genetic datasets that enables quality control steps to be performed on mass spectrometry data. </span><span class="c2">Genoppi</span><span class="c0">&nbsp;allows interactive overlaying of genetic datasets, such as GWAS, exome sequencing, the Exome Aggregation Consortium</span><span class="c18 c2">1</span><span class="c0">&nbsp;(</span><span class="c2 c15"><a class="c11" href="https://www.google.com/url?q=http://exac.broadinstitute.org&amp;sa=D&amp;ust=1559064348749000">ExAC</a></span><span class="c0">) or other user-defined inputs </span><span class="c2">with</span><span class="c0">&nbsp;</span><span class="c2">protein-protein interaction data</span><span class="c0">.</span><span class="c2">&nbsp;</span><span class="c0">Genoppi can be fully deployed across Mac OS and Linux distributions</span><span class="c2">&nbsp;for local installation for personal use. To do so:</span></p><ul class="c5 lst-kix_list_7-0 start"><li class="c23 c26"><span class="c0">Download and install </span><span class="c2 c10 c17"><a class="c11" href="https://www.google.com/url?q=https://www.rstudio.com/products/rstudio/download/&amp;sa=D&amp;ust=1559064348750000">RStudio</a></span><span class="c0">&nbsp;and </span><span class="c2 c10 c17"><a class="c11" href="https://www.google.com/url?q=https://www.r-project.org/&amp;sa=D&amp;ust=1559064348750000">R version 3.3.x</a></span><span class="c0">. </span></li><li class="c23 c26"><span class="c0">Begin RStudio and run the following commands to install necessary R packages</span></li></ul><p class="c23" id="h.gjdgxs"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 576.00px; height: 126.07px;"><img alt="" src="images/image21.png" style="width: 576.00px; height: 126.07px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><ul class="c5 lst-kix_list_8-0 start"><li class="c23 c26"><span class="c0">To start the local version of Genoppi, </span><span class="c2">download the </span><span class="c2 c17"><a class="c11" href="https://www.google.com/url?q=https://github.com/lagelab/Genoppi&amp;sa=D&amp;ust=1559064348751000">github repository</a></span><span class="c2">, unzip and move the folder to a directory of your choice. R</span><span class="c0">un the following commands in RStudio.</span></li></ul><p class="c23" id="h.30j0zll"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 576.00px; height: 37.53px;"><img alt="" src="images/image23.png" style="width: 576.00px; height: 37.53px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span><span class="c0">&nbsp;</span></p><p class="c23"><span class="c0">Please direct your questions and comments to </span><span class="c2 c10 c17"><a class="c11" href="mailto:aprilkim@broadinstitute.org">aprilkim@broadinstitute.org</a></span><span class="c0">, </span><span class="c2 c10 c17"><a class="c11" href="mailto:edytam@broadinstitute.org">edytam@broadinstitute.org</a></span><span class="c0">&nbsp;and/or </span><span class="c2 c10 c17"><a class="c11" href="mailto:lage.kasper@mgh.harvard.edu">lage.kasper@mgh.harvard.edu</a></span><span class="c0">.</span><span class="c0">&nbsp;</span></p><p class="c1 c4"><span class="c6"></span></p><p class="c1 c4"><span class="c6"></span></p><p class="c1 c4"><span class="c6"></span></p><p class="c1 c4"><span class="c6"></span></p><p class="c1 c4"><span class="c6"></span></p><p class="c1 c4"><span class="c6"></span></p><p class="c1 c4"><span class="c6"></span></p><p class="c1 c4"><span class="c6"></span></p><p class="c1 c4"><span class="c6"></span></p><p class="c1 c4"><span class="c6"></span></p><p class="c1 c4"><span class="c6"></span></p><p class="c1 c4"><span class="c6"></span></p><p class="c1 c4"><span class="c6"></span></p><p class="c1 c4"><span class="c6"></span></p><hr style="page-break-before:always;display:none;"><p class="c32"><span class="c6">Basics</span></p><p class="c1"><span class="c0">Genoppi accepts i) experimental isobaric</span><span class="c2">&nbsp;tag-based (iTRAQ or TMT)</span><span class="c0">&nbsp;quantitative interaction proteomics data obtained from tandem mass spectrometry (MS/MS)</span><span class="c2">&nbsp;analysis of immunoprecipitations (IPs)</span><span class="c0">&nbsp;or ii) the MS/MS results from a whole proteome analysis. The input file must be a tab delimited plain text file. At minimum, the file must contain three columns specifying protein identifiers (either as gene </span><span class="c2">symbols</span><span class="c0">&nbsp;or as UniProt</span><span class="c18 c2">3</span><span class="c0">&nbsp;IDs [</span><span class="c2 c13"><a class="c11" href="https://www.google.com/url?q=http://www.uniprot.org&amp;sa=D&amp;ust=1559064348754000">www.uniprot.org</a></span><span class="c0">]), and two columns with</span><span class="c2">&nbsp;</span><span class="c0">protein enrichment values between studied conditions (as log2 fold change [FC]), one for each replicate. </span></p><p class="c1 c4"><span class="c0"></span></p><ul class="c5 lst-kix_list_4-0 start"><li class="c1 c22"><span class="c12">Column 1:</span><span class="c2">&nbsp;protein identifiers as either HUGO Gene Nomenclature Committee</span><span class="c2 c24">2</span><span class="c2">&nbsp;[</span><span class="c15 c2"><a class="c11" href="https://www.google.com/url?q=http://www.genenames.org&amp;sa=D&amp;ust=1559064348755000">HGNC</a></span><span class="c2">]</span><span class="c0">&nbsp;approved symbols, or </span><span class="c2">UniProt</span><span class="c2 c24">3</span><span class="c2">&nbsp;[</span><span class="c15 c2"><a class="c11" href="https://www.google.com/url?q=http://www.uniprot.org&amp;sa=D&amp;ust=1559064348756000">www.uniprot.org</a></span><span class="c2">]</span><span class="c0">&nbsp;accession numbers</span></li><li class="c1 c22"><span class="c12">Columns 2, 3:</span><span class="c2">&nbsp;logFC values for two replicates observed in the IP experiment </span><span class="c3">OR</span></li><li class="c1 c22"><span class="c12">Columns 2, 3, 4:</span><span class="c2">&nbsp;</span><span class="c0">results of 1-sample moderated t-test on two replicates</span><span class="c2 c14 c9">:</span><span class="c0">&nbsp;average logFC, </span><span class="c2 c14 c9">P</span><span class="c0">&nbsp;and false discovery rate (FDR) values</span></li><li class="c1 c22"><span class="c0">All fields must be filled with no missing values. Any rows with missing values will be disregarded with no error message.</span></li></ul><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c0">Examples of correct column headers and accepted values:</span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c3">1. HGNC symbol and logFC for two replicates</span></p><p class="c1"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 269.49px; height: 105.60px;"><img alt="" src="images/image22.png" style="width: 269.49px; height: 105.60px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span><span class="c0">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp;</span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c3">2. UniProt accession numbers and logFC for two replicates</span></p><p class="c1"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 314.70px; height: 105.60px;"><img alt="" src="images/image25.png" style="width: 314.70px; height: 105.60px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span><span class="c0">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp;</span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c3">3. HGNC symbol and results of statistical analyses</span></p><p class="c1"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 374.40px; height: 124.80px;"><img alt="" src="images/image24.png" style="width: 374.40px; height: 124.80px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span><span class="c0">&nbsp; &nbsp; &nbsp; &nbsp;</span></p><p class="c1"><span class="c0">&nbsp; &nbsp;</span></p><p class="c1 c4"><span class="c3"></span></p><p class="c1 c4"><span class="c3"></span></p><p class="c1"><span class="c3">4. UniProt accession numbers and results of statistical analyses</span></p><p class="c1"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 430.88px; height: 124.80px;"><img alt="" src="images/image26.png" style="width: 430.88px; height: 124.80px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span><span class="c0">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;</span></p><p class="c1"><span class="c0">&nbsp; </span></p><p class="c1"><span class="c2">If the protein enrichment values from two replicates are given, a moderated </span><span class="c2 c9">t-</span><span class="c2">test from the limma</span><span class="c2 c24">15</span><span class="c2">&nbsp;package is applied t</span><span class="c0">o calculate </span><span class="c2 c14 c9">P</span><span class="c0">-values</span><span class="c0">&nbsp;</span><span class="c2">for the significance</span><span class="c0">&nbsp;of a protein enrichment</span><span class="c2">. </span><span class="c0">Limma was originally introduced to robustly identify differentially expressed genes in microarray experiments and has </span><span class="c2">since </span><span class="c0">been used on a variety of data types, including proteomics results</span><span class="c2 c24">4,5</span><span class="c0">. The empirical Bayes moderated t-test is used in Genoppi, </span><span class="c2">because it</span><span class="c0">&nbsp;</span><span class="c2">is less sensitive to underestimated sample variances and performs best on small sample sizes compared to the classical </span><span class="c2 c9">t</span><span class="c2">-test</span><span class="c2 c24">4</span><span class="c0">.</span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c0">If your data is in Excel format and you are a Mac user, please convert it to text file by selecting &lsquo;File&rsquo; -&gt; &lsquo;Save As&hellip;&rsquo; and choose &lsquo;</span><span class="c2">Tab delimited Text</span><span class="c0">&nbsp;(.txt)&rsquo; or &lsquo;Comma Separated Values (.csv)&rsquo; as shown below. All other formats create a file that terminates each line with a carriage return character, which is incompatible with subsequent analysis in Genoppi. </span></p><p class="c1 c4"><span class="c0"></span></p><p class="c20"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 358.00px; height: 188.00px;"><img alt="" src="images/image3.png" style="width: 358.00px; height: 188.00px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c1 c4"><span class="c6"></span></p><p class="c1 c4"><span class="c6"></span></p><p class="c1"><span class="c6">Quality control</span></p><p class="c1 c4"><span class="c6"></span></p><p class="c1"><span class="c0">Upon data upload, interactive volcano and scatter plots will be generated as shown below. The scatter plot is only </span><span class="c2">generated</span><span class="c0">&nbsp;if the user input file includes rep1 and rep2 values. </span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 576.00px; height: 577.33px;"><img alt="" src="images/image2.png" style="width: 576.00px; height: 577.33px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c0">The initial FDR threshold for color scale is 0.1. Results are shown as a volcano plot where logFC values of the bait versus control conditions are plotted on the x-axis and the negative lo</span><span class="c2 c10">g10 transformed </span><span class="c2 c9 c28">P-</span><span class="c2 c10">values </span><span class="c2">of enrichment</span><span class="c2 c10">&nbsp;are plotted on the y-axis. </span><span class="c2">Enriched i</span><span class="c2 c10">nteraction partners with an FDR </span><span class="c2">&le;</span><span class="c2 c10">&nbsp;0.1 are in green, and </span><span class="c2">detected proteins</span><span class="c2 c10">&nbsp;with an FDR </span><span class="c2">&gt;</span><span class="c2 c10">&nbsp;0.1 are in blue. </span><span class="c2 c16">Changing the</span><span class="c2 c16">&nbsp;FDR value threshold in the left panel will adapt colors in the interactive plot to reflect that change.</span><span class="c2 c10">&nbsp;</span><span class="c2">The u</span><span class="c2 c10">ser is able</span><span class="c0">&nbsp;to change the colors determined by the chosen FDR threshold using &ldquo;</span><span class="c2">C</span><span class="c0">olor selection for markers.&rdquo;</span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c0">Hovering over each dot, which represents a protein in </span><span class="c2">the</span><span class="c0">&nbsp;</span><span class="c2">data</span><span class="c0">, will show the corresponding HGNC gene symbol. </span><span class="c0">The table to the right</span><span class="c0">&nbsp;of the volcano plot shows the total count of enriched interactors in regions specified by the threshold controllers on the left panel.</span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c8 c2">Color by user value</span></p><p class="c1 c4"><span class="c8 c2"></span></p><p class="c1"><span class="c2">The </span><span class="c0">&ldquo;</span><span class="c2">u</span><span class="c0">ser </span><span class="c2">value</span><span class="c0">&rdquo; color scheme allows the user to upload genes and their associated values</span><span class="c2">&nbsp;to be mapped on a color scale</span><span class="c0">. </span><span class="c2">The u</span><span class="c0">ser can upload external datasets such as Project Achilles</span><span class="c18 c2">8,9,10,11</span><span class="c0">, or RNA-Seq expression values associated with genes as a text file. The file must be in tabular format as delimited text with &lsquo;gene&rsquo; and &lsquo;score&rsquo; as column headers.</span></p><p class="c1"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 185.76px; height: 163.20px;"><img alt="" src="images/image27.png" style="width: 185.76px; height: 163.20px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c2">A </span><span class="c0">discrete color scale assigns different colors to each score reported in the user file and a continuous color scale creates a gradient between the </span><span class="c2">reported </span><span class="c0">minimum and maximum scores. The color theme can then be changed using &lsquo;color selection for markers.&rsquo; The example plot shown below is plotted using the YlGnBu (YellowGreenBlue) color</span><span class="c2">map</span><span class="c0">.</span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c8 c2">Overlay InWeb</span></p><p class="c1 c4"><span class="c8 c2"></span></p><p class="c1"><span class="c2">The user can further </span><span class="c2">QC</span><span class="c2">&nbsp;proteomics data by identifying known human protein interaction partners from InWeb_InBioMap</span><span class="c2 c24">7 </span><span class="c2">(www.intomics.com/inbio/map.html#downloads), which contains known protein-protein interactions (PPI) compiled from &gt;40,000 articles. Overlaying this information enables users to easily distinguish new interactions from those already reported in the literature. InWeb_InBioMap interactors of the bait protein are highlighted in yellow in the figure below.</span><span class="c0">&nbsp;</span></p><p class="c1"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 576.00px; height: 353.00px;"><img alt="" src="images/image19.png" style="width: 576.00px; height: 391.00px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1 c4"><span class="c8 c2"></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c6">Overlay with genetics</span></p><p class="c1 c4"><span class="c8 c2"></span></p><p class="c1"><span class="c0">Genoppi allows </span><span class="c2">overlaying</span><span class="c0">&nbsp;experimental proteomics data with</span></p><ul class="c5 lst-kix_list_6-0 start"><li class="c1 c22"><span class="c0">SNPs from genetic studies,</span></li><li class="c1 c22"><span class="c0">user</span><span class="c2">-</span><span class="c0">uploaded genes (e.g.&nbsp;set of interesting genes from </span><span class="c2">Genotype Tissue Expression</span><span class="c2 c24 c31">12</span><span class="c2 c10">&nbsp;(GTE</span><span class="c2">x, </span><span class="c2 c10">http://www.gtexportal.org/home/</span><span class="c0">) to highlight corresponding proteins and</span></li><li class="c1 c22"><span class="c0">proteins encoded by genes that are intolerant to loss-of-function mutations as determined by </span><span class="c2 c10">ExAC</span><span class="c0">&nbsp;data ana</span><span class="c2">lysis</span><span class="c0">. &nbsp;</span></li></ul><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c8 c2">SNP</span><span class="c2 c21">-</span><span class="c8 c2">to</span><span class="c2 c21">-</span><span class="c8 c2">gene mapping</span></p><p class="c1 c4"><span class="c2 c8"></span></p><p class="c1"><span class="c0">Genoppi hosts a pre-calculated pairwise measure of linkage disequilibrium (LD), which is used to perform SNP-to-gene mapping using the 1000 Genomes Project</span><span class="c18 c2">13</span><span class="c0">&nbsp;(</span><span class="c15 c2"><a class="c11" href="https://www.google.com/url?q=http://www.internationalgenome.org&amp;sa=D&amp;ust=1559064348768000">http://www.internationalgenome.org</a></span><span class="c2">) </span><span class="c0">i</span><span class="c0">mputed single-nucleotide polymorphism (SNP) results. </span><span class="c2">The u</span><span class="c0">ser may provide a list of SNPs, found in the 1000 Genomes Project, that are significantly associated </span><span class="c2">with a</span><span class="c0">&nbsp;trait of interest to the experiment as a plain text file in the format shown below, with no column header. </span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 81.90px; height: 58.56px;"><img alt="" src="images/image28.png" style="width: 81.90px; height: 58.56px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c23 c4"><span class="c0"></span></p><p class="c23"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 576.00px; height: 513.33px;"><img alt="" src="images/image12.png" style="width: 576.00px; height: 513.33px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c23 c4"><span class="c0"></span></p><p class="c1"><span class="c2">The u</span><span class="c0">ser may change the color theme for SNP markers by using the drop-down menu titled &ldquo;SNPs&rdquo;.</span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1 c4"><span class="c8 c2"></span></p><p class="c1"><span class="c8 c2">Genes of interest</span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c0">This feature allows highlighting proteins corresponding to the user&rsquo;s uploaded genes of interest (e.g. a set of genes from GTEx project</span><span class="c2">, a set of genes from </span><span class="c0">permutation-based </span><span class="c2 c14 c9">P</span><span class="c0">-value calculations). </span><span class="c2">The u</span><span class="c0">ser must upload the ge</span><span class="c2">ne list with HGNC symbols as a plain text file</span><span class="c0">. </span><span class="c2">The c</span><span class="c0">olumn header must contain the name describing the group of genes as shown below. </span><span class="c0">There </span><span class="c2">is</span><span class="c0">&nbsp;no limit</span><span class="c2">&nbsp;to the number of columns</span><span class="c0">. </span></p><p class="c1"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 193.37px; height: 143.04px;"><img alt="" src="images/image29.png" style="width: 193.37px; height: 143.04px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c0">A detailed </span><span class="c2">count of overlaps</span><span class="c0">&nbsp;and hypergeometric </span><span class="c2 c14 c9">P</span><span class="c0">-value is shown to the right of the volcano plot. </span><span class="c2">The u</span><span class="c0">ser may change the color </span><span class="c2">scheme</span><span class="c0">&nbsp;by using the drop-down menu titled &lsquo;</span><span class="c0">Genes of interest</span><span class="c0">&rsquo;.</span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c2">Th</span><span class="c2">e u</span><span class="c0">ser may upload </span><span class="c2">any combination of </span><span class="c0">input files for SNP</span><span class="c2">-</span><span class="c0">to</span><span class="c2">-</span><span class="c0">gene mapping</span><span class="c2">&nbsp;and </span><span class="c0">genes of interest, </span><span class="c2">and</span><span class="c0">&nbsp;HGNC symbol</span><span class="c0">&nbsp;for InWeb_InBi</span><span class="c2">oMap</span><span class="c0">&nbsp;</span><span class="c2">interaction partners</span><span class="c0">. </span><span class="c2">The r</span><span class="c0">esulting plot will highlight all </span><span class="c2">identified</span><span class="c0">&nbsp;proteins from user-input data. </span><span class="c0">Overlaying multiple datasets could </span><span class="c2">result in a densely overloaded plot, </span><span class="c0">in which case the user can choose to </span><span class="c2">remove </span><span class="c0">labels or </span><span class="c2">temporary hide</span><span class="c0">&nbsp;markers by </span><span class="c0">clicking</span><span class="c0">&nbsp;on the legend </span><span class="c2">to be hidden</span><span class="c0">.</span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c36"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 266.00px; height: 57.00px;"><img alt="" src="images/image11.png" style="width: 560.00px; height: 579.00px; margin-left: -294.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c1"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 560.00px; height: 397.00px;"><img alt="" src="images/image11.png" style="width: 560.00px; height: 579.00px; margin-left: 0.00px; margin-top: -182.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1 c4"><span class="c8 c2"></span></p><p class="c1"><span class="c8 c2">Intolerance to LoF mutations determined by ExAC data</span></p><p class="c1 c4"><span class="c8 c2"></span></p><p class="c1"><span class="c0">If the user would like to identify proteins encoded by genes that are likely intolerant to loss-of-function variation using ExAC data, the user can </span><span class="c2">select</span><span class="c0">&nbsp;the &ldquo;ExAC&rdquo; </span><span class="c2">color scheme </span><span class="c0">on the left panel. </span><span class="c2">G</span><span class="c0">enes likely to be extremely intolerant to loss-of-function (LoF) variation </span><span class="c2">are computed using</span><span class="c0">&nbsp;ExAC pLI scores. Genes marked as pLI </span><span class="c2">&ge; </span><span class="c0">0.9 are defined as extremely intolerant </span><span class="c2">to LoF variations whereby g</span><span class="c0">enes marked as pLI </span><span class="c2 c31 c35">&le;</span><span class="c0">&nbsp;0.</span><span class="c2">1</span><span class="c0">&nbsp;are defined as tolerant </span><span class="c2">to LoF variations</span><span class="c0">.</span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 585.60px; height: 424.59px;"><img alt="" src="images/image30.png" style="width: 585.60px; height: 424.59px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1 c4"><span class="c6"></span></p><p class="c1 c4"><span class="c6"></span></p><p class="c1"><span class="c6">Significance of overlaps</span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c0">When genetics datasets are integrated with proteomics data, the overlaps are statistically tested and the results are made available to users. </span><span class="c2">S</span><span class="c0">tatistical significance can help determine, for example, the overlap of interactors and genes genetically linked to a particular disease. Hypergeometric distribution</span><span class="c2 c18">14</span><span class="c0">&nbsp;is used to perform gene </span><span class="c2">set</span><span class="c0">&nbsp;enrichment analys</span><span class="c2">e</span><span class="c0">s between experimental proteomics data and 1) uploaded gene list(s) and/or 2) known protein-protein interactions from InWeb_InBioMap. The input file format used in the &lsquo;Integrated Plots&rsquo; for genes of interest and InWeb_InBio</span><span class="c2">Map</span><span class="c0">&nbsp;input are identical to those used in the &lsquo;Venn Diagrams&rsquo; tab.</span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c0">Using genes of interest as an example, the user can adjust FDR, </span><span class="c2 c9 c14">P</span><span class="c0">-value and logFC sliders to select the values for significance threshold as shown in the figure below. </span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 576.00px; height: 96.00px;"><img alt="Macintosh HD:Users:aprilkim:Desktop:my-files:Shiny:documentation:goi_vd_subset.png" src="images/image20.png" style="width: 576.00px; height: 289.00px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c1"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 576.00px; height: 154.00px;"><img alt="Macintosh HD:Users:aprilkim:Desktop:my-files:Shiny:documentation:goi_vd_subset.png" src="images/image20.png" style="width: 576.00px; height: 289.00px; margin-left: 0.00px; margin-top: -135.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c0">Upon threshold values selection, the overlap between experimental data and the InWeb database is identified to determine the </span><span class="c0">population size</span><span class="c2">. Successes in a population are determined as </span><span class="c2">the</span><span class="c2">&nbsp;proteins that passed</span><span class="c0">&nbsp;the user</span><span class="c2">-</span><span class="c0">defined significance threshold. </span><span class="c2">Next</span><span class="c0">, the list of known interacting proteins of the user-defined bait is identified from the InWeb database and overlapped with the experimental data to determine the sample size. Lastly, proteins that are successes in the population and identified as InWeb interactors are determined as successes in the sample. Using this distribution, a </span><span class="c0">hypergeometric test</span><span class="c0">&nbsp;is run to calculate the statistical significance and the Venn diagram is generated to visualize the overlap </span><span class="c2">between successes in</span><span class="c0">&nbsp;population and sample. </span><span class="c2">The u</span><span class="c0">ser can adjust the analysis thresholds and have the changes (hypergeometric </span><span class="c2 c14 c9">P</span><span class="c0">-value and counts) reflected in the Venn diagram. The counts in each field of the Venn diagram are based on unique gene names. </span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 288.00px; height: 240.76px;"><img alt="" src="images/image31.png" style="width: 288.00px; height: 240.76px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c0">1. </span><span class="c14 c12 c9">P-</span><span class="c3">value</span><span class="c0">: hypergeometric </span><span class="c2 c14 c9">P</span><span class="c0">-value</span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c0">2. </span><span class="c3">population (N)</span><span class="c0">: subset of gene </span><span class="c2">symbols</span><span class="c0">&nbsp;from proteomics data found in HUGO Gene Nomenclature Committee (guidelines for Human Gene Nomenclature) (total number of genes)</span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c0">3. </span><span class="c3">number of successes in population (n)</span><span class="c0">: </span><span class="c0">category</span><span class="c0">&nbsp;A represents proteomics data present in HGNC database that passes the user</span><span class="c2">-</span><span class="c0">defined significance threshold (all significant bait interactors)</span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c0">4. </span><span class="c3">sample (S)</span><span class="c0">: category B represents user</span><span class="c2">-</span><span class="c0">uploaded genes found in the HGNC database OR InWeb_InBioMap interactors found in proteomics data &nbsp;</span></p><p class="c1 c4"><span class="c8 c2"></span></p><p class="c1"><span class="c0">If the user uploads multiple gene lists, denoted by different columns in one text file, Venn diagram representations and hypergeometric </span><span class="c2 c14 c9">P</span><span class="c0">-values for each list of genes will be generated and calculated, respectively. </span><span class="c2">The u</span><span class="c0">ser is able to </span><span class="c2">browse</span><span class="c0">&nbsp;through the multiple </span><span class="c0">gene lists</span><span class="c0">&nbsp;by clicking on the down arrow and selecting the category name, which corresponds to the column header uploaded by the user. </span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 573.00px; height: 88.00px;"><img alt="" src="images/image4.png" style="width: 573.00px; height: 470.00px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c1"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 573.00px; height: 309.00px;"><img alt="" src="images/image4.png" style="width: 573.00px; height: 470.00px; margin-left: 0.00px; margin-top: -161.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c20 c4"><span class="c0"></span></p><p class="c1 c4"><span class="c6"></span></p><p class="c1 c4"><span class="c6"></span></p><p class="c1 c4"><span class="c6"></span></p><p class="c1 c4"><span class="c6"></span></p><p class="c1"><span class="c6">Protein families and </span><span class="c12 c21">subcellular localizations</span></p><p class="c1 c4"><span class="c6"></span></p><p class="c1"><span class="c2">Genoppi provides an option </span><span class="c2">to overlay protein families or subcellular localizations</span><span class="c2">&nbsp;to explore the diversity of protein functions in the proteomics results. An example of how protein families are displayed on a volcano plot is shown below..</span></p><p class="c1 c4"><span class="c6"></span></p><p class="c1"><span class="c0">To identify which protein families the </span><span class="c2">identified proteins</span><span class="c0">&nbsp;belong to, experimental pull</span><span class="c2">-</span><span class="c0">down data represented on a volcano and scatter plots can be overlaid with information on protein families of enriched genes based on user-defined significance (2). </span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 629.64px; height: 576.00px;"><img alt="Macintosh HD:Users:aprilkim:Desktop:my-files:Shiny:documentation:pf_vp.png" src="images/image8.png" style="width: 629.64px; height: 576.00px; margin-left: -0.00px; margin-top: -0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c0">By default, the &lsquo;Turn on/off marker sizing&rsquo; option is set </span><span class="c2">on</span><span class="c0">&nbsp;</span><span class="c2">&lsquo;</span><span class="c0">on</span><span class="c2">&rsquo;</span><span class="c0">&nbsp;and the user </span><span class="c2">can change the markers size using the</span><span class="c0">&nbsp;&lsquo;Marker size&rsquo; slider</span><span class="c2">, </span><span class="c0">as shown below. </span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 609.88px; height: 556.80px;"><img alt="Macintosh HD:Users:aprilkim:Desktop:my-files:Shiny:documentation:pf_vp_diff.png" src="images/image9.png" style="width: 609.88px; height: 556.80px; margin-left: -0.00px; margin-top: -0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c0">When </span><span class="c2">&ldquo;</span><span class="c0">marker siz</span><span class="c2">ing&rdquo;</span><span class="c0">&nbsp;is turned on, </span><span class="c2">it</span><span class="c0">&nbsp;is a direct representation of how many proteins belong to the same protein family</span><span class="c2">&nbsp;as </span><span class="c0">indicated by different colors. Hovering over the marker will display protein identifiers using HGNC symbols, the protein family identified, and the number of other proteins belonging to the same protein family in the user-defined region. </span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c0">Turning off </span><span class="c2">&ldquo;</span><span class="c0">mark</span><span class="c2">er</span><span class="c0">&nbsp;siz</span><span class="c2">ing&rdquo;</span><span class="c0">&nbsp;will </span><span class="c2">display</span><span class="c0">&nbsp;uniform</span><span class="c2">-</span><span class="c0">sized markers as shown in the figure below. </span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 619.47px; height: 547.20px;"><img alt="Macintosh HD:Users:aprilkim:Desktop:my-files:Shiny:documentation:pf_vp_uniform.png" src="images/image10.png" style="width: 619.47px; height: 547.20px; margin-left: -0.00px; margin-top: -0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c1 c4"><span class="c8 c2"></span></p><p class="c1 c4"><span class="c8 c2"></span></p><p class="c1"><span class="c8 c2">Protein families as scatter plot</span></p><p class="c1 c4"><span class="c8 c2"></span></p><p class="c1"><span class="c0">This scatter plot is </span><span class="c2">only </span><span class="c0">applicable if the uploaded proteomics data contains two replicates, rep1 and rep2. Visualizing protein families in a scatter plot is especially useful when identifying general outliers and background contaminants</span><span class="c2">, such as</span><span class="c0">&nbsp;proteins that interact with affinity reagents or epitope tags. These outliers and contaminants often have poor correlation between two replicates and are plotted outside of the diagonal as shown below. </span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 305.07px; height: 281.60px;"><img alt="Macintosh HD:Users:aprilkim:Desktop:my-files:Shiny:documentation:poor_quality.png" src="images/image13.png" style="width: 305.07px; height: 281.60px; margin-left: -0.00px; margin-top: -0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 136.06px; height: 141.07px;"><img alt="" src="images/image1.png" style="width: 136.06px; height: 141.07px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c0">One method to troubleshoot and identify background contaminants from </span><span class="c2 c14 c9">bona fide</span><span class="c0">&nbsp;interactors is to identify which protein families the outliers (red box region above) belong to. Users can define the limits of rep1 and rep2 values using the appropriate sliders and use the preview area on the right to select a region of interest, as shown in the figure below. </span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 614.40px; height: 324.27px;"><img alt="Macintosh HD:Users:aprilkim:Desktop:my-files:Shiny:documentation:pf_sp.png" src="images/image14.png" style="width: 614.40px; height: 324.27px; margin-left: -0.00px; margin-top: -0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1 c4"><span class="c0"></span></p><hr style="page-break-before:always;display:none;"><p class="c20 c4"><span class="c0"></span></p><p class="c1 c4"><span class="c6"></span></p><p class="c1"><span class="c8 c2">Protein families removal</span></p><p class="c1 c4"><span class="c6"></span></p><p class="c1"><span class="c0">The &ldquo;Protein families&rdquo; field (1) in the Quality Control panel is a cleanup function designed to QC proteomics data by removing protein family members considered as contaminants or </span><span class="c2">&lsquo;</span><span class="c0">sticky</span><span class="c2">&rsquo;</span><span class="c0">&nbsp;proteins. This feature is only available if the input data contains either HGNC approved symbols or UniProt accession numbers with logFC values for two replicates observed in the IP experiment.</span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c0">The drop</span><span class="c2">-</span><span class="c0">down menu</span><span class="c2">&nbsp;</span><span class="c0">shows all protein families present in the HGNC database and </span><span class="c2">more than one</span><span class="c0">&nbsp;protein famil</span><span class="c2">y</span><span class="c0">&nbsp;can be selected. Any proteins in </span><span class="c2">the dataset matching </span><span class="c0">the selected protein family will be highlighted in the plot and appear </span><span class="c2">i</span><span class="c0">n the legend. The matching protein families coloring can be changed using the &lsquo;protein families color&rsquo; drop</span><span class="c2">-</span><span class="c0">down menu.</span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c0">If there are any matches, the &ldquo;Remove selected PF from input&rdquo; button will appear, through which the user </span><span class="c2">may</span><span class="c0">&nbsp;download an updated </span><span class="c2">input</span><span class="c0">&nbsp;data </span><span class="c2">where</span><span class="c0">&nbsp;the proteins belonging to the </span><span class="c2">selected</span><span class="c0">&nbsp;protein families are removed.</span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 585.60px; height: 407.06px;"><img alt="" src="images/image15.png" style="width: 585.60px; height: 407.06px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1 c4"><span class="c6"></span></p><p class="c1 c4"><span class="c6"></span></p><p class="c1 c4"><span class="c6"></span></p><p class="c1 c4"><span class="c6"></span></p><p class="c1 c4"><span class="c6"></span></p><p class="c1 c4"><span class="c6"></span></p><p class="c1"><span class="c6">Multiple files comparison (MFC)</span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c0">Users can upload up to three proteomics experiment results to perform comparative analys</span><span class="c2">e</span><span class="c0">s. The QC panel is similar to that of the single file upload and users can visualize up to three results as volcano and scatter plots, search for a protein of interest using its corresponding gene name, and search for and/or remove groups of genes belonging to </span><span class="c2">specific</span><span class="c0">&nbsp;protein families. Multiple files comparison features </span><span class="c2">require a minimum</span><span class="c0">&nbsp;of two input files. Users can change </span><span class="c2">t</span><span class="c0">he marker colors </span><span class="c2">by selecting</span><span class="c0">&nbsp;the drop-down option titled &lsquo;Color selection for markers.&rsquo; Genoppi interface with three </span><span class="c2">loaded </span><span class="c0">sample files is shown in the figure below. </span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 595.20px; height: 414.51px;"><img alt="" src="images/image16.png" style="width: 595.20px; height: 414.51px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c8 c2">Protein comparison</span></p><p class="c1 c4"><span class="c8 c2"></span></p><p class="c1"><span class="c0">FDR, </span><span class="c2 c14 c9">P</span><span class="c0">-value and logFC controllers are generated for each uploaded proteomics dataset in the &lsquo;Protein Comparison&rsquo; tab as shown below. </span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 574.93px; height: 238.93px;"><img alt="Macintosh HD:Users:aprilkim:Desktop:my-files:Shiny:documentation:protein_comaparison copy.png" src="images/image17.png" style="width: 574.93px; height: 238.93px; margin-left: -0.00px; margin-top: -0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c0">After setting the significance thresholds, clicking on the &lsquo;Generate plots&rsquo; button will generate comparative volcano plots of all user-uploaded data. The comparison is performed using </span><span class="c0">HGNC symbols</span><span class="c0">&nbsp;as </span><span class="c2">protein identifiers </span><span class="c0">and the color scheme </span><span class="c2">is</span><span class="c0">&nbsp;as explained below.</span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 213.33px; height: 213.33px;"><img alt="Macintosh HD:Users:aprilkim:Documents:lagelab_github:Genoppi:www:pc_3.png" src="images/image18.png" style="width: 213.33px; height: 213.33px; margin-left: -0.00px; margin-top: -0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c3">f1</span><span class="c0">: marked in red are proteins unique to file1</span></p><p class="c1"><span class="c3">f2</span><span class="c0">: marked in yellow are proteins unique to file2</span></p><p class="c1"><span class="c3">f3</span><span class="c0">: marked in blue are proteins unique to file3</span></p><p class="c1"><span class="c3">f12</span><span class="c0">: marked in orange are proteins identified in both file1 and file2</span></p><p class="c1"><span class="c3">f13</span><span class="c0">: marked in purple are proteins identified in both file1 and file3</span></p><p class="c1"><span class="c3">f23</span><span class="c0">: marked in green are proteins identified in both file2 and file3</span></p><p class="c1"><span class="c3">f123</span><span class="c0">: marked in gray are proteins identified in file1, file2 and file3</span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c2">V</span><span class="c0">olcano plots show </span><span class="c2">the</span><span class="c0">&nbsp;uploaded data as background markers in white/gray and colored markers belonging to different protein groups as explained above (i.e. unique to file1, in both file2 and file3 etc.)</span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 604.80px; height: 209.02px;"><img alt="" src="images/image5.png" style="width: 604.80px; height: 209.02px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c0">To easily identify what proteins and how many of them are unique to each group, a Venn diagram </span><span class="c2">represents protein distributions across files</span><span class="c0">&nbsp;and the table to the right lists the proteins using HGNC symbols </span><span class="c0">unique to each </span><span class="c2">file</span><span class="c0">. </span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 604.80px; height: 365.96px;"><img alt="" src="images/image6.png" style="width: 604.80px; height: 365.96px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c2 c21">MFC</span><span class="c8 c2">&nbsp;integrative analysis</span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c0">The integrative analysis using genetics data and identifying protein families is performed </span><span class="c2">similarly to</span><span class="c0">&nbsp;the single file analysis. Please refer to the </span><span class="c2">integrative analysis </span><span class="c0">on the single file tab </span><span class="c2">f</span><span class="c0">or </span><span class="c2">a </span><span class="c0">detailed description. </span></p><p class="c20 c4"><span class="c0"></span></p><p class="c20 c4"><span class="c0"></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c6">Downloads</span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c0">Users can download all generated data as plain text files and interactive plots as HTML reports under the &lsquo;Download&rsquo; tab. Download buttons are only active once the relevant data and/or plots have been generated. </span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1" id="h.1fob9te"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 576.00px; height: 565.33px;"><img alt="Macintosh HD:Users:aprilkim:Desktop:my-files:Shiny:documentation:download.png" src="images/image7.png" style="width: 576.00px; height: 565.33px; margin-left: -0.00px; margin-top: -0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c20 c4"><span class="c0"></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1 c4"><span class="c0"></span></p><p class="c1"><span class="c3">References</span></p><p class="c1 c4"><span class="c0"></span></p><ol class="c5 lst-kix_tgkyxxtz27uw-0 start" start="1"><li class="c1 c26"><span class="c2">Lek, M. </span><span class="c2 c9">et al.</span><span class="c2">&nbsp;</span><span class="c2 c9">Nature</span><span class="c2">&nbsp;</span><span class="c12">536</span><span class="c0">, 285&ndash;291 (2016).</span></li></ol><p class="c1 c4 c27"><span class="c0"></span></p><ol class="c5 lst-kix_tgkyxxtz27uw-0" start="2"><li class="c19"><span class="c2">Yates, B. </span><span class="c2 c9">et al.</span><span class="c2">&nbsp;</span><span class="c2 c9">Nucleic Acids Res.</span><span class="c2">&nbsp;</span><span class="c12">45</span><span class="c0">, D619-D625 (2017)</span></li><li class="c19"><span class="c2">The UniProt Consortium, </span><span class="c2 c9">Nucleic Acids Res</span><span class="c2">. </span><span class="c12">45</span><span class="c0">, D158-D169 (2017).</span></li><li class="c19"><span class="c2">Smyth, G. K. </span><span class="c2 c9">Stat. Appl. Genet. Mol. Biol.</span><span class="c2">&nbsp;</span><span class="c12">3</span><span class="c0">, 1&ndash;25 (2004).</span></li><li class="c19"><span class="c0">Agnetti, G., Lindsey, M. L. &amp; Foster, D. B. (Springer, 2016).</span></li><li class="c19"><span class="c2">Markowski, C. A. &amp; Markowski, E. P. </span><span class="c2 c9">Am. Stat.</span><span class="c2">&nbsp;</span><span class="c12">44</span><span class="c0">, 322&ndash;326 (1990).</span></li><li class="c19"><span class="c2">Li, T. </span><span class="c2 c9">et al.</span><span class="c2">&nbsp;</span><span class="c2 c9">Nat. Methods</span><span class="c0">&nbsp;064535 (2016). doi:10.1038/nmeth.4083</span></li><li class="c19"><span class="c2">Meyers, R. M. </span><span class="c2 c9">et al.</span><span class="c2">&nbsp;</span><span class="c2 c9">Nat. Genet.</span><span class="c2">&nbsp;</span><span class="c12">49</span><span class="c0">, 1779&ndash;1784 (2017).</span></li><li class="c19"><span class="c2">Cowley, G. S. </span><span class="c2 c9">et al.</span><span class="c2">&nbsp;</span><span class="c2 c9">Sci. Data</span><span class="c2">&nbsp;</span><span class="c12">1</span><span class="c0">, 140035 (2014).</span></li><li class="c19"><span class="c2">Aguirre, A. J. </span><span class="c2 c9">et al.</span><span class="c2">&nbsp;</span><span class="c2 c9">Cancer Discov.</span><span class="c2">&nbsp;</span><span class="c12">6</span><span class="c0">, 914&ndash;29 (2016).</span></li><li class="c19"><span class="c2">Tsherniak, A. </span><span class="c2 c9">et al.</span><span class="c2">&nbsp;</span><span class="c2 c9">Cell</span><span class="c2">&nbsp;</span><span class="c12">170</span><span class="c0">, 564&ndash;576.e16 (2017).</span></li><li class="c19"><span class="c2">Carithers, L. J. </span><span class="c2 c9">et al.</span><span class="c2">&nbsp;</span><span class="c2 c9">Biopreserv. Biobank.</span><span class="c2">&nbsp;</span><span class="c12">13</span><span class="c0">, 311&ndash;319 (2015).</span></li><li class="c19"><span class="c2">Gibbs, R. A. </span><span class="c2 c9">et al.</span><span class="c2">&nbsp;</span><span class="c2 c9">Nature</span><span class="c2">&nbsp;</span><span class="c12">526</span><span class="c0">, 68&ndash;74 (2015).</span></li><li class="c19"><span class="c0">Feller, W. (Wiley, New York, 1968).</span></li><li class="c19"><span class="c2">Ritchie, M. E. </span><span class="c2 c9">et al.</span><span class="c2">&nbsp;</span><span class="c2 c9">Nucleic Acids Res.</span><span class="c2">&nbsp;</span><span class="c12">43</span><span class="c0">, e47&ndash;e47 (2015).</span></li></ol><p class="c1 c4"><span class="c0"></span></p><div><p class="c4 c30"><span class="c14 c29"></span></p></div></body></html>