# shiny server
shinyServer(function(input, output, session){
  #supress warnings
  storeWarn<- getOption("warn")
  options(warn = -1) 
  
  observe({
    input$filetype
    updateTabsetPanel(session, "basic", selected = "p1")
  })
  
  ##### VISUALIZATIONS START ##### 
  output$a_file <- renderUI({
    fileInput('a_file_pulldown_r', 'Upload user input file',
              accept = c(
                'text/csv',
                'text/comma-separated-values',
                'text/tab-separated-values',
                'text/plain',
                '.csv',
                '.tsv')
    )
  })
  
  output$a_color_scheme <- renderUI({
    radioButtons("colorscheme", "Color scheme:",
                 c("FDR" = "fdr", 
                   "ExAC" = "exac",
                   "Grayscale" = "cbf",
                   "User value" = "user"),
                 inline = T)
  })
  
  output$a_logfc_direction_ui <- renderUI({
    radioButtons("a_logfc_direction", "LogFC direction",
                 c("Neg" = "negative", 
                   "Both" = "both",
                   "Pos" = "positive"),
                 selected = 'positive',
                 inline = T)
  })
  
  
  # 
  output$a_significance_type_ui <- renderUI({
    radioButtons("a_significance_type", "Significance",
                 c("FDR" = "fdr", "P-value" = "pvalue"),
                 inline = T)
  })
  
  output$FDR_thresh <- renderUI({
    #validate(need(input$a_significance_type == 'fdr', ''))
    sliderInput("a_fdr_thresh", "FDR threshold",
                min = 0, max = 1, value = 0.1, step = 0.01)
  })
  
  output$PVal_thresh <- renderUI({
    #validate(need(input$a_significance_type == 'pvalue', ''))
    sliderInput("a_pval_thresh", "P-value threshold",
                min = 0, max = 1, value = 0.05, step = 0.001)
  })
  
  
  
  # based on a_pulldown(), create slider for logFC
  output$logFC_thresh <- renderUI({
    if(!is.null(input$a_file_pulldown_r)){
      df <- a_pulldown()
      if (input$a_logfc_direction == 'negative'){
        limit <- abs(min(df$logFC))
        limit <- round(limit+0.5, 1)
      } else if (input$a_logfc_direction == 'positive'){
        limit <- max(df$logFC)
        limit <- round(limit+0.5, 1)
      } else {
        limit <- max(max(df$logFC), abs(min(df$logFC)))
        limit <- round(limit+0.5, 1)
      }
      sliderInput("a_logFC_thresh", "logFC threshold",
                  min = 0, max = limit, value = 0, step = 0.1)
    }else
      sliderInput("a_logFC_thresh", "logFC threshold",
                  min = 0, max = 1, value = 0, step = 0.1)
  })  
  
  # based on a_pulldown(), create slider for logFC
  output$logFC_thresh_old <- renderUI({
    if(!is.null(input$a_file_pulldown_r)){
      df <- a_pulldown()
      min_logFC <- min(df$logFC)
      min_logFC <- round(min_logFC-0.5, 1)
      max_logFC <- max(df$logFC)
      max_logFC <- round(max_logFC+0.5, 1)
      sliderInput("a_logFC_thresh", "logFC threshold",
                  min = min_logFC, max = max_logFC, value = c(0, max_logFC), step = 0.1)
    }else
      sliderInput("a_logFC_thresh", "logFC threshold",
                  min = -1, max = 1, value = c(0, 1), step = 0.1)
  })  
  
  
  #output$a_color_theme <- renderUI({
  #  validate(
  #    need(input$a_file_pulldown_r != '', ""),
  #    need(input$colorscheme == "user", "")
  #  )
  #  selectInput("colorbrewer_theme", "", 
  #              c("YlOrRd", "YlOrBr", "YlGnBu", "YlGn", "Reds", "RdPu", "Purples", "PuRd", "PuBuGn", "PuBu", "OrRd", 
  #                "Oranges", "Greys", "Greens", "GnBu", "BuPu", "BuGn", "Blues", "Set3", "Set2", "Set1", "Pastel2", "Pastel1",
  #                "Paired", "Dark2", "Accent", "Spectral", "RdYlGn", "RdYlBu", "RdGy", "RdBu", "PuOr", "PRGn", "PiYG", "BrBG"))
  #})
  
  #output$a_color_theme_integrated <- renderUI({
  #  validate(
  #    need(input$colorscheme == "user", "")
  #  )
  #  selectInput("colorbrewer_theme_integrated", "User value", 
  #              c("YlOrRd", "YlOrBr", "YlGnBu", "YlGn", "Reds", "RdPu", "Purples", "PuRd", "PuBuGn", "PuBu", "OrRd", 
  #                "Oranges", "Greys", "Greens", "GnBu", "BuPu", "BuGn", "Blues", "Set3", "Set2", "Set1", "Pastel2", "Pastel1",
  #                "Paired", "Dark2", "Accent", "Spectral", "RdYlGn", "RdYlBu", "RdGy", "RdBu", "PuOr", "PRGn", "PiYG", "BrBG"))
  #})
  
  #output$a_color_theme_pf <- renderUI({
  #  validate(
  #    need(input$a_file_pulldown_r != '', "")
  #  )
  #  selectInput("colorbrewer_theme_pf", "Protein families color", 
  #              c("YlOrRd", "YlOrBr", "YlGnBu", "YlGn", "Reds", "RdPu", "Purples", "PuRd", "PuBuGn", "PuBu", "OrRd", 
  #                "Oranges", "Greys", "Greens", "GnBu", "BuPu", "BuGn", "Blues", "Set3", "Set2", "Set1", "Pastel2", "Pastel1",
  #                "Paired", "Dark2", "Accent", "Spectral", "RdYlGn", "RdYlBu", "RdGy", "RdBu", "PuOr", "PRGn", "PiYG", "BrBG"))
  #})
  
  #output$a_color_setting_indv_text <- renderUI({
  #  validate(
  #    need(input$a_file_pulldown_r != '', ""),
  #    need(input$colorscheme == "fdr" || input$colorscheme == "user", "")
  #  )
  #  HTML("<b>Color selection for markers:</b>")
  #})
  
  
  #a_significance_tresholds <- reactive({
  #  sigtype = ifelse(input$a_significance_type == 'fdr', 'FDR', 'P-value')
  #  
  #})
  
  
  # track significance threshols for FDR and P-value
  monitor_significance_tresholds <- reactive({
    sig_type = ifelse(input$a_significance_type == 'fdr', 'FDR', 'P-value')
    sig_value = ifelse(sig_type == 'FDR', input$a_fdr_thresh, input$a_pval_thresh)
    fc_sign = ifelse(input$a_logfc_direction, '<', '≥')
    region_l <- c(paste0(sig_type,"<", sig_value))
    region_ge <- c(paste0(sig_type,"≥", sig_value))
    return(list(sig=region_l, insig=region_ge, sig_type=sig_type, sig_value=sig_value))
  })
  
  # track significance threshols for logFC
  monitor_logfc_threshold <- reactive({
    fc_dir = input$a_logfc_direction
    fc = input$a_logFC_thresh
    if (fc_dir == 'negative') {fc_sig = paste("logFC&lt;", -fc); fc_insig =  paste("logFC&ge;", -fc)}
    if (fc_dir == 'positive') {fc_sig = paste("logFC&ge;", fc); fc_insig =  paste("logFC&lt;", fc)}
    if (fc_dir == 'both') {fc_sig = paste("|logFC|&ge;", fc); fc_insig = paste("|logFC|&lt;", fc) }
    return(list(sig=fc_sig, insig=fc_insig))
  })
  

  #----------------------------------------------------------
  # sliders for selecting color, symbols and labels of plots
  
  # basic plot
  output$a_color_theme_indv_sig <- renderUI({
    validate(need(input$a_file_pulldown_r != '', ""))
    label = HTML(paste(c(monitor_significance_tresholds()$sig, monitor_logfc_threshold()$sig), collapse =', '))
    colourpicker::colourInput('a_color_indv_sig', label, value = '#41AB5D', showColour = 'text', 
                  palette = c( "limited"), allowedCols = allowed_colors)
  })
    
  # basic plot
  output$a_color_theme_indv_insig <- renderUI({
    validate(need(input$a_file_pulldown_r != '', ""))
    label = HTML(paste(c(monitor_significance_tresholds()$insig, monitor_logfc_threshold()$insig), collapse =', '))
    colourpicker::colourInput('a_color_indv_insig', label, value = '#808080', showColour = 'text', 
                              palette = c( "limited"), allowedCols = allowed_colors)
  })
  
  # intgrated plot, snp
  output$a_color_snp_sig_ui <- renderUI({
    validate(need(input$a_file_pulldown_r != '', ""))
    label = HTML(paste(c(monitor_significance_tresholds()$sig, monitor_logfc_threshold()$sig), collapse =', '))
    colourpicker::colourInput('a_color_snp_sig', label, value = 'blue', showColour = 'both', 
                              palette = c( "limited"), allowedCols = allowed_colors)
  })
  # intgrated plot, snp
  output$a_color_snp_insig_ui <- renderUI({
    validate(need(input$a_file_pulldown_r != '', ""))
    label = HTML(paste(c(monitor_significance_tresholds()$insig, monitor_logfc_threshold()$insig), collapse =', '))
    colourpicker::colourInput('a_color_snp_insig', label, value = '#808080', showColour = 'both', 
                              palette = c( "limited"), allowedCols = allowed_colors)
  })
  # integrated plot, snp
  output$a_symbol_snp_ui <- renderUI({
    validate(need(input$a_file_pulldown_r != '', ""))
    selectInput('a_symbol_snp', 'Symbol', choices = allowed_plotly_symbols)
  })
  # integrated plot, snp
  output$a_label_snp_ui <- renderUI({
    validate(need(input$a_file_pulldown_r != '', ""))
    checkboxInput("a_label_snp", label = "Toggle labels", value = TRUE)
  })
  
  # integrated plot, snp,
  output$a_reset_snp_ui <- renderUI({
    actionButton('a_reset_snp','clear')
  })
  
  
  # intgrated plot, genes upload
  output$a_color_genes_upload_sig_ui <- renderUI({
    validate(need(input$a_file_pulldown_r != '', ""))
    label = HTML(paste(c(monitor_significance_tresholds()$sig, monitor_logfc_threshold()$sig), collapse =', '))
    colourpicker::colourInput('a_color_genes_upload_sig', label, value = 'blue', showColour = 'both', 
                              palette = c( "limited"), allowedCols = allowed_colors)
  })
  # intgrated plot, genes upload
  output$a_color_genes_upload_insig_ui <- renderUI({
    validate(need(input$a_file_pulldown_r != '', ""))
    label = HTML(paste(c(monitor_significance_tresholds()$insig, monitor_logfc_threshold()$insig), collapse =', '))
    colourpicker::colourInput('a_color_genes_upload_insig', label, value = '#808080', showColour = 'both', 
                              palette = c( "limited"), allowedCols = allowed_colors)
  })
  
  # integrated plot, genes uplaod
  output$a_symbol_genes_upload_ui <- renderUI({
    validate(need(input$a_file_pulldown_r != '', ""))
    selectInput('a_symbol_genes_upload', 'Symbol', choices = allowed_plotly_symbols)
  })
  
  # integrated plot, genes upload
  output$a_label_genes_upload_ui <- renderUI({
    validate(need(input$a_file_pulldown_r != '', ""))
    checkboxInput("a_label_genes_upload", label = "Toggle labels", value = TRUE)
  })
  
  # integrated plot, reset
  output$a_reset_genes_upload_ui <- renderUI({
    validate(need(input$a_file_pulldown_r != '', ""))
    actionButton('a_reset_genes_upload', 'Reset')
  })
  observeEvent(input$a_reset_genes_upload, {
    reset("a_file_genes_rep")
  })
  
  
  # intgrated plot, inweb
  output$a_color_inweb_sig_ui <- renderUI({
    validate(need(input$a_file_pulldown_r != '', ""))
    label = HTML(paste(c(monitor_significance_tresholds()$sig, monitor_logfc_threshold()$sig), collapse =', '))
    colourpicker::colourInput('a_color_inweb_sig', label, value = 'yellow', showColour = 'both', 
                              palette = c( "limited"), allowedCols = allowed_colors)
  })
  # intgrated plot, inweb
  output$a_color_inweb_insig_ui <- renderUI({
    validate(need(input$a_file_pulldown_r != '', ""))
    label = HTML(paste(c(monitor_significance_tresholds()$insig, monitor_logfc_threshold()$insig), collapse =', '))
    colourpicker::colourInput('a_color_inweb_insig', label, value = '#808080', showColour = 'both', 
                              palette = c( "limited"), allowedCols = allowed_colors)
  })
  # integrated plot, inweb
  output$a_symbol_inweb_ui <- renderUI({
    validate(need(input$a_file_pulldown_r != '', ""))
    selectInput('a_symbol_inweb', 'Symbol', choices = allowed_plotly_symbols)
  })
  # integrated plot, inweb
  output$a_label_inweb_ui <- renderUI({
    validate(need(input$a_file_pulldown_r != '', ""))
    checkboxInput("a_label_inweb", label = "Toggle labels", value = TRUE)
  })
  
  
  # intgrated plot, gwas catalogue
  output$a_color_gwas_cat_sig_ui <- renderUI({
    validate(need(input$a_file_pulldown_r != '', ""))
    label = HTML(paste(c(monitor_significance_tresholds()$sig, monitor_logfc_threshold()$sig), collapse =', '))
    colourpicker::colourInput('a_color_gwas_cat_sig', label, value = 'cyan', showColour = 'both', 
                              palette = c( "limited"), allowedCols = allowed_colors)
  })
  # intgrated plot, gwas catalogue
  output$a_color_gwas_cat_insig_ui <- renderUI({
    validate(need(input$a_file_pulldown_r != '', ""))
    label = HTML(paste(c(monitor_significance_tresholds()$insig, monitor_logfc_threshold()$insig), collapse =', '))
    colourpicker::colourInput('a_color_gwas_cat_insig', label, value = '#808080', showColour = 'both', 
                              palette = c( "limited"), allowedCols = allowed_colors)
  })
  # integrated plot, gwas catalogue
  output$a_symbol_gwas_cat_ui <- renderUI({
    validate(need(input$a_file_pulldown_r != '', ""))
    selectInput('a_symbol_gwas_cat', 'Symbol', choices = allowed_plotly_symbols)
  })
  # integrated plot, genes upload
  output$a_label_gwas_cat_ui <- renderUI({
    validate(need(input$a_file_pulldown_r != '', ""))
    checkboxInput("a_label_gwas_cat", label = "Toggle labels", value = TRUE)
  })
  
  
  
  #output$a_color_setting_text <- renderUI({
  #  validate(
  #    need(input$a_file_pulldown_r != '', "")
  #  )
  #  HTML("<b>Color selection for markers:</b>")
  #})
  
  # output$a_color_theme_multi_sig <- renderUI({
  #  validate(
  #    need(input$a_file_pulldown_r != '', ""),
  #    need(input$colorscheme == "fdr", "")
  #  )
  #  region <- c(paste0("FDR<", input$a_fdr_thresh))
  #  selectInput('a_color_multi_sig', region, marker_cols$V1, multiple=F, selectize=TRUE, selected = "seagreen3")
  #})
  
  #output$a_color_theme_multi_insig <- renderUI({
  #  validate(
  #    need(input$a_file_pulldown_r != '', ""),
  #    need(input$colorscheme == "fdr", "")
  #  )
  #  region <- c(paste0("FDR≥", input$a_fdr_thresh))
  #  selectInput('a_color_multi_insig', region, marker_cols$V1, multiple=F, selectize=TRUE, selected = "royalblue2")
  #})
  
  #output$a_color_theme_goi <- renderUI({
  #  validate(
  #    need(input$a_file_pulldown_r != '', ""), 
  #    need(!is.null(input$a_file_genes_rep) && input$a_file_genes_rep != "", "")
  #  )
  #  selectInput("colorbrewer_theme_goi", "Genes of interest", 
  #              c("YlOrRd", "YlOrBr", "YlGnBu", "YlGn", "Reds", "RdPu", "Purples", "PuRd", "PuBuGn", "PuBu", "OrRd", 
  #                "Oranges", "Greys", "Greens", "GnBu", "BuPu", "BuGn", "Blues", "Set3", "Set2", "Set1", "Pastel2", "Pastel1",
  #                "Paired", "Dark2", "Accent", "Spectral", "RdYlGn", "RdYlBu", "RdGy", "RdBu", "PuOr", "PRGn", "PiYG", "BrBG"),
  #              selected = "Set2")
  #})
  
  #output$a_color_theme_snp <- renderUI({
  #  validate(
  #    need(input$a_file_pulldown_r != '', ""),
  #    need(!is.null(input$a_file_SNP_rep) && input$a_file_SNP_rep != "", "")
  #  )
  #  selectInput("colorbrewer_theme_snp", "SNPs", 
  #              c("YlOrRd", "YlOrBr", "YlGnBu", "YlGn", "Reds", "RdPu", "Purples", "PuRd", "PuBuGn", "PuBu", "OrRd", 
  #                "Oranges", "Greys", "Greens", "GnBu", "BuPu", "BuGn", "Blues", "Set3", "Set2", "Set1", "Pastel2", "Pastel1",
  #                "Paired", "Dark2", "Accent", "Spectral", "RdYlGn", "RdYlBu", "RdGy", "RdBu", "PuOr", "PRGn", "PiYG", "BrBG"),
  #              selected = "PuOr")
  #})
  
  #output$a_color_theme_inweb <- renderUI({
  #  validate(
  #    need(input$a_file_pulldown_r != '', ""), 
  #    need(!is.null(input$a_bait_rep) && input$a_bait_rep != "", "")
  #  )
  #  selectInput('a_color_inweb', 'InWeb', add_marker_cols$V1, multiple=F, selectize=TRUE, selected = "yellow")
  #})
  
  output$a_bait_layer <- renderUI({
    textInput("a_bait_rep", value = "", "Input HGNC symbol to search for InWeb protein interactors (e.g. ZBTB7A)")
  })
  
  output$a_bait_search <- renderUI({
    textInput("a_bait_search_rep", "Input bait (e.g. BCL2)")
  })
  
  output$a_GOI_search <- renderUI({
    textInput("a_goi_search_rep", "Search")
  })
  
  output$a_gwas_catalogue_ui <- renderUI({
    selectInput('a_gwas_catalogue', 'GWAS Catalogue', unique(as.character(as.vector(gwas_table$DISEASE.TRAIT))), multiple=T, selectize=TRUE, selected = "grey")
  })
  
  
  ### snp labels
  #output$a_toggle_snp_label_ui <- renderUI({
  #  actionButton("a_toggle_snp_label", label = "Toggle label")
  #})
  

  

  ###
  
  
  # Search for replicates in data
  available_replicates <- reactive({
    d <- a_pulldown()
    if(!is.null(d)){
      reps = sum(grepl('^rep[0-9]+$',colnames(d)))
      enum = enumerate_replicate_combinations(reps)
      enum = as.data.frame(apply(enum, 2, function(x) paste0('rep', x)))
      enum = paste0(enum[,1],'.',enum[,2])
    } else {
      enum = 'rep1.rep2'
    }
    return(enum)
  })
  
  output$a_select_scatterplot_ui <- renderUI({
    selectInput('a_select_scatterplot','Scatterplot', choices = available_replicates())
  })
 #
  
  #output$a_text_label <- renderUI({
  #  radioButtons('a_marker_text', 'Turn on/off labels',
  #               c(On = 'yes_label',
  #                 Off = 'no_label'),
  #               inline = T)
  #})
  
  output$a_SNP_file <- renderUI({
    req(a_pulldown_significant())
    fileInput('a_file_SNP_rep', 'File containing list of SNPs, one ID per line',
              accept = c(
                'text/csv',
                'text/comma-separated-values',
                'text/tab-separated-values',
                'text/plain',
                '.csv',
                '.tsv')
    )
  })
  
  output$a_SNP_file_vennd <- renderUI({
    fileInput('a_file_SNP_vennd', 'File containing list of SNPs, one ID per line (e.g. rs11172113)',
              accept = c(
                'text/csv',
                'text/comma-separated-values',
                'text/tab-separated-values',
                'text/plain',
                '.csv',
                '.tsv')
    )
  })
  
  output$a_genes_file <- renderUI({
    fileInput('a_file_genes_rep', 'File containing at least 2 genes with header, one HGNC symbol per line (e.g. TINMAN)',
              accept = c(
                'text/csv',
                'text/comma-separated-values',
                'text/tab-separated-values',
                'text/plain',
                '.csv',
                '.tsv'), multiple = T
    )
  })
  
  output$a_genes_file_vennd <- renderUI({
    fileInput('a_file_genes_vennd', 'File containing at least 2 genes with header, one HGNC symbol per line (e.g. TINMAN)',
              accept = c(
                'text/csv',
                'text/comma-separated-values',
                'text/tab-separated-values',
                'text/plain',
                '.csv',
                '.tsv')
    )
  })
  
  output$a_plot_button <- renderUI({
    validate(
      need(input$a_file_pulldown_r != '', "")
    )
    all_inputs <- c(input$a_file_SNP_rep, input$a_file_genes_rep, input$a_bait_rep)
    
    if(any(!is.null(all_inputs)) && any(all_inputs != "")){
      actionButton("a_make_plot", "Generate plots")
    }
  })
  
  output$a_vennd_button3 <- renderUI({
    validate(
      need(input$a_file_pulldown_r != '', ""),
      need(input$a_bait_vennd != '', "")
    )
    if(!is.null(a_bait_gene_vennd())){
      actionButton("a_make_vennd_bait", "Bait")
    }
  })
  
  output$a_vennd_button1 <- renderUI({
    validate(
      need(input$a_file_pulldown_r != '', "")
    )
    if(!is.null(a_snp_vennd())){
      actionButton("a_make_vennd_snp", "SNP")
    }
  })
  
  output$a_vennd_button2 <- renderUI({
    validate(
      need(input$a_file_pulldown_r != '', "")
    )
    if(!is.null(a_upload_genes_vennd())){
      actionButton("a_make_vennd_goi", "GOI")
    }
  })
  
  # based on a_pulldown(), create slider for logFC
  output$a_logFC_slider <- renderUI({
    validate(
      need(input$a_file_pulldown_r != '', "")
    )
    if(!is.null(a_pulldown())){
      input_file <- a_pulldown()
      df <- input_file
      min_logFC <- min(df$logFC)
      min_logFC <- round(min_logFC-0.5, 1)
      max_logFC <- max(df$logFC)
      max_logFC <- round(max_logFC+0.5, 1)
      sliderInput("a_logFC_range", "logFC",
                  min = min_logFC, max = max_logFC, value = c(0, max_logFC), step = 0.1)
    }
  })
  
  # based on a_pulldown(), create slider for logFC
  output$a_FDR_slider <- renderUI({
    validate(
      need(input$a_file_pulldown_r != '', "")
    )
    sliderInput("a_FDR_range", "FDR",
                min = 0, max = 1, value = c(0, 0.1), step = 0.01)
  })
  
  # based on a_pulldown(), create slider for logFC
  output$a_pvalue_slider <- renderUI({
    validate(
      need(input$a_file_pulldown_r != '', "")
    )
    sliderInput("a_pvalue_range", "pvalue",
                min = 0, max = 1, value = c(0, 1), step = 0.01)
  })
  
  output$a_pf_loc_selection <- renderUI({
    #req(input$a_file_pulldown_r())
    selectInput('a_pf_loc_option', 'Data options', c("Protein family" = 'pf', "Localization_GO" = 'loc_go', "Localization_UniProt" = 'loc_uniprot'), selectize=FALSE)
  })
  
  a_pf_data_selection <- reactive({
    inDselection <- input$a_pf_loc_option
    if (inDselection == "pf") {
      pf_data <- PF_gene_indexed
    } else if (inDselection == "loc_go"){
      pf_data <- go_location_gene_indexed
    } else {
      pf_data <- uniprot_location_gene_indexed
    }
  })
  
  output$a_pf_plot_selection <- renderUI({
    validate(
      need(input$a_file_pulldown_r != '', "")
    )
    d <- a_orig_pulldown()
    d_col <- colnames(d)
    if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col &
       "rep1" %in% d_col & "rep2" %in% d_col){
      # radioButtons('a_BPF_option_sp', 'Turn on/off marker sizing',
      #              c(On = 'change_m_BPF',
      #                Off = 'static_m_BPF'),
      #              inline = T
      # )
      radioButtons('a_pf_plot_option', 'Plot options', c("Volcano plot" = 'vp', "Scatter plot" = 'sp'), inline=T)
    } else if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col){
      # validate(
      #   return(NULL)
      # )
      radioButtons('a_pf_plot_option', 'Plot options', c("Volcano plot" = 'vp'), inline=T)
    } else if("rep1" %in% d_col & "rep2" %in% d_col){
      # radioButtons('a_BPF_option_sp', 'Turn on/off marker sizing',
      #              c(On = 'change_m_BPF',
      #                Off = 'static_m_BPF'),
      #              inline = T
      # )
      radioButtons('a_pf_plot_option', 'Plot options', c("Volcano plot" = 'vp', "Scatter plot" = 'sp'), inline=T)
    }
  })
  
  a_pf_viz_selection <- reactive({
    inDselection <- input$a_pf_plot_option
    if (inDselection == "vp") {
      pf_viz <- "volcano_viz"
    } else {
      pf_viz <- "scatter_viz"
    }
  })
  
  output$a_PF_sort_col <- renderUI({
    validate(
      need(input$a_file_pulldown_r != '', "")
    )
    radioButtons('a_PF_sort_option', 'Sort',
                 c("Alphabetically" = 'sort_alph',
                   "PF Frequency" = 'sort_freq'),
                 inline = T
    )
  })
  
  a_PF_sorting <- reactive({
    inPFsort <- input$a_PF_sort_option
    if (inPFsort == "sort_alph") {
      marker <- "sort_a"
    } else {
      marker <- "sort_f"
    }
  })
  
  #create slider for FDR
  output$a_BPF_FDR_slider <- renderUI({
    validate(
      need(input$a_file_pulldown_r != '', ""),
      need(input$a_pf_plot_option != '', "")
    )
    if(a_pf_viz_selection() == "volcano_viz"){
      sliderInput("a_BPF_FDR_range", "FDR",
                  min = 0, max = 1, value = c(0, 1), step = 0.01, sep = '', pre = NULL, post = NULL)
    } else if(a_pf_viz_selection() == "scatter_viz"){
      d <- a_pulldown()
      d_col <- colnames(d)
      if("rep1" %in% d_col & "rep2" %in% d_col){
        min_rep1 <- min(d$rep1)
        min_rep1 <- round(min_rep1-0.5, 1)
        max_rep1 <- max(d$rep1)
        max_rep1 <- round(max_rep1+0.5, 1)
        sliderInput("a_BPF_rep1_range", "rep1",
                    min = min_rep1, max = max_rep1, value = c(-1, 1), step = 0.01, sep = '', pre = NULL, post = NULL)
      }
    }
  })
  
  #create slider for pvalue
  output$a_BPF_pvalue_slider <- renderUI({
    validate(
      need(input$a_file_pulldown_r != '', ""),
      need(input$a_pf_plot_option != '', "")
    )
    if(a_pf_viz_selection() == "volcano_viz"){
      sliderInput("a_BPF_pvalue_range", "pvalue",
                  min = 0, max = 1, value = c(0, 1), step = 0.01, sep = '', pre = NULL, post = NULL)
    } else if(a_pf_viz_selection() == "scatter_viz"){
      d <- a_pulldown()
      d_col <- colnames(d)
      if("rep1" %in% d_col & "rep2" %in% d_col){
        min_rep2 <- min(d$rep2)
        min_rep2 <- round(min_rep2-0.5, 1)
        max_rep2 <- max(d$rep2)
        max_rep2 <- round(max_rep2+0.5, 1)
        sliderInput("a_BPF_rep2_range", "rep2",
                    min = min_rep2, max = max_rep2, value = c(-1, 1), step = 0.01, sep = '', pre = NULL, post = NULL)
      }
    }
  })
  
  # based on a_pulldown(), create slider for BPF logFC
  output$a_BPF_logFC_slider <- renderUI({
    validate(
      need(input$a_file_pulldown_r != '', ""),
      need(input$a_pf_plot_option != '', "")
    )
    if(a_pf_viz_selection() == "volcano_viz"){
      if(!is.null(a_pulldown())){
        input_file <- a_pulldown()
        df <- input_file
        min_logFC <- min(df$logFC)
        min_logFC <- round(min_logFC-0.5, 1)
        max_logFC <- max(df$logFC)
        max_logFC <- round(max_logFC+0.5, 1)
        sliderInput("a_BPF_logFC_range", "logFC",
                    min = min_logFC, max = max_logFC, value = c(1, max_logFC), step = 0.1)
      }
    } else if(a_pf_viz_selection() == "scatter_viz"){
      return(NULL)
    }
  })
  
  output$a_BPF_marker_size <- renderUI({
    validate(
      need(input$a_file_pulldown_r != '', "")
    )
    radioButtons('a_BPF_option', 'Turn on/off marker sizing',
                 c(On = 'change_m_BPF',
                   Off = 'static_m_BPF'),
                 inline = T
    )
  })
  
  output$a_BPF_freq <- renderUI({
    validate(
      need(input$a_file_pulldown_r != '', "")
    )
    if(!is.null(input$a_BPF_option)){
      inBPFmarker <- input$a_BPF_option
      if (inBPFmarker == "change_m_BPF") {
        sliderInput("a_BPF_marker_freq", "Marker size",
                    min = 1, max = 40, value = 1, step = 1)     
      }
    }
  })
  
  output$a_BPF_text <- renderUI({
    validate(
      need(input$a_file_pulldown_r != '', "")
    )
    HTML("<b>User-defined significance thresholds:</b>")
  })
  
  output$a_BPF_button <- renderUI({
    validate(
      need(input$a_file_pulldown_r != '', "")
    )
    if(!is.null(a_pulldown())){
      actionButton("a_make_bpf", "Generate PF plot")
    }
  })
  
  output$a_prot_fam_db <- renderUI({
    validate(
      need(input$a_file_pulldown_r != '', "")
    )
    selectInput('a_pfam_db', 'Protein families', colnames(prot_fam), multiple=TRUE, selectize=TRUE)
  })
  
  output$a_text_prot_fam_db <- renderUI({
    validate(
      need(input$a_file_pulldown_r != '', "")
    )
    radioButtons('a_marker_text_prot_fam_db', 'Turn on/off labels',
                 c(On = 'yes_label',
                   Off = 'no_label'),
                 inline = T
    )
  })
  
  #observeEvent(input$basic,{
  #  if(input$basic == "p3" | input$basic == "p4" | input$basic == "p5"){
  #    shinyjs::hide("colorscheme")
  #    shinyjs::hide("file_color")
  #    shinyjs::hide("a_fdr_thresh")
  #    shinyjs::hide("a_pval_thresh")
  #    shinyjs::hide("a_logFC_thresh")
  #    shinyjs::hide("a_logfc_direction_ui")
  #  }else {
  #    shinyjs::hide("colorscheme") # <---- note todo ... this is still here. should be removed!
  #   shinyjs::show("file_color")
  #    shinyjs::show("a_logFC_thresh")
  #    shinyjs::show("a_pval_thresh")
  #    shinyjs::show("a_fdr_thresh")
  #    shinyjs::show("a_logfc_direction_ui")
  #  }
  #})
  
  # show/hide the fdr/pvalue bar
  observeEvent(input$a_significance_type,{
    if (input$a_significance_type == 'fdr'){
        shinyjs::hide("a_pval_thresh")
        shinyjs::show("a_fdr_thresh")
      } else {
        shinyjs::show("a_pval_thresh")
        shinyjs::hide("a_fdr_thresh")
      }
  })
  
  # loading the data and getting the pulldown
  a_in_pulldown <- reactive({
    req(input$a_file_pulldown_r)
    d <- read_input(input$a_file_pulldown_r$datapath)
    return(d)
  })
  
  # map accession_numbers to gene ids if needed
  a_orig_pulldown <- reactive({
    pulldown <- a_in_pulldown()
    if (pulldown$format$accession_rep == TRUE){
      pulldown$data <- map_gene_id(pulldown$data)} # this may also be in another namespace. check!
    pulldown$data$gene = toupper(pulldown$data$gene)
    return(pulldown$data)
  })  
  
  # final pulldown formatted data.frame
  a_pulldown <- reactive({
    pulldown <- a_orig_pulldown()
    format <- a_in_pulldown()$format
    if (format$gene_rep | format$accession_rep){
      result = calc_mod_ttest(pulldown) }
    result$all_gene_names <- result$gene
    return(result)
  })
  
  # id the enriched proteins
  a_pulldown_significant <- reactive({
    d = a_pulldown()
    if (input$a_significance_type == 'fdr'){
      d1 = id_enriched_proteins(d, fdr_cutoff = input$a_fdr_thresh, logfc_dir = input$a_logfc_direction, 
                                logfc_cutoff = input$a_logFC_thresh)
    } else {
      d1 = id_enriched_proteins(d, fdr_cutoff = NULL, p_cutoff = input$a_pval_thresh, logfc_dir = input$a_logfc_direction, 
                                logfc_cutoff = input$a_logFC_thresh)
    }
  return(d1)
  })
  

  
  # read in the snps from a file
  #a_snp <- reactive({
  #  snp <- input$a_file_SNP_rep
  #  if(is.null(snp)){
  #    return(NULL)
  #  } else{
  #    df <- read.csv(snp$datapath, header = FALSE)
  #    df
  #  }
  #})
  
  a_snp_vennd <- reactive({
    snp <- input$a_file_SNP_vennd
    if(is.null(snp)){
      return(NULL)
    } else{
      df <- read.csv(snp$datapath, header = FALSE)
      df
    }
  })
  
  # function for handling uploaded genes
  a_genes_upload <- reactive({
    filepath = input$a_file_genes_rep$datapath
    if (!is.null(filepath)){
      genes = get_gene_lists(filepath)
      genes$data$dataset = 'genes upload'
      genes$data$col_significant = input$a_color_genes_upload_sig
      genes$data$col_other = input$a_color_genes_upload_insig
      genes$data$symbol = input$a_symbol_genes_upload
      genes$data$label = input$a_label_genes_upload
      return(genes$data)
    }
  })
  
  #snp to gene using LD r^2>0.6±user defined extension
  #a_snp <- reactive({
  #  req(input$a_file_SNP_rep)
  #  dsnp = read.table(input$a_file_SNP_rep$datapath, header = F)
  #  return(dsnp)
  #})
  
  # read in the snps from a file
  #a_snp_mapping <- reactive({
  #  mapping = read_snp_list(infile = a_snp(), a_pulldown()$gene)
  #  mapping$alt_label = mapping$SNP
  #  mapping$col_significant = 'blue'
  #  mapping$col_other = 'grey'
  #  mapping$dataset = 'SNP'
  #  return(mapping)
  #})
  
  
  #a_upload_genes <- reactive({
  #  genes <- input$a_file_genes_rep
  #  if (is.null(genes)){
  #    return(NULL)
  #  } else{
  #    df <- fread(genes$datapath, header = T, fill = T,
  #                sep="auto", na.strings=c(""," ","NA"), stringsAsFactors = FALSE, data.table = FALSE, blank.lines.skip = T)
  #  }
  #})
  
  a_genes_uploaded <- reactive({
    if(!is.null(a_upload_genes())){
      genes <- a_upload_genes()
      genes <- as.data.frame(sapply(genes, toupper)) 
      genes
    }
  })
  
  a_upload_genes_vennd <- reactive({
    genes <- input$a_file_genes_vennd
    if (is.null(genes)){
      return(NULL)
    } else{
      df <- fread(genes$datapath, header = T, fill = T,
                  sep="auto", na.strings=c(""," ","NA"), stringsAsFactors = FALSE, data.table = FALSE, blank.lines.skip = T)
    }
  })
  
  a_genes_uploaded_vennd <- eventReactive(input$a_make_vennd_goi, { 
    if(!is.null(a_upload_genes_vennd())){
      genes <- a_upload_genes_vennd()
      genes <- as.data.frame(sapply(genes, toupper)) 
      genes
    }
  })
  
  #output$a_goi_num_inputs <- renderUI({
  #  validate(
  #    need(!is.null(a_genes_uploaded_vennd()), "")
  #  )
  #  d <- a_pulldown()
  #  gene_interest <- a_genes_uploaded_vennd()
  #  d_g2s <- lapply(gene_interest, function(x) subset(d, gene %in% x) )  
  #  choices <- names(d_g2s)
  #  choices <- append(choices, "total")
  #  list_count <- length(d_g2s)
  #  if(list_count>0){
  #    selectInput("a_goi_num_inputs", "GOI list",
  #                choices = choices, multiple = F)
  #  }
  #})
  
  #population_bait -- pull down overlap inweb
  population_bait <- reactive({
    present_in_inweb <- a_pulldown()
    population_bait <- subset(present_in_inweb, present_in_inweb$gene %in% inweb_combined$V1)
    population_bait <- population_bait[!duplicated(population_bait[,c("gene")]),]
    population_bait
  })
  
  population_GOI <- reactive({
    present_in_hum_genome <- a_pulldown()
    hgnc <- toupper(human_genome$HGNCsymbol)
    population_hum_genome <- subset(present_in_hum_genome, present_in_hum_genome$gene %in% hgnc)
  })
  
  population_SNP <- reactive({
    present_in_hum_genome <- a_pulldown()
    hgnc <- toupper(human_genome$HGNCsymbol)
    population_hum_genome <- subset(present_in_hum_genome, present_in_hum_genome$gene %in% hgnc)
  })
  
  #success in population_bait -- within user FDR and logFC cutoff in population_bait
  success_population_bait <- reactive({ #eventReactive(input$a_logFC_range | input$a_FDR_range, {
    pullDown_in_inweb <- population_bait()
    success_population <- subset(pullDown_in_inweb, logFC <= input$a_logFC_range[2] & logFC >= input$a_logFC_range[1] &
                                   FDR <= input$a_FDR_range[2] & FDR >= input$a_FDR_range[1] &
                                   pvalue <= input$a_pvalue_range[2] & pvalue >= input$a_pvalue_range[1])
    rownames(success_population) <- NULL 
    success_population <- success_population[!duplicated(success_population[,c("gene")]),]
    success_population
  })
  
  success_population_GOI <- reactive({ #eventReactive(input$a_logFC_range | input$a_FDR_range, {
    pullDown_in_hum_genome <- population_GOI()
    success_population <- subset(pullDown_in_hum_genome, logFC <= input$a_logFC_range[2] & logFC >= input$a_logFC_range[1] &
                                   FDR <= input$a_FDR_range[2] & FDR >= input$a_FDR_range[1] &
                                   pvalue <= input$a_pvalue_range[2] & pvalue >= input$a_pvalue_range[1])
    rownames(success_population) <- NULL 
    success_population
  })
  
  success_population_SNP <- reactive({ #eventReactive(input$a_logFC_range | input$a_FDR_range | input$a_pvalue_range, {
    pullDown_in_hum_genome <- population_SNP()
    success_population <- subset(pullDown_in_hum_genome, logFC <= input$a_logFC_range[2] & logFC >= input$a_logFC_range[1] &
                                   FDR <= input$a_FDR_range[2] & FDR >= input$a_FDR_range[1] &
                                   pvalue <= input$a_pvalue_range[2] & pvalue >= input$a_pvalue_range[1])
    rownames(success_population) <- NULL 
    success_population
  })
  
  a_bait_gene_layer <- reactive({
    bait_in <- input$a_bait_rep
    if (is.null(bait_in) | bait_in == "" ){
      return(NULL)
    } else{
      bait <- bait_in
      bait <- toupper(bait)
    }
  })
  
  a_bait_gene_vennd <- reactive({
    bait_in <- input$a_bait_vennd
    if (is.null(bait_in) | bait_in == "" ){
      return(NULL)
    } else{
      bait <- bait_in
      bait <- toupper(bait)
    }
  })
  
  a_search_gene <- reactive({
    gene_in <- input$a_goi_search_rep
    if (is.null(gene_in) | gene_in == "" ){
      return(NULL)
    } else{
      gene <- gene_in
      gene <- toupper(gene)
    }
  })
  
  #a_bait_friends <- eventReactive(input$a_make_plot, {#reactive({
  #  if(!is.null(a_bait_gene_layer())){
  #    withProgress(message = 'Finding bait interactors in InWeb', 
  #                 detail = 'Hold please', value = 0, {
  #                   bait <- a_bait_gene_layer()
  #                   bait_interactors <- inweb_hash[[bait]]
  #                 })
  #    bait_interactors
  #  }
  #})
  
  #a_bait_friends_vennd <- eventReactive(input$a_make_vennd_bait, { 
  #  if(!is.null(a_bait_gene_vennd())){
  #    withProgress(message = 'Finding bait interactors in InWeb', 
  #                 detail = 'Hold please', value = 0, {
  #                   bait <- a_bait_gene_vennd()
  #                   bait_interactors <- inweb_hash[[bait]]
  #                   incProgress(0.8)
  #                 })
  #    bait_interactors
  #  } else{
  #    return(NULL)
  #  }
  #})
  
  #set sample size 
  #used for volcano and scatter plot
  sample <- reactive({
    if(!is.null(a_bait_friends())){
      bait_interactors <- a_bait_friends()
      pop <- population_bait()
      sample <- subset(pop, pop$gene %in% bait_interactors)
      rownames(sample) <- NULL
      sample
    }
  })
  
  sample_bait <- reactive({
    if(!is.null(a_bait_friends_vennd())){
      bait_interactors <- a_bait_friends_vennd()
      pop <- population_bait()
      sample <- subset(pop, pop$gene %in% bait_interactors)
      rownames(sample) <- NULL 
      sample
    }
  })
  
  sample_GOI <- reactive({
    if(!is.null(a_genes_uploaded_vennd())){
      genes_upload <- a_genes_uploaded_vennd()
      pop <- population_GOI()
      sample <- lapply(genes_upload, function(x) subset(pop, gene %in% x) )  
      sample
    }
  })
  
  sample_SNP <- reactive({
    if(!is.null(SNP_to_gene_vennd())){
      snp2gene <- SNP_to_gene_vennd()
      pop <- population_GOI()
      sample <- subset(pop, pop$gene %in% snp2gene$gene)
      rownames(sample) <- NULL 
      sample
    }
  })
  
  sample_SNP_SGL <- reactive({
    if(!is.null(SNP_to_gene_vennd())){
      snp2gene <- SNP_to_gene_vennd()
      snp_sgl <- subset(snp2gene, Freq == 1)
      pop <- population_GOI()
      sample <- subset(pop, pop$gene %in% snp_sgl$gene)
      rownames(sample) <- NULL 
      sample
    }
  })
  
  sample_SNP_MGL <- reactive({
    if(!is.null(SNP_to_gene_vennd())){
      snp2gene <- SNP_to_gene_vennd()
      snp_mgl <- subset(snp2gene, Freq != 1)
      pop <- population_GOI()
      sample <- subset(pop, pop$gene %in% snp_mgl$gene)
      rownames(sample) <- NULL 
      sample
    }
  })
  
  #success_sample_bait <- reactive({
  #  if(!is.null(sample_bait()) & !is.null(success_population_bait())){
  #    samp <- sample_bait()
  #    success_pop <- success_population_bait()
  #    success_samp <- subset(samp, samp$gene %in% success_pop$gene)
  #    rownames(success_samp) <- NULL
  #    success_samp
  #  }
  #})
  
  #success_sample_GOI <- reactive({
  #  if(!is.null(sample_GOI()) & !is.null(success_population_GOI())){
  #    samp <- sample_GOI()
  #    samp_total <- as.data.frame(data.table::rbindlist(samp))
  #    samp <- c(samp, list(total = samp_total))
  #    list_num <- input$a_goi_num_inputs
  #    success_pop <- success_population_GOI()
  #    success_samp <- subset(samp[[list_num]], samp[[list_num]]$gene %in% success_pop$gene)
  #    rownames(success_samp) <- NULL
  #    success_samp
  #  }
  #})
  
  #success_sample_SNP <- reactive({
  #  if(!is.null(sample_SNP()) & !is.null(success_population_GOI())){
  #    samp <- sample_SNP()
  #    success_pop <- success_population_GOI()
  #    success_samp <- subset(samp, samp$gene %in% success_pop$gene)
  #    rownames(success_samp) <- NULL
  #    success_samp
  #  }
  #})
  
  #success_sample_SNP_SGL <- reactive({
  #  if(!is.null(sample_SNP_SGL()) & !is.null(success_population_GOI())){
  #    samp <- sample_SNP_SGL()
  #    success_pop <- success_population_GOI()
  #    success_samp <- subset(samp, samp$gene %in% success_pop$gene)
  #    rownames(success_samp) <- NULL
  #    success_samp
  #  }
  #})
  
  #success_sample_SNP_MGL <- reactive({
  #  if(!is.null(sample_SNP_MGL()) & !is.null(success_population_GOI())){
  #    samp <- sample_SNP_MGL()
  #    success_pop <- success_population_GOI()
  #    success_samp <- subset(samp, samp$gene %in% success_pop$gene)
  #    rownames(success_samp) <- NULL
  #    success_samp
  #  }
  #})
  
  #hypergeometric_test_list <- reactive({
  #  if(!is.null(sample_bait()) & !is.null(success_population_bait()) & !is.null(success_sample_bait())){
  #    samp <- sample_bait()
  #    success_pop <- success_population_bait()
  #    success_samp <- success_sample_bait()
  #    bait <- a_bait_gene_vennd()
  #    all_interactor <- a_bait_friends_vennd()
  #    inweb_bait <- paste0("All known interactors of ", bait)
  #    x <- list()
  #    x[["A"]] <- as.character(success_pop$gene)
  #    x[["B"]] <- as.character(samp$gene)
  #    x[["Overlap"]] <- as.character(success_samp$gene)
  #    x[[inweb_bait]] <- as.character(all_interactor)
  #    n.obs <- sapply(x, length)
  #    seq.max <- seq_len(max(n.obs))
  #    mat <- sapply(x, "[", i = seq.max)
  #    mat[is.na(mat)] <- " "
  #    as.data.frame(mat)
  #  }
  #})
  
  #hypergeometric_test_list_GOI <- reactive({
  #  if(!is.null(sample_GOI()) & !is.null(success_population_GOI()) & !is.null(success_sample_GOI())){
  #    samp <- sample_GOI()
  #    samp_total <- as.data.frame(data.table::rbindlist(samp))
  #    samp <- c(samp, list(total = samp_total))
  #    list_num <- input$a_goi_num_inputs
  #    success_pop <- success_population_GOI()
  #    success_samp <- success_sample_GOI()
  #    genes_upload <- a_genes_uploaded_vennd()
  #    x <- list()
  #    x[["A"]] <- as.character(success_pop$gene)
  #    x[["B"]] <- as.character(samp[[list_num]]$gene)
  #    x[["Overlap"]] <- as.character(success_samp$gene)
  #    x[["Uploaded list of genes"]] <- as.character(genes_upload)
  #    n.obs <- sapply(x, length)
  #    seq.max <- seq_len(max(n.obs))
  #    mat <- sapply(x, "[", i = seq.max)
  #    mat[is.na(mat)] <- " "
  #    as.data.frame(mat)
  #  }
  #})
  
  hypergeometric_test_list_SNP <- reactive({
    if(!is.null(sample_SNP()) & !is.null(success_population_SNP()) & !is.null(success_sample_SNP())){
      samp <- sample_SNP()
      success_pop <- success_population_SNP()
      success_samp <- success_sample_SNP()
      snp_upload <- a_snp_vennd()
      x <- list()
      x[["A"]] <- as.character(success_pop$gene)
      x[["B"]] <- as.character(samp$gene)
      x[["Overlap"]] <- as.character(success_samp$gene)
      n.obs <- sapply(x, length)
      seq.max <- seq_len(max(n.obs))
      mat <- sapply(x, "[", i = seq.max)
      mat[is.na(mat)] <- " "
      as.data.frame(mat)
    }
  })
  
  #snp to gene using LD r^2>0.6±user defined extension
  a_snp <- reactive({
    req(input$a_file_SNP_rep)
    dsnp = data.table::fread(input$a_file_SNP_rep$datapath, header=T)
    return(dsnp)
  })
  
  # read in the snps from a file
  a_snp_mapping <- reactive({
    mapping = get_snp_lists(infile = a_snp(), a_pulldown()$gene)
    mapping$alt_label = mapping$SNP
    mapping$col_significant = input$a_color_snp_sig
    mapping$col_other = input$a_color_snp_insig
    mapping$symbol = input$a_symbol_snp
    mapping$label = input$a_label_snp
    mapping$dataset = mapping$listName 
    return(mapping)
  })
  
  # map inweb prorteins
  a_inweb_mapping <- reactive({
    req(input$a_bait_rep)
    mapping = get_inweb_list(input$a_bait_rep)
    if (!is.null(mapping)){
        mapping = mapping[mapping$significant, ]
        mapping$col_significant = input$a_color_inweb_sig
        mapping$col_other = input$a_color_inweb_insig
        mapping$symbol = input$a_symbol_inweb
        mapping$label = input$a_label_inweb
        mapping$dataset = 'InWeb'
        return(mapping)
    } 
  })
  
  # map gwas catalouge
  a_gwas_catalogue_mapping <- reactive({
    req(input$a_gwas_catalogue, a_pulldown())
    genes = a_pulldown()$gene
    mapping = get_gwas_lists(input$a_gwas_catalogue, genes)
    if (!is.null(mapping)){
      mapping$col_significant = input$a_color_gwas_cat_sig
      mapping$col_other = input$a_color_gwas_cat_insig
      mapping$symbol = input$a_symbol_gwas_cat
      mapping$label = input$a_label_gwas_cat
      mapping$dataset = mapping$trait
      mapping$alt_label = mapping$SNP
      print(mapping)
      return(mapping)
    }
  })
  
  #a_gnomad_mapping <- reactive({
  #  req(a_pulldown())
  #  pulldown = a_pulldown_significant()
  #  gnomad = merge(pulldown, gnomad_table, by = 'gene')
  #  gnomad$significant = ifelse(gnomad$pLI > 0.9, TRUE, FALSE)
  #  gnomad$col_significant = 'orange'
  #  gnomad$col_other = 'grey'
  #  gnomad$dataset = 'gnomAD'
  #  return(gnomad)
  #})
  
  
  #---------------------------------------------------------------
  # download different mappings
  
  # download moderated t-test
  output$a_mttest_mapping_download <- downloadHandler(
    filename = function() {
      paste("pulldown-mttest",".csv", sep="")
    },
    content = function(file) {
      write.csv(a_pulldown_significant(), file, row.names = F)
    }
  )
  
  # download snp mapping
  output$a_snp_mapping_download <- downloadHandler(
    filename = function() {
      paste("snps-mapping",".csv", sep="")
    },
    content = function(file) {
      write.csv(a_snp_mapping()[,c('gene', 'SNP')], file, row.names = F)
    }
  )
  
  # download gwas catalogue mapping
  output$a_gwas_catalogue_mapping_download <- downloadHandler(
    filename = function() {
      paste("gwas-catalogue-mapping",".csv", sep="")
    },
    content = function(file) {
      write.csv(a_gwas_catalogue_mapping(), file, row.names = F)
    }
  )
  
  #--------------------------------------------------------
  # venn digrams and hypergeometric testing
  
  # inweb hypergeometric overlap
  a_inweb_calc_hyper <- reactive({
    req(input$a_bait_rep, a_pulldown_significant())
    inweb = data.frame(listName="InWeb", get_inweb_list(input$a_bait_rep))
    inweb_intersect = data.frame(listName="InWeb", intersectN=T)
    data = a_pulldown_significant()
    # compile venn diagram information
    hyper = calc_hyper(data, inweb, inweb_intersect)
    hyper[['venn']][['Pulldown']] <- data$gene
    hyper[['venn']][['InWeb']] <- inweb$gene[inweb$significant]
    hyper
  })
  
  # hypergeometric overlap gene upload
  a_genes_upload_calc_hyper <- reactive({
    req(a_genes_upload(), a_pulldown_significant())
    genes = data.frame(listName="genelist", a_genes_upload())
    genes_intersect = data.frame(listName="genelist", intersectN=T)
    data = a_pulldown_significant()
    # compile venn diagram information
    hyper = calc_hyper(data, genes, genes_intersect)
    hyper[['venn']][['Pulldown']] <- data$gene
    hyper[['venn']][['Genelist']] <- genes$gene[genes$significant]
    hyper
  })
  
  # hypergeometric overlap of SNPs, keep p-value
  # in case we need it later
  a_snp_calc_overlap <- reactive({
    req(a_genes_upload(), a_pulldown_significant())
    genes = data.frame(listName="genelist", a_snp_mapping())
    
    ## to be determined..
  })
  
  # draw venn diagram
  output$a_inweb_venn_ui <- renderPlot({
    req(input$a_bait_rep, a_pulldown_significant(), a_inweb_calc_hyper())
    hyper = a_inweb_calc_hyper()
    v = draw_genoppi_venn(hyper$venn, main = paste0('P-value = ', format(hyper$statistics$pvalue, digits = 3)))
    grid::grid.newpage()
    grid::grid.draw(v)
  })
  
  # draw venn diagram
  output$a_genes_upload_venn_ui <- renderPlot({
    req(a_genes_upload(), a_pulldown_significant())
    catf('ensure that p-value of genes_upload_calc_hyper is correct!')
    hyper = a_genes_upload_calc_hyper()
    v = draw_genoppi_venn(hyper$venn, c('blue', 'red'), main = paste0('P-value = ', format(hyper$statistics$pvalue, digits = 3)))
    grid::grid.newpage()
    grid::grid.draw(v)
  })
  
  
  # what is to be plotted below venn diagram
  a_inweb_venn_verbatim <- reactive({
    req(a_pulldown_significant(), a_inweb_calc_hyper())
    tresholds = a_vp_count_text()
    population = a_inweb_calc_hyper()
    A <- paste0("A = pull down subset of ", tresholds, " &#40;", '(inweb_population)', "&#41;")
    B <- paste0("B = Genes of interest in HGNC database", " &#40;", '(goi)', "&#41;")
    total <- paste0("pull down &cap; HGNC database &#40;", "(total)", "&#41;")
    return(list(A=A, B=B, total=total))
  })
  
  # print to ui
  output$a_inweb_venn_verbatim_ui <- renderUI({
    output <- a_inweb_venn_verbatim()
    HTML(paste(output$total, output$A, output$B, sep = "<br/>"))
  })
  
  #---------------------------------------------------------------
  # gnomad integration
  
  a_table_gnomad_constraints <- reactive({
    hover_index = event_data("plotly_click", source = "Multi_VolcanoPlot")
    if (!is.null(hover_index)){
      if (hover_index$key %in% gnomad_table$gene){
        tabl = get_gnomad_constraints(hover_index$key)
        return(tabl)
      }
    }
  })
  
  # render text for gnomad status
  output$a_gnomad_constraints_available_ui <- renderUI({
    gene = event_data("plotly_click", source = "Multi_VolcanoPlot")$key
    if (!is.null(gene)){
      if (gene %in% gnomad_table$gene){
        return(HTML(paste(bold(gene),'constraint info from gnomAD 2.1.1.')))
      } else {
        return(HTML(paste('No constraint info for', bold(gene), 'in gnomAD 2.1.1.')))
      }
    }
    return('Nothing selected. Click a plot point.')
  })
  
  # render table
  output$a_table_gnomad_constraints_ui <- renderTable(a_table_gnomad_constraints())
  
  
  ## this function is now redundant.
  #SNP_to_gene <- eventReactive(input$a_make_plot, {#reactive({
  #  
  #  stop('deprecated!')
  #  
  #  if(!is.null(a_snp())){
  #    withProgress(message = 'Finding genes in SNPs loci', 
  #                 detail = "Hold please", value = 0, {
  #                   d <- a_pulldown()
  #                   snp_data <- a_snp()
  #                   incProgress(0.2)
  #                   res <- lapply(d$gene, function(s) {c(s, subset(snp_data, V1 %in%genes_snps[[s]]))})
  #                   # Convert each list to a data.table
  #                   dt_list <- map(res, as.data.table)
  #                   # Harness the power of rbind list
  #                   dt <- rbindlist(dt_list, fill = TRUE)
  #                   if(ncol(dt)<2){
  #                     dt$V2 <- NA
  #                   }
  #                   names(dt) <- c("gene", "snpid")
  #                   dt <- dt[!is.na(dt$snpid)]
  #                   incProgress(0.6)
  #                   incProgress(0.8)
  #                   if(nrow(dt) >= 1){
  #                     n_occur <- data.frame(table(dt$snpid))
  #                     incProgress(0.9)
  #                     SNP_n_occur <- merge(dt, n_occur, by.x = 'snpid', by.y = 'Var1')
  #                   } else if(nrow(dt) == 0){
  #                     SNP_n_occur <- data.frame(snpid = character(), gene = character(), Freq = integer())
  #                   }
  #                 })
  #    snpList <- SNP_n_occur
  #  }
  #  
  #
  #})
  
  # output$a_SNP_extend2 <- renderUI({
  #   sliderInput("a_SNP_ext2", "Gene extension (±Kb)",
  #               min = 0, max = 100, value = 50, step = 10)
  # })
  
  # eventReactive(input$a_make_vennd, 
  #SNP_to_gene_vennd <- eventReactive(input$a_make_vennd_snp,{
  #  if(!is.null(a_snp_vennd())){
  #    withProgress(message = 'Finding genes in SNPs loci', 
  #                 detail = "Hold please", value = 0, {
  #                   d <- a_pulldown()
  #                   snp_data <- a_snp_vennd()
  #                   incProgress(0.2)
  #                   res <- lapply(d$gene, function(s) {c(s, subset(snp_data, V1 %in%genes_snps[[s]]))})
  #                   # Convert each list to a data.table
  #                   dt_list <- map(res, as.data.table)
  #                   # Harness the power of rbind list
  #                   dt <- rbindlist(dt_list, fill = TRUE)
  #                   if(ncol(dt)<2){
  #                     dt$V2 <- NA
  #                   }
  #                   names(dt) <- c("gene", "snpid")
  #                   dt <- dt[!is.na(dt$snpid)]
  #                   incProgress(0.6)
  #                   incProgress(0.8)
  #                   if(nrow(dt) >= 1){
  #                     n_occur <- data.frame(table(dt$snpid))
  #                     incProgress(0.9)
  #                     SNP_n_occur <- merge(dt, n_occur, by.x = 'snpid', by.y = 'Var1')
  #                   } else if(nrow(dt) == 0){
  #                     SNP_n_occur <- data.frame(snpid = character(), gene = character(), Freq = integer())
  #                   }
  #                 })
  #    snpList <- SNP_n_occur
  #  } else{
  #    return(NULL)
  #  }
  #})
  
  #a_protein_family <- function(x){
  #  pf_db <- input$a_pfam_db
  #  
  #}
  
  ## there is a dependency on this somewhere..do not remove for now.
  a_pf_db <- reactive({
    if(!is.null(input$a_pfam_db)){
      pf_db <- input$a_pfam_db
    }
  })
  
  #a_pf_db_search <- reactive({
  #  pf_db <- a_pf_db()
  #  selected_pf <- prot_fam[grep(paste(pf_db,collapse='|'), names(prot_fam))]
  #  selected_pf <- lapply(selected_pf, function(x) x[!is.na(x)])
  #  selected_pf
  #})
  
  #a_pf_cleanup <- reactive({
  #  validate(
  #    need(input$a_file_pulldown_r != '', ""),
  #    need(!is.null(a_pf_db()), "")
  #  )
  #  d <- a_orig_pulldown()
  #  d_col <- colnames(d)
  #  if("rep1" %in% d_col & "rep2" %in% d_col){
  #    prot_remove <- unique(unlist(a_pf_db_search()))
  #    prot_cleaned <- subset(d, gene %!in% prot_remove)
  #    prot_cleaned
  #  }
  #})
  
  ## ggplot automatically generated and reactive color bars
  a_vp_colorbar <- reactive({
    req(input$a_color_indv_sig, input$a_color_indv_insig)
    thresholds = monitor_significance_tresholds()
    
    # generate matrix with colors
    d <- data.frame(limit = rep('x', 101), value = seq(0, 1, 0.01))
    if (thresholds$sig_type == 'FDR'){
      d$col <- ifelse(d$value < input$a_fdr_thresh, input$a_color_indv_sig, input$a_color_indv_insig)
    } else {
      d$col <- ifelse(d$value < input$a_pval_thresh, input$a_color_indv_sig, input$a_color_indv_insig)
    }

    # plot result
    bar <- ggplot(d, aes(xmin = 0, xmax = 0.1, ymin = d$value-0.01, ymax = d$value)) + geom_rect(fill = d$col) +      
      scale_y_continuous(trans = "reverse", breaks = seq(0, 1, 0.1)) +
      labs(title = ifelse(thresholds$sig_type == 'P-value', '  P', 'FDR')) + theme_genoppi_bar() + coord_fixed()
    bar
  })
  
  # dont know if this is needed?
  #a_vp_colorbar_dl <- reactive({
  #  a_vp_colorbar()
  #})
  
  #a_vp <- function(){ # can tjhis function be removed
  #  d <- a_pulldown_significant()
  #  p <- plot_volcano_basic(d, col_signficant = input$a_color_indv_sig, col_other = a_color_indv_insig)
  #  p
  #}
  
  a_vp_gg <- reactive({
    d <- a_pulldown_significant()
    req(input$a_color_indv_sig, input$a_color_indv_insig)
    p <- plot_volcano_basic(d, col_signficant = input$a_color_indv_sig, col_other = input$a_color_indv_insig)
    p <- plot_overlay(p, as.bait(input$a_bait_search_rep), volcano = T) # add bait
    return(p)
  })
  
  a_vp_layerx <- reactive({
    p <- a_vp_gg()
    p <- make_interactive(p, volcano = T)
    if (input$a_goi_search_rep != '') p <- add_markers_search(p, a_search_gene(), volcano = T)
    p <- add_hover_lines_volcano(p, line_pvalue = input$a_pval_thresh, line_logfc = input$a_logFC_thresh, logfc_direction = input$a_logfc_direction)
    p <- add_layout_html_axes_volcano(p)
    return(p)
  })
  
  #a_vp_plus_rep_gg <- function(){
  #  validate(
  #    need(!is.null(a_search_gene()), "")
  #  )
  #  browser()
  #  p <- a_vp_gg()
  #  goi <- a_search_gene()
  #  orig_data <- a_pulldown()
  #  searchgene <- orig_data[grepl(goi,orig_data$gene),]
  #  p1 <- search_volcano_gg(p, searchgene)
  #  p1
  #}
  
  #a_vp_pf_db_gg <- function(){
  #  validate(
  #    need(!is.null(a_pulldown()), ""),
  #    need(!is.null(a_pf_db()), "")
  #  )
  #  p <- a_vp_gg()
  #  a_pf_db <- a_pf_db_vp()
  #  pf_col <- input$colorbrewer_theme_pf
  #  label <- input$a_marker_text_prot_fam_db
  #  
  #  #if(input$colorscheme == "fdr" | input$colorscheme == "exac" | input$colorscheme == "user"){
  #    df <- ldply(a_pf_db$d_g2s, data.frame)
  #    if(nrow(df) != 0){
  #      vp_layer_genes <- vp_layer_for_uploaded_genes_gg(p, df, pf_col, label)
  #      p1 <- vp_layer_genes
  #    } else{
  #      p1 <- p
  #    }
    #}
    #else if(input$colorscheme == "cbf"){
    #  df <- ldply(a_pf_db$d_g2s, data.frame)
    #  if(nrow(df) != 0){
    #    vp_layer_genes <- vp_layer_for_uploaded_genes_cbf_gg(p, df, label)
    #    p1 <- vp_layer_genes
    #  } else{
    #    p1 <- p
    #  }
    #}
  #  p1
  #}
  
  #a_vp_pf_db_plus_gg <- function(){
  #  validate(
  #    need(!is.null(a_search_gene()), "")
  #  )
  #  p <- a_vp_pf_db_gg()
  #  goi <- a_search_gene()
  #  orig_data <- a_pulldown()
  #  searchgene <- orig_data[grepl(goi,orig_data$gene),]
  #  p1 <- search_volcano_gg(p, searchgene)
  #  p1
  #}
  
  #output$download_vp_gg <- downloadHandler(
  #  filename = function() { paste("vp_gg", '.png', sep='') },
  #  content = function(file) {
  #    if(is.null(a_pf_db())){
  #      if(is.null(a_search_gene())){
  #        ggsave(file, plot = a_vp_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
  #      } else {
  #        ggsave(file, plot = a_vp_plus_rep_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
  #      }
  #    } else if(!is.null(a_pf_db())){
  #      if(is.null(a_search_gene())){
  #        ggsave(file, plot = a_vp_pf_db_gg(), device = "png", width = 5, height = 4, units = "in", dpi = 300)
  #      } else{
  #        ggsave(file, plot = a_vp_pf_db_plus_gg(), device = "png", width = 5, height = 4, units = "in", dpi = 300)
  #      }
  #    }
  #  }
  #)
  
  #a_sp_gg <- function(){
  #  d <- a_pulldown()
    #if(input$colorscheme == "fdr"){
  #    req(input$a_color_indv_sig, input$a_color_indv_insig)
  #    data <- separate_to_groups_for_color_integrated(d, input$a_fdr_thresh, input$a_color_indv_sig, input$a_color_indv_insig)
  #    p <- plot_scatter_qc_gg(data)
    #} else if(input$colorscheme == "exac"){
    #  d$s <- exac$em_p_hi[match(d$gene, exac$GENE_NAME)]
    #  d$s[is.na(d$s)] <- 2
    #  below_thresh <- subset(d, s < 0.9)
    #  above_thresh <- subset(d, s >= 0.9)
    #  no_exist <- subset(d, s == 2)
    #  p <- plot_scatter_exac_gg(below_thresh, above_thresh, no_exist)
    #} else if(input$colorscheme == "cbf"){
    #  data <- separate_to_groups_for_cbf_integrated(d, input$a_fdr_thresh)
    #  p <- plot_scatter_qc_gg(data)
    #} else if(input$colorscheme == "user"){
    #  validate(
    #    need(!is.null(input$file_color), "Please upload file with gene and score")
    #  )
    #  d1 <- a_in_file_color()
    #  req(input$colorbrewer_theme)
    #  col_theme <- input$colorbrewer_theme
    #  if(input$colorscheme_style == "cont"){
    #    df <- separate_to_groups_for_color_continuous(d, d1, col_theme)
    #    data <- df$df1
    #    data1 <- df$no_exist
    #  } else if(input$colorscheme_style == "disc"){
    #    df <- separate_to_groups_for_color_discrete(d, d1, col_theme)
    #    data <- df$df1
    #    data1 <- df$no_exist
    #    data2 <- rbind(data, data1)
    #    p <- ggplot(data = data2, aes(x = logFC, y = -log10(pvalue), text = gene)) +
    #      geom_point(data = data1, alpha = 0.5, size = 1.5, colour = "#f7f7f7") + 
    #      geom_point(data = data, alpha = 0.5, size = 1.5, colour = data$col) + 
    #      xlab("log2FC") + ylab("-log10(P)") +
    #      theme_minimal() +
    #      theme(axis.text.x = element_text(size=7),
    #            axis.title.x=element_text(size=8),
    #            axis.text.y=element_text(size=7),
    #            axis.title.y=element_text(size=8),
    #            panel.grid.minor = element_blank(),
    #            plot.margin = unit(c(1,1,1,1), "pt"))
    #  }
    #  p
    #}
  #  p
  #}
  
  #a_sp_plus_rep_gg <- function(){
  #  validate(
  #    need(!is.null(a_search_gene()), "")
  #  )
  #  p <- a_sp_gg()
  #  goi <- a_search_gene()
  #  orig_data <- a_pulldown()
  #  searchgene <- orig_data[grepl(goi,orig_data$gene),]
  #  p1 <- search_scatter_gg(p, searchgene)
  #  p1
  #}
  
  #a_sp_pf_db_gg <- function(){
  #  validate(
  #    need(!is.null(a_pulldown()), ""),
  #    need(!is.null(a_pf_db()), "")
  #  )
  #  p <- a_sp_gg()
  #  a_pf_db <- a_pf_db_vp()
  #  pf_col <- input$colorbrewer_theme_pf
  #  label <- input$a_marker_text_prot_fam_db
  #  
  #  #if(input$colorscheme == "fdr" | input$colorscheme == "exac" | input$colorscheme == "user"){
  #    df <- ldply(a_pf_db$d_g2s, data.frame)
  #    if(nrow(df) != 0){
  #      sp_layer_genes <- sp_layer_for_uploaded_genes_gg(p, df, pf_col, label)
  #      p1 <- sp_layer_genes
  #    } else{
  #      p1 <- p
  #    }
    #} else if(input$colorscheme == "cbf"){
    #  df <- ldply(a_pf_db$d_g2s, data.frame)
    #  if(nrow(df) != 0){
    #    sp_layer_genes <- sp_layer_for_uploaded_genes_cbf_gg(p, df, label)
    #    p1 <- sp_layer_genes
    #  } else{
    #    p1 <- p
    #  }
    #}
  #  p1
  #}
  
  #a_sp_pf_db_plus_gg <- function(){
  #  validate(
  #    need(!is.null(a_search_gene()), "")
  #  )
  #  p <- a_sp_pf_db_gg()
  #  goi <- a_search_gene()
  #  orig_data <- a_pulldown()
  #  searchgene <- orig_data[grepl(goi,orig_data$gene),]
  #  p1 <- search_scatter_gg(p, searchgene)
  #  p1
  #}
  
  #output$download_sp_gg <- downloadHandler(
  #  filename = function() { paste("sp_gg", '.png', sep='') },
  #  content = function(file) {
  #    if(is.null(a_pf_db())){
  #      if(is.null(a_search_gene())){
  #        ggsave(file, plot = a_sp_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
  #      } else {
  #        ggsave(file, plot = a_sp_plus_rep_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
  #      }
  #    } else if(!is.null(a_pf_db())){
  #      if(is.null(a_search_gene())){
  #        ggsave(file, plot = a_sp_pf_db_gg(), device = "png", width = 5, height = 4, units = "in", dpi = 300)
  #      } else{
  #        ggsave(file, plot = a_sp_pf_db_plus_gg(), device = "png", width = 5, height = 4, units = "in", dpi = 300)
  #      }
  #    }
  #  }
  #)
  
  
  #a_vp1 <- reactive({
  #  browser()
  #  d <- a_pulldown()
  #  q <- a_vp_gg()
  #  gg_p <- ggplotly(p = q, tooltip="text") #, tooltip = c("gene", "FDR")
  #  gg_p <- gg_p %>%
  #    layout(xaxis = list(title = "log<sub>2</sub>(Fold change)",
  #                        yaxis = list(title = "-log<sub>10</sub>(<i>P</i>-value)")))
  #  
  #})
  #
  #a_vp1_plus_rep <- reactive({
  #  q <- a_vp_plus_rep_gg_reg_label()
  #  gg_p <- ggplotly(p = q, tooltip="text") #, tooltip = c("gene", "FDR")
  #  gg_p <- gg_p %>%
  #    layout(xaxis = list(title = "log<sub>2</sub>(Fold change)",
  #                        yaxis = list(title = "-log<sub>10</sub>(<i>P</i>-value)"))) 
  #})
  
  #a_vp_plus_rep_gg_reg_label <- function(){
  #  validate(
  #    need(!is.null(a_search_gene()), "")
  #  )
  #  p <- a_vp_gg()
  #  goi <- a_search_gene()
  #  orig_data <- a_pulldown()
  #  searchgene <- orig_data[grepl(goi,orig_data$gene),]
  #  p1 <- search_volcano_gg_reg_label(p, searchgene)
  #  p1
  #}
  
  #a_vp_plus_rep <- reactive({
  #  validate(
  #    need(!is.null(a_search_gene()), "")
  #  )
  #  p <- a_vp_layerx()
  #  goi <- a_search_gene()
  #  orig_data <- a_pulldown()
  #  searchgene <- orig_data[grepl(goi,orig_data$gene),]
  #  p1 <- search_volcano(p, searchgene)
  #  p1
  #})
  

  
  #a_pf_db_vp <- reactive({
  #  validate(
  #    need(input$a_file_pulldown_r != '', "")
  #  )
  #  if(!is.null(a_pf_db_search())){
  #    d <- a_pulldown()
  #    gene_interest <- a_pf_db_search()
  #    d_g2s <- lapply(gene_interest, function(x) subset(d, gene %in% x) )
  #    list(d_g2s=d_g2s)
  #  }
  #})
  
  #a_vp_pf_db <- reactive({
  #  validate(
  #    need(!is.null(a_pulldown()), ""),
  #    need(!is.null(a_pf_db()), "")
  #  )
  #  d <- a_pulldown()
  #  a_pf_db <- a_pf_db_vp()
  #  pf_col <- input$colorbrewer_theme_pf
  #  #if(input$colorscheme == "fdr"){
  #    req(input$a_color_indv_sig, input$a_color_indv_insig)
  #    data <- separate_to_groups_for_color_integrated(d, input$a_fdr_thresh, input$a_color_indv_sig, input$a_color_indv_insig)
  #    p <- plot_ly(colors = pf_col, showlegend = T, width = 650, height = 550)
  #    for(i in nrow(data)){
  #      p <- add_markers(p, data = data, x = ~logFC, y = ~-log10(pvalue),
  #                       marker = list(size = 6, cmin = 0, cmax = 1, color = ~col), 
  #                       opacity = 0.8, 
  #                       text = ~paste(gene), hoverinfo = "text", name = "pull down")
  #    }
  #    p 
    #} else if(input$colorscheme == "exac"){
    #  d$s <- exac$em_p_hi[match(d$gene, exac$GENE_NAME)]
    #  d$s[is.na(d$s)] <- 2
    #  below_thresh <- subset(d, s < 0.9)
    #  above_thresh <- subset(d, s >= 0.9)
    #  no_exist <- subset(d, s == 2)
    #  p <- plot_ly(colors = pf_col, showlegend = T, width = 650, height = 550)
    #  p <- add_markers(p, data = below_thresh, x = ~logFC, y = ~-log10(pvalue),
    #                   marker = list(size = 8, line = list(width=0.1, color = 'black'), cmin = 0, cmax = 1, color = "#66c2a5"),
    #                   opacity = 0.8, 
    #                   text = ~paste(gene), hoverinfo = "text", name = paste0("pLI<0.9 (", nrow(below_thresh), ")"))
    #  p <- add_markers(p, data = above_thresh, x = ~logFC, y = ~-log10(pvalue),
    #                   marker = list(size = 8, line = list(width=0.1, color = "black"), cmin = 0, cmax = 1, color = "#fc8d62"),
    #                   opacity = 0.8, 
    #                   text = ~paste(gene), hoverinfo = "text", name = paste0("pLI>=0.9 (", nrow(above_thresh), ")"))
    #  p <- add_markers(p, data = no_exist, x = ~logFC, y = ~-log10(pvalue),
    #                   marker = list(size = 8, line = list(width=0.1, color = "black"), cmin = 0, cmax = 1, color = "#8da0cb"),
    #                   opacity = 0.8, 
    #                   text = ~paste(gene), hoverinfo = "text", name = paste0("not in ExAC (", nrow(no_exist), ")"))
    #  p
    #} else if(input$colorscheme == "cbf"){
    #  data <- separate_to_groups_for_cbf_integrated(d, input$a_fdr_thresh)
    #  p <- plot_ly(colors = "Greys", showlegend = T, width = 650, height = 550)
    #  for(i in nrow(data)){
    #    p <- add_markers(p, data = data, x = ~logFC, y = ~-log10(pvalue),
    #                     marker = list(size = 6, cmin = 0, cmax = 1, color = ~col), 
    #                     opacity = 0.6, 
    #                     text = ~paste(gene), hoverinfo = "text", name = "pull down")
    #  }
    #} else if(input$colorscheme == "user"){
    #  validate(
    #    need(!is.null(input$file_color), "Please upload file with gene and score")
    #  )
    #  d1 <- a_in_file_color()
    #  col_theme <- input$colorbrewer_theme
    #  if(input$colorscheme_style == "cont"){
    #    df <- separate_to_groups_for_color_continuous(d, d1, col_theme)
    #    data <- df$df1
    #    data1 <- df$no_exist
    #  } else if(input$colorscheme_style == "disc"){
    #    df <- separate_to_groups_for_color_discrete(d, d1, col_theme)
    #    data <- df$df1
    #    data1 <- df$no_exist
    #  }
    #  p <- plot_ly(colors = pf_col, showlegend = T, width = 650, height = 550)
    #  p <- add_markers(p, data = data1, x = ~logFC, y = ~-log10(pvalue),
    #                   marker = list(size = 7, line = list(width=0.1, color = "grey89"), cmin = 0, cmax = 1, color = "#f7f7f7"),
    #                   opacity = 0.8,
    #                   text = ~paste(gene), hoverinfo = "text", name = paste0("Not in user data (", nrow(df$no_exist), ")"))
    #  for(i in nrow(data)){
    #    p <- add_markers(p, data = data, x = ~logFC, y = ~-log10(pvalue),
    #                     marker = list(size = 7, cmin = 0, cmax = 1, color = ~col, line = list(width=0.2, color = "grey89")),
    #                     opacity = 1,
    #                     text = ~paste0(gene, ", FDR=", signif(FDR, digits = 3)), hoverinfo = "text",
    #                     name = paste0("Found in user data (", nrow(data), ")")) #, name = "pull down"
    #  }
    #  p
    #}
    #p <- p %>%
    #  layout(xaxis = list(range=~c(min(d$logFC)-0.5, max(d$logFC)+0.5)),
    #         yaxis = list(range=~c(min(-log10(d$pvalue)-0.5), max(-log10(d$pvalue))+0.5)))
  #  
  #  #if(input$colorscheme == "fdr" | input$colorscheme == "exac" | input$colorscheme == "user"){
  #    df <- ldply(a_pf_db$d_g2s, data.frame)
  #    if(nrow(df) != 0){
  #      if(input$a_marker_text_prot_fam_db == "yes_label"){
  #        vp_layer_genes <- vp_layer_for_uploaded_genes(p, df)
  #      } else if(input$a_marker_text_prot_fam_db == "no_label"){
  #        vp_layer_genes <- vp_layer_for_uploaded_genes_no_text(p, df)
  #      }
  #      p <- vp_layer_genes
  #    } else{
  #      vp_layer_no_genes <- vp_layer_for_uploaded_genes_none(p, d)
  #      p <- vp_layer_no_genes
  #    }
    #} else if(input$colorscheme == "cbf"){
    #  df <- ldply(a_pf_db$d_g2s, data.frame)
    #  if(nrow(df) != 0){
    #    if(input$a_marker_text_prot_fam_db == "yes_label"){
    #      vp_layer_genes <- vp_layer_for_uploaded_genes_cbf(p, df)
    #    } else if(input$a_marker_text_prot_fam_db == "no_label"){
    #      vp_layer_genes <- vp_layer_for_uploaded_genes_cbf_no_text(p, df)
    #    }
    #    p <- vp_layer_genes
    #  } else{
    #    vp_layer_no_genes <- vp_layer_for_uploaded_genes_none_cbf(p, d)
    #    p <- vp_layer_no_genes
    #  }
    #}
    #p <- p %>%
    #  add_lines(x = ~c(min(d$logFC)-0.5, max(d$logFC)+0.5), y = ~-log10(input$a_pval_thresh), line = list(dash = "dash", width = 0.5, color = "#2b333e"), 
    #            name = '', hoverinfo = "text", text = paste0("pvalue = ", input$a_pval_thresh), showlegend = F) %>%
    #  add_lines(x = input$a_logFC_thresh[1], y = ~c(min(-log10(d$pvalue)-0.5), max(-log10(d$pvalue))+0.5), line = list(dash = "dash", width = 0.5, color = "#252525"), 
    #            name = '', hoverinfo = "text", text = paste0("logFC = ", input$a_logFC_thresh[1]), showlegend = F) %>%
    #  add_lines(x = input$a_logFC_thresh[2], y = ~c(min(-log10(d$pvalue)-0.5), max(-log10(d$pvalue))+0.5), line = list(dash = "dash", width = 0.5, color = "#252525"), 
    #            name = '', hoverinfo = "text", text = paste0("logFC = ", input$a_logFC_thresh[2]), showlegend = F)
  #})
  
  #a_vp_pf_db_plus_rep <- reactive({
  #  validate(
  #    need(!is.null(a_search_gene()), "")
  #  )
  #  p <- a_vp_pf_db()
  #  goi <- a_search_gene()
  #  orig_data <- a_pulldown()
  #  searchgene <- orig_data[grepl(goi,orig_data$gene),]
  #  p1 <- search_volcano(p, searchgene)
  #  p1
  #})
  
  #a_vp_count <- reactive({
  #  d <- a_pulldown_significant()
  #  count <- data.frame(paste0(sum(d$significant), " (total = ", nrow(d), ")"))
  #  row.names(count) <- c("Pull down")
  #  colnames(count) <- c(paste0("FDR<", input$a_fdr_thresh, ", pvalue<", input$a_pval_thresh, ", ", input$a_logFC_thresh[1], "<logFC<", input$a_logFC_thresh[2]))
  #  count
  #})
  
  a_verbatim_count <- reactive({
    d <- a_pulldown_significant()
    HTML(paste(bold(sum(d$significant)), 'proteins out of', bold(nrow(d)), 'proteins enriched.'))
  })
  
  
  a_vp_count_text <- reactive({
    if(!is.null(a_verbatim_count())){
      enriched <- paste(c(monitor_significance_tresholds()$sig, monitor_logfc_threshold()$sig), collapse = ', ')
      enriched
    }
  })
  
  #a_sp_cor <- reactive({
  #  input_file <- a_pulldown()
  #  if ("rep1" %in% colnames(input_file) & "rep2" %in% colnames(input_file)){
  #    cor <- signif(cor(input_file$rep1, input_file$rep2), 4)
  #    cor <- paste0("correlation coefficient: ", cor)
  #  } else if ("logFC" %in% colnames(input_file) & "FDR" %in% colnames(input_file) & "pvalue" %in% colnames(input_file)){
  #    return(NULL)
  #  }
  #})
  
  
  #a_sp_gg <- reactive({
  #  stop('error, deprecated')
  #})
  
  a_sp_gg <- reactive({
    
    # what replicates are inputted
    req(input$a_select_scatterplot)
    rep = unlist(strsplit(input$a_select_scatterplot,'\\.'))
    
    # handle all plots
    d = a_pulldown_significant()
    p = plot_scatter_basic(d, col_signficant = input$a_color_indv_sig, col_other = input$a_color_indv_insig)

    # handle individual plot
    p1 = p[[input$a_select_scatterplot]]$ggplot
    p1 = plot_overlay(p1, as.bait(input$a_bait_search_rep), x=rep[1], y=rep[2])
    p1$r = format(p[[input$a_select_scatterplot]]$correlation, digits = 3)
    p1

    # add markers 
    #p1 = make_interactive(p1, x=rep[1], y=rep[2])
    #if (input$a_goi_search_rep != '') p1 <- add_markers_search(p1, a_search_gene(), x=rep[1], y=rep[2], volcano = F)
    #p1 = add_layout_html_axes_scatterplot(p1, rep[1], rep[2], paste0('r=',correlation))
    #p1

  })
  
  a_sp <- reactive({
    
    # get basic stats
    p1 = a_sp_gg()
    rep = unlist(strsplit(input$a_select_scatterplot,'\\.'))
    r = p1$r
    
    # concert into interactive graphics
    p1 = make_interactive(p1, x=rep[1], y=rep[2])
    if (input$a_goi_search_rep != '') p1 <- add_markers_search(p1, a_search_gene(), x=rep[1], y=rep[2], volcano = F)
    p1 = add_layout_html_axes_scatterplot(p1, rep[1], rep[2], paste0('r=',r))
    p1
    
  })
  
  
  #a_sp_plus <- reactive({
  #  validate(
  #    need(!is.null(a_search_gene()), "")
  #  )
  #  p <- a_sp()
  #  goi <- a_search_gene()
  #  orig_data <- a_pulldown()
  #  searchgene <- orig_data[grepl(goi,orig_data$gene),]
  #  p1 <- search_scatter(p, searchgene)
  #  p1
  #})
  
  #a_sp_pf_db <- reactive({
  #  validate(
  #    need(!is.null(a_pulldown()), ""),
  #    need(!is.null(a_pf_db()), "")
  #  )
  #  d <- a_pulldown()
  #  a_pf_db <- a_pf_db_vp()
  #  cc <- a_sp_cor()
  #  pf_col <- input$colorbrewer_theme_pf
  #  if(input$colorscheme == "fdr"){
  #    req(input$a_color_indv_sig, input$a_color_indv_insig)
  #    data <- separate_to_groups_for_color_integrated(d, input$a_fdr_thresh, input$a_color_indv_sig, input$a_color_indv_insig)
  #    p <- plot_ly(colors = pf_col, showlegend = F, width = 550, height = 550) 
  #    p <- add_lines(p, data = d, x = ~c(ceiling(min(rep1, rep2)), floor(max(rep1, rep2))), y = ~c(ceiling(min(rep1, rep2)), floor(max(rep1, rep2))),
  #                   line = list(dash = "dash", width = 1, color = "#252525"), showlegend = FALSE)
  #    for(i in nrow(data)){
  #      p <- add_markers(p, data = data, x = ~rep1, y = ~rep2, 
  #                       marker = list(size = 6, cmin = 0, cmax = 1, color = ~col), #line = list(width=0.1, color = "black"),
  #                       opacity = 0.8, 
  #                       text = ~paste(gene), hoverinfo = "text", name = "pull down")
  #    }
  #  } else if(input$colorscheme == "exac"){
  #    d$s <- exac$em_p_hi[match(d$gene, exac$GENE_NAME)]
  #    d$s[is.na(d$s)] <- 2
  #    below_thresh <- subset(d, s < 0.9)
  #    above_thresh <- subset(d, s >= 0.9)
  #    no_exist <- subset(d, s == 2)
  #    p <- plot_ly(colors = pf_col, showlegend = F, width = 550, height = 550) 
  #    p <- add_lines(p, data = d, x = ~c((min(rep1, rep2)), (max(rep1, rep2))), y = ~c((min(rep1, rep2)), (max(rep1, rep2))),
  #                   line = list(dash = "dash", width = 1, color = "#252525"), showlegend = FALSE)
  #    p <- add_markers(p, data = below_thresh, x = ~rep1, y = ~rep2, 
  #                     marker = list(size = 8, line = list(width=0.1, color = 'black'), cmin = 0, cmax = 1, color = "#66c2a5"),
  #                     opacity = 0.7, 
  #                     text = ~paste(gene), hoverinfo = "text") #, name = paste0("pLI<0.9 (", nrow(below_thresh), ")")
  #    p <- add_markers(p, data = above_thresh, x = ~rep1, y = ~rep2, 
  #                     marker = list(size = 8, line = list(width=0.1, color = "black"), cmin = 0, cmax = 1, color = "#fc8d62"),
  #                     opacity = 0.7, 
  #                     text = ~paste(gene), hoverinfo = "text") #, name = paste0("pLI>=0.9 (", nrow(above_thresh), ")")
  #    p <- add_markers(p, data = no_exist, x = ~rep1, y = ~rep2, 
  #                     marker = list(size = 8, line = list(width=0.1, color = "black"), cmin = 0, cmax = 1, color = "#8da0cb"),
  #                     opacity = 0.7, 
  #                     text = ~paste(gene), hoverinfo = "text") #, name = paste0("not in ExAC (", nrow(no_exist), ")")
  #    p
  #  } else if(input$colorscheme == "cbf"){
  #    data <- separate_to_groups_for_cbf_integrated(d, input$a_fdr_thresh)
  #    p <- plot_ly(colors = "Greys", showlegend = F, width = 550, height = 550) 
  #    p <- add_lines(p, data = d, x = ~c(ceiling(min(rep1, rep2)), floor(max(rep1, rep2))), y = ~c(ceiling(min(rep1, rep2)), floor(max(rep1, rep2))),
  #                   line = list(dash = "dash", width = 1, color = "#252525"), showlegend = FALSE)
  #    for(i in nrow(data)){
  #      p <- add_markers(p, data = data, x = ~rep1, y = ~rep2, 
  #                       marker = list(size = 6, cmin = 0, cmax = 1, color = ~col), #line = list(width=0.1, color = "black"), 
  #                       opacity = 0.6, 
  #                       text = ~paste(gene), hoverinfo = "text", name = "pull down")
  #    }
  #  } else if(input$colorscheme == "user"){
  #    validate(
  #      need(!is.null(input$file_color), "Please upload file with gene and score")
  #    )
  #    d1 <- a_in_file_color()
  #    col_theme <- input$colorbrewer_theme
  #    if(input$colorscheme_style == "cont"){
  #      df <- separate_to_groups_for_color_continuous(d, d1, col_theme)
  #      data <- df$df1
  #      data1 <- df$no_exist
  #    } else if(input$colorscheme_style == "disc"){
  #      df <- separate_to_groups_for_color_discrete(d, d1, col_theme)
  #      data <- df$df1
  #      data1 <- df$no_exist
  #    }
  #    p <- plot_ly(colors = pf_col, showlegend = FALSE, width = 550, height = 550) 
  #    p <- add_lines(p, data = d, x = ~c((min(rep1, rep2)), (max(rep1, rep2))), y = ~c((min(rep1, rep2)), (max(rep1, rep2))),
  #                   text = "x=y", hoverinfo = "text",
  #                   line = list(dash = "dash", width = 1, color = "#252525"), showlegend = FALSE)
  #    p <- add_markers(p, data = data1, x = ~rep1, y = ~rep2,
  #                     marker = list(size = 7, line = list(width=0.1, color = "grey89"), cmin = 0, cmax = 1, color = "#f7f7f7"),
  #                     opacity = 0.8,
  #                     text = ~paste(gene), hoverinfo = "text", name = paste0("Not in user data (", nrow(df$no_exist), ")"))
  #    for(i in nrow(data)){
  #      p <- add_markers(p, data = data, x = ~rep1, y = ~rep2,
  #                       marker = list(size = 7, cmin = 0, cmax = 1, color = ~col, line = list(width=0.2, color = "grey89")),
  #                       opacity = 1,
  #                       text = ~paste0(gene, ", FDR=", signif(FDR, digits = 3)), hoverinfo = "text",
  #                       name = paste0("Found in user data (", nrow(data), ")")) #, name = "pull down"
  #    }
  #    p
  #  }
  #  if(input$colorscheme == "fdr" | input$colorscheme == "exac" | input$colorscheme == "user"){
  ##    df <- ldply(a_pf_db$d_g2s, data.frame)
  #    if(nrow(df) != 0){
  #      if(input$a_marker_text_prot_fam_db == "yes_label"){
  #        sp_layer_genes <- sp_layer_for_uploaded_genes(p, df)
  #      } else if(input$a_marker_text_prot_fam_db == "no_label"){
  #        sp_layer_genes <- sp_layer_for_uploaded_genes_no_text(p, df)
  #      }
  #      p <- sp_layer_genes
  #    } else{
  #      sp_layer_no_genes <- sp_layer_for_uploaded_genes_none(p, d)
  #      p <- sp_layer_no_genes
  #    }
  #  } else if(input$colorscheme == "cbf"){
  #    df <- ldply(a_pf_db$d_g2s, data.frame)
  #    if(nrow(df) != 0){
  #      if(input$a_marker_text_prot_fam_db == "yes_label"){
  #        sp_layer_genes <- sp_layer_for_uploaded_genes_cbf(p, df)
  #      } else if(input$a_marker_text_prot_fam_db == "no_label"){
  #        sp_layer_genes <- sp_layer_for_uploaded_genes_cbf_no_text(p, df)
  #      }
  #      p <- sp_layer_genes
  #    } else{
  #      sp_layer_no_genes <- sp_layer_for_uploaded_genes_none_cbf(p, d)
  #      p <- sp_layer_no_genes
  #    }
  #  }
  #  p <- p %>%
  #    layout(xaxis = list(title = "rep1", range=~c((min(d$rep1, d$rep2))-1, (max(d$rep1, d$rep2))+1)), 
  #           yaxis = list(title = "rep2", range=~c((min(d$rep1, d$rep2))-1, (max(d$rep1, d$rep2))+1)), 
  #           title = cc, titlefont = list(size=12))
  #})
  
  
  
  ### volcano plot for multiple overlays?
  ### seems to be only compiling stats
  #a_multi_vp <- eventReactive(input$a_make_plot, {
  #  validate(
  #    need(!is.null(a_pulldown()), "")
  #  )
  #  d <- a_pulldown_significant()
  #  
  #  # InWeb, SNP to gene, and genes upload
  #  if(!is.null(a_bait_gene_layer()) & !is.null(a_snp()) & !is.null(a_upload_genes())){
  #    inwebFile <- sample()
  #    gene_interest <- a_genes_uploaded()
  #    snp_interest <- SNP_to_gene()
  #    bait <- a_bait_gene_layer()
  #    d_in <- subset(d, gene %in% inwebFile$gene & gene != bait)
  #    d_snp <- merge(d, snp_interest, by.x = 'gene', by.y = 'gene')
  #    d_g2s <- lapply(gene_interest, function(x) subset(d, gene %in% x) )  
  #    list(d_in=d_in, d_snp=d_snp, d_g2s=d_g2s)
  #  } else if(!is.null(a_bait_gene_layer()) & is.null(a_snp()) & is.null(a_upload_genes())){
  #    # InWeb only 
  #    inwebFile <- sample()
  #    bait <- a_bait_gene_layer()
  #    d_in <- subset(d, gene %in% inwebFile$gene & gene != bait)
  #    list(d_in=d_in)
  #  } else if(is.null(a_bait_gene_layer()) & is.null(a_snp()) & is.null(a_upload_genes())){
  #    # None
  #  } else if(!is.null(a_bait_gene_layer()) & is.null(a_snp()) & !is.null(a_upload_genes())){
  #    # InWeb, and genes upload
  #    inwebFile <- sample()
  #    gene_interest <- a_genes_uploaded()
  #    bait <- a_bait_gene_layer()
  #    d_in <- subset(d, gene %in% inwebFile$gene & gene != bait)
  #    d_g2s <- lapply(gene_interest, function(x) subset(d, gene %in% x) )  
  #    list(d_in=d_in, d_g2s=d_g2s)
  #  } else if(is.null(a_bait_gene_layer()) & !is.null(a_snp()) & !is.null(a_upload_genes())){
  #    # SNP to gene, and genes upload
  #    gene_interest <- a_genes_uploaded()
  #    snp_interest <- SNP_to_gene()
  #    d_snp <- merge(d, snp_interest, by.x = 'gene', by.y = 'gene')
  #    d_g2s <- lapply(gene_interest, function(x) subset(d, gene %in% x) )
  #    list(d_snp=d_snp, d_g2s=d_g2s)
  #  } else if(is.null(a_bait_gene_layer()) & is.null(a_snp()) & !is.null(a_upload_genes())){
  #    # Genes upload only
  #    gene_interest <- a_genes_uploaded()
  #    d_g2s <- lapply(gene_interest, function(x) subset(d, gene %in% x) )  
  #    list(d_g2s=d_g2s)
  #  } else if(is.null(a_bait_gene_layer()) & !is.null(a_snp()) & is.null(a_upload_genes())){
  #    # SNP to gene only
  #    snp_interest <- SNP_to_gene()
  #    d_snp <- merge(d, snp_interest, by.x = 'gene', by.y = 'gene')
  #    list(d_snp=d_snp)
  #  } else if(!is.null(a_bait_gene_layer()) & !is.null(a_snp()) & is.null(a_upload_genes())){
  #    # InWeb, and SNP to gene
  #    inwebFile <- sample()
  #    snp_interest <- SNP_to_gene()
  #    bait <- a_bait_gene_layer()
  #    d_in <- subset(d, gene %in% inwebFile$gene & gene != bait)
  #    d_snp <- merge(d, snp_interest, by.x = 'gene', by.y = 'gene')
  #    list(d_in=d_in, d_snp=d_snp)
  #  }
  #})
  
  #a_multi_vp_colorbar <- reactive({
    #wait until multi_vp is created
  #  multi_vp <- a_multi_vp()
  #  FDR <- seq(0, 1, 0.01)
  #  limit <- rep("FDR", 101)
  #  d <- data.frame(limit, FDR)
  #  if(input$colorscheme == "fdr"){
  #    req(input$a_color_multi_sig, input$a_color_multi_insig)
  #    d1 <- separate_to_groups_for_color_integrated(d, input$a_fdr_thresh, input$a_color_multi_sig, input$a_color_multi_insig)
  #    mycol <- as.vector(d1$col)
  #    bar <- ggplot(d1, aes(xmin = 0, xmax = 0.1, ymin = d1$FDR-0.01, ymax = d1$FDR)) + geom_rect(fill = mycol) +      
  #      scale_y_continuous(trans = "reverse", breaks = seq(0, 1, 0.1)) +
  #      labs(title = "FDR") +
  #      theme(axis.title.x=element_blank(),
  #            axis.text.x=element_blank(),
  #            axis.ticks.x=element_blank(),
  #            panel.background=element_blank(),
  #            plot.title = element_text(size = rel(1))) + coord_fixed()
  #    bar
  #  } else if(input$colorscheme == "exac"){
  #    d1 <- separate_to_groups_for_exac_bar(d)
  #    mycol <- as.vector(d1$col)
  #    bar <- ggplot(d1, aes(xmin = 0, xmax = 0.1, ymin = d1$FDR-0.01, ymax = d1$FDR)) + geom_rect(fill = mycol) +
  #      labs(y = " pLI < 0.9     pLI >= 0.9    not in ExAC") +
  #      theme(axis.title.x=element_blank(),
  #            axis.text.y=element_blank(),
  #            axis.ticks.y=element_blank(),
  #            axis.text.x=element_blank(),
  #            axis.ticks.x=element_blank(),
  #            panel.background=element_blank(),
  #            axis.title = element_text(size = rel(1))) + coord_fixed()
  #    bar
  #  } else if(input$colorscheme == "cbf"){
  #    d1 <- separate_to_groups_for_cbf_integrated(d, input$a_fdr_thresh)
  #    mycol <- as.vector(d1$col)
  #    bar <- ggplot(d1, aes(xmin = 0, xmax = 0.1, ymin = d1$FDR-0.01, ymax = d1$FDR)) + geom_rect(fill = mycol) +      
  #      scale_y_continuous(trans = "reverse", breaks = seq(0, 1, 0.1)) +
  #      labs(title = "FDR") +
  #      theme(axis.title.x=element_blank(),
  #            axis.text.x=element_blank(),
  #            axis.ticks.x=element_blank(),
  #            panel.background=element_blank(),
  #            plot.title = element_text(size = rel(1))) + coord_fixed()
  #    bar
  #  }
  #})
  
  # generate plot in ggformat
  a_integrated_plot_gg <- reactive({
    p = a_vp_gg()
    if (!is.null(input$a_gwas_catalogue)) if (input$a_gwas_catalogue != '') p = plot_overlay(p, list(gwas=a_gwas_catalogue_mapping()), volcano = T)
    if (!is.null(input$a_bait_rep)) if (input$a_bait_rep != '') p = plot_overlay(p, list(inweb=a_inweb_mapping()), volcano = T)
    if (!is.null(input$a_file_SNP_rep)){p = plot_overlay(p, list(snps=a_snp_mapping()), volcano = T)}
    if (!is.null(input$a_file_genes_rep)){p = plot_overlay(p, list(snps=a_genes_upload()), volcano = T)}
    p$overlay <- collapse_labels(p$overlay)
    p
  })
  
  # convert into plotly graphics
  a_integrated_plot <- reactive({
    p <- a_integrated_plot_gg()
    p <- make_interactive(p, volcano = T, source = "Multi_VolcanoPlot")
    p <- add_hover_lines_volcano(p, line_pvalue = input$a_pval_thresh, line_logfc = input$a_logFC_thresh, logfc_direction = input$a_logfc_direction)
    if (input$a_goi_search_rep != '') p <- add_markers_search(p, a_search_gene(), volcano = T)
    p <- add_layout_html_axes_volcano(p, 500, 500) # error in searching overlay here when layout width/height supplied. 
    p
  })
  
  # download integrated plot graphics.
  input_integrated_plot_gg <- function(){a_integrated_plot_gg()}
  output$a_integrated_plot_download = downloadHandler(
    filename = 'integrated-plot.png',
    content = function(file) {
      device <- function(..., width, height) {
        grDevices::png(..., width = width, height = height,
                       res = 300, units = "in")
      }
      ggsave(file, plot =  input_integrated_plot_gg(), device = device)
    })
  
  # download basic scatter plot
  input_sp_gg <- function(){a_sp_gg()}
  output$a_scatter_plot_download = downloadHandler(
    filename = 'scatter-plot.png',
    content = function(file) {
      device <- function(..., width, height) {
        grDevices::png(..., width = width, height = height,
                       res = 300, units = "in")
      }
      ggsave(file, plot =  input_sp_gg(), device = device)
    })
  
  # download basic scatter plot
  input_vp_gg <- function(){a_vp_gg()}
  output$a_volcano_plot_download = downloadHandler(
    filename = 'basic-volcano-plot.png',
    content = function(file) {
      device <- function(..., width, height) {
        grDevices::png(..., width = width, height = height,
                       res = 300, units = "in")
      }
      ggsave(file, plot =  input_vp_gg(), device = device)
    })
  
  
  
  #a_multi_vp_layer <- reactive({
  #  stop('a_multi_vp_layer is deprecated!!')
  #})
  # 
  #
  #a_multi_vp_layer_gg <- function(){
  #  multi_vp <- a_multi_vp()
  #  d <- a_pulldown()
  #  p_col <- input$colorbrewer_theme_goi
  #  label <- input$a_marker_text
  #  
  #  if(input$colorscheme == "fdr"){
  #    req(input$a_color_multi_sig, input$a_color_multi_insig)
  #    data <- separate_to_groups_for_color_integrated(d, input$a_fdr_thresh, input$a_color_multi_sig, input$a_color_multi_insig)
  #    p <- plot_volcano_qc_gg(data)
  #  } else if(input$colorscheme == "exac"){
  #    d$s <- exac$em_p_hi[match(d$gene, exac$GENE_NAME)]
  #    d$s[is.na(d$s)] <- 2
  #    below_thresh <- subset(d, s < 0.9)
  #    above_thresh <- subset(d, s >= 0.9)
  #    no_exist <- subset(d, s == 2)
  #    p <- plot_volcano_exac_gg(below_thresh, above_thresh, no_exist)
  #  } else if(input$colorscheme == "cbf"){
  #    data <- separate_to_groups_for_cbf_integrated(d, input$a_fdr_thresh)
  #    p <- plot_volcano_qc_gg(data)
  #  } else if(input$colorscheme == "user"){
  ##    validate(
  ##      need(!is.null(input$file_color), "Please upload file with gene and score")
  #    )
  #    d1 <- a_in_file_color()
  #    req(input$colorbrewer_theme)
  #    col_theme <- input$colorbrewer_theme
  ##    if(input$colorscheme_style == "cont"){
  #      df <- separate_to_groups_for_color_continuous(d, d1, col_theme)
  #      data <- df$df1
  #      data1 <- df$no_exist
  #    } else if(input$colorscheme_style == "disc"){
  #      df <- separate_to_groups_for_color_discrete(d, d1, col_theme)
  #      data <- df$df1
  #      data1 <- df$no_exist
  #      data2 <- rbind(data, data1)
  #      p <- ggplot(data = data2, aes(x = logFC, y = -log10(pvalue), text = gene)) +
  #        geom_point(data = data1, alpha = 0.5, size = 1.5, colour = "#f7f7f7") + 
  #        geom_point(data = data, alpha = 0.5, size = 1.5, colour = data$col) + 
  #        xlab("log2FC") + ylab("-log10(P)") +
  #        theme_minimal() +
  #        theme(axis.text.x = element_text(size=7),
  #              axis.title.x=element_text(size=8),
  #              axis.text.y=element_text(size=7),
  #              axis.title.y=element_text(size=8),
  #              panel.grid.minor = element_blank(),
  #              plot.margin = unit(c(1,1,1,1), "pt"))
  #    }
  #    p
  #  }
  #  if(input$colorscheme == "fdr" | input$colorscheme == "exac" | input$colorscheme == "user"){
  #    # InWeb, SNP to gene, and genes upload
  #    if(!is.null(a_bait_gene_layer())){
  #      col <- input$a_color_inweb
  #      vp_layer_inweb <- vp_layer_for_inweb_sf_gg(p, multi_vp$d_in, col, label)
  #      p <- vp_layer_inweb
  #    }
  #    if(!is.null(SNP_to_gene())){
  #      if(nrow(multi_vp$d_snp) != 0){
  #        snp_sgl <- subset(multi_vp$d_snp, Freq == 1)
  #        snp_mgl <- subset(multi_vp$d_snp, Freq != 1)
  #        col <- c(color = colorRampPalette(brewer.pal(3, input$colorbrewer_theme_snp))(2))
  #        if(nrow(snp_sgl) != 0){
  #          vp_layer_snp2gene_sgl <- vp_layer_for_snp_to_gene_sgl_gg(p, snp_sgl, col[1], label)
  #          p <- vp_layer_snp2gene_sgl
  #        }
  #        if(nrow(snp_mgl) != 0){
  #          vp_layer_snp2gene_mgl <- vp_layer_for_snp_to_gene_mgl_gg(p, snp_mgl, col[2], label)
  #          p <- vp_layer_snp2gene_mgl
  #        }
  #      } else{
  #        p
  #      }
  #    }
  #    if(!is.null(a_upload_genes())){
  #      df <- ldply(multi_vp$d_g2s, data.frame)
  #      if(nrow(df) != 0){
  #        vp_layer_genes <- vp_layer_for_uploaded_genes_gg(p, df, p_col, label)
  #        p <- vp_layer_genes
  #      } else{
  #        p
  #      }
  #    }
  #    p
  #  } else if(input$colorscheme == "cbf"){
  #    # InWeb, SNP to gene, and genes upload
  #    if(!is.null(a_bait_gene_layer())){
  #      vp_layer_inweb <- vp_layer_for_inweb_cbf_sf_gg(p, multi_vp$d_in, label)
  #      p <- vp_layer_inweb
  #    }
  #    if(!is.null(SNP_to_gene())){
  #      if(nrow(multi_vp$d_snp) != 0){
  #        snp_sgl <- subset(multi_vp$d_snp, Freq == 1)
  #        snp_mgl <- subset(multi_vp$d_snp, Freq != 1)
  #        if(nrow(snp_sgl) != 0){
  #          vp_layer_snp2gene_sgl <- vp_layer_for_snp_to_gene_sgl_cbf_gg(p, snp_sgl, label)
  #          p <- vp_layer_snp2gene_sgl
  #        }
  #        if(nrow(snp_mgl) != 0){
  #          vp_layer_snp2gene_mgl <- vp_layer_for_snp_to_gene_mgl_cbf_gg(p, snp_mgl, label)
  #          p <- vp_layer_snp2gene_mgl
  #        }
  #      } else{
  #        p
  #      }
  #    }
  #    if(!is.null(a_upload_genes())){
  #      df <- ldply(multi_vp$d_g2s, data.frame)
  #      if(nrow(df) != 0){
  #        vp_layer_genes <- vp_layer_for_uploaded_genes_cbf_gg(p, df, label)
  #        p <- vp_layer_genes
  #      } else{
  #        p
  #      }
  #    }
  #    p
  #  }
  #}
  
 # a_multi_vp_layer_plus_gg <- function(){
#    
#    stop('a_multi_vp_layer_plus_gg is deprecated!')
#    
#    validate(
#      need(!is.null(a_search_gene()), "")
#    )
#    p <- a_multi_vp_layer_gg()
#    goi <- a_search_gene()
#    orig_data <- a_pulldown()
#    searchgene <- orig_data[grepl(goi,orig_data$gene),]
#    p1 <- search_volcano_gg(p, searchgene)
#    p1
#  }
  
  #output$download_multi_vp_gg <- downloadHandler(
  #  filename = function() { paste("multi_vp_gg", '.png', sep='') },
  #  content = function(file) {
  #    if(is.null(a_search_gene())){
  #      ggsave(file, plot = a_multi_vp_layer_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
  #    } else {
  #      ggsave(file, plot = a_multi_vp_layer_plus_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
  #    }
  #    
  #  }
  #)
  
  a_multi_vp_plus <- reactive({
    validate(
      need(!is.null(a_search_gene()), "")
    )
    p <- a_multi_vp_layer()
    goi <- a_search_gene()
    orig_data <- a_pulldown()
    searchgene <- orig_data[grepl(goi,orig_data$gene),]
    p1 <- search_volcano(p, searchgene)
    p1
  })
  
  a_multi_vp_count <- reactive({
    
    stop('a_multi_vp_count is deprecated!')
    multi_vp <- a_multi_vp()
    vp_data <- a_pulldown()
    current_subset <- nrow(subset(vp_data, (FDR <= input$a_fdr_thresh) & (pvalue <= input$a_pval_thresh) & (logFC >= input$a_logFC_thresh[1]) & (logFC <= input$a_logFC_thresh[2])))
    orig_count <- data.frame(paste0(current_subset, " (total = ", nrow(vp_data), ")"))
    row.names(orig_count) <- c("pull down")
    colnames(orig_count) <- c(paste0("FDR<", input$a_fdr_thresh, ", pvalue<", input$a_pval_thresh, ", ", input$a_logFC_thresh[1], "<logFC<", input$a_logFC_thresh[2]))
    
    # InWeb, SNP to gene, and genes upload
    if(!is.null(a_bait_gene_layer())){
      if(nrow(multi_vp$d_in)>0){
        pop <- subset(vp_data, vp_data$gene %in% inweb_combined$V1)
        sample <- sample() 
        inweb_count <- nrow(subset(multi_vp$d_in, (FDR <= input$a_fdr_thresh) & (pvalue <= input$a_pval_thresh) & (logFC >= input$a_logFC_thresh[1]) & (logFC <= input$a_logFC_thresh[2])))
        success_pop <- subset(pop, (FDR <= input$a_fdr_thresh) & (pvalue <= input$a_pval_thresh) & (logFC >= input$a_logFC_thresh[1]) & (logFC <= input$a_logFC_thresh[2]))
        success_samp <- subset(sample, sample$gene %in% success_pop$gene)
        rownames(success_samp) <- NULL
        samp_l <- length(unique(sample$gene))
        pop_l <- length(unique(pop$gene))
        success_pop_l <- length(unique(success_pop$gene))
        success_samp_l <- length(unique(success_samp$gene))
        pvalue <- phyper((success_samp_l-1), samp_l, (pop_l-samp_l), success_pop_l, lower.tail = F)
        pvalue <- signif(pvalue, 4)
        i_count <- data.frame(paste0(inweb_count, " (total = ", nrow(multi_vp$d_in), ", pvalue = ", pvalue, ")")) #, nrow(multi_vp$d_in)
        row.names(i_count) <- c("InWeb")
        colnames(i_count) <- c(paste0("FDR<", input$a_fdr_thresh, ", pvalue<", input$a_pval_thresh, ", ", input$a_logFC_thresh[1], "<logFC<", input$a_logFC_thresh[2]))
        orig_count <- do.call("rbind", list(orig_count, i_count))
      } else if(nrow(multi_vp$d_in)==0){
        orig_count
      }
    }
    
    if(!is.null(SNP_to_gene())){
      if(nrow(multi_vp$d_snp)>0){
        snp_count <- nrow(subset(multi_vp$d_snp, (FDR <= input$a_fdr_thresh) & (pvalue <= input$a_pval_thresh) & (logFC >= input$a_logFC_thresh[1]) & (logFC <= input$a_logFC_thresh[2])))
        s_count <- data.frame(snp_count) 
        row.names(s_count) <- c("SNP")
        colnames(s_count) <- c(paste0("FDR<", input$a_fdr_thresh, ", pvalue<", input$a_pval_thresh, ", ", input$a_logFC_thresh[1], "<logFC<", input$a_logFC_thresh[2]))
        orig_count <- do.call("rbind", list(orig_count, s_count))
      } else {
        orig_count
      }
    }
    
    if(!is.null(a_upload_genes())){
      if(length(multi_vp$d_g2s)>0){
        df <- ldply(multi_vp$d_g2s, data.frame)
        df1 <- split(df, df$.id)
        pop <- subset(vp_data, vp_data$gene %in% toupper(human_genome$HGNCsymbol)) 
        sample <- lapply(a_genes_uploaded(), function(x) subset(pop, gene %in% x) )  
        success_pop <- subset(pop, (FDR <= input$a_fdr_thresh) & (pvalue <= input$a_pval_thresh) & (logFC >= input$a_logFC_thresh[1]) & (logFC <= input$a_logFC_thresh[2]))
        success_samp <- lapply(sample, function(x) subset(x, gene %in% success_pop$gene))
        rownames(success_samp) <- NULL
        samp_l <- lapply(sample, function(x) length(unique(x$gene)))
        pop_l <- length(unique(pop$gene))
        success_pop_l <- length(unique(success_pop$gene))
        success_samp_l <- lapply(success_samp, function(x) length(unique(x$gene)))
        p_l <- rep(list(pop_l), length(samp_l))
        sp_l <- rep(list(success_pop_l), length(samp_l))
        pvalue <- mapply(function(p, sp, s, ss) {signif(phyper((ss-1), s, (p-s), sp, lower.tail = F), 4)}, p=p_l, sp=sp_l, s=samp_l, ss=success_samp_l)
        
        gene_count <- lapply(df1, function(x) subset(x, (FDR <= input$a_fdr_thresh) & (pvalue <= input$a_pval_thresh) & (logFC >= input$a_logFC_thresh[1]) & (logFC <= input$a_logFC_thresh[2])))  
        gene_count1 <- lapply(gene_count, function(x) nrow(x))
        df1_count <- lapply(df1, function(x) nrow(x))
        gene_count_fin <- ldply(gene_count1, data.frame)
        df1_total <- ldply(df1_count, data.frame)
        pval_total <- ldply(pvalue, data.frame)
        g_count <- data.frame(paste0(gene_count_fin$X..i.., " (total = ", df1_total$X..i.., ", pvalue = ", pval_total$X..i.., ")")) #, df1_total$X..i..
        row.names(g_count) <- df1_total$.id
        colnames(g_count) <- c(paste0("FDR<", input$a_fdr_thresh, ", pvalue<", input$a_pval_thresh, ", ", input$a_logFC_thresh[1], "<logFC<", input$a_logFC_thresh[2]))
        orig_count <- do.call("rbind", list(orig_count, g_count))
      } else {
        orig_count
      }
    }
    orig_count
  })
  
  a_multi_vp_count_text <- reactive({
    if(!is.null(a_multi_vp_count())){
      enriched <- c(paste0("FDR&le;", input$a_fdr_thresh, ", pvalue&le;", input$a_pval_thresh, ", ", input$a_logFC_thresh[1], "&le;logFC&le;", input$a_logFC_thresh[2]))
      enriched
    }
  })
  
  # a_multi_sp_layer <- reactive({
  #   multi_sp <- a_multi_vp()
  #   d <- a_pulldown()
  #   cc <- a_sp_cor()
  #   if(input$colorscheme == "fdr"){
  #     data <- separate_to_groups_for_color_integrated(d, input$a_fdr_thresh)
  #     p <- plot_ly(colors = "RdPu", showlegend = T, width = 650, height = 550) 
  #     p <- add_lines(p, data = d, x = ~c(ceiling(min(rep1, rep2)), floor(max(rep1, rep2))), y = ~c(ceiling(min(rep1, rep2)), floor(max(rep1, rep2))),
  #                    line = list(dash = "dash", width = 1, color = "#252525"), showlegend = FALSE)
  #     for(i in nrow(data)){
  #       p <- add_markers(p, data = data, x = ~rep1, y = ~rep2, 
  #                        marker = list(size = 6, cmin = 0, cmax = 1, color = ~col), #line = list(width=0.1, color = "black"),
  #                        opacity = 0.8, 
  #                        text = ~paste(gene), hoverinfo = "text", name = "pull down")
  #     }
  #   } else if(input$colorscheme == "exac"){
  #     d$s <- exac$em_p_hi[match(d$gene, exac$GENE_NAME)]
  #     d$s[is.na(d$s)] <- 2
  #     below_thresh <- subset(d, s < 0.9)
  #     above_thresh <- subset(d, s >= 0.9)
  #     no_exist <- subset(d, s == 2)
  #     p <- plot_ly(colors = "RdPu", showlegend = T, width = 650, height = 550) 
  #     p <- add_lines(p, data = d, x = ~c((min(rep1, rep2)), (max(rep1, rep2))), y = ~c((min(rep1, rep2)), (max(rep1, rep2))),
  #                    line = list(dash = "dash", width = 1, color = "#252525"), showlegend = FALSE)
  #     p <- add_markers(p, data = below_thresh, x = ~rep1, y = ~rep2, 
  #                      marker = list(size = 8, line = list(width=0.1, color = 'black'), cmin = 0, cmax = 1, color = "#66c2a5"),
  #                      opacity = 0.7, 
  #                      text = ~paste(gene), hoverinfo = "text", name = paste0("pLI<0.9 (", nrow(below_thresh), ")"))
  #     p <- add_markers(p, data = above_thresh, x = ~rep1, y = ~rep2, 
  #                      marker = list(size = 8, line = list(width=0.1, color = "black"), cmin = 0, cmax = 1, color = "#fc8d62"),
  #                      opacity = 0.7, 
  #                      text = ~paste(gene), hoverinfo = "text", name = paste0("pLI>=0.9 (", nrow(above_thresh), ")"))
  #     p <- add_markers(p, data = no_exist, x = ~rep1, y = ~rep2, 
  #                      marker = list(size = 8, line = list(width=0.1, color = "black"), cmin = 0, cmax = 1, color = "#8da0cb"),
  #                      opacity = 0.7, 
  #                      text = ~paste(gene), hoverinfo = "text", name = paste0("not in ExAC (", nrow(no_exist), ")"))
  #     p
  #   } else if(input$colorscheme == "cbf"){
  #     data <- separate_to_groups_for_cbf_integrated(d, input$a_fdr_thresh)
  #     p <- plot_ly(colors = "Greys", showlegend = T, width = 650, height = 550) 
  #     p <- add_lines(p, data = d, x = ~c(ceiling(min(rep1, rep2)), floor(max(rep1, rep2))), y = ~c(ceiling(min(rep1, rep2)), floor(max(rep1, rep2))),
  #                    line = list(dash = "dash", width = 1, color = "#252525"), showlegend = FALSE)
  #     for(i in nrow(data)){
  #       p <- add_markers(p, data = data, x = ~rep1, y = ~rep2, 
  #                        marker = list(size = 6, cmin = 0, cmax = 1, color = ~col), #line = list(width=0.1, color = "black"), 
  #                        opacity = 0.6, 
  #                        text = ~paste(gene), hoverinfo = "text", name = "pull down")
  #     }
  #   }
  #   p <- p %>%
  #     layout(xaxis = list(title = "rep1", range=~c((min(rep1, rep2))-1, (max(rep1, rep2))+1)), 
  #            yaxis = list(title = "rep2", range=~c((min(rep1, rep2))-1, (max(rep1, rep2))+1)), 
  #            title = cc, titlefont = list(size=15))
  # 
  #   if(input$colorscheme == "fdr" | input$colorscheme == "exac"){
  #     # InWeb, SNP to gene, and genes upload
  #     if(!is.null(a_bait_gene_layer())){
  #       if(input$a_marker_text == "yes_label"){
  #         sp_layer_inweb <- sp_layer_for_inweb(p, multi_sp$d_in)
  #       } else if(input$a_marker_text == "no_label"){
  #         sp_layer_inweb <- sp_layer_for_inweb_no_text(p, multi_sp$d_in)
  #       }
  #       p <- sp_layer_inweb
  #     }
  #     if(!is.null(a_snp())){
  #       if(nrow(multi_sp$d_snp) != 0){
  #         snp_sgl <- subset(multi_sp$d_snp, Freq == 1)
  #         snp_mgl <- subset(multi_sp$d_snp, Freq != 1)
  #         if(nrow(snp_sgl) != 0){
  #           if(input$a_marker_text == "yes_label"){
  #             sp_layer_snp2gene_sgl <- sp_layer_for_snp_to_gene_sgl(p, snp_sgl)
  #           } else if(input$a_marker_text == "no_label"){
  #             sp_layer_snp2gene_sgl <- sp_layer_for_snp_to_gene_sgl_no_text(p, snp_sgl)
  #           }
  #           p <- sp_layer_snp2gene_sgl
  #         }
  #         if(nrow(snp_mgl) != 0){
  #           if(input$a_marker_text == "yes_label"){
  #             sp_layer_snp2gene_mgl <- sp_layer_for_snp_to_gene_mgl(p, snp_mgl)
  #           } else if(input$a_marker_text == "no_label"){
  #             sp_layer_snp2gene_mgl <- sp_layer_for_snp_to_gene_mgl_no_text(p, snp_mgl)
  #           }
  #           p <- sp_layer_snp2gene_mgl
  #         }
  #       } else{
  #         sp_layer_no_snp2gene <- sp_layer_for_snp_to_gene_none(p, d)
  #         p <- sp_layer_no_snp2gene
  #       }
  #     }
  #     if(!is.null(a_upload_genes())){
  #       df <- ldply(multi_sp$d_g2s, data.frame)
  #       if(nrow(df) != 0){
  #         if(input$a_marker_text == "yes_label"){
  #           sp_layer_genes <- sp_layer_for_uploaded_genes(p, df)
  #         } else if(input$a_marker_text == "no_label"){
  #           sp_layer_genes <- sp_layer_for_uploaded_genes_no_text(p, df)
  #         }
  #         p <- sp_layer_genes
  #       } else{
  #         sp_layer_no_genes <- sp_layer_for_uploaded_genes_none(p, d)
  #         p <- sp_layer_no_genes
  #       }
  #     }
  #     p
  #   } else if(input$colorscheme == "cbf"){
  #     # InWeb, SNP to gene, and genes upload
  #     if(!is.null(a_bait_gene_layer())){
  #       if(input$a_marker_text == "yes_label"){
  #         sp_layer_inweb <- sp_layer_for_inweb_cbf(p, multi_sp$d_in)
  #       } else if(input$a_marker_text == "no_label"){
  #         sp_layer_inweb <- sp_layer_for_inweb_cbf_no_text(p, multi_sp$d_in)
  #       }
  #       p <- sp_layer_inweb
  #     }
  #     if(!is.null(a_snp())){
  #       if(nrow(multi_sp$d_snp) != 0){
  #         snp_sgl <- subset(multi_sp$d_snp, Freq == 1)
  #         snp_mgl <- subset(multi_sp$d_snp, Freq != 1)
  #         if(nrow(snp_sgl) != 0){
  #           if(input$a_marker_text == "yes_label"){
  #             sp_layer_snp2gene_sgl <- sp_layer_for_snp_to_gene_sgl_cbf(p, snp_sgl)
  #           } else if(input$a_marker_text == "no_label"){
  #             sp_layer_snp2gene_sgl <- sp_layer_for_snp_to_gene_sgl_cbf_no_text(p, snp_sgl)
  #           }
  #           p <- sp_layer_snp2gene_sgl
  #         }
  #         if(nrow(snp_mgl) != 0){
  #           if(input$a_marker_text == "yes_label"){
  #             sp_layer_snp2gene_mgl <- sp_layer_for_snp_to_gene_mgl_cbf(p, snp_mgl)
  #           } else if(input$a_marker_text == "no_label"){
  #             sp_layer_snp2gene_mgl <- sp_layer_for_snp_to_gene_mgl_cbf_no_text(p, snp_mgl)
  #           }
  #           p <- sp_layer_snp2gene_mgl
  #         }
  #       } else{
  #         sp_layer_no_snp2gene <- sp_layer_for_snp_to_gene_none_cbf(p, d)
  #         p <- sp_layer_no_snp2gene
  #       }
  #     }
  #     if(!is.null(a_upload_genes())){
  #       df <- ldply(multi_sp$d_g2s, data.frame)
  #       if(nrow(df) != 0){
  #         if(input$a_marker_text == "yes_label"){
  #           sp_layer_genes <- sp_layer_for_uploaded_genes_cbf(p, df)
  #         } else if(input$a_marker_text == "no_label"){
  #           sp_layer_genes <- sp_layer_for_uploaded_genes_cbf_no_text(p, df)
  #         }
  #         p <- sp_layer_genes
  #       } else{
  #         sp_layer_no_genes <- sp_layer_for_uploaded_genes_none_cbf(p, d)
  #         p <- sp_layer_no_genes
  #       }
  #     }
  #     p
  #   }
  #   p
  # })
  # 
  # a_multi_sp_plus <- reactive({
  #   validate(
  #     need(!is.null(a_search_gene()), "")
  #   )
  #   p <- a_multi_sp_layer()
  #   goi <- a_search_gene()
  #   orig_data <- a_pulldown()
  #   searchgene <- orig_data[grepl(goi,orig_data$gene),]
  #   p1 <- search_scatter(p, searchgene)
  #   p1
  # })
  
  output$a_VD_sig_text <- renderUI({
    validate(
      need(input$a_file_pulldown_r != '', "")
    )
    HTML("<b>User-defined significance thresholds:</b>")
  })
  
  a_venndiagram <- reactive({
    bait <- a_bait_gene_vennd()
    success_pop <- success_population_bait()
    s_p <- unique(sort(success_pop$gene))
    pop <- population_bait()
    p <- unique(sort(pop$gene))
    samp <- sample_bait()
    s <- unique(sort(samp$gene))
    success_samp <- success_sample_bait()
    s_s <- unique(sort(success_samp$gene))
    success_pop_l <- length(s_p)
    pop_l <- length(p)
    samp_l <- length(s)
    success_samp_l <- length(s_s)
    pvalue <- phyper((success_samp_l-1), samp_l, (pop_l-samp_l), success_pop_l, lower.tail = F)
    mycolours3 = c("cornflowerblue", "yellow1")
    
    x <- list()
    x[["A"]] <- success_pop$gene
    x[["B"]] <- samp$gene
    
    v0 <- venn.diagram(x,
                       col = mycolours3, margin=0.05, filename = NULL, resolution = 900, height = 400, force.unique = T,
                       main = paste("p value = ", format(pvalue, digits = 4)),
                       sub = " ", sub.pos = c(0, 0), scaled = F,
                       cat.cex = 1.1, cex = 2, cat.pos = c(180,180), cat.dist = c(0.05,0.05),
                       fontfamily = 'sans', cat.fontfamily = 'sans', main.fontfamily = 'sans')
    
    v0
  })
  
  a_venndiagram_text <- reactive({
    bait <- a_bait_gene_vennd()
    pop <- population_bait()
    success_pop <- success_population_bait()
    samp <- sample_bait()
    subset_limit <- paste0("A = pull down subset of ", input$a_FDR_range[1], "&le;FDR&le;", input$a_FDR_range[2], 
                           ", ", input$a_logFC_range[1], "&le;logFC&le;", input$a_logFC_range[2], 
                           " and ", input$a_pvalue_range[1], "&le;pvalue&le;", input$a_pvalue_range[2],
                           " &#40;", length(unique(sort(success_pop$gene))), "&#41;")
    inweb_name <- paste0("B = InWeb_IM interactors of ", bait, " &#40;", length(unique(sort(samp$gene))), "&#41;")
    total <- paste0("pull down &cap; InWeb_IM database &#40;", length(unique(sort(pop$gene))), "&#41;")
    list(a=subset_limit, b=inweb_name, c=total)
  })
  
  output$a_vd_text <- renderUI({
    validate(
      need(input$a_file_pulldown_r != '', ""),
      need(input$a_bait_vennd != '', "")
    )
    output <- a_venndiagram_text()
    HTML(paste(output$c, output$a, output$b, sep = "<br/>"))
  })
  
  a_venndiagram_GOI <- reactive({
    success_pop <- success_population_GOI()
    s_p <- unique(sort(success_pop$gene))
    pop <- population_GOI()
    p <- unique(sort(pop$gene))
    samp <- sample_GOI()
    samp_total <- as.data.frame(data.table::rbindlist(samp))
    samp <- c(samp, list(total = samp_total))
    list_num <- input$a_goi_num_inputs
    s <- unique(sort(samp[[list_num]]$gene))
    success_samp <- success_sample_GOI()
    s_s <- unique(sort(success_samp$gene))
    success_pop_l <- length(s_p)
    pop_l <- length(p)
    samp_l <- length(s)
    success_samp_l <- length(s_s)
    pvalue <- phyper((success_samp_l-1), samp_l, (pop_l-samp_l), success_pop_l, lower.tail = F)
    mycolours3 = c("cornflowerblue", "#fdb462")
    
    x <- list()
    x[["A"]] <- success_pop$gene
    x[["B"]] <- samp[[list_num]]$gene
    
    v0 <- venn.diagram(x, 
                       col = mycolours3, margin=0.05, filename = NULL, resolution = 900, height = 400, force.unique = T,
                       main = paste("p value = ", format(pvalue, digits = 4)), 
                       sub = " ", sub.pos = c(0, 0), scaled = F,
                       cat.cex = 1.1, cex = 2, cat.pos = c(180,180), cat.dist = c(0.05,0.05))
    for (i in 1:length(v0)){ v0[[i]]$gp$fontfamily <- 'Arial' }
    
    v0
  })
  
  a_venndiagram_GOI_text <- reactive({ 
    pop <- population_GOI()
    success_pop <- success_population_GOI()
    samp <- sample_GOI()
    samp_total <- as.data.frame(data.table::rbindlist(samp))
    samp <- c(samp, list(total = samp_total))
    list_num <- input$a_goi_num_inputs
    subset_limit <- paste0("A = pull down subset of ", input$a_FDR_range[1], "&le;FDR&le;", input$a_FDR_range[2], 
                           ", ", input$a_logFC_range[1], "&le;logFC&le;", input$a_logFC_range[2], 
                           " and ", input$a_pvalue_range[1], "&le;pvalue&le;", input$a_pvalue_range[2],
                           " &#40;", length(unique(sort(success_pop$gene))), "&#41;")
    inweb_name <- paste0("B = Genes of interest in HGNC database", " &#40;", length(unique(sort(samp[[list_num]]$gene))), "&#41;") #   nrow(samp[[list_num]])
    total <- paste0("pull down &cap; HGNC database &#40;", length(unique(sort(pop$gene))), "&#41;")
    list(a=subset_limit, b=inweb_name, c=total)
  })
  
  output$a_vd_GOI_text <- renderUI({
    validate(
      need(input$a_file_pulldown_r != '', ""),
      need(!is.null(input$a_file_genes_vennd), "")
    )
    output <- a_venndiagram_GOI_text()
    HTML(paste(output$c, output$a, output$b, sep = "<br/>"))
  })
  
  a_venndiagram_SNP <- reactive({
    success_pop <- success_population_GOI()
    s_p <- unique(sort(success_pop$gene))
    pop <- population_GOI()
    p <- unique(sort(pop$gene))
    samp <- sample_SNP()
    s <- unique(sort(samp$gene))
    success_samp <- success_sample_SNP()
    s_s <- unique(sort(success_samp$gene))
    success_pop_l <- length(s_p)
    pop_l <- length(p)
    samp_l <- length(s)
    success_samp_l <- length(s_s)
    mycolours3 = c("cornflowerblue", "salmon")
    
    x <- list()
    x[["A"]] <- success_pop$gene
    x[["B"]] <- samp$gene
    
    v0 <- venn.diagram(x, 
                       col = mycolours3, margin=0.05, filename = NULL, resolution = 900, height = 400, force.unique = T,
                       main = "all SNPs", 
                       sub = " ", sub.pos = c(0, 0), scaled = F,
                       cat.cex = 1.1, cex = 2, cat.pos = c(180,180), cat.dist = c(0.05,0.05))
    for (i in 1:length(v0)){ v0[[i]]$gp$fontfamily <- 'Arial' }
    v0
  })
  
  a_venndiagram_SNP_text <- reactive({
    pop <- population_GOI()
    success_pop <- success_population_GOI()
    samp <- sample_SNP()
    subset_limit <- paste0("A = pull down subset of ", input$a_FDR_range[1], "&le;FDR&le;", input$a_FDR_range[2], 
                           ", ", input$a_logFC_range[1], "&le;logFC&le;", input$a_logFC_range[2], 
                           " and ", input$a_pvalue_range[1], "&le;pvalue&le;", input$a_pvalue_range[2],
                           " &#40;", length(unique(sort(success_pop$gene))), "&#41;")
    inweb_name <- paste0("B = SNPs in 1K Genome database", " &#40;", length(unique(sort(samp$gene))), "&#41;")
    total <- paste0("pull down &cap; HGNC database &#40;", length(unique(sort(pop$gene))), "&#41;")
    list(a=subset_limit, b=inweb_name, c=total)
  })
  
  output$a_vd_SNP_text <- renderUI({
    validate(
      need(input$a_file_pulldown_r != '', ""),
      need(!is.null(input$a_file_SNP_vennd), "")
    )
    output <- a_venndiagram_SNP_text()
    HTML(paste(output$a, output$b, sep = "<br/>"))
  })
  
  a_venndiagram_SNP_SGL <- reactive({
    success_pop <- success_population_GOI()
    s_p <- unique(sort(success_pop$gene))
    pop <- population_GOI()
    p <- unique(sort(pop$gene))
    samp <- sample_SNP_SGL()
    s <- unique(sort(samp$gene))
    success_samp <- success_sample_SNP_SGL()
    s_s <- unique(sort(success_samp$gene))
    success_pop_l <- length(s_p)
    pop_l <- length(p)
    samp_l <- length(s)
    success_samp_l <- length(s_s)
    mycolours3 = c("cornflowerblue", "#80cdc1")
    
    x <- list()
    x[["A"]] <- success_pop$gene
    x[["B"]] <- samp$gene
    
    v0 <- venn.diagram(x, 
                       col = mycolours3, margin=0.05, filename = NULL, resolution = 900, height = 400, force.unique = T,
                       main = "Single-gene-loci SNPs", 
                       sub = " ", sub.pos = c(0, 0), scaled = F,
                       cat.cex = 1.1, cex = 2, cat.pos = c(180,180), cat.dist = c(0.05,0.05))
    for (i in 1:length(v0)){ v0[[i]]$gp$fontfamily <- 'Arial' }
    v0
  })
  
  a_venndiagram_SNP_SGL_text <- reactive({
    pop <- population_GOI()
    success_pop <- success_population_GOI()
    samp <- sample_SNP_SGL()
    subset_limit <- paste0("A = pull down subset of ", input$a_FDR_range[1], "&le;FDR&le;", input$a_FDR_range[2], 
                           ", ", input$a_logFC_range[1], "&le;logFC&le;", input$a_logFC_range[2], 
                           " and ", input$a_pvalue_range[1], "&le;pvalue&le;", input$a_pvalue_range[2],
                           " &#40;", length(unique(sort(success_pop$gene))), "&#41;")
    inweb_name <- paste0("B = SGL SNPs in 1K Genome database", " &#40;", length(unique(sort(samp$gene))), "&#41;")
    total <- paste0("pull down &cap; HGNC database &#40;", length(unique(sort(pop$gene))), "&#41;")
    list(a=subset_limit, b=inweb_name, c=total)
  })
  
  output$a_vd_SNP_SGL_text <- renderUI({
    validate(
      need(input$a_file_pulldown_r != '', ""),
      need(!is.null(input$a_file_SNP_vennd), "")
    )
    output <- a_venndiagram_SNP_SGL_text()
    HTML(paste(output$a, output$b, sep = "<br/>"))
  })
  
  a_venndiagram_SNP_MGL <- reactive({ 
    success_pop <- success_population_GOI()
    s_p <- unique(sort(success_pop$gene))
    pop <- population_GOI()
    p <- unique(sort(pop$gene))
    samp <- sample_SNP_MGL()
    s <- unique(sort(samp$gene))
    success_samp <- success_sample_SNP_MGL()
    s_s <- unique(sort(success_samp$gene))
    success_pop_l <- length(s_p)
    pop_l <- length(p)
    samp_l <- length(s)
    success_samp_l <- length(s_s)
    mycolours3 = c("cornflowerblue", "#c2a5cf")
    
    x <- list()
    x[["A"]] <- success_pop$gene
    x[["B"]] <- samp$gene
    
    v0 <- venn.diagram(x, 
                       col = mycolours3, margin=0.05, filename = NULL, resolution = 900, height = 400, force.unique = T,
                       main = "Multi-genes-loci SNPs", 
                       sub = " ", sub.pos = c(0, 0), scaled = F,
                       cat.cex = 1.1, cex = 2, cat.pos = c(180,180), cat.dist = c(0.05,0.05))
    for (i in 1:length(v0)){ v0[[i]]$gp$fontfamily <- 'Arial' }
    v0
  })
  
  a_venndiagram_SNP_MGL_text <- reactive({
    pop <- population_GOI()
    success_pop <- success_population_GOI()
    samp <- sample_SNP_MGL()
    subset_limit <- paste0("A = pull down subset of ", input$a_FDR_range[1], "&le;FDR&le;", input$a_FDR_range[2], 
                           ", ", input$a_logFC_range[1], "&le;logFC&le;", input$a_logFC_range[2], 
                           " and ", input$a_pvalue_range[1], "&le;pvalue&le;", input$a_pvalue_range[2],
                           " &#40;", length(unique(sort(success_pop$gene))), "&#41;")
    inweb_name <- paste0("B = MGL SNPs in 1K Genome database", " &#40;", length(unique(sort(samp$gene))), "&#41;")
    total <- paste0("pull down &cap; HGNC database &#40;", length(unique(sort(pop$gene))), "&#41;")
    list(a=subset_limit, b=inweb_name, c=total)
  })
  
  output$a_vd_SNP_MGL_text <- renderUI({
    validate(
      need(input$a_file_pulldown_r != '', ""),
      need(!is.null(input$a_file_SNP_vennd), "")
    )
    output <- a_venndiagram_SNP_MGL_text()
    HTML(paste(output$a, output$b, sep = "<br/>"))
  })
  
  a_BPF_marker <- reactive({
    inBPFmarker <- input$a_BPF_option
    if (inBPFmarker == "change_m_BPF") {
      marker <- "change"
    } else {
      marker <- "no_change"
    }
  })
  
  a_bpf_data <- eventReactive(input$a_make_bpf, {
    validate(
      need(input$a_file_pulldown_r != '', "Upload file")
    )
    data_selection <- a_pf_data_selection()
    filter_option <- a_pf_viz_selection()
    withProgress(message = 'This may take a while', 
                 detail = 'Hold please', value = 0, {
                   bpf <- a_pulldown()
                   
                   makePlotFamilies_1quadrant <- function(data, bait){
                     if(filter_option == "volcano_viz"){
                       im <- subset(data, 
                                    (pvalue <= input$a_BPF_pvalue_range[2] & pvalue >= input$a_BPF_pvalue_range[1]) & 
                                      (FDR <= input$a_BPF_FDR_range[2] & FDR >= input$a_BPF_FDR_range[1]) &
                                      (logFC <= input$a_BPF_logFC_range[2] & logFC >= input$a_BPF_logFC_range[1]))
                     } else if(filter_option == "scatter_viz"){
                       im <- subset(data, 
                                    (rep1 <= input$a_BPF_rep1_range[2] & rep1 >= input$a_BPF_rep1_range[1]) & 
                                      (rep2 <= input$a_BPF_rep2_range[2] & rep2 >= input$a_BPF_rep2_range[1]))
                     }
                     incProgress(0.6)
                     enM_families <- assignFamily_inc_doubles(im, data_selection)
                     # enM_gna <- addNames(im)
                     incProgress(0.8)
                     ### replace logFC and pvalue in di* by these from dpm
                     # enM_families$logFC <- data$logFC[match(enM_families$gene, data$gene)]
                     # enM_families$pvalue <- data$pvalue[match(enM_families$gene, data$gene)]
                     # enM_gna$logFC <- data$logFC[match(enM_gna$gene, data$gene)]
                     # enM_gna$pvalue <- data$pvalue[match(enM_gna$gene, data$gene)]
                     # 
                     # mybait <- data[grepl(bait,data$gene),]
                     
                     howmany <- length(unique(enM_families$overlay))
                     bpfmsizing <- a_BPF_marker()
                     increase_size <- input$a_BPF_marker_freq
                     
                     incProgress(0.9)
                     bpf_list <- list("list" = bpf, "list" = enM_families, "integer" = howmany, bpfmsizing, increase_size) #"list" = enM_gna, 
                     return(bpf_list)
                   }
                   makePlotFamilies_1quadrant(bpf)
                 })
  }) 
  
  a_bpf_families <- reactive({
    bpf_data <- a_bpf_data()
    enM_families <- bpf_data[[2]]
    families <- cbind(enM_families$gene, enM_families$overlay, enM_families$frequency)
    colnames(families) <- c("gene", "overlay", "frequency")
    families
  })
  
  a_bpf_families1 <- reactive({
    bpf_data <- a_bpf_data()
    enM_families <- bpf_data[[2]]
    enM_families
  })
  
  a_bpf <- reactive({
    bpf_data <- a_bpf_data()
    pfsort <- a_PF_sorting()
    viz_type <- a_pf_viz_selection()
    bpf <- bpf_data[[1]]
    enM_families <- bpf_data[[2]]
    howmany <- bpf_data[[3]]
    bpfmsizing <- bpf_data[[4]]
    increase_size <- bpf_data[[5]]
    
    if(pfsort == "sort_f"){
      enM_families$overlay <- paste0(enM_families$frequency, "-", enM_families$overlay)
      enM_families$frequency <- as.numeric(enM_families$frequency)
      enM_families <- enM_families[with(enM_families, order(-frequency)), ]
    } else if(pfsort == "sort_a"){
      enM_families$overlay <- paste0(enM_families$overlay, " (", enM_families$frequency, ")")
      enM_families <- enM_families[with(enM_families, order(overlay, decreasing = F)),]
    }
    
    if (bpfmsizing == "change"){
      if(nrow(enM_families)==0){
        validate(
          need(nrow(enM_families)>0, "No match in protein families assignment")
        )
      } else if(nrow(enM_families)>0){
        if(viz_type == "volcano_viz"){
          p <- plot_ly(colors = rainbow(howmany), width = 1200, height = 800) %>%
            #background
            add_markers(data = bpf, x = ~logFC, y = ~-log10(pvalue), opacity = 0.6,
                        marker = list(color = 'rgba(176,196,222,08)'),
                        text = ~paste(gene), hoverinfo = "text", showlegend = FALSE)
          # if(nrow(enM_gna)>0){
          #   p <- add_markers(p, data = enM_gna, x = ~logFC, y = ~-log10(pvalue),
          #                    marker = list(size = 6, symbol = 2, color = c('white'), opacity = 0.8, line = list(width=0.9, color = "black")),
          #                    text = ~paste(gene, name, sep = "  "), hoverinfo="text",
          #                    name="Unassigned genes")
          # }
          if(nrow(enM_families)>0){
            p <- add_markers(p, data = enM_families, x = ~logFC, y = ~-log10(pvalue),
                             marker = list(symbol = c('square'), opacity = 0.8, line = list(width=0.6, color = "black"), size = ~increase_size*frequency),
                             color = ~factor(overlay),
                             text = ~paste(gene, overlay, frequency, sep = "  "), hoverinfo = "text")
          }
          p
        } else if(viz_type == "scatter_viz"){
          p <- plot_ly(colors = rainbow(howmany), width = 1200, height = 800) %>%
            #background
            add_markers(data = bpf, x = ~rep1, y = ~rep2, opacity = 0.6,
                        marker = list(color = 'rgba(176,196,222,08)'),
                        text = ~paste(gene), hoverinfo = "text", showlegend = FALSE)
          # if(nrow(enM_gna)>0){
          #   p <- add_markers(p, data = enM_gna, x = ~logFC, y = ~-log10(pvalue),
          #                    marker = list(size = 6, symbol = 2, color = c('white'), opacity = 0.8, line = list(width=0.9, color = "black")),
          #                    text = ~paste(gene, name, sep = "  "), hoverinfo="text",
          #                    name="Unassigned genes")
          # }
          if(nrow(enM_families)>0){
            p <- add_markers(p, data = enM_families, x = ~rep1, y = ~rep2,
                             marker = list(symbol = c('square'), opacity = 0.8, line = list(width=0.6, color = "black"), size = ~increase_size*frequency),
                             color = ~factor(overlay),
                             text = ~paste(gene, overlay, frequency, sep = "  "), hoverinfo = "text")
          }
          p
        }
        
        
      }
    } else{
      if(nrow(enM_families)==0){
        validate(
          need(nrow(enM_families)>0, "No match in protein families assignment")
        )
      } else if(nrow(enM_families)>0){
        if(viz_type == "volcano_viz"){
          p <- plot_ly(colors = rainbow(howmany), width = 1200, height = 800) %>%
            #background
            add_markers(data = bpf, x = ~logFC, y = ~-log10(pvalue), opacity = 0.6,
                        marker = list(color = 'rgba(176,196,222,08)'),
                        text = ~paste(gene), hoverinfo = "text", showlegend = FALSE)
          # if(nrow(enM_gna)>0){
          #   p <- add_markers(p, data = enM_gna, x = ~logFC, y = ~-log10(pvalue), 
          #                    marker = list(size = 6, symbol = 2, color = c('white'), opacity = 0.8, line = list(width=0.9, color = "black")),
          #                    text = ~paste(gene, name, sep = "  "), hoverinfo="text", 
          #                    name="Unassigned genes")
          # }
          if(nrow(enM_families)>0){
            p <- add_markers(p, data = enM_families, x = ~logFC, y = ~-log10(pvalue),
                             marker = list(symbol = c('square'), opacity = 0.8, line = list(width=0.6, color = "black"), size = 12),
                             color = ~factor(overlay), 
                             text = ~paste(gene, overlay, frequency, sep = "  "), hoverinfo = "text")
          }
          p
        } else if(viz_type == "scatter_viz"){
          p <- plot_ly(colors = rainbow(howmany), width = 1200, height = 800) %>%
            #background
            add_markers(data = bpf, x = ~rep1, y = ~rep2, opacity = 0.6,
                        marker = list(color = 'rgba(176,196,222,08)'),
                        text = ~paste(gene), hoverinfo = "text", showlegend = FALSE)
          # if(nrow(enM_gna)>0){
          #   p <- add_markers(p, data = enM_gna, x = ~logFC, y = ~-log10(pvalue), 
          #                    marker = list(size = 6, symbol = 2, color = c('white'), opacity = 0.8, line = list(width=0.9, color = "black")),
          #                    text = ~paste(gene, name, sep = "  "), hoverinfo="text", 
          #                    name="Unassigned genes")
          # }
          if(nrow(enM_families)>0){
            p <- add_markers(p, data = enM_families, x = ~rep1, y = ~rep2,
                             marker = list(symbol = c('square'), opacity = 0.8, line = list(width=0.6, color = "black"), size = 12),
                             color = ~factor(overlay), 
                             text = ~paste(gene, overlay, frequency, sep = "  "), hoverinfo = "text")
          }
          p
        }
      }
    }
    p
  })
  
  a_bpf_plus <- reactive({
    validate(
      need(!is.null(a_search_gene()), "")
    )
    p <- a_bpf()
    goi <- a_search_gene()
    orig_data <- a_pulldown()
    viz_type <- a_pf_viz_selection()
    searchgene <- orig_data[grepl(goi,orig_data$gene),]
    if(viz_type == "volcano_viz"){
      p1 <- search_volcano(p, searchgene)
    } else if(viz_type == "scatter_viz"){
      p1 <- search_scatter(p, searchgene)
    }
    p1
  })
  
  # a_BPF_marker_sp <- reactive({
  #   inBPFmarker <- input$a_BPF_option_sp
  #   if (inBPFmarker == "change_m_BPF") {
  #     marker <- "change"
  #   } else {
  #     marker <- "no_change"
  #   }
  # })
  
  # a_bpf_plus_sp <- reactive({
  #   validate(
  #     need(!is.null(a_search_gene()), "")
  #   )
  #   p <- a_bpf_sp()
  #   goi <- a_search_gene()
  #   orig_data <- a_pulldown()
  #   searchgene <- orig_data[grepl(goi,orig_data$gene),]
  #   p1 <- search_scatter(p, searchgene)
  #   p1
  # })
  
  a_bpf_sp_preview <- reactive({
    d <- a_pulldown()
    p <- plot_ly(showlegend = FALSE) 
    req(input$a_BPF_rep1_range[1], input$a_BPF_rep1_range[2], input$a_BPF_rep2_range[1], input$a_BPF_rep2_range[2])
    line1 <- data.frame("a" = c(input$a_BPF_rep1_range[1], input$a_BPF_rep1_range[2], input$a_BPF_rep1_range[2]), 
                        "b" = c(input$a_BPF_rep2_range[2], input$a_BPF_rep2_range[2], input$a_BPF_rep2_range[1]))
    line2 <- data.frame("a" = c(input$a_BPF_rep1_range[1], input$a_BPF_rep1_range[1], input$a_BPF_rep1_range[2]), 
                        "b" = c(input$a_BPF_rep2_range[2], input$a_BPF_rep2_range[1], input$a_BPF_rep2_range[1]))
    p <- add_lines(p, data = d, x = ~c((min(rep1, rep2)), (max(rep1, rep2))), y = ~c((min(rep1, rep2)), (max(rep1, rep2))),
                   text = "x=y", hoverinfo = "text",
                   line = list(dash = "dash", width = 1, color = "#252525"), showlegend = FALSE)
    p <- add_markers(p, data = d, x = ~rep1, y = ~rep2, 
                     marker = list(color = 'rgba(176,196,222,08)', size = 7, cmin = 0, cmax = 1, line = list(width=0.2, color = "grey89")), 
                     opacity = 0.9, 
                     text = ~paste0(gene, ", rep1=", rep1, ", rep2=", rep2), hoverinfo = "text", name = "pull down")
    p <- add_lines(p, data = line1, x = ~a, y = ~b, line = list(width = 0.7, color = "#e41a1c"),
                   name = '', showlegend = F)
    p <- add_lines(p, data = line2, x = ~a, y = ~b, line = list(width = 0.7, color = "#e41a1c"),
                   name = '', showlegend = F)
    p <- p %>%
      layout(xaxis = list(title = "rep1", range=~c((min(d$rep1, d$rep2))-1, (max(d$rep1, d$rep2))+1)), 
             yaxis = list(title = "rep2", range=~c((min(d$rep1, d$rep2))-1, (max(d$rep1, d$rep2))+1))) 
    
  })
  
  output$FDR_colorbar <- renderPlot({
    validate(need(input$a_file_pulldown_r != '', ""))
    a_vp_colorbar()
  })
  
  output$FDR_colorbar_integrated <- renderPlot({
    validate(need(input$a_file_pulldown_r != '', ""))
    a_vp_colorbar()
  })
  
  
  
  
  # the actual volcano plot outputted to the user
  output$VolcanoPlot <- renderPlotly({
    validate(
      need(input$a_file_pulldown_r != '', "Upload file")
    )
    if(is.null(a_pf_db())){
      a_vp_layerx()
    } else if(!is.null(a_pf_db())){
      if(is.null(a_search_gene())){
        a_vp_pf_db()
      } else{
        a_vp_pf_db_plus_rep()
      }
    }
  })
  
  output$a_verbatim_count_ui <- renderUI({
    validate(need(input$a_file_pulldown_r != '', " "))
    a_verbatim_count()
  })
  
  # Print counts to the table
  #output$VP_count <- renderTable({
  #  validate(need(input$a_file_pulldown_r != '', " "))
  #  a_vp_count()
  #}, rownames = T, bordered = T, colnames = F)
  
  output$VP_count_text <- renderUI({
    validate(need(input$a_file_pulldown_r != '', " "))
    output <- a_vp_count_text()
    HTML(output)
  })
  
  output$ScatterPlot <- renderPlotly({
    validate(need(input$a_file_pulldown_r != '', "Upload file"))
    a_sp()
  })
  
  
  output$ScatterPlotOld <- renderPlotly({
    validate(
      need(input$a_file_pulldown_r != '', "Upload file")
    )
    input_file <- a_in_pulldown()
    d_col <- colnames(input_file)
    if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col &
       "rep1" %in% d_col & "rep2" %in% d_col){
      if(is.null(a_pf_db())){
        if(is.null(a_search_gene())){
          a_sp()
        } else{
          a_sp_plus()
        }
      } else if(!is.null(a_pf_db())){
        a_sp_pf_db()
      }
    } else if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col){
      validate(
        need("rep1" %in% d_col & "rep2" %in% d_col, "Must have rep1 and rep2 values.")
      )
    } else if("rep1" %in% d_col & "rep2" %in% d_col){
      if(is.null(a_pf_db())){
        if(is.null(a_search_gene())){
          a_sp()
        } else{
          a_sp_plus()
        }
      } else if(!is.null(a_pf_db())){
        a_sp_pf_db()
      }
    }
  })
  
  output$multi_FDR_colorbar <- renderPlot({
    validate(
      need(input$a_file_pulldown_r != '', "")
    )
    a_multi_vp_colorbar()
  })
  
  output$Multi_VolcanoPlot <- renderPlotly({
    #validate(need(input$a_file_pulldown_r != '', "Upload file"))
    req(a_pulldown_significant)
    a_integrated_plot()
  })
  
  output$Multi_VP_count <- renderTable({
    validate(
      need(input$a_file_pulldown_r != '', " ")
    )
    a_multi_vp_count()
  }, rownames = T, bordered = T, colnames = F)
  
  output$Multi_VP_count_text <- renderUI({
    validate(
      need(input$a_file_pulldown_r != '', " ")
    )
    output <- a_multi_vp_count_text()
    HTML(output)
  })
  
  output$Multi_ScatterPlot <- renderPlotly({
    validate(
      need(input$a_file_pulldown_r != '', "Upload file")
    )
    input_file <- a_in_pulldown()
    d_col <- colnames(input_file)
    if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col &
       "rep1" %in% d_col & "rep2" %in% d_col){
      if(is.null(a_search_gene())){
        a_multi_sp_layer()
      } else{
        a_multi_sp_plus()
      }
    } else if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col){
      validate(
        need("rep1" %in% d_col & "rep2" %in% d_col, "Must have rep1 and rep2 values.")
      )
    } else if("rep1" %in% d_col & "rep2" %in% d_col){
      if(is.null(a_search_gene())){
        a_multi_sp_layer()
      } else{
        a_multi_sp_plus()
      }
    }
  })
  
  output$Venn_Diagram_bait <- renderPlot({
    validate(
      need(input$a_file_pulldown_r != '', "Upload file"),
      need(input$a_bait_vennd != '', "Requires bait")
    )
    v0 <- a_venndiagram()
    grid.newpage()
    grid.draw(v0)
  })
  
  output$Venn_Diagram_GOI <- renderPlot({
    validate(
      need(input$a_file_pulldown_r != '', "Upload file"),
      need(!is.null(a_upload_genes_vennd()), "Requires GOI")
    )
    v0 <- a_venndiagram_GOI()
    grid.newpage()
    grid.draw(v0)
    
  })
  
  output$Venn_Diagram_SNP <- renderPlot({
    validate(
      need(input$a_file_pulldown_r != '', "Upload file"),
      need(!is.null(a_snp_vennd()), "Requires SNPs")
    )
    v0 <- a_venndiagram_SNP()
    grid.newpage()
    grid.draw(v0)
  })
  
  output$Venn_Diagram_SNP_SGL <- renderPlot({
    validate(
      need(input$a_file_pulldown_r != '', "Upload file"),
      need(!is.null(a_snp_vennd()), "")
    )
    v0 <- a_venndiagram_SNP_SGL()
    grid.newpage()
    grid.draw(v0)
  })
  
  output$Venn_Diagram_SNP_MGL <- renderPlot({
    validate(
      need(input$a_file_pulldown_r != '', "Upload file"),
      need(!is.null(a_snp_vennd()), "")
    )
    v0 <- a_venndiagram_SNP_MGL()
    grid.newpage()
    grid.draw(v0)
  })
  
  output$Basic_Protein_Family <- renderPlotly({
    validate(
      need(input$a_file_pulldown_r != '', "Upload file")
    )
    if(is.null(a_search_gene())){
      a_bpf()
    } else{
      a_bpf_plus()
    }
  })
  
  output$Basic_Protein_Family_sp_prev <- renderPlotly({
    validate(
      need(input$a_file_pulldown_r != '', "Upload file"),
      need(input$a_pf_plot_option != '', "")
    )
    d <- a_orig_pulldown()
    d_col <- colnames(d)
    validate(
      need("rep1" %in% d_col & "rep2" %in% d_col, "")
    )
    viz_type <- a_pf_viz_selection()
    validate(
      need(viz_type == "scatter_viz", "")
    )
    a_bpf_sp_preview()
  })
  
  # output$Basic_Protein_Family_sp <- renderPlotly({
  #   validate(
  #     need(input$a_file_pulldown_r != '', "Upload file")
  #   )
  #   if(is.null(a_search_gene())){
  #     a_bpf_sp()
  #   } else{
  #     a_bpf_plus_sp()
  #   }
  # })
  
  observe({
    if(!is.null(input$a_file_pulldown_r)){
      df <- a_in_pulldown()
      if("rep1" %in% colnames(df) & "rep2" %in% colnames(df)){
        shinyjs::enable("download_replications_calculated")
      }
    }
  })
  
  observe({
    if(!is.null(input$a_file_pulldown_r)){
      df <- a_in_pulldown()
      if("accession_number" %in% colnames(df)){
        shinyjs::enable("download_mapped_uniprot")
      }
    }
  })
  
  observe({
    if (is.null(input$a_file_pulldown_r)){
      shinyjs::disable("download_basic_plots")
      shinyjs::disable("download_layered_plots")
      shinyjs::disable("download_venn_diagram_plot")
      shinyjs::disable("download_replications_calculated")
      shinyjs::disable("download_bpf_vp_plot")
      shinyjs::disable("download_bpf_sp_plot")
      shinyjs::disable("download_enriched_families_bpf")
      shinyjs::disable("download_venn_diagram_hypergeometric")
      shinyjs::disable("download_snp_to_genes")
      shinyjs::disable("download_goi_overlap")
      shinyjs::disable("download_venn_diagram_GOI_plot")
      shinyjs::disable("download_venn_diagram_SNP_plot")
      shinyjs::disable("download_mapped_uniprot")
    } else {
      shinyjs::enable("download_basic_plots")
    }
  })
  
  observe({
    if(!is.null(input$a_file_pulldown_r)){
      df <- a_in_pulldown()
      if("logFC" %in% colnames(df) & "FDR" %in% colnames(df) & "pvalue" %in% colnames(df)){
        shinyjs::disable("download_replications_calculated")
      }
    }
  })
  
  observe({
    if(!is.null(input$a_file_pulldown_r)){
      df <- a_in_pulldown()
      if("gene" %in% colnames(df)){
        shinyjs::disable("download_mapped_uniprot")
      }
    }
  })
  
  observe({
    if (input$a_make_bpf == 0 || is.null(input$a_make_bpf)){
      shinyjs::disable("download_bpf_vp_plot")
      shinyjs::disable("download_enriched_families_bpf")
    } else {
      shinyjs::enable("download_bpf_vp_plot")
      shinyjs::enable("download_enriched_families_bpf")
    }
  })
  
  observe({
    if (input$a_make_bpf_sp == 0 || is.null(input$a_make_bpf_sp)){
      shinyjs::disable("download_bpf_sp_plot")
    } else {
      shinyjs::enable("download_bpf_sp_plot")
    }
  })
  
  observe({
    if (input$a_make_vennd_bait == 0 || is.null(input$a_make_vennd_bait)){
      shinyjs::disable("download_venn_diagram_plot")
      shinyjs::hide("download_venn_diagram_inweb_genes")
    } else {
      shinyjs::enable("download_venn_diagram_plot")
      shinyjs::show("download_venn_diagram_inweb_genes")
    }
  })
  
  observe({
    if (input$a_make_vennd_snp == 0 || is.null(input$a_make_vennd_snp)){
      shinyjs::disable("download_venn_diagram_SNP_plot")
      shinyjs::hide("download_venn_diagram_SNP_genes")
    } else {
      shinyjs::enable("download_venn_diagram_SNP_plot")
      shinyjs::show("download_venn_diagram_SNP_genes")
    }
  })
  
  observe({
    if (input$a_make_vennd_goi == 0 || is.null(input$a_make_vennd_goi)){
      shinyjs::disable("download_venn_diagram_GOI_plot")
      shinyjs::hide("download_venn_diagram_GOI_genes")
    } else {
      shinyjs::enable("download_venn_diagram_GOI_plot")
      shinyjs::show("download_venn_diagram_GOI_genes")
    }
  })
  
  observe({
    if (is.null(input$a_file_SNP_rep)){
      shinyjs::disable("download_snp_to_genes")
    } else {
      shinyjs::enable("download_snp_to_genes")
    }
  })
  
  observe({
    if (is.null(input$a_file_genes_rep)){
      shinyjs::disable("download_goi_overlap")
    } else {
      shinyjs::enable("download_goi_overlap")
    }
  })
  
  observe({
    if (input$a_make_plot == 0 || is.null(input$a_make_plot)){
      shinyjs::disable("download_layered_plots") 
    } else {
      shinyjs::enable("download_layered_plots") 
    }
  })
  
  observe({
    if(!is.null(a_pf_db())){
      if(!is.null(a_pulldown())){
        d <- a_orig_pulldown() 
        d_col <- colnames(d)
        if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col &
           "rep1" %in% d_col & "rep2" %in% d_col){
          shinyjs::hide("download_pf_cleaned_input")
        } else if("rep1" %in% d_col & "rep2" %in% d_col){
          shinyjs::show("download_pf_cleaned_input")
        } else if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col) {
          shinyjs::hide("download_pf_cleaned_input")
        }
      }
    } else if(is.null(a_pf_db())){
      shinyjs::hide("download_pf_cleaned_input")
    }
  })
  
  output$download_pf_cleaned_input <- downloadHandler(
    filename = function() {
      paste("input-pf-removed", ".txt", sep = "")
    },
    content = function(file) {
      write.table(a_pf_cleanup(), file, sep = "\t", col.names = T, row.names = F, quote = F)
    }
  )
  
  output$download_mapped_uniprot <- downloadHandler(
    filename = function() {
      paste("protein-to-gene-names", ".txt", sep = "")
    },
    content = function(file) {
      write.table(a_converted(), file, sep = "\t", col.names = T, row.names = F, quote = F)
    }
  )
  
  output$download_replications_calculated <- downloadHandler(
    filename = function() {
      paste("input_calculated", ".txt", sep = "")
    },
    content = function(file) {
      write.table(a_pulldown(), file, sep = "\t", col.names = T, row.names = F, quote = F)
    }
  )
  
  basic_plot_dl <- reactive({
    browser()
    df <- a_in_pulldown()
    if("logFC" %in% colnames(df) & "FDR" %in% colnames(df) & "pvalue" %in% colnames(df) &
       "rep1" %in% colnames(df) & "rep2" %in% colnames(df)){
      if(is.null(a_search_gene())){
        params <- list(a = a_vp(), b = a_sp(), c = a_vp_colorbar_dl(), d = a_vp_count())
      } else{
        params <- list(a = a_vp_plus_rep(), b = a_sp_plus(), c = a_vp_colorbar_dl(), d = a_vp_count())
      }
    } else if("logFC" %in% colnames(df) & "FDR" %in% colnames(df) & "pvalue" %in% colnames(df)){
      if(is.null(a_search_gene())){
        params <- list(a = a_vp(), c = a_vp_colorbar_dl(), d = a_vp_count())
      } else{
        params <- list(a = a_vp_plus_rep(), c = a_vp_colorbar_dl(), d = a_vp_count())
      }
    } else if("rep1" %in% colnames(df) & "rep2" %in% colnames(df)){
      if(is.null(a_search_gene())){
        params <- list(a = a_vp(), b = a_sp(), c = a_vp_colorbar_dl(), d = a_vp_count())
      } else{
        params <- list(a = a_vp_plus_rep(), b = a_sp_plus(), c = a_vp_colorbar_dl(), d = a_vp_count())
      }
    }
    rmarkdown::render("scripts/basic.Rmd", 
                      output_file = file,
                      params = params
    )
  })
  
  output$download_basic_plots <- downloadHandler(
    filename = "basic-plots.html",
    content = function(file) {
      basic_plot_dl()
    }
  )
  
  output$download_layered_plots <- downloadHandler(
    filename = "integrated-plots.html",
    content = function(file) {
      df <- a_in_pulldown()
      if(is.null(a_search_gene())){
        params <- list(a = a_multi_vp_layer(), c = a_multi_vp_colorbar_dl(), d = a_multi_vp_count())
      } else{
        params <- list(a = a_multi_vp_plus(), c = a_multi_vp_colorbar_dl(), d = a_multi_vp_count())
      }
      rmarkdown::render("scripts/integrated.Rmd", 
                        output_file = file,
                        params = params
      )
    }
  )
  
  output$download_venn_diagram_plot <- downloadHandler(
    filename = "venn-diagram-InWeb.html",
    content = function(file) {
      params <- list(a = a_venndiagram(), b = hypergeometric_test_list(), c = input$a_FDR_range[1], d = input$a_FDR_range[2], 
                     e = a_bait_gene_vennd(), f = population_bait(), g = success_population_bait(), h = sample_bait(),
                     i = input$a_logFC_range[1], j = input$a_logFC_range[2])
      rmarkdown::render("scripts/vennd-inweb.Rmd", 
                        output_file = file,
                        params = params
                        # envir = new.env(parent = globalenv())
      )
    }
  )
  
  output$download_venn_diagram_inweb_genes <- downloadHandler(
    filename = function() {
      paste("inweb-overlap-genes", ".txt", sep = "")
    },
    content = function(file) {
      write.table(hypergeometric_test_list(), file, sep = "\t", col.names = T, row.names = F, quote = F)
    }
  )
  
  output$download_venn_diagram_GOI_plot <- downloadHandler(
    filename = "venn-diagram-GOI.html",
    content = function(file) {
      params <- list(a = a_venndiagram_GOI(), b = hypergeometric_test_list_GOI(), c = input$a_FDR_range[1], d = input$a_FDR_range[2], 
                     f = population_GOI(), g = success_population_GOI(), h = sample_GOI(),
                     i = input$a_logFC_range[1], j = input$a_logFC_range[2])
      rmarkdown::render("scripts/vennd-GOI.Rmd", 
                        output_file = file,
                        params = params
                        # envir = new.env(parent = globalenv())
      )
    }
  )
  
  output$download_venn_diagram_GOI_genes <- downloadHandler(
    filename = function() {
      paste("GOI-overlap-genes", ".txt", sep = "")
    },
    content = function(file) {
      write.table(hypergeometric_test_list_GOI(), file, sep = "\t", col.names = T, row.names = F, quote = F)
    }
  )
  
  output$download_venn_diagram_SNP_plot <- downloadHandler(
    filename = "venn-diagram-SNP.html",
    content = function(file) {
      params <- list(a = a_venndiagram_SNP(), b = hypergeometric_test_list_SNP(), c = input$a_FDR_range[1], d = input$a_FDR_range[2], 
                     f = population_SNP(), g = success_population_SNP(), h = sample_SNP(),
                     i = input$a_logFC_range[1], j = input$a_logFC_range[2], k = a_venndiagram_SNP_SGL(), l = a_venndiagram_SNP_MGL(),
                     m = success_sample_SNP_SGL(), n = success_sample_SNP_MGL(), o = sample_SNP_SGL(), p = sample_SNP_MGL())
      rmarkdown::render("scripts/vennd-SNP.Rmd", 
                        output_file = file,
                        params = params
                        # envir = new.env(parent = globalenv())
      )
    }
  )
  
  output$download_venn_diagram_SNP_genes <- downloadHandler(
    filename = function() {
      paste("SNP-overlap-genes", ".txt", sep = "")
    },
    content = function(file) {
      write.table(hypergeometric_test_list_SNP(), file, sep = "\t", col.names = T, row.names = F, quote = F)
    }
  )
  
  output$download_bpf_vp_plot <- downloadHandler(
    filename = "basic-protein-family-vp.html",
    content = function(file) {
      if(is.null(a_search_gene())){
        params <- list(a = a_bpf())
      } else{
        params <- list(a = a_bpf_plus())
      }
      rmarkdown::render("scripts/pf.Rmd", 
                        output_file = file,
                        params = params
      )
    }
  )
  
  # output$download_bpf_sp_plot <- downloadHandler(
  #   filename = "basic-protein-family-sp.html",
  #   content = function(file) {
  #     if(is.null(a_search_gene())){
  #       params <- list(a = a_bpf_sp())
  #     } else{
  #       params <- list(a = a_bpf_plus_sp())
  #     }
  #     rmarkdown::render("scripts/pf.Rmd", 
  #                       output_file = file,
  #                       params = params
  #     )
  #   }
  # )
  
  output$download_snp_to_genes <- downloadHandler(
    filename = function() {
      paste("snp-to-gene", ".txt", sep = "")
    },
    content = function(file) {
      write.table(SNP_to_gene(), file, sep = "\t", col.names = T, row.names = F, quote = F)
    }
  )
  
  ##START HERE FOR GOI OVERLAP DOWNLOAD##
  output$download_goi_overlap <- downloadHandler(
    filename = function() {
      paste("goi-overlap", ".txt", sep = "")
    },
    content = function(file) {
      multi_vp <- a_multi_vp()
      df <- ldply(multi_vp$d_g2s, data.frame)
      write.table(df, file, sep = "\t", col.names = T, row.names = F, quote = F)
    }
  )
  
  output$download_enriched_families_bpf <- downloadHandler(
    filename = function() {
      paste("enriched_protein_families", ".txt", sep = "")
    },
    content = function(file) {
      write.table(a_bpf_families1(), file, sep = "\t", col.names = T, row.names = F, quote = F)
    }
  )
  
  ##### VISUALIZATIONS END ##### 
  
  ##### ADVANCED FEATURES START ##### 
  output$b_pf_loc_selection <- renderUI({
    validate(
      need(input$c_file_pulldown3 != '', "")
    )
    selectInput('b_pf_loc_option', 'Data options', c("Protein family" = 'pf', "Localization_GO" = 'loc_go', "Localization_UniProt" = 'loc_uniprot'), selectize=FALSE)
  })
  
  b_pf_data_selection <- reactive({
    inDselection <- input$b_pf_loc_option
    if (inDselection == "pf") {
      pf_data <- PF_gene_indexed
    } else if (inDselection == "loc_go"){
      pf_data <- go_location_gene_indexed
    } else {
      pf_data <- uniprot_location_gene_indexed
    }
  })
  
  #create slider for FDR
  output$b_PF_FDR_slider <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_file_pulldown2 != '', ""),
      need(input$c_file_pulldown3 != '', "")
    )
    sliderInput("b_PF_FDR_range", "Range for FDR",
                min = 0, max = 1, value = c(0, 1), step = 0.01, sep = '', pre = NULL, post = NULL)
  })
  
  #create slider for pvalue
  output$b_PF_pvalue_slider <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_file_pulldown2 != '', ""),
      need(input$c_file_pulldown3 != '', "")
    )
    sliderInput("b_PF_pvalue_range", "Range for pvalue",
                min = 0, max = 1, value = c(0, 1), step = 0.01, sep = '', pre = NULL, post = NULL)
  })
  
  # based on a_pulldown(), create slider for BPF logFC
  output$b_PF_logFC_slider <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_file_pulldown2 != '', ""),
      need(input$c_file_pulldown3 != '', "")
    )
    if(!is.null(c_pd1()) & !is.null(c_pd2())  & !is.null(c_pd3())){
      d1 <- c_pd1()
      d2 <- c_pd2()
      max_logFC <- max(max(d1$logFC), max(d2$logFC))
      max_logFC <- round(max_logFC+0.5, 1)
      sliderInput("b_PF_logFC_range", "Range for logFC",
                  min = 0, max = max_logFC, value = c(0, max_logFC), step = 0.1)
    }
  })
  
  output$b_PF_marker_size <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_file_pulldown2 != '', ""),
      need(input$c_file_pulldown3 != '', "")
    )
    radioButtons('b_PF_option', 'Turn on/off marker sizing',
                 c(On = 'change_m_PF',
                   Off = 'static_m_PF'),
                 inline = T
    )
  })
  
  output$b_PF_freq <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_file_pulldown2 != '', ""),
      need(input$c_file_pulldown3 != '', "")
    )
    if(!is.null(input$b_PF_option)){
      inBPFmarker <- input$b_PF_option
      if (inBPFmarker == "change_m_PF") {
        sliderInput("b_PF_marker_freq", "Marker size",
                    min = 1, max = 40, value = 1, step = 1)
      }
    }
  })
  
  output$b_PF_button <- renderUI({
    # if (input$file_pfam1 != 0 & input$file_pfam2 != 0 & input$file_pfam3 != 0){
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_file_pulldown2 != '', ""),
      need(input$c_file_pulldown3 != '', "")
    )
    if(!is.null(c_pd1()) & !is.null(c_pd2()) & !is.null(c_pd3())){
      actionButton("b_make_pf", "Generate PF VP")
    }
  })
  
  # output$b_GOI_search <- renderUI({
  #   textInput("b_goi_search", "Search for gene (e.g. SPOCK)")
  # })
  # 
  # b_search_gene <- reactive({
  #   gene_in <- input$b_goi_search
  #   if (is.null(gene_in) | gene_in == "" ){
  #     return(NULL)
  #   } else{
  #     gene <- gene_in
  #     gene <- toupper(gene)
  #   }
  # })
  
  output$b_PF_sort_col <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_file_pulldown2 != '', ""),
      need(input$c_file_pulldown3 != '', "")
    )
    radioButtons('b_PF_sort_option', 'Sort',
                 c("Alphabetically" = 'sort_alph',
                   "PF Frequency" = 'sort_freq'),
                 inline = T
    )
  })
  
  b_PF_sorting <- reactive({
    inPFsort <- input$b_PF_sort_option
    if (inPFsort == "sort_alph") {
      marker <- "sort_a"
    } else {
      marker <- "sort_f"
    }
  })
  
  # b_in_pulldown1 <- reactive({
  #   if(!is.null(input$c_file_pulldown1)){
  #     pulldown <- input$c_file_pulldown1
  #     d <- fread(pulldown$datapath, header = TRUE,
  #                sep="auto", na.strings=c(""," ","NA"), stringsAsFactors = FALSE, data.table = FALSE, blank.lines.skip = T)
  #     d <- na.omit(d)
  #   }
  # })
  # 
  # b_orig_pulldown1 <- reactive({
  #   if(!is.null(b_in_pulldown1())){
  #     d <- b_in_pulldown1()
  #     d_col <- colnames(d)
  #     if("gene" %in% d_col & "accession_number" %in% d_col){
  #       df <- d
  #       df$gene <- toupper(df$gene)
  #     } else if("accession_number" %in% d_col){
  #       withProgress(message = 'Mapping UniProt IDs to HGNC symbols',
  #                    detail = 'Hold please', value = 0, {
  #                      write.table(d$accession_number, "scripts/gene-tools-master/map/in.txt", append = F, quote = F,
  #                                  row.names = F, col.names = F)
  #                      incProgress(0.6)
  #                      system("python scripts/gene-tools-master/map/map.py")
  #                      incProgress(0.8)
  #                      d1 <- fread("scripts/gene-tools-master/map/results.txt", header = FALSE,
  #                                  sep="auto", na.strings=c(""," ","NA"), stringsAsFactors = FALSE, data.table = FALSE)
  #                      colnames(d1) <- c("uniprot_id", "gene", "method")
  #                      df <- merge(d, d1, by.x = "accession_number", by.y = "uniprot_id")
  #                      df <- subset(df, select=-c(method))
  #                      incProgress(0.9)
  #                      df
  #                    })
  #     } else if("gene" %in% d_col){
  #       df <- d
  #       df$gene <- toupper(df$gene)
  #     }
  #     df
  #   }
  # })
  # 
  # b_pulldown1 <- reactive({
  #   if(!is.null(b_orig_pulldown1())){
  #     d <- b_orig_pulldown1()
  #     d_col <- colnames(d)
  #     if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col &
  #        "rep1" %in% d_col & "rep2" %in% d_col){
  #       df1 <- d
  #     }else if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col){
  #       df1 <- d
  #     }else if("rep1" %in% d_col & "rep2" %in% d_col){
  #       df <- d
  #       if("accession_number" %in% d_col){
  #         df1 <- calculate_moderated_ttest_uniprot(df)
  #       }
  #       else if("accession_number" %!in% d_col){
  #         df1 <- calculate_moderated_ttest_hgnc(df) 
  #       }
  #     }
  #     df1
  #   }
  # })
  # 
  # b_in_pulldown2 <- reactive({
  #   if(!is.null(input$c_file_pulldown2)){
  #     pulldown <- input$c_file_pulldown2
  #     d <- fread(pulldown$datapath, header = TRUE,
  #                sep="auto", na.strings=c(""," ","NA"), stringsAsFactors = FALSE, data.table = FALSE, blank.lines.skip = T)
  #     d <- na.omit(d)
  #   }
  # })
  # 
  # b_orig_pulldown2 <- reactive({
  #   if(!is.null(b_in_pulldown2())){
  #     d <- b_in_pulldown2()
  #     d_col <- colnames(d)
  #     if("gene" %in% d_col & "accession_number" %in% d_col){
  #       df <- d
  #       df$gene <- toupper(df$gene)
  #     } else if("accession_number" %in% d_col){
  #       withProgress(message = 'Mapping UniProt IDs to HGNC symbols',
  #                    detail = 'Hold please', value = 0, {
  #                      write.table(d$accession_number, "scripts/gene-tools-master/map/in.txt", append = F, quote = F,
  #                                  row.names = F, col.names = F)
  #                      incProgress(0.6)
  #                      system("python scripts/gene-tools-master/map/map.py")
  #                      incProgress(0.8)
  #                      d1 <- fread("scripts/gene-tools-master/map/results.txt", header = FALSE,
  #                                  sep="auto", na.strings=c(""," ","NA"), stringsAsFactors = FALSE, data.table = FALSE)
  #                      colnames(d1) <- c("uniprot_id", "gene", "method")
  #                      df <- merge(d, d1, by.x = "accession_number", by.y = "uniprot_id")
  #                      df <- subset(df, select=-c(method))
  #                      incProgress(0.9)
  #                      df
  #                    })
  #     } else if("gene" %in% d_col){
  #       df <- d
  #       df$gene <- toupper(df$gene)
  #     }
  #     df
  #   }
  # })
  # 
  # b_pulldown2 <- reactive({
  #   if(!is.null(b_orig_pulldown2())){
  #     d <- b_orig_pulldown2()
  #     d_col <- colnames(d)
  #     if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col &
  #        "rep1" %in% d_col & "rep2" %in% d_col){
  #       df1 <- d
  #     }else if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col){
  #       df1 <- d
  #     }else if("rep1" %in% d_col & "rep2" %in% d_col){
  #       df <- d
  #       if("accession_number" %in% d_col){
  #         df1 <- calculate_moderated_ttest_uniprot(df)
  #       }
  #       else if("accession_number" %!in% d_col){
  #         df1 <- calculate_moderated_ttest_hgnc(df) 
  #       }
  #     }
  #     df1
  #   }
  # })
  # 
  # b_in_pulldown3 <- reactive({
  #   if(!is.null(input$c_file_pulldown3)){
  #     pulldown <- input$c_file_pulldown3
  #     d <- fread(pulldown$datapath, header = TRUE,
  #                sep="auto", na.strings=c(""," ","NA"), stringsAsFactors = FALSE, data.table = FALSE, blank.lines.skip = T)
  #     d <- na.omit(d)
  #   }
  # })
  # 
  # b_orig_pulldown3 <- reactive({
  #   if(!is.null(b_in_pulldown3())){
  #     d <- b_in_pulldown3()
  #     d_col <- colnames(d)
  #     if("gene" %in% d_col & "accession_number" %in% d_col){
  #       df <- d
  #       df$gene <- toupper(df$gene)
  #     } else if("accession_number" %in% d_col){
  #       withProgress(message = 'Mapping UniProt IDs to HGNC symbols',
  #                    detail = 'Hold please', value = 0, {
  #                      write.table(d$accession_number, "scripts/gene-tools-master/map/in.txt", append = F, quote = F,
  #                                  row.names = F, col.names = F)
  #                      incProgress(0.6)
  #                      system("python scripts/gene-tools-master/map/map.py")
  #                      incProgress(0.8)
  #                      d1 <- fread("scripts/gene-tools-master/map/results.txt", header = FALSE,
  #                                  sep="auto", na.strings=c(""," ","NA"), stringsAsFactors = FALSE, data.table = FALSE)
  #                      colnames(d1) <- c("uniprot_id", "gene", "method")
  #                      df <- merge(d, d1, by.x = "accession_number", by.y = "uniprot_id")
  #                      df <- subset(df, select=-c(method))
  #                      incProgress(0.9)
  #                      df
  #                    })
  #     } else if("gene" %in% d_col){
  #       df <- d
  #       df$gene <- toupper(df$gene)
  #     }
  #     df
  #   }
  # })
  # 
  # b_pulldown3 <- reactive({
  #   if(!is.null(b_orig_pulldown3())){
  #     d <- b_orig_pulldown3()
  #     d_col <- colnames(d)
  #     if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col &
  #        "rep1" %in% d_col & "rep2" %in% d_col){
  #       df1 <- d
  #     }else if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col){
  #       df1 <- d
  #     }else if("rep1" %in% d_col & "rep2" %in% d_col){
  #       df <- d
  #       if("accession_number" %in% d_col){
  #         df1 <- calculate_moderated_ttest_uniprot(df)
  #       }
  #       else if("accession_number" %!in% d_col){
  #         df1 <- calculate_moderated_ttest_hgnc(df) 
  #       }
  #     }
  #     df1
  #   }
  # })
  
  b_bait_gene <- reactive({
    bait_in <- input$b_bait_PF
    if (is.null(bait_in) | bait_in == "" ){
      return(NULL)
    } else{
      bait <- bait_in
      bait <- toupper(bait)
    }
  })
  
  b_PF_marker <- reactive({
    inBPFmarker <- input$b_PF_option
    if (inBPFmarker == "change_m_PF") {
      marker <- "change"
    } else {
      marker <- "no_change"
    }
  })
  
  b_pf_data <- eventReactive(input$b_make_pf, {
    data_selection <- b_pf_data_selection()
    withProgress(message = 'This may take a while', 
                 detail = 'Hold please', value = 0, {
                   pf1 <- c_pd1()
                   pf2 <- c_pd2()
                   pf3 <- c_pd3()
                   pfmsizing <- b_PF_marker()
                   incProgress(0.2)
                   
                   makePlotFamilies <- function(dim, dip, dpm, increase_size = input$b_PF_marker_freq){ #, bait
                     im <- subset(dim, 
                                  (pvalue <= input$b_PF_pvalue_range[2] & pvalue >= input$b_PF_pvalue_range[1]) & 
                                    (FDR <= input$b_PF_FDR_range[2] & FDR >= input$b_PF_FDR_range[1]) &
                                    (logFC <= input$b_PF_logFC_range[2] & logFC >= input$b_PF_logFC_range[1]))
                     ip <- subset(dip, 
                                  (pvalue <= input$b_PF_pvalue_range[2] & pvalue >= input$b_PF_pvalue_range[1]) & 
                                    (FDR <= input$b_PF_FDR_range[2] & FDR >= input$b_PF_FDR_range[1]) &
                                    (logFC <= input$b_PF_logFC_range[2] & logFC >= input$b_PF_logFC_range[1]))
                     pm <- dpm
                     
                     enM <- subset(im, gene %!in% ip$gene)
                     enP <- subset(ip, gene %!in% im$gene)
                     
                     incProgress(0.4)
                     enM_families <- assignFamily_inc_doubles(enM, data_selection)
                     # enM_gna <- addNames(enM)
                     enP_families <- assignFamily_inc_doubles(enP, data_selection)
                     # enP_gna <- addNames(enP)
                     incProgress(0.6)
                     ### replace logFC and pvalue in di* by these from dpm
                     enM_families$logFC <- dpm$logFC[match(enM_families$gene, dpm$gene)]
                     enM_families$pvalue <- dpm$pvalue[match(enM_families$gene, dpm$gene)]
                     enP_families$logFC <- dpm$logFC[match(enP_families$gene, dpm$gene)]
                     enP_families$pvalue <- dpm$pvalue[match(enP_families$gene, dpm$gene)]
                     # enM_gna$logFC <- dpm$logFC[match(enM_gna$gene, dpm$gene)]
                     # enM_gna$pvalue <- dpm$pvalue[match(enM_gna$gene, dpm$gene)]
                     # enP_gna$logFC <- dpm$logFC[match(enP_gna$gene, dpm$gene)]
                     # enP_gna$pvalue <- dpm$pvalue[match(enP_gna$gene, dpm$gene)]
                     
                     # mybait <- dpm[grepl(bait,dpm$gene),]
                     
                     incProgress(0.8)
                     howmany1 <- length(unique(enM_families$overlay))
                     howmany2 <- length(unique(enP_families$overlay))
                     
                     # enM_families$family <- paste(enM_families$family, " ", sep = '')
                     
                     increase_size <- input$b_PF_marker_freq
                     
                     incProgress(0.9)
                     pf_list <- list("list" = dpm, "list" = enM_families, "list" = enP_families,
                                     "integer" = howmany1, "integer" = howmany2, pfmsizing, increase_size)
                     return(pf_list)
                   }
                   makePlotFamilies(pf1, pf2, pf3)
                 })
  })
  
  b_pf_families <- reactive({
    pf_data <- b_pf_data()
    enM_families <- pf_data[[2]]
    enP_families <- pf_data[[3]]
    families1 <- cbind(enM_families$gene, enM_families$overlay, enM_families$frequency)
    colnames(families1) <- c("gene", "overlay", "frequency")
    families2 <- cbind(enP_families$gene, enP_families$overlay, enP_families$frequency)
    colnames(families2) <- c("gene", "overlay", "frequency")
    families <- rbind(families1, families2)
    families
  })
  
  b_pf <- reactive({
    pf_data <- b_pf_data()
    pfsort <- b_PF_sorting()
    dpm <- pf_data[[1]]
    enM_families <- pf_data[[2]]
    # enM_gna <- pf_data[[3]]
    enP_families <- pf_data[[3]]
    # enP_gna <- pf_data[[5]]
    howmany1 <- pf_data[[4]]
    howmany2 <- pf_data[[5]]
    pfmsizing <- pf_data[[6]]
    increase_size <- pf_data[[7]]
    
    if(pfsort == "sort_f"){
      enM_families <- enM_families[order(-enM_families$frequency), ]
      if(nrow(enM_families)>=1){
        enM_families <- cbind(enM_families, new_f = paste0("(", enM_families$frequency, ") ", enM_families$overlay))
      } else {
        enM_families <- data.frame("id" = character(0), "rep1" = numeric(0), "rep2" = numeric(0), 
                                   "logFC" = numeric(0), "pvalue" = numeric(0), "FDR" = numeric(0), 
                                   "gene" = character(0), "overlay" = character(0), "frequency" = numeric(0),
                                   "new_f" = character(0))
      }
      enP_families <- enP_families[order(-enP_families$frequency), ]
      if(nrow(enP_families)>=1){
        enP_families <- cbind(enP_families, new_f = paste0("(", enP_families$frequency, ") ", enP_families$overlay))
      } else {
        enP_families <- data.frame("id" = character(0), "rep1" = numeric(0), "rep2" = numeric(0), 
                                   "logFC" = numeric(0), "pvalue" = numeric(0), "FDR" = numeric(0), 
                                   "gene" = character(0), "overlay" = character(0), "frequency" = numeric(0),
                                   "new_f" = character(0))
      }
    } else if(pfsort == "sort_a"){
      enM_families 
      if(nrow(enM_families)>=1){
        enM_families <- cbind(enM_families, new_f = paste0(enM_families$overlay, " (", enM_families$frequency, ")"))
      } else {
        enM_families <- data.frame("id" = character(0), "rep1" = numeric(0), "rep2" = numeric(0),
                                   "logFC" = numeric(0), "pvalue" = numeric(0), "FDR" = numeric(0),
                                   "gene" = character(0), "overlay" = character(0), "frequency" = numeric(0),
                                   "new_f" = character(0))
      }
      enP_families 
      if(nrow(enP_families)>=1){
        enP_families <- cbind(enP_families, new_f = paste0(enP_families$overlay, " (", enP_families$frequency, ")"))
      } else {
        enP_families <- data.frame("id" = character(0), "rep1" = numeric(0), "rep2" = numeric(0),
                                   "logFC" = numeric(0), "pvalue" = numeric(0), "FDR" = numeric(0),
                                   "gene" = character(0), "overlay" = character(0), "frequency" = numeric(0),
                                   "new_f" = character(0))
      }
    }
    
    enM_families$new_f <- paste(enM_families$new_f, " ", sep = '')
    enM_families$new_f <- factor(enM_families$new_f, levels = unique(enM_families$new_f))
    enP_families$new_f <- factor(enP_families$new_f, levels = unique(enP_families$new_f))
    
    if (pfmsizing == "change"){
      
      p <- plot_ly(colors = rainbow(howmany1+howmany2, start = 0, end = 0.85), width = 1300, height = 900) %>% # ,visible='legendonly' width = 1300, height = 900
        #background
        add_markers(data = dpm, x = ~logFC, y = ~-log10(pvalue), opacity = 0.6,
                    marker = list(color = 'rgba(176,196,222,08)'),
                    text = ~paste(gene), hoverinfo = "text", showlegend = FALSE)
      #families1
      if(nrow(enM_families)>0){
        p <- add_markers(p, data = enM_families, x = ~logFC, y = ~-log10(pvalue),
                         marker = list(size = ~frequency*increase_size, symbol = c('square'), opacity = 0.8, line = list(width=0.6, color = "black")),
                         color = ~new_f,
                         text = ~paste(gene, overlay, frequency, sep = "  "), hoverinfo="text")
      }
      #families2
      if(nrow(enP_families)>0){
        p <- add_markers(p, data = enP_families, x = ~logFC, y = ~-log10(pvalue),
                         marker = list(size = ~frequency*increase_size, symbol = c('circle'), opacity = 0.8, line = list(width = 0.6, color = "black")),
                         color = ~new_f,
                         text = ~paste(gene, overlay, frequency, sep = "  "), hoverinfo="text")
      }
      # ### not assigned genes1
      # add_markers(data = enM_gna, x = ~logFC, y = ~-log10(pvalue),
      #             marker = list(size = 8, symbol = 14, color = c('white'), opacity = 0.8, line = list(width=0.9, color = "black")),
      #             hoverinfo="text", text = ~paste(gene, name, sep = "  "), name = "Unassigned genes") %>%
      ### not assigned genes2
      # add_markers(data = enP_gna, x = ~logFC, y = ~-log10(pvalue),
      #             marker = list(size = 6, symbol = 2, color = c('white'), opacity = 0.8, line = list(width = 0.9, color = "black")),
      #             hoverinfo="text", text = ~paste(gene, name, sep = "  "), name="Unassigned genes") %>%
      p
    } else{
      p <- plot_ly(colors = rainbow(howmany1+howmany2, start = 0, end = 0.85), width = 1300, height = 900) %>%
        #background
        add_markers(data = dpm, x = ~logFC, y = ~-log10(pvalue), opacity = 0.6,
                    marker = list(color = 'rgba(176,196,222,08)'),
                    text = ~paste(gene), hoverinfo = "text", showlegend = FALSE)
      ### not assigned genes1
      # add_markers(data = enM_gna, x = ~logFC, y = ~-log10(pvalue), 
      #             marker = list(size = 8, symbol = 14, color = c('white'), opacity = 0.8, line = list(width=0.9, color = "black")), 
      #             hoverinfo="text", text = ~paste(gene, name, sep = "  "), name = "Unassigned genes") %>%
      #families1
      if(nrow(enM_families)>0){
        p <- add_markers(p, data = enM_families, x = ~logFC, y = ~-log10(pvalue), 
                         marker = list(symbol = c('square'), opacity = 1, line = list(width=0.6, color = "black"), size = 12), 
                         color = ~new_f,
                         text = ~paste(gene, overlay, frequency, sep = "  "), hoverinfo="text")
      }
      ### not assigned genes2
      # add_markers(data = enP_gna, x = ~logFC, y = ~-log10(pvalue), 
      #             marker = list(size = 6, symbol = 2, color = c('white'), opacity = 0.8, line = list(width = 0.9, color = "black")), 
      #             hoverinfo="text", text = ~paste(gene, name, sep = "  "), name="Unassigned genes") %>%
      #families2
      if(nrow(enP_families)>0){
        p <- add_markers(p, data = enP_families, x = ~logFC, y = ~-log10(pvalue), 
                         marker = list(symbol = c('circle'), opacity = 1, line = list(width = 0.6, color = "black"), size = 12), 
                         color = ~new_f,
                         text = ~paste(gene, overlay, frequency, sep = "  "), hoverinfo="text") 
      }
      p
    }
    p
  })
  
  b_pf_plus <- reactive({
    validate(
      need(!is.null(c_search_gene()), "")
    )
    p <- b_pf()
    goi <- c_search_gene()
    orig_data <- c_pd3()
    searchgene <- orig_data[grepl(goi,orig_data$gene),]
    
    p %>%
      add_trace(data = searchgene, x = ~logFC, y = ~-log10(pvalue), 
                mode = "markers+text", hoverinfo="text+x+y", text = ~paste(gene), 
                marker = list(color = "#f7f4f9", size = 10, line = list(width=1.3, color = "#3f007d")),
                textposition = ~ifelse(logFC>0, "middle right", "middle left"), textfont = list(color='black', size = 10), 
                type = 'scatter', showlegend = FALSE)
  })
  
  output$Protein_Family <- renderPlotly({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_file_pulldown2 != '', ""),
      need(input$c_file_pulldown3 != '', "")
    )
    if(is.null(c_search_gene())){
      b_pf()
    } else{
      b_pf_plus()
    }
  })
  
  observe({
    # if (is.null(input$input$file_pfam1) & is.null(input$file_pfam2) & is.null(input$file_pfam3)){
    if (input$b_make_pf == 0 || is.null(input$b_make_pf)){
      shinyjs::disable("download_pf_plot")
      shinyjs::disable("download_enriched_families_pf")
    } else {
      shinyjs::enable("download_pf_plot")
      shinyjs::enable("download_enriched_families_pf")
    }
  })
  
  output$download_pf_plot <- downloadHandler(
    filename = "advanced-protein-family.html",
    content = function(file) {
      if(is.null(c_search_gene())){
        params <- list(a = b_pf())
      } else{
        params <- list(a = b_pf_plus())
      }
      rmarkdown::render("scripts/pf.Rmd", 
                        output_file = file,
                        params = params
      )
    }
  )
  
  output$download_enriched_families_pf <- downloadHandler(
    filename = function() {
      paste("advanced_enriched_families", ".txt", sep = "")
    },
    content = function(file) {
      write.table(b_pf_families(), file, sep = "\t", col.names = T, row.names = F, quote = F)
    }
  )
  
  ##### ADVANCED FEATURES END ##### 
  
  ##### MULTIPLE PD COMPARISONS START #####
  
  output$c_file1 <- renderUI({
    fileInput('c_file_pulldown1', 'Upload user input file1',
              accept = c(
                'text/csv',
                'text/comma-separated-values',
                'text/tab-separated-values',
                'text/plain',
                '.csv',
                '.tsv')
    )
  })
  
  output$c_file2 <- renderUI({
    fileInput('c_file_pulldown2', 'Upload user input file2',
              accept = c(
                'text/csv',
                'text/comma-separated-values',
                'text/tab-separated-values',
                'text/plain',
                '.csv',
                '.tsv')
    )
  })
  
  output$c_file3 <- renderUI({
    fileInput('c_file_pulldown3', 'Upload user input file3',
              accept = c(
                'text/csv',
                'text/comma-separated-values',
                'text/tab-separated-values',
                'text/plain',
                '.csv',
                '.tsv')
    )
  })
  
  output$c_file_color <- renderUI({
    validate(
      need(input$c_colorscheme == "user", "")
    )
    fileInput('c_file_color', 'Upload user score input',
              accept = c(
                'text/csv',
                'text/comma-separated-values',
                'text/tab-separated-values',
                'text/plain',
                '.csv',
                '.tsv')
    )
  })
  
  c_in_file_color <- reactive({
    if(!is.null(input$c_file_color)){
      color_file <- input$c_file_color
      d <- fread(color_file$datapath, header = TRUE, 
                 sep="auto", na.strings=c(""," ","NA"), stringsAsFactors = FALSE, data.table = FALSE, blank.lines.skip = T)
      d <- na.omit(d)
    }
  })
  
  output$c_color_style <- renderUI({
    validate(
      need(input$c_colorscheme == "user", "")
    )
    radioButtons("c_colorscheme_style", "Color scale:",
                 c("Continuous" = "cont", 
                   "Discrete" = "disc"),
                 inline = T)
  })
  
  output$c_color_theme <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_colorscheme == "user", "")
    )
    selectInput("c_colorbrewer_theme", "User value", 
                c("YlOrRd", "YlOrBr", "YlGnBu", "YlGn", "Reds", "RdPu", "Purples", "PuRd", "PuBuGn", "PuBu", "OrRd", 
                  "Oranges", "Greys", "Greens", "GnBu", "BuPu", "BuGn", "Blues", "Set3", "Set2", "Set1", "Pastel2", "Pastel1",
                  "Paired", "Dark2", "Accent", "Spectral", "RdYlGn", "RdYlBu", "RdGy", "RdBu", "PuOr", "PRGn", "PiYG", "BrBG"))
  })
  
  output$c_color_theme_tab3 <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_file_pulldown2 != '', ""),
      need(input$c_colorscheme == "user", "")
    )
    selectInput("c_colorbrewer_theme_tab3", "User value", 
                c("YlOrRd", "YlOrBr", "YlGnBu", "YlGn", "Reds", "RdPu", "Purples", "PuRd", "PuBuGn", "PuBu", "OrRd", 
                  "Oranges", "Greys", "Greens", "GnBu", "BuPu", "BuGn", "Blues", "Set3", "Set2", "Set1", "Pastel2", "Pastel1",
                  "Paired", "Dark2", "Accent", "Spectral", "RdYlGn", "RdYlBu", "RdGy", "RdBu", "PuOr", "PRGn", "PiYG", "BrBG"))
  })
  
  output$c_color_theme_tab4 <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_file_pulldown2 != '', ""),
      need(input$c_colorscheme == "user", "")
    )
    selectInput("c_colorbrewer_theme_tab4", "User value", 
                c("YlOrRd", "YlOrBr", "YlGnBu", "YlGn", "Reds", "RdPu", "Purples", "PuRd", "PuBuGn", "PuBu", "OrRd", 
                  "Oranges", "Greys", "Greens", "GnBu", "BuPu", "BuGn", "Blues", "Set3", "Set2", "Set1", "Pastel2", "Pastel1",
                  "Paired", "Dark2", "Accent", "Spectral", "RdYlGn", "RdYlBu", "RdGy", "RdBu", "PuOr", "PRGn", "PiYG", "BrBG"))
  })
  
  output$c_color_theme_tab5 <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_file_pulldown2 != '', ""),
      need(input$c_colorscheme == "user", "")
    )
    selectInput("c_colorbrewer_theme_tab5", "User value", 
                c("YlOrRd", "YlOrBr", "YlGnBu", "YlGn", "Reds", "RdPu", "Purples", "PuRd", "PuBuGn", "PuBu", "OrRd", 
                  "Oranges", "Greys", "Greens", "GnBu", "BuPu", "BuGn", "Blues", "Set3", "Set2", "Set1", "Pastel2", "Pastel1",
                  "Paired", "Dark2", "Accent", "Spectral", "RdYlGn", "RdYlBu", "RdGy", "RdBu", "PuOr", "PRGn", "PiYG", "BrBG"))
  })
  
  output$c_color_theme_indv_sig <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_colorscheme == "fdr", "")
    )
    region <- c(paste0("FDR<", input$a_fdr_thresh))
    selectInput('c_color_indv_sig', region, marker_cols$V1, multiple=F, selectize=TRUE, selected = "seagreen3")
  })
  
  output$c_color_theme_indv_insig <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_colorscheme == "fdr", "")
    )
    region <- c(paste0("FDR≥", input$a_fdr_thresh))
    selectInput('c_color_indv_insig', region, marker_cols$V1, multiple=F, selectize=TRUE, selected = "royalblue2")
  })
  
  output$c_color_theme_inweb_sig <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_file_pulldown2 != '', ""),
      need(input$c_colorscheme == "fdr", "")
    )
    region <- c(paste0("FDR<", input$a_fdr_thresh))
    selectInput('c_color_inweb_sig', region, marker_cols$V1, multiple=F, selectize=TRUE, selected = "seagreen3")
  })
  
  output$c_color_theme_inweb_insig <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_file_pulldown2 != '', ""),
      need(input$c_colorscheme == "fdr", "")
    )
    region <- c(paste0("FDR≥", input$a_fdr_thresh))
    selectInput('c_color_inweb_insig', region, marker_cols$V1, multiple=F, selectize=TRUE, selected = "royalblue2")
  })
  
  # output$c_color_theme_snp_sig <- renderUI({
  #   validate(
  #     need(input$c_file_pulldown1 != '', ""),
  #     need(input$c_file_pulldown2 != '', ""),
  #     need(input$c_colorscheme == "fdr", "")
  #   )
  #   region <- c(paste0("FDR<", input$a_fdr_thresh))
  #   selectInput('c_color_snp_sig', region, marker_cols$V1, multiple=F, selectize=TRUE, selected = "seagreen3")
  # })
  # 
  # output$c_color_theme_snp_insig <- renderUI({
  #   validate(
  #     need(input$c_file_pulldown1 != '', ""),
  #     need(input$c_file_pulldown2 != '', ""),
  #     need(input$c_colorscheme == "fdr", "")
  #   )
  #   region <- c(paste0("FDR≥", input$a_fdr_thresh))
  #   selectInput('c_color_snp_insig', region, marker_cols$V1, multiple=F, selectize=TRUE, selected = "royalblue2")
  # })
  
  output$c_color_theme_goi_sig <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_file_pulldown2 != '', ""),
      need(input$c_colorscheme == "fdr", "")
    )
    region <- c(paste0("FDR<", input$a_fdr_thresh))
    selectInput('c_color_goi_sig', region, marker_cols$V1, multiple=F, selectize=TRUE, selected = "seagreen3")
  })
  
  output$c_color_theme_goi_insig <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_file_pulldown2 != '', ""),
      need(input$c_colorscheme == "fdr", "")
    )
    region <- c(paste0("FDR≥", input$a_fdr_thresh))
    selectInput('c_color_goi_insig', region, marker_cols$V1, multiple=F, selectize=TRUE, selected = "royalblue2")
  })
  
  output$c_marker_theme_goi <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', ""), 
      need(input$c_file_pulldown2 != '', ""),
      need(!is.null(input$c_file_genes) && input$c_file_genes != "", "")
    )
    selectInput("c_colorbrewer_theme_goi", "Genes of interest", 
                c("YlOrRd", "YlOrBr", "YlGnBu", "YlGn", "Reds", "RdPu", "Purples", "PuRd", "PuBuGn", "PuBu", "OrRd", 
                  "Oranges", "Greys", "Greens", "GnBu", "BuPu", "BuGn", "Blues", "Set3", "Set2", "Set1", "Pastel2", "Pastel1",
                  "Paired", "Dark2", "Accent", "Spectral", "RdYlGn", "RdYlBu", "RdGy", "RdBu", "PuOr", "PRGn", "PiYG", "BrBG"),
                selected = "Set2")
  })
  
  # output$c_marker_theme_snp <- renderUI({
  #   validate(
  #     need(input$c_file_pulldown1 != '', ""), 
  #     need(input$c_file_pulldown2 != '', ""),
  #     need(!is.null(input$c_file_SNP) && input$c_file_SNP != "", "")
  #   )
  #   selectInput("c_colorbrewer_theme_snp", "SNPs", 
  #               c("YlOrRd", "YlOrBr", "YlGnBu", "YlGn", "Reds", "RdPu", "Purples", "PuRd", "PuBuGn", "PuBu", "OrRd", 
  #                 "Oranges", "Greys", "Greens", "GnBu", "BuPu", "BuGn", "Blues", "Set3", "Set2", "Set1", "Pastel2", "Pastel1",
  #                 "Paired", "Dark2", "Accent", "Spectral", "RdYlGn", "RdYlBu", "RdGy", "RdBu", "PuOr", "PRGn", "PiYG", "BrBG"),
  #               selected = "PuOr")
  # })
  
  output$c_marker_theme_inweb <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_file_pulldown2 != '', ""),
      need(input$c_bait != "", "")
    )
    selectInput('c_color_inweb', 'InWeb', add_marker_cols$V1, multiple=F, selectize=TRUE, selected = "yellow")
  })
  
  output$c_color_scheme <- renderUI({
    radioButtons("c_colorscheme", "Color scheme:",
                 c("FDR" = "fdr", 
                   "ExAC" = "exac",
                   "Grayscale" = "cbf",
                   "User value" = "user"),
                 inline = T)
  })
  
  output$c_color_theme_pf <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', "")
    )
    selectInput("c_colorbrewer_theme_pf", "Protein families color", 
                c("YlOrRd", "YlOrBr", "YlGnBu", "YlGn", "Reds", "RdPu", "Purples", "PuRd", "PuBuGn", "PuBu", "OrRd", 
                  "Oranges", "Greys", "Greens", "GnBu", "BuPu", "BuGn", "Blues", "Set3", "Set2", "Set1", "Pastel2", "Pastel1",
                  "Paired", "Dark2", "Accent", "Spectral", "RdYlGn", "RdYlBu", "RdGy", "RdBu", "PuOr", "PRGn", "PiYG", "BrBG"))
  })
  
  output$c_GOI_search <- renderUI({
    textInput("c_goi_search", "Search for gene (e.g. SHH)")
  })
  
  output$c_FDR_thresh <- renderUI({
    sliderInput("c_fdr_thresh", "FDR threshold",
                min = 0, max = 1, value = 0.1, step = 0.01)
  })
  
  output$c_PVal_thresh <- renderUI({
    sliderInput("c_pval_thresh", "p-value threshold",
                min = 0, max = 1, value = 0.05, step = 0.001)
  })
  
  output$c_logFC_thresh_combined <- renderUI({
    if(!is.null(c_in_pd1()) & is.null(c_in_pd2()) & is.null(c_in_pd3())){
      c1 <- c_pd1()
      min_logFC <- min(c1$logFC)
      min_logFC <- round(min_logFC-0.5, 1)
      max_logFC <- max(c1$logFC)
      max_logFC <- round(max_logFC+0.5, 1)
      # combined_max <- max(-(min_logFC), max_logFC)
      sliderInput("c_logfc_thresh_comb", "logFC threshold",
                  min = min_logFC, max = max_logFC, value = c(0, max_logFC), step = 0.1)
    } else if(!is.null(c_in_pd1()) & !is.null(c_in_pd2()) & is.null(c_in_pd3())){
      c1 <- c_pd1()
      c2 <- c_pd2()
      min_logFC <- min(min(c1$logFC), min(c2$logFC))
      min_logFC <- round(min_logFC-0.5, 1)
      max_logFC <- max(max(c1$logFC), max(c2$logFC))
      max_logFC <- round(max_logFC+0.5, 1)
      # combined_max <- max(-(min_logFC), max_logFC)
      sliderInput("c_logfc_thresh_comb", "logFC threshold",
                  min = min_logFC, max = max_logFC, value = c(0, max_logFC), step = 0.1)
    } else if(!is.null(c_in_pd1()) & !is.null(c_in_pd2()) & !is.null(c_in_pd3())){
      c1 <- c_pd1()
      c2 <- c_pd2()
      c3 <- c_pd3()
      min_logFC <- min(min(c1$logFC), min(c2$logFC), min(c3$logFC))
      min_logFC <- round(min_logFC-0.5, 1)
      max_logFC <- max(max(c1$logFC), max(c2$logFC), max(c3$logFC))
      max_logFC <- round(max_logFC+0.5, 1)
      # combined_max <- max(-(min_logFC), max_logFC)
      sliderInput("c_logfc_thresh_comb", "logFC threshold",
                  min = min_logFC, max = max_logFC, value = c(0, max_logFC), step = 0.1)
    } else
      sliderInput("c_logfc_thresh_comb", "logFC threshold",
                  min = -1, max = 1, value = c(0, 1), step = 0.1)
  })
  
  #create slider for FDR f1
  output$c_comparison1_FDR_slider <- renderUI({
    validate(
      need(input$c_file_pulldown2 != '', "")
    )
    sliderInput("c_compare1_FDR_range", "FDR",
                min = 0, max = 1, value = c(0, 1), step = 0.01, sep = '', pre = NULL, post = NULL)
  })
  
  #create slider for FDR f2
  output$c_comparison2_FDR_slider <- renderUI({
    validate(
      need(input$c_file_pulldown2 != '', "")
    )
    sliderInput("c_compare2_FDR_range", "FDR",
                min = 0, max = 1, value = c(0, 1), step = 0.01, sep = '', pre = NULL, post = NULL)
  })
  
  #create slider for FDR f3
  output$c_comparison3_FDR_slider <- renderUI({
    validate(
      need(input$c_file_pulldown3 != '', "")
    )
    sliderInput("c_compare3_FDR_range", "FDR",
                min = 0, max = 1, value = c(0, 1), step = 0.01, sep = '', pre = NULL, post = NULL)
  })
  
  #create slider for pvalue f1
  output$c_comparison1_pvalue_slider <- renderUI({
    validate(
      need(input$c_file_pulldown2 != '', "")
    )
    sliderInput("c_compare1_pvalue_range", "pvalue",
                min = 0, max = 1, value = c(0, 1), step = 0.01, sep = '', pre = NULL, post = NULL)
  })
  
  #create slider for pvalue f2
  output$c_comparison2_pvalue_slider <- renderUI({
    validate(
      need(input$c_file_pulldown2 != '', "")
    )
    sliderInput("c_compare2_pvalue_range", "pvalue",
                min = 0, max = 1, value = c(0, 1), step = 0.01, sep = '', pre = NULL, post = NULL)
  })
  
  #create slider for pvalue f3
  output$c_comparison3_pvalue_slider <- renderUI({
    validate(
      need(input$c_file_pulldown3 != '', "")
    )
    sliderInput("c_compare3_pvalue_range", "pvalue",
                min = 0, max = 1, value = c(0, 1), step = 0.01, sep = '', pre = NULL, post = NULL)
  })
  
  #create slide for logFC FILE1
  output$c_comparison_logFC_slider1 <- renderUI({
    validate(
      need(input$c_file_pulldown2 != '', "")
    )
    df <- c_pd1()
    min_logFC <- min(df$logFC)
    min_logFC <- round(min_logFC-0.5, 1)
    max_logFC <- max(df$logFC)
    max_logFC <- round(max_logFC+0.5, 1)
    sliderInput("c_f1_logFC_range", "logFC",
                min = min_logFC, max = max_logFC, value = c(1, max_logFC), step = 0.1)
  })
  
  #create slide for logFC FILE2
  output$c_comparison_logFC_slider2 <- renderUI({
    validate(
      need(input$c_file_pulldown2 != '', "")
    )
    df <- c_pd2()
    min_logFC <- min(df$logFC)
    min_logFC <- round(min_logFC-0.5, 1)
    max_logFC <- max(df$logFC)
    max_logFC <- round(max_logFC+0.5, 1)
    sliderInput("c_f2_logFC_range", "logFC",
                min = min_logFC, max = max_logFC, value = c(1, max_logFC), step = 0.1)
  })
  
  #create slide for logFC FILE3
  output$c_comparison_logFC_slider3 <- renderUI({
    validate(
      need(input$c_file_pulldown3 != '', "")
    )
    df <- c_pd3()
    min_logFC <- min(df$logFC)
    min_logFC <- round(min_logFC-0.5, 1)
    max_logFC <- max(df$logFC)
    max_logFC <- round(max_logFC+0.5, 1)
    sliderInput("c_f3_logFC_range", "logFC",
                min = min_logFC, max = max_logFC, value = c(1, max_logFC), step = 0.1)
  })
  
  output$c_bait_layer <- renderUI({
    textInput("c_bait", "Input HGNC symbol to search for InWeb protein interactors (e.g. ZBTB7A)")
  })
  
  output$c_text_inweb <- renderUI({
    radioButtons('c_marker_text_inweb', 'Turn on/off labels',
                 c(On = 'yes_label',
                   Off = 'no_label'),
                 inline = T
    )
  })
  
  output$c_genes_file <- renderUI({
    fileInput('c_file_genes', 'File containing at least 2 genes with header, one HGNC symbol per line (e.g. TINMAN)',
              accept = c(
                'text/csv',
                'text/comma-separated-values',
                'text/tab-separated-values',
                'text/plain',
                '.csv',
                '.tsv'), multiple = T
    )
  })
  
  # output$c_SNP_file <- renderUI({
  #   fileInput('c_file_SNP', 'File containing list of SNPs, one ID per line (e.g. rs12493885)',
  #             accept = c(
  #               'text/csv',
  #               'text/comma-separated-values',
  #               'text/tab-separated-values',
  #               'text/plain',
  #               '.csv',
  #               '.tsv')
  #   )
  # })
  
  output$c_text_goi <- renderUI({
    radioButtons('c_marker_text_goi', 'Turn on/off labels',
                 c(On = 'yes_label',
                   Off = 'no_label'),
                 inline = T
    )
  })
  
  # output$c_text_snp <- renderUI({
  #   radioButtons('c_marker_text_snp', 'Turn on/off labels',
  #                c(On = 'yes_label',
  #                  Off = 'no_label'),
  #                inline = T
  #   )
  # })
  
  output$c_button_inweb <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_file_pulldown2 != '', "")
    )
    if(any(input$c_bait != "") | !is.null(input$c_bait)){
      actionButton("c_make_plot_inweb", "Generate plots")
    }
  })
  
  # c_bait_in <- reactive({
  #   bait_in <- input$c_bait
  #   if (bait_in == "" ){
  #     return(NULL)
  #   } else{
  #     bait <- bait_in
  #     bait <- toupper(bait)
  #   }
  # })
  # 
  # 
  # output$a_plot_button <- renderUI({
  #   validate(
  #     need(input$a_file_pulldown_r != '', "")
  #   )
  #   all_inputs <- c(input$a_file_SNP_rep, input$a_file_genes_rep, input$a_bait_rep)
  #   
  #   if(any(!is.null(all_inputs)) && any(all_inputs != "")){
  #     actionButton("a_make_plot", "Generate plots")
  #   }
  # })
  
  
  
  output$c_button_goi <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_file_pulldown2 != '', "")
    )
    if(!is.null(c_upload_genes())){
      actionButton("c_make_plot_goi", "Generate plots")
    }
  })
  
  # output$c_button_snp <- renderUI({
  #   validate(
  #     need(input$c_file_pulldown1 != '', ""), 
  #     need(input$c_file_pulldown2 != '', "")
  #   )
  #   if(!is.null(c_snp())){
  #     actionButton("c_make_plot_snp", "Generate plots")
  #   }
  # })
  
  #create slider for FDR
  output$c_pf_FDR_slider1 <- renderUI({
    validate(
      need(input$c_file_pulldown2 != '', "")
    )
    sliderInput("c_f1_pf_FDR", "FDR",
                min = 0, max = 1, value = c(0, 1), step = 0.01, sep = '', pre = NULL, post = NULL)
  })
  
  #create slider for FDR file2
  output$c_pf_FDR_slider2 <- renderUI({
    validate(
      need(input$c_file_pulldown2 != '', "")
    )
    sliderInput("c_f2_pf_FDR", "FDR",
                min = 0, max = 1, value = c(0, 1), step = 0.01, sep = '', pre = NULL, post = NULL)
  })
  
  #create slider for FDR file3
  output$c_pf_FDR_slider3 <- renderUI({
    validate(
      need(input$c_file_pulldown3 != '', "")
    )
    sliderInput("c_f3_pf_FDR", "FDR",
                min = 0, max = 1, value = c(0, 1), step = 0.01, sep = '', pre = NULL, post = NULL)
  })
  
  #create slider for pvalue
  output$c_pf_pvalue_slider1 <- renderUI({
    validate(
      need(input$c_file_pulldown2 != '', "")
    )
    sliderInput("c_f1_pf_pvalue", "pvalue",
                min = 0, max = 1, value = c(0, 1), step = 0.01, sep = '', pre = NULL, post = NULL)
  })
  
  #create slider for pvalue file2
  output$c_pf_pvalue_slider2 <- renderUI({
    validate(
      need(input$c_file_pulldown2 != '', "")
    )
    sliderInput("c_f2_pf_pvalue", "pvalue",
                min = 0, max = 1, value = c(0, 1), step = 0.01, sep = '', pre = NULL, post = NULL)
  })
  
  #create slider for pvalue file3
  output$c_pf_pvalue_slider3 <- renderUI({
    validate(
      need(input$c_file_pulldown3 != '', "")
    )
    sliderInput("c_f3_pf_pvalue", "pvalue",
                min = 0, max = 1, value = c(0, 1), step = 0.01, sep = '', pre = NULL, post = NULL)
  })
  
  #create slide for logFC FILE1
  output$c_pf_logFC_slider1 <- renderUI({
    validate(
      need(input$c_file_pulldown2 != '', "")
    )
    df <- c_pd1()
    min_logFC <- min(df$logFC)
    min_logFC <- round(min_logFC-0.5, 1)
    max_logFC <- max(df$logFC)
    max_logFC <- round(max_logFC+0.5, 1)
    sliderInput("c_f1_pf_logFC_range", "logFC",
                min = min_logFC, max = max_logFC, value = c(1, max_logFC), step = 0.1)
  })
  
  #create slide for logFC FILE2
  output$c_pf_logFC_slider2 <- renderUI({
    validate(
      need(input$c_file_pulldown2 != '', "")
    )
    df <- c_pd2()
    min_logFC <- min(df$logFC)
    min_logFC <- round(min_logFC-0.5, 1)
    max_logFC <- max(df$logFC)
    max_logFC <- round(max_logFC+0.5, 1)
    sliderInput("c_f2_pf_logFC_range", "logFC",
                min = min_logFC, max = max_logFC, value = c(1, max_logFC), step = 0.1)
  })
  
  #create slide for logFC FILE3
  output$c_pf_logFC_slider3 <- renderUI({
    validate(
      need(input$c_file_pulldown3 != '', "")
    )
    df <- c_pd3()
    min_logFC <- min(df$logFC)
    min_logFC <- round(min_logFC-0.5, 1)
    max_logFC <- max(df$logFC)
    max_logFC <- round(max_logFC+0.5, 1)
    sliderInput("c_f3_pf_logFC_range", "logFC",
                min = min_logFC, max = max_logFC, value = c(1, max_logFC), step = 0.1)
  })
  
  output$c_PF_button <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_file_pulldown2 != '', "")
    )
    if(!is.null(c_compare1_pf()) & !is.null(c_compare2_pf())){
      actionButton("c_make_bpf", "Generate PF plot")
    }
  })
  
  output$c_PF_marker_size <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_file_pulldown2 != '', "")
    )
    radioButtons('c_PF_option', 'Turn on/off marker sizing',
                 c(On = 'change_m_PF',
                   Off = 'static_m_PF'),
                 inline = T
    )
  })
  
  c_PF_marker <- reactive({
    inPFmarker <- input$c_PF_option
    if (inPFmarker == "change_m_PF") {
      marker <- "change"
    } else {
      marker <- "no_change"
    }
  })
  
  output$c_PF_sort_col <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_file_pulldown2 != '', "")
    )
    radioButtons('c_PF_sort_option', 'Sort',
                 c("Alphabetically" = 'sort_alph',
                   "PF Frequency" = 'sort_freq'),
                 inline = T
    )
  })
  
  c_PF_sorting <- reactive({
    inPFsort <- input$c_PF_sort_option
    if (inPFsort == "sort_alph") {
      marker <- "sort_a"
    } else {
      marker <- "sort_f"
    }
  })
  
  output$c_PF_freq <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_file_pulldown2 != '', "")
    )
    if(!is.null(input$c_PF_option)){
      inPFmarker <- input$c_PF_option
      if (inPFmarker == "change_m_PF") {
        sliderInput("c_PF_marker_freq", "Marker size",
                    min = 1, max = 40, value = 1, step = 1)     
      }
    }
  })
  
  output$c_pf_loc_selection <- renderUI({
    validate(
      need(input$c_file_pulldown2 != '', "")
    )
    selectInput('c_pf_loc_option', 'Data options', c("Protein family" = 'pf', "Localization_GO" = 'loc_go', "Localization_UniProt" = 'loc_uniprot'), selectize=FALSE)
  })
  
  c_pf_data_selection <- reactive({
    inDselection <- input$c_pf_loc_option
    if (inDselection == "pf") {
      pf_data <- PF_gene_indexed
    } else if (inDselection == "loc_go"){
      pf_data <- go_location_gene_indexed
    } else {
      pf_data <- uniprot_location_gene_indexed
    }
  })
  
  output$c_prot_fam_db <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', "")
    )
    selectInput('c_pfam_db', 'Protein families', colnames(prot_fam), multiple=TRUE, selectize=TRUE)
  })
  
  output$c_color_setting_text <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_colorscheme == "fdr" || input$c_colorscheme == "user", "")
    )
    HTML("<b>Color selection for markers:</b>")
  })
  
  output$c_color_setting_text_inweb <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_file_pulldown2 != '', ""),
      need(input$c_colorscheme == "fdr" || input$c_colorscheme == "user", "")
    )
    HTML("<b>Color selection for markers:</b>")
  })
  
  # output$c_color_setting_text_snp <- renderUI({
  #   validate(
  #     need(input$c_file_pulldown1 != '', ""),
  #     need(input$c_file_pulldown2 != '', ""),
  #     need(input$c_colorscheme == "fdr" || input$c_colorscheme == "user", "")
  #   )
  #   HTML("<b>Color selection for markers:</b>")
  # })
  
  output$c_color_setting_text_goi <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_file_pulldown2 != '', ""),
      need(input$c_colorscheme == "fdr" || input$c_colorscheme == "user", "")
    )
    HTML("<b>Color selection for markers:</b>")
  })
  
  output$c_text_prot_fam_db <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', "")
    )
    radioButtons('c_marker_text_prot_fam_db', 'Turn on/off labels',
                 c(On = 'yes_label',
                   Off = 'no_label'),
                 inline = T
    )
  })
  
  observe({
    if(input$comparison == "3_p2" | input$comparison == "3_p6"){
      shinyjs::hide("c_colorscheme")
      shinyjs::hide("c_fdr_thresh")
      shinyjs::hide("c_pval_thresh")
      shinyjs::hide("c_logfc_thresh_comb")
    } else {
      shinyjs::show("c_colorscheme")
      shinyjs::show("c_fdr_thresh")
      shinyjs::show("c_pval_thresh")
      shinyjs::show("c_logfc_thresh_comb")
    }
  })
  
  c_search_gene <- reactive({
    gene_in <- input$c_goi_search
    if (is.null(gene_in) | gene_in == "" ){
      return(NULL)
    } else{
      gene <- gene_in
      gene <- toupper(gene)
    }
  })
  
  c_bait_in <- reactive({
    bait_in <- input$c_bait
    if (bait_in == "" ){ #is.null(bait_in) | 
      return(NULL)
    } else{
      bait <- bait_in
      bait <- toupper(bait)
    }
  })
  
  c_bait_friends <- reactive({
    if(!is.null(c_bait_in())){
      withProgress(message = 'Finding bait interactors in InWeb', 
                   detail = 'Hold please', value = 0, {
                     bait <- c_bait_in()
                     bait_interactors <- inweb_hash[[bait]]
                     # bait_inweb3 <-
                     #   system(paste("grep -w", bait, "data/InWeb3_interactions.txt | awk -F\"\t\" '{print $1, $2}' | tr ' ' '\n' | sort -u | grep -v", bait, sep = " "), intern = TRUE)
                     # incProgress(0.5)
                     # bait_inweb_im <- 
                     #   system(paste("zgrep -w", bait, "data/core.psimitab.gz | awk -v FS=\"(uniprotkb:|gene name)\" '{print $6, $9}' | sed -e 's/(//g' | tr ' ' '\\n' | sort -u | grep -v", bait, sep = " "), intern = TRUE)
                     incProgress(0.8)
                   })
      # bait_interactors <- c(bait_inweb_im, bait_inweb3)
      # bait_interactors <- unique(sort(bait_interactors))
      bait_interactors
    }
  })
  
  c_upload_genes <- reactive({
    genes <- input$c_file_genes
    if (is.null(genes)){
      return(NULL)
    } else{
      df <- fread(genes$datapath, header = T, fill = T,
                  sep="auto", na.strings=c(""," ","NA"), stringsAsFactors = FALSE, data.table = FALSE, blank.lines.skip = T)
    }
  })
  
  c_genes_uploaded <- reactive({
    if(!is.null(c_upload_genes())){
      genes <- c_upload_genes()
      genes <- as.data.frame(sapply(genes, toupper)) 
      genes
    }
  })
  
  output$c_goi_num_inputs <- renderUI({
    validate(
      need(!is.null(c_vp1_goi_layer()), ""),
      need(!is.null(c_vp2_goi_layer()), "")
    )
    d <- c_pd1()
    gene_interest <- c_genes_uploaded()
    d_g2s <- lapply(gene_interest, function(x) subset(d, gene %in% x) )  
    choices <- names(d_g2s)
    choices <- append(choices, "total")
    list_count <- length(d_g2s)
    if(list_count>0){
      selectInput("c_goi_num_inputs", "GOI list",
                  choices = choices, multiple = F)
    }
  })
  
  # output$snp_num_inputs <- renderUI({
  #   validate(
  #     need(!is.null(c_vp1_snp_layer()), ""),
  #     need(!is.null(c_vp2_snp_layer()), "")
  #   )
  #   print("this is starting")
  #   c1 <- c_snp_vp1()
  #   c2 <- c_snp_vp2()
  #   c_total <- data.frame(rbind(c1$d_snp, c2$d_snp))
  #   i <- sapply(c_total, is.factor)
  #   c_total[i] <- lapply(c_total[i], as.character)
  #   print(str(c_total))
  #   # c_total <- rbind(c_pd1(), c_pd2())
  #   # if(is.null(c_vp3_snp_layer())){
  #   #   print("i am here")
  #   #   c_total <- rbind(c_pd1(), c_pd2())
  #   # } else {
  #   #   print("or here")
  #   #   c_total <- rbind(c_pd1(), c_pd2(), c_pd3())
  #   # }
  #   # snp2gene <- c_SNP_to_gene(c_total)
  #   print("is this working")
  #   # print(snp2gene)
  #   # snp_interest <- split(snp2gene, snp2gene$snpid)
  #   choices <- c_total$snpid
  #   # choices <- names(snp_interest)
  #   choices <- append(choices, "total")
  #   print("these are choices")
  #   print(choices)
  #   list_count <- length(unique(c_total$snpid))
  #   print("list count")
  #   print(list_count)
  #   if(list_count>0){
  #     selectInput("c_snp_num_inputs", "SNP input",
  #                 choices = choices, multiple = F)
  #   }
  # })
  # 
  # c_snp <- reactive({
  #   snp <- input$c_file_SNP
  #   if(is.null(snp)){
  #     return(NULL)
  #   } else{
  #     df <- read.table(snp$datapath, header = FALSE, stringsAsFactors = F)
  #     df
  #   }
  # })
  
  # output$c_SNP_extend <- renderUI({
  #   sliderInput("c_SNP_ext", "Gene extension (±Kb)",
  #               min = 0, max = 100, value = 50, step = 10)
  # })
  
  #snp to gene using LD r^2>0.6±user defined extension
  # c_SNP_to_gene <- function(d){
  #   if(!is.null(c_snp())){
  #     withProgress(message = 'Finding genes in SNPs loci', 
  #                  detail = "Hold please", value = 0, {
  #                    snp_data <- c_snp()
  #                    incProgress(0.2)
  #                    res <- lapply(d$gene, function(s) {c(s, subset(snp_data, V1 %in%genes_snps[[s]]))})
  #                    # Convert each list to a data.table
  #                    dt_list <- map(res, as.data.table)
  #                    # Harness the power of rbind list
  #                    dt <- rbindlist(dt_list, fill = TRUE)
  #                    if(ncol(dt)<2){
  #                      dt$V2 <- NA
  #                    }
  #                    names(dt) <- c("gene", "snpid")
  #                    dt <- dt[!is.na(dt$snpid)]
  #                    incProgress(0.6)
  #                    incProgress(0.8)
  #                    if(nrow(dt) >= 1){
  #                      n_occur <- data.frame(table(dt$snpid))
  #                      incProgress(0.9)
  #                      SNP_n_occur <- merge(dt, n_occur, by.x = 'snpid', by.y = 'Var1')
  #                    } else if(nrow(dt) == 0){
  #                      SNP_n_occur <- data.frame(snpid = character(), gene = character(), Freq = integer())
  #                    }
  #                  })
  #     snpList <- SNP_n_occur
  #   }
  # }
  
  c_inweb_pd1 <- eventReactive(input$c_make_plot_inweb, {
    if(!is.null(c_bait_friends())){
      bait_interactors <- c_bait_friends()
      d <- c_pd1()
      d_subset <- subset(d, gene %in% bait_interactors)
      rownames(d_subset) <- NULL
      d_subset
    }
  })
  
  c_inweb_pd2 <- eventReactive(input$c_make_plot_inweb, {
    if(!is.null(c_bait_friends())){
      bait_interactors <- c_bait_friends()
      d <- c_pd2()
      d_subset <- subset(d, gene %in% bait_interactors)
      rownames(d_subset) <- NULL
      d_subset
    }
  })
  
  c_inweb_pd3 <- eventReactive(input$c_make_plot_inweb, { 
    if(!is.null(c_bait_friends())){
      bait_interactors <- c_bait_friends()
      d <- c_pd3()
      d_subset <- subset(d, gene %in% bait_interactors)
      rownames(d_subset) <- NULL
      d_subset
    }
  })
  
  c_in_pd1 <-reactive({
    if(!is.null(input$c_file_pulldown1)){
      pulldown <- input$c_file_pulldown1
      d <- fread(pulldown$datapath, header = TRUE,
                 sep="auto", na.strings=c(""," ","NA"), stringsAsFactors = FALSE, data.table = FALSE, blank.lines.skip = T)
      d <- na.omit(d)
    }
  })
  
  c_orig_pd1 <- reactive({
    if(!is.null(c_in_pd1())){
      d <- c_in_pd1()
      d_col <- colnames(d)
      if("gene" %in% d_col & "accession_number" %in% d_col){
        df <- d
        df$gene <- toupper(df$gene)
      } else if("accession_number" %in% d_col){
        withProgress(message = 'Mapping UniProt IDs to HGNC symbols',
                     detail = 'Hold please', value = 0, {
                       d$uniprot_new <- vapply(strsplit(d$accession_number,"-"), `[`, 1, FUN.VALUE=character(1))
                       incProgress(0.6)
                       df <- join(d, up_to_hgnc, type="left")
                       df$gene[is.na(df$gene)] <- df$accession_number[is.na(df$gene)]
                       df$all_gene_names[is.na(df$all_gene_names)] <- df$accession_number[is.na(df$all_gene_names)]
                       incProgress(0.8)
                       df <- df[ , !(names(df) %in% c("uniprot_new"))]
                       incProgress(0.9)
                       df
                     })
      } else if("gene" %in% d_col){
        df <- d
        df$gene <- toupper(df$gene)
      }
      df
    }
  })
  
  c_pd1 <- reactive({
    if(!is.null(c_orig_pd1())){
      d <- c_orig_pd1()
      d_col <- colnames(d)
      if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col &
         "rep1" %in% d_col & "rep2" %in% d_col){
        df1 <- d
      }else if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col){
        df1 <- d
      }else if("rep1" %in% d_col & "rep2" %in% d_col){
        df <- d
        if("accession_number" %in% d_col){
          df1 <- calculate_moderated_ttest_uniprot(df)
        }
        else if("accession_number" %!in% d_col){
          df1 <- calculate_moderated_ttest_hgnc(df) 
        }
      }
      df1
    }
  })
  
  c_pd1_converted <- reactive({
    if(!is.null(c_orig_pd1())){
      d <- fread("scripts/gene-tools-master/map/f1_res.txt", header = TRUE,
                 sep="auto", na.strings=c(""," ","NA"), stringsAsFactors = FALSE, data.table = FALSE)
    }
  })
  
  c_in_pd2 <-reactive({
    if(!is.null(input$c_file_pulldown2)){
      pulldown <- input$c_file_pulldown2
      d <- fread(pulldown$datapath, header = TRUE,
                 sep="auto", na.strings=c(""," ","NA"), stringsAsFactors = FALSE, data.table = FALSE, blank.lines.skip = T)
      d <- na.omit(d)
    }
  })
  
  c_orig_pd2 <- reactive({
    if(!is.null(c_in_pd2())){
      d <- c_in_pd2()
      d_col <- colnames(d)
      if("gene" %in% d_col & "accession_number" %in% d_col){
        df <- d
        df$gene <- toupper(df$gene)
      } else if("accession_number" %in% d_col){
        withProgress(message = 'Mapping UniProt IDs to HGNC symbols',
                     detail = 'Hold please', value = 0, {
                       d$uniprot_new <- vapply(strsplit(d$accession_number,"-"), `[`, 1, FUN.VALUE=character(1))
                       incProgress(0.6)
                       df <- join(d, up_to_hgnc, type="left")
                       df$gene[is.na(df$gene)] <- df$accession_number[is.na(df$gene)]
                       df$all_gene_names[is.na(df$all_gene_names)] <- df$accession_number[is.na(df$all_gene_names)]
                       incProgress(0.8)
                       df <- df[ , !(names(df) %in% c("uniprot_new"))]
                       incProgress(0.9)
                       df
                     })
      } else if("gene" %in% d_col){
        df <- d
        df$gene <- toupper(df$gene)
      }
      df
    }
  })
  
  c_pd2 <- reactive({
    if(!is.null(c_orig_pd2())){
      d <- c_orig_pd2()
      d_col <- colnames(d)
      if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col &
         "rep1" %in% d_col & "rep2" %in% d_col){
        df1 <- d
      }else if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col){
        df1 <- d
      }else if("rep1" %in% d_col & "rep2" %in% d_col){
        df <- d
        if("accession_number" %in% d_col){
          df1 <- calculate_moderated_ttest_uniprot(df)
        }
        else if("accession_number" %!in% d_col){
          df1 <- calculate_moderated_ttest_hgnc(df) 
        }
      }
      df1
    }
  })
  
  c_pd2_converted <- reactive({
    if(!is.null(c_orig_pd2())){
      d <- fread("scripts/gene-tools-master/map/f2_res.txt", header = TRUE,
                 sep="auto", na.strings=c(""," ","NA"), stringsAsFactors = FALSE, data.table = FALSE)
    }
  })
  
  c_in_pd3 <-reactive({
    if(!is.null(input$c_file_pulldown3)){
      pulldown <- input$c_file_pulldown3
      d <- fread(pulldown$datapath, header = TRUE,
                 sep="auto", na.strings=c(""," ","NA"), stringsAsFactors = FALSE, data.table = FALSE, blank.lines.skip = T)
      d <- na.omit(d)
    }
  })
  
  c_orig_pd3 <- reactive({
    if(!is.null(c_in_pd3())){
      d <- c_in_pd3()
      d_col <- colnames(d)
      if("gene" %in% d_col & "accession_number" %in% d_col){
        df <- d
        df$gene <- toupper(df$gene)
      } else if("accession_number" %in% d_col){
        withProgress(message = 'Mapping UniProt IDs to HGNC symbols',
                     detail = 'Hold please', value = 0, {
                       d$uniprot_new <- vapply(strsplit(d$accession_number,"-"), `[`, 1, FUN.VALUE=character(1))
                       incProgress(0.6)
                       df <- join(d, up_to_hgnc, type="left")
                       df$gene[is.na(df$gene)] <- df$accession_number[is.na(df$gene)]
                       df$all_gene_names[is.na(df$all_gene_names)] <- df$accession_number[is.na(df$all_gene_names)]
                       incProgress(0.8)
                       df <- df[ , !(names(df) %in% c("uniprot_new"))]
                       incProgress(0.9)
                       df
                     })
      } else if("gene" %in% d_col){
        df <- d
        df$gene <- toupper(df$gene)
      }
      df
    }
  })
  
  c_pd3 <- reactive({
    if(!is.null(c_orig_pd3())){
      d <- c_orig_pd3()
      d_col <- colnames(d)
      if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col &
         "rep1" %in% d_col & "rep2" %in% d_col){
        df1 <- d
      }else if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col){
        df1 <- d
      }else if("rep1" %in% d_col & "rep2" %in% d_col){
        df <- d
        if("accession_number" %in% d_col){
          df1 <- calculate_moderated_ttest_uniprot(df)
        }
        else if("accession_number" %!in% d_col){
          df1 <- calculate_moderated_ttest_hgnc(df) 
        }
      }
      df1
    }
  })
  
  c_pd3_converted <- reactive({
    if(!is.null(c_pd3())){
      d <- fread("scripts/gene-tools-master/map/f3_res.txt", header = TRUE,
                 sep="auto", na.strings=c(""," ","NA"), stringsAsFactors = FALSE, data.table = FALSE)
    }
  })
  
  c_min_x <- reactive({
    if(!is.null(c_in_pd1()) & is.null(c_in_pd2()) & is.null(c_in_pd3())){
      c1 <- c_pd1()
      min_x_axis <- min(c1$logFC)
    } else if(!is.null(c_in_pd1()) & !is.null(c_in_pd2()) & is.null(c_in_pd3())){
      c1 <- c_pd1()
      c2 <- c_pd2()
      min_x_axis <- min(min(c1$logFC), min(c2$logFC))
    } else if(!is.null(c_in_pd1()) & !is.null(c_in_pd2()) & !is.null(c_in_pd3())){
      c1 <- c_pd1()
      c2 <- c_pd2()
      c3 <- c_pd3()
      min_x_axis <- min(min(c1$logFC), min(c2$logFC), min(c3$logFC))
    }
    min_x_axis
  })
  
  c_max_x <- reactive({
    if(!is.null(c_in_pd1()) & is.null(c_in_pd2()) & is.null(c_in_pd3())){
      c1 <- c_pd1()
      max_x_axis <- max(c1$logFC)
    } else if(!is.null(c_in_pd1()) & !is.null(c_in_pd2()) & is.null(c_in_pd3())){
      c1 <- c_pd1()
      c2 <- c_pd2()
      max_x_axis <- max(max(c1$logFC), max(c2$logFC))
    } else if(!is.null(c_in_pd1()) & !is.null(c_in_pd2()) & !is.null(c_in_pd3())){
      c1 <- c_pd1()
      c2 <- c_pd2()
      c3 <- c_pd3()
      max_x_axis <- max(max(c1$logFC), max(c2$logFC), max(c3$logFC))
    }
    max_x_axis
  })
  
  c_min_y <- reactive({
    if(!is.null(c_in_pd1()) & is.null(c_in_pd2()) & is.null(c_in_pd3())){
      c1 <- c_pd1()
      min_y_axis <- min(-log10(c1$pvalue))
    } else if(!is.null(c_in_pd1()) & !is.null(c_in_pd2()) & is.null(c_in_pd3())){
      c1 <- c_pd1()
      c2 <- c_pd2()
      min_y_axis <- min(min(-log10(c1$pvalue)), min(-log10(c2$pvalue)))
    } else if(!is.null(c_in_pd1()) & !is.null(c_in_pd2()) & !is.null(c_in_pd3())){
      c1 <- c_pd1()
      c2 <- c_pd2()
      c3 <- c_pd3()
      min_y_axis <- min(min(-log10(c1$pvalue)), min(-log10(c2$pvalue)), min(-log10(c3$pvalue)))
    }
    min_y_axis
  })
  
  c_max_y <- reactive({
    if(!is.null(c_in_pd1()) & is.null(c_in_pd2()) & is.null(c_in_pd3())){
      c1 <- c_pd1()
      max_y_axis <- max(-log10(c1$pvalue))
    } else if(!is.null(c_in_pd1()) & !is.null(c_in_pd2()) & is.null(c_in_pd3())){
      c1 <- c_pd1()
      c2 <- c_pd2()
      max_y_axis <- max(max(-log10(c1$pvalue)), max(-log10(c2$pvalue)))
    } else if(!is.null(c_in_pd1()) & !is.null(c_in_pd2()) & !is.null(c_in_pd3())){
      c1 <- c_pd1()
      c2 <- c_pd2()
      c3 <- c_pd3()
      max_y_axis <- max(max(-log10(c1$pvalue)), max(-log10(c2$pvalue)), max(-log10(c3$pvalue)))
    }
    max_y_axis
  })
  
  c_vp_gg <- function(d, sig_col, insig_col){
    if(input$c_colorscheme == "fdr"){
      data <- separate_to_groups_for_color_integrated(d, input$c_fdr_thresh, sig_col, insig_col)
      p <- plot_volcano_qc_gg(data)
    } else if(input$c_colorscheme == "exac"){
      d$s <- exac$em_p_hi[match(d$gene, exac$GENE_NAME)]
      d$s[is.na(d$s)] <- 2
      below_thresh <- subset(d, s < 0.9)
      above_thresh <- subset(d, s >= 0.9)
      no_exist <- subset(d, s == 2)
      p <- plot_volcano_exac_gg(below_thresh, above_thresh, no_exist)
    } else if(input$c_colorscheme == "cbf"){
      data <- separate_to_groups_for_cbf_integrated(d, input$c_fdr_thresh)
      p <- plot_volcano_qc_gg(data)
    } else if(input$c_colorscheme == "user"){
      validate(
        need(!is.null(input$c_file_color), "Please upload file with gene and score")
      )
      d1 <- c_in_file_color()
      req(input$c_colorbrewer_theme)
      col_theme <- input$c_colorbrewer_theme
      if(input$c_colorscheme_style == "cont"){
        df <- separate_to_groups_for_color_continuous(d, d1, col_theme)
        data <- df$df1
        data1 <- df$no_exist
      } else if(input$c_colorscheme_style == "disc"){
        df <- separate_to_groups_for_color_discrete(d, d1, col_theme)
        data <- df$df1
        data1 <- df$no_exist
        data2 <- rbind(data, data1)
        p <- ggplot(data = data2, aes(x = logFC, y = -log10(pvalue), text = gene)) +
          geom_point(data = data1, alpha = 0.5, size = 1.5, colour = "#f7f7f7") + 
          geom_point(data = data, alpha = 0.5, size = 1.5, colour = data$col) + 
          xlab("log2FC") + ylab("-log10(P)") +
          theme_minimal() +
          theme(axis.text.x = element_text(size=7),
                axis.title.x=element_text(size=8),
                axis.text.y=element_text(size=7),
                axis.title.y=element_text(size=8),
                panel.grid.minor = element_blank(),
                plot.margin = unit(c(1,1,1,1), "pt"))
      }
      p
    }
    p
  }
  
  c1_vp_gg <- function(){
    data <- c_pd1()
    print(head(data))
    p <- c_vp_gg(data, input$c_color_indv_sig, input$c_color_indv_insig)
    p
  }
  
  c2_vp_gg <- function(){
    data <- c_pd2()
    p <- c_vp_gg(data, input$c_color_indv_sig, input$c_color_indv_insig)
    p
  }
  
  c3_vp_gg <- function(){
    data <- c_pd3()
    p <- c_vp_gg(data, input$c_color_indv_sig, input$c_color_indv_insig)
    p
  }
  
  
  c_vp_plus_rep_gg <- function(p, orig_data){
    validate(
      need(!is.null(c_search_gene()), "")
    )
    goi <- c_search_gene()
    searchgene <- orig_data[grepl(goi,orig_data$gene),]
    p1 <- search_volcano_gg(p, searchgene)
    p1
  }
  
  c1_vp_plus_rep_gg <- function(){
    vp <- c1_vp_gg()
    d <- c_pd1()
    p <- c_vp_plus_rep_gg(vp, d)
    p
  }
  
  c2_vp_plus_rep_gg <- function(){
    vp <- c2_vp_gg()
    d <- c_pd2()
    p <- c_vp_plus_rep_gg(vp, d)
    p
  }
  
  c3_vp_plus_rep_gg <- function(){
    vp <- c3_vp_gg()
    d <- c_pd3()
    p <- c_vp_plus_rep_gg(vp, d)
    p
  }
  
  c_vp_pf_db_gg <- function(p, pf){
    validate(
      need(!is.null(c_pf_db()), "")
    )
    pf_col <- input$c_colorbrewer_theme_pf
    label <- input$c_marker_text_prot_fam_db
    
    if(input$c_colorscheme == "fdr" | input$c_colorscheme == "exac" | input$c_colorscheme == "user"){
      df <- ldply(pf$d_g2s, data.frame)
      if(nrow(df) != 0){
        vp_layer_genes <- vp_layer_for_uploaded_genes_gg(p, df, pf_col, label)
        p1 <- vp_layer_genes
      } else{
        p1 <- p
      }
    } else if(input$c_colorscheme == "cbf"){
      df <- ldply(pf$d_g2s, data.frame)
      if(nrow(df) != 0){
        vp_layer_genes <- vp_layer_for_uploaded_genes_cbf_gg(p, df, label)
        p1 <- vp_layer_genes
      } else{
        p1 <- p
      }
    }
    p1
  }
  
  c1_vp_pf_db_gg <- function(){
    validate(
      need(!is.null(c1_vp_gg()), "")
    )
    vp <- c1_vp_gg()
    pf_db <- c_pf_db_vp1()
    p <- c_vp_pf_db_gg(vp, pf_db)
    p
  }
  
  c2_vp_pf_db_gg <- function(){
    validate(
      need(!is.null(c2_vp_gg()), "")
    )
    vp <- c2_vp_gg()
    pf_db <- c_pf_db_vp2()
    p <- c_vp_pf_db_gg(vp, pf_db)
    p
  }
  
  c3_vp_pf_db_gg <- function(){
    validate(
      need(!is.null(c3_vp_gg()), "")
    )
    vp <- c3_vp_gg()
    pf_db <- c_pf_db_vp3()
    p <- c_vp_pf_db_gg(vp, pf_db)
    p
  }
  
  
  c_vp_pf_db_plus_gg <- function(p, orig_data){
    validate(
      need(!is.null(c_search_gene()), "")
    )
    goi <- c_search_gene()
    searchgene <- orig_data[grepl(goi,orig_data$gene),]
    p1 <- search_volcano_gg(p, searchgene)
    p1
  }
  
  c1_vp_pf_db_plus_gg <- function(){
    vp <- c1_vp_pf_db_gg()
    d <- c_pd1()
    p <- c_vp_pf_db_plus_gg(vp, d)
    p
  }
  
  c2_vp_pf_db_plus_gg <- function(){
    vp <- c2_vp_pf_db_gg()
    d <- c_pd2()
    p <- c_vp_pf_db_plus_gg(vp, d)
    p
  }
  
  c3_vp_pf_db_plus_gg <- function(){
    vp <- c3_vp_pf_db_gg()
    d <- c_pd3()
    p <- c_vp_pf_db_plus_gg(vp, d)
    p
  }
  
  output$download_c_vp1_gg <- downloadHandler(
    filename = function() { paste("vp1_gg", '.png', sep='') },
    content = function(file) {
      if(is.null(c_pf_db())){
        if(is.null(c_search_gene())){
          ggsave(file, plot = c1_vp_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
        } else {
          ggsave(file, plot = c1_vp_plus_rep_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
        }
      } else if(!is.null(c_pf_db())){
        if(is.null(c_search_gene())){
          ggsave(file, plot = c1_vp_pf_db_gg(), device = "png", width = 5, height = 4, units = "in", dpi = 300)
        } else{
          ggsave(file, plot = c1_vp_pf_db_plus_gg(), device = "png", width = 5, height = 4, units = "in", dpi = 300)
        }
      }
    }
  )
  
  output$download_c_vp2_gg <- downloadHandler(
    filename = function() { paste("vp2_gg", '.png', sep='') },
    content = function(file) {
      if(is.null(c_pf_db())){
        if(is.null(c_search_gene())){
          ggsave(file, plot = c2_vp_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
        } else {
          ggsave(file, plot = c2_vp_plus_rep_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
        }
      } else if(!is.null(c_pf_db())){
        if(is.null(c_search_gene())){
          ggsave(file, plot = c2_vp_pf_db_gg(), device = "png", width = 5, height = 4, units = "in", dpi = 300)
        } else{
          ggsave(file, plot = c2_vp_pf_db_plus_gg(), device = "png", width = 5, height = 4, units = "in", dpi = 300)
        }
      }
    }
  )
  
  output$download_c_vp3_gg <- downloadHandler(
    filename = function() { paste("vp3_gg", '.png', sep='') },
    content = function(file) {
      if(is.null(c_pf_db())){
        if(is.null(c_search_gene())){
          ggsave(file, plot = c3_vp_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
        } else {
          ggsave(file, plot = c3_vp_plus_rep_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
        }
      } else if(!is.null(c_pf_db())){
        if(is.null(c_search_gene())){
          ggsave(file, plot = c3_vp_pf_db_gg(), device = "png", width = 5, height = 4, units = "in", dpi = 300)
        } else{
          ggsave(file, plot = c3_vp_pf_db_plus_gg(), device = "png", width = 5, height = 4, units = "in", dpi = 300)
        }
      }
    }
  )
  
  c_vp_colorbar_dl <- reactive({
    validate(
      need(input$c_file_pulldown1 != '', "")
    )
    FDR <- seq(0, 1, 0.01)
    limit <- rep("FDR", 101)
    d <- data.frame(limit, FDR)
    if(input$c_colorscheme == "fdr"){
      req(input$c_color_indv_sig, input$c_color_indv_insig)
      d1 <- separate_to_groups_for_color_integrated(d, input$c_fdr_thresh, input$c_color_indv_sig, input$c_color_indv_insig)
      mycol <- as.vector(d1$col)
      bar <- ggplot(d1, aes(xmin = d1$FDR-0.01, xmax = d1$FDR, ymin = 0, ymax = 0.1)) + geom_rect(fill = mycol) +
        scale_x_continuous(breaks = seq(0, 1, 0.1)) +
        labs(x = "FDR") +
        theme(axis.title.y=element_blank(),
              axis.text.y=element_blank(),
              axis.ticks.y=element_blank(),
              panel.background=element_blank(),
              axis.title = element_text(size = rel(1))) + coord_fixed()
      bar
    } else if(input$c_colorscheme == "exac"){
      d1 <- separate_to_groups_for_exac_bar(d)
      mycol <- as.vector(d1$col)
      bar <- ggplot(d1, aes(xmin = d1$FDR-0.01, xmax = d1$FDR, ymin = 0, ymax = 0.1)) + geom_rect(fill = mycol) +
        labs(x = " pLI < 0.9          pLI >= 0.9       not in ExAC") +
        theme(axis.title.y=element_blank(),
              axis.text.y=element_blank(),
              axis.ticks.y=element_blank(),
              axis.text.x=element_blank(),
              axis.ticks.x=element_blank(),
              panel.background=element_blank(),
              axis.title = element_text(size = rel(1))) + coord_fixed()
      bar
    } else if(input$c_colorscheme == "cbf"){
      d1 <- separate_to_groups_for_cbf_integrated(d, input$c_fdr_thresh)
      mycol <- as.vector(d1$col)
      bar <- ggplot(d1, aes(xmin = d1$FDR-0.01, xmax = d1$FDR, ymin = 0, ymax = 0.1)) + geom_rect(fill = mycol) +
        scale_x_continuous(breaks = seq(0, 1, 0.1)) +
        labs(x = "FDR") +
        theme(axis.title.y=element_blank(),
              axis.text.y=element_blank(),
              axis.ticks.y=element_blank(),
              panel.background=element_blank(),
              axis.title = element_text(size = rel(1))) + coord_fixed()
      bar
    }
  })
  
  c_vp_colorbar_inweb <- reactive({
    FDR <- seq(0, 1, 0.01)
    limit <- rep("FDR", 101)
    d <- data.frame(limit, FDR)
    if(input$c_colorscheme == "fdr"){
      req(input$c_color_inweb_sig, input$c_color_inweb_insig)
      d1 <- separate_to_groups_for_color_integrated(d, input$c_fdr_thresh, input$c_color_inweb_sig, input$c_color_inweb_insig)
      mycol <- as.vector(d1$col)
      bar <- ggplot(d1, aes(xmin = d1$FDR-0.01, xmax = d1$FDR, ymin = 0, ymax = 0.1)) + geom_rect(fill = mycol) +
        scale_x_continuous(breaks = seq(0, 1, 0.1)) +
        labs(x = "FDR") +
        theme(axis.title.y=element_blank(),
              axis.text.y=element_blank(),
              axis.ticks.y=element_blank(),
              panel.background=element_blank(),
              axis.title = element_text(size = rel(1))) + coord_fixed()
      bar
    } else if(input$c_colorscheme == "exac"){
      d1 <- separate_to_groups_for_exac_bar(d)
      mycol <- as.vector(d1$col)
      bar <- ggplot(d1, aes(xmin = d1$FDR-0.01, xmax = d1$FDR, ymin = 0, ymax = 0.1)) + geom_rect(fill = mycol) +
        labs(x = " pLI < 0.9          pLI >= 0.9       not in ExAC") +
        theme(axis.title.y=element_blank(),
              axis.text.y=element_blank(),
              axis.ticks.y=element_blank(),
              axis.text.x=element_blank(),
              axis.ticks.x=element_blank(),
              panel.background=element_blank(),
              axis.title = element_text(size = rel(1))) + coord_fixed()
      bar
    } else if(input$c_colorscheme == "cbf"){
      d1 <- separate_to_groups_for_cbf_integrated(d, input$c_fdr_thresh)
      mycol <- as.vector(d1$col)
      bar <- ggplot(d1, aes(xmin = d1$FDR-0.01, xmax = d1$FDR, ymin = 0, ymax = 0.1)) + geom_rect(fill = mycol) +
        scale_x_continuous(breaks = seq(0, 1, 0.1)) +
        labs(x = "FDR") +
        theme(axis.title.y=element_blank(),
              axis.text.y=element_blank(),
              axis.ticks.y=element_blank(),
              panel.background=element_blank(),
              axis.title = element_text(size = rel(1))) + coord_fixed()
      bar
    }
  })
  
  c_vp_colorbar_goi <- reactive({
    FDR <- seq(0, 1, 0.01)
    limit <- rep("FDR", 101)
    d <- data.frame(limit, FDR)
    if(input$c_colorscheme == "fdr"){
      req(input$c_color_goi_sig, input$c_color_goi_insig)
      d1 <- separate_to_groups_for_color_integrated(d, input$c_fdr_thresh, input$c_color_goi_sig, input$c_color_goi_insig)
      mycol <- as.vector(d1$col)
      bar <- ggplot(d1, aes(xmin = d1$FDR-0.01, xmax = d1$FDR, ymin = 0, ymax = 0.1)) + geom_rect(fill = mycol) +
        scale_x_continuous(breaks = seq(0, 1, 0.1)) +
        labs(x = "FDR") +
        theme(axis.title.y=element_blank(),
              axis.text.y=element_blank(),
              axis.ticks.y=element_blank(),
              panel.background=element_blank(),
              axis.title = element_text(size = rel(1))) + coord_fixed()
      bar
    } else if(input$c_colorscheme == "exac"){
      d1 <- separate_to_groups_for_exac_bar(d)
      mycol <- as.vector(d1$col)
      bar <- ggplot(d1, aes(xmin = d1$FDR-0.01, xmax = d1$FDR, ymin = 0, ymax = 0.1)) + geom_rect(fill = mycol) +
        labs(x = " pLI < 0.9          pLI >= 0.9       not in ExAC") +
        theme(axis.title.y=element_blank(),
              axis.text.y=element_blank(),
              axis.ticks.y=element_blank(),
              axis.text.x=element_blank(),
              axis.ticks.x=element_blank(),
              panel.background=element_blank(),
              axis.title = element_text(size = rel(1))) + coord_fixed()
      bar
    } else if(input$c_colorscheme == "cbf"){
      d1 <- separate_to_groups_for_cbf_integrated(d, input$c_fdr_thresh)
      mycol <- as.vector(d1$col)
      bar <- ggplot(d1, aes(xmin = d1$FDR-0.01, xmax = d1$FDR, ymin = 0, ymax = 0.1)) + geom_rect(fill = mycol) +
        scale_x_continuous(breaks = seq(0, 1, 0.1)) +
        labs(x = "FDR") +
        theme(axis.title.y=element_blank(),
              axis.text.y=element_blank(),
              axis.ticks.y=element_blank(),
              panel.background=element_blank(),
              axis.title = element_text(size = rel(1))) + coord_fixed()
      bar
    }
  })
  
  # c_vp_colorbar_snp <- reactive({
  #   FDR <- seq(0, 1, 0.01)
  #   limit <- rep("FDR", 101)
  #   d <- data.frame(limit, FDR)
  #   if(input$c_colorscheme == "fdr"){
  #     req(input$c_color_snp_sig, input$c_color_snp_insig)
  #     d1 <- separate_to_groups_for_color_integrated(d, input$c_fdr_thresh, input$c_color_snp_sig, input$c_color_snp_insig)
  #     mycol <- as.vector(d1$col)
  #     bar <- ggplot(d1, aes(xmin = d1$FDR-0.01, xmax = d1$FDR, ymin = 0, ymax = 0.1)) + geom_rect(fill = mycol) +
  #       scale_x_continuous(breaks = seq(0, 1, 0.1)) +
  #       labs(x = "FDR") +
  #       theme(axis.title.y=element_blank(),
  #             axis.text.y=element_blank(),
  #             axis.ticks.y=element_blank(),
  #             panel.background=element_blank(),
  #             axis.title = element_text(size = rel(1))) + coord_fixed()
  #     bar
  #   } else if(input$c_colorscheme == "exac"){
  #     d1 <- separate_to_groups_for_exac_bar(d)
  #     mycol <- as.vector(d1$col)
  #     bar <- ggplot(d1, aes(xmin = d1$FDR-0.01, xmax = d1$FDR, ymin = 0, ymax = 0.1)) + geom_rect(fill = mycol) +
  #       labs(x = " pLI < 0.9          pLI >= 0.9       not in ExAC") +
  #       theme(axis.title.y=element_blank(),
  #             axis.text.y=element_blank(),
  #             axis.ticks.y=element_blank(),
  #             axis.text.x=element_blank(),
  #             axis.ticks.x=element_blank(),
  #             panel.background=element_blank(),
  #             axis.title = element_text(size = rel(1))) + coord_fixed()
  #     bar
  #   } else if(input$c_colorscheme == "cbf"){
  #     d1 <- separate_to_groups_for_cbf_integrated(d, input$c_fdr_thresh)
  #     mycol <- as.vector(d1$col)
  #     bar <- ggplot(d1, aes(xmin = d1$FDR-0.01, xmax = d1$FDR, ymin = 0, ymax = 0.1)) + geom_rect(fill = mycol) +
  #       scale_x_continuous(breaks = seq(0, 1, 0.1)) +
  #       labs(x = "FDR") +
  #       theme(axis.title.y=element_blank(),
  #             axis.text.y=element_blank(),
  #             axis.ticks.y=element_blank(),
  #             panel.background=element_blank(),
  #             axis.title = element_text(size = rel(1))) + coord_fixed()
  #     bar
  #   }
  # })
  
  c_vp <- function(d){
    if(input$c_colorscheme == "fdr"){
      req(input$c_color_indv_sig, input$c_color_indv_insig)
      data <- separate_to_groups_for_color_integrated(d, input$c_fdr_thresh, input$c_color_indv_sig, input$c_color_indv_insig)
      p <- plot_volcano_multiple_cond(data)
    } else if(input$c_colorscheme == "exac"){
      d$s <- exac$em_p_hi[match(d$gene, exac$GENE_NAME)]
      d$s[is.na(d$s)] <- 2
      below_thresh <- subset(d, s < 0.9)
      above_thresh <- subset(d, s >= 0.9)
      no_exist <- subset(d, s == 2)
      p <- plot_volcano_exac_multi(below_thresh, above_thresh, no_exist)
      p
    } else if(input$c_colorscheme == "cbf"){
      data <- separate_to_groups_for_cbf_integrated(d, input$c_fdr_thresh)
      p <- plot_volcano_multiple_cond(data)
    } else if(input$c_colorscheme == "user"){
      validate(
        need(!is.null(input$c_file_color), "Please upload file with gene and score")
      )
      d1 <- c_in_file_color()
      col_theme <- input$c_colorbrewer_theme
      if(input$c_colorscheme_style == "cont"){
        df <- separate_to_groups_for_color_continuous(d, d1, col_theme)
        data <- df$df1
        data1 <- df$no_exist
      } else if(input$c_colorscheme_style == "disc"){
        df <- separate_to_groups_for_color_discrete(d, d1, col_theme)
        data <- df$df1
        data1 <- df$no_exist
      }
      p <- plot_ly(showlegend = T, width = 320, height = 390)
      p <- add_markers(p, data = data1, x = ~logFC, y = ~-log10(pvalue),
                       marker = list(size = 7, line = list(width=0.1, color = "grey89"), cmin = 0, cmax = 1, color = "#f7f7f7"),
                       opacity = 0.8,
                       text = ~paste(gene), hoverinfo = "text", name = paste0("Not in user data (", nrow(df$no_exist), ")"))
      for(i in nrow(data)){
        p <- add_markers(p, data = data, x = ~logFC, y = ~-log10(pvalue),
                         marker = list(size = 7, cmin = 0, cmax = 1, color = ~col, line = list(width=0.2, color = "grey89")),
                         opacity = 1,
                         text = ~paste0(gene, ", FDR=", signif(FDR, digits = 3)), hoverinfo = "text",
                         name = paste0("Found in user data (", nrow(data), ")")) #, name = "pull down"
      }
      p
    }
    min_x <- c_min_x()
    min_y <- c_min_y()
    max_x <- c_max_x()
    max_y <- c_max_y()
    p <- p %>%
      layout(xaxis = list(range=c(min_x-0.5, max_x+0.5), showgrid = F), yaxis = list(range=c(min_y-0.5, max_y+0.5), showgrid = F), 
             legend = list(orientation = 'h', y = -0.23)) %>%
      add_lines(x = c(min_x-0.5, max_x+0.5), y = -log10(input$c_pval_thresh), line = list(dash = "dash", width = 0.5, color = "#2b333e"), 
                name = '', hoverinfo = "text", text = paste0("pvalue = ", input$c_pval_thresh), showlegend = F) %>%
      add_lines(x = input$c_logfc_thresh_comb[1], y = c(min_y-0.5, max_y+0.5), line = list(dash = "dash", width = 0.5, color = "#252525"), 
                name = '', hoverinfo = "text", text = paste0("logFC = ", input$c_logfc_thresh_comb[1]), showlegend = F) %>%
      add_lines(x = input$c_logfc_thresh_comb[2], y = c(min_y-0.5, max_y+0.5), line = list(dash = "dash", width = 0.5, color = "#252525"), 
                name = '', hoverinfo = "text", text = paste0("logFC = ", input$c_logfc_thresh_comb[2]), showlegend = F)
  }
  
  c_vp1 <- reactive({
    p <- c_vp(c_pd1())
  })
  
  c_vp1_plus <- reactive({
    validate(
      need(!is.null(c_search_gene()), "")
    )
    p <- c_vp1()
    goi <- c_search_gene()
    orig_data <- c_pd1()
    searchgene <- orig_data[grepl(goi,orig_data$gene),]
    p1 <- search_volcano(p, searchgene)
    p1
  })
  
  c_vp1_inweb <- eventReactive(input$c_make_plot_inweb, {
    validate(
      need(!is.null(c_pd1()), "")
    )
    d <- c_pd1()
    if(!is.null(c_inweb_pd1())){
      inwebFile <- c_inweb_pd1()
      bait <- c_bait_in()
      d_in <- subset(d, gene %in% inwebFile$gene & gene != bait)
      list(d_in=d_in)
    }
  })
  
  c_vp_inweb_layer <- function(d, inweb_d, vp_title){
    validate(
      need(!is.null(c_bait_in()), "")
    )
    min_x <- c_min_x()
    min_y <- c_min_y()
    max_x <- c_max_x()
    max_y <- c_max_y()
    if(input$c_colorscheme == "fdr"){
      req(input$c_color_inweb_sig, input$c_color_inweb_insig)
      data <- separate_to_groups_for_color_integrated(d, input$c_fdr_thresh, input$c_color_inweb_sig, input$c_color_inweb_insig)
      p <- plot_volcano_multiple_cond(data)
    } else if(input$c_colorscheme == "exac"){
      d$s <- exac$em_p_hi[match(d$gene, exac$GENE_NAME)]
      d$s[is.na(d$s)] <- 2
      below_thresh <- subset(d, s < 0.9)
      above_thresh <- subset(d, s >= 0.9)
      no_exist <- subset(d, s == 2)
      p <- plot_volcano_exac_multi(below_thresh, above_thresh, no_exist)
      p
    } else if(input$c_colorscheme == "cbf"){
      data <- separate_to_groups_for_cbf_integrated(d, input$c_fdr_thresh)
      p <- plot_volcano_multiple_cond(data)
    } else if(input$c_colorscheme == "user"){
      validate(
        need(!is.null(input$c_file_color), "Please upload file with gene and score")
      )
      d1 <- c_in_file_color()
      col_theme <- input$c_colorbrewer_theme_tab3
      col_type <- input$c_colorscheme_style
      df <- separate_to_groups_for_color(d, d1, col_theme, col_type)
      data <- df$df1
      data1 <- df$no_exist
      p <- plot_ly(showlegend = T, width = 320, height = 390)
      if(nrow(data1)>0){
        p <- add_markers(p, data = data1, x = ~logFC, y = ~-log10(pvalue),
                         marker = list(size = 7, line = list(width=0.1, color = "grey89"), cmin = 0, cmax = 1, color = "#f7f7f7"),
                         opacity = 0.8,
                         text = ~paste(gene), hoverinfo = "text", name = paste0("Not in user data (", nrow(df$no_exist), ")"))
      }
      if(nrow(data)>0){
        for(i in nrow(data)){
          p <- add_markers(p, data = data, x = ~logFC, y = ~-log10(pvalue),
                           marker = list(size = 7, cmin = 0, cmax = 1, color = ~col, line = list(width=0.2, color = "grey89")),
                           opacity = 1,
                           text = ~paste0(gene, ", FDR=", signif(FDR, digits = 3)), hoverinfo = "text",
                           name = paste0("Found in user data (", nrow(data), ")")) #, name = "pull down"
        }
      }
      p
    }
    p <- p %>%
      layout(xaxis = list(range=c(min_x-0.5, max_x+0.5), showgrid = F), yaxis = list(range=c(min_y-0.5, max_y+0.5), showgrid = F),
             title = paste0("p-value = ", vp_title), titlefont = list(size=15), legend = list(orientation = 'h', y = -0.23))
    label <- input$c_marker_text_inweb
    if(input$c_colorscheme == "fdr" | input$c_colorscheme == "exac" | input$c_colorscheme == "user"){
      if(!is.null(c_bait_in())){
        col <- input$c_color_inweb
        p <- vp_layer_for_inweb(p, inweb_d$d_in, col, label)
      }
    } else if(input$c_colorscheme == "cbf"){
      if(!is.null(c_bait_in())){
        p <- vp_layer_for_inweb_cbf(p, inweb_d$d_in, label)
      }
    }
    p <- p %>% 
      layout(xaxis = list(range=c(min_x-0.5, max_x+0.5), showgrid = F), yaxis = list(range=c(min_y-0.5, max_y+0.5), showgrid = F)) %>%
      add_lines(x = c(min_x-0.5, max_x+0.5), y = -log10(input$c_pval_thresh), line = list(dash = "dash", width = 0.5, color = "#2b333e"), 
                name = '', hoverinfo = "text", text = paste0("pvalue = ", input$c_pval_thresh), showlegend = F) %>%
      add_lines(x = input$c_logfc_thresh_comb[1], y = c(min_y-0.5, max_y+0.5), line = list(dash = "dash", width = 0.5, color = "#252525"), 
                name = '', hoverinfo = "text", text = paste0("logFC = ", input$c_logfc_thresh_comb[1]), showlegend = F) %>%
      add_lines(x = input$c_logfc_thresh_comb[2], y = c(min_y-0.5, max_y+0.5), line = list(dash = "dash", width = 0.5, color = "#252525"), 
                name = '', hoverinfo = "text", text = paste0("logFC = ", input$c_logfc_thresh_comb[2]), showlegend = F)
  }
  
  c_vp1_inweb_layer <- reactive({
    validate(
      need(!is.null(c_pd1()), "")
    )
    data <- c_pd1()
    c1_multi_vp <- c_vp1_inweb()
    vp_title <- c_inweb_hypergeometric(data, c_inweb_pd1())
    p <- c_vp_inweb_layer(data, c1_multi_vp, vp_title)
  })
  
  c_vp1_inweb_plus <- reactive({
    validate(
      need(!is.null(c_search_gene()), "")
    )
    p <- c_vp1_inweb_layer()
    goi <- c_search_gene()
    orig_data <- c_pd1()
    searchgene <- orig_data[grepl(goi,orig_data$gene),]
    p1 <- search_volcano(p, searchgene)
    p1
  })
  
  c_goi_vp1 <- eventReactive(input$c_make_plot_goi, {
    if(!is.null(c_upload_genes())){
      d <- c_pd1()
      gene_interest <- c_genes_uploaded()
      d_g2s <- lapply(gene_interest, function(x) subset(d, gene %in% x) )
      list(d_g2s=d_g2s)
    }
  })
  
  c_vp1_goi_layer <- reactive({
    validate(
      need(!is.null(c_pd1()), "")
    )
    d <- c_pd1()
    c1_goi <- c_goi_vp1()
    min_x <- c_min_x()
    min_y <- c_min_y()
    max_x <- c_max_x()
    max_y <- c_max_y()
    col_goi <- input$c_colorbrewer_theme_goi
    # vp_title <- c_pd3_inweb_hypergeometric()
    if(input$c_colorscheme == "fdr"){
      req(input$c_color_goi_sig, input$c_color_goi_insig)
      data <- separate_to_groups_for_color_integrated(d, input$c_fdr_thresh, input$c_color_goi_sig, input$c_color_goi_insig)
      p <- plot_ly(colors = col_goi, showlegend = T, width = 300, height = 390)
      for(i in nrow(data)){
        p <- add_markers(p, data = data, x = ~logFC, y = ~-log10(pvalue),
                         marker = list(size = 6, cmin = 0, cmax = 1, color = ~col), 
                         opacity = 0.8, 
                         text = ~paste(gene), hoverinfo = "text", name = "pull down")
      }
      p 
    } else if(input$c_colorscheme == "exac"){
      d$s <- exac$em_p_hi[match(d$gene, exac$GENE_NAME)]
      d$s[is.na(d$s)] <- 2
      below_thresh <- subset(d, s < 0.9)
      above_thresh <- subset(d, s >= 0.9)
      no_exist <- subset(d, s == 2)
      p <- plot_ly(colors = col_goi, showlegend = T, width = 300, height = 390)
      p <- add_markers(p, data = below_thresh, x = ~logFC, y = ~-log10(pvalue),
                       marker = list(size = 8, line = list(width=0.1, color = 'black'), cmin = 0, cmax = 1, color = "#66c2a5"),
                       opacity = 0.8, 
                       text = ~paste(gene), hoverinfo = "text", name = paste0("pLI<0.9 (", nrow(below_thresh), ")"))
      p <- add_markers(p, data = above_thresh, x = ~logFC, y = ~-log10(pvalue),
                       marker = list(size = 8, line = list(width=0.1, color = "black"), cmin = 0, cmax = 1, color = "#fc8d62"),
                       opacity = 0.8, 
                       text = ~paste(gene), hoverinfo = "text", name = paste0("pLI>=0.9 (", nrow(above_thresh), ")"))
      p <- add_markers(p, data = no_exist, x = ~logFC, y = ~-log10(pvalue),
                       marker = list(size = 8, line = list(width=0.1, color = "black"), cmin = 0, cmax = 1, color = "#8da0cb"),
                       opacity = 0.8, 
                       text = ~paste(gene), hoverinfo = "text", name = paste0("not in ExAC (", nrow(no_exist), ")"))
      p 
    } else if(input$c_colorscheme == "cbf"){
      data <- separate_to_groups_for_cbf_integrated(d, input$c_fdr_thresh)
      p <- plot_ly(colors = "Greys", showlegend = T, width = 300, height = 390)
      for(i in nrow(data)){
        p <- add_markers(p, data = data, x = ~logFC, y = ~-log10(pvalue),
                         marker = list(size = 6, cmin = 0, cmax = 1, color = ~col), 
                         opacity = 0.6, 
                         text = ~paste(gene), hoverinfo = "text", name = "pull down")
      }
      p 
    } else if(input$c_colorscheme == "user"){
      validate(
        need(!is.null(input$c_file_color), "Please upload file with gene and score")
      )
      d1 <- c_in_file_color()
      col_theme <- input$c_colorbrewer_theme_tab4
      if(input$c_colorscheme_style == "cont"){
        df <- separate_to_groups_for_color_continuous(d, d1, col_theme)
        data <- df$df1
        data1 <- df$no_exist
      } else if(input$c_colorscheme_style == "disc"){
        df <- separate_to_groups_for_color_discrete(d, d1, col_theme)
        data <- df$df1
        data1 <- df$no_exist
      }
      p <- plot_ly(showlegend = T, width = 320, height = 390)
      if(nrow(data1)>0){
        p <- add_markers(p, data = data1, x = ~logFC, y = ~-log10(pvalue),
                         marker = list(size = 7, line = list(width=0.1, color = "grey89"), cmin = 0, cmax = 1, color = "#f7f7f7"),
                         opacity = 0.8,
                         text = ~paste(gene), hoverinfo = "text", name = paste0("Not in user data (", nrow(df$no_exist), ")"))
      }
      if(nrow(data)>0){
        for(i in nrow(data)){
          p <- add_markers(p, data = data, x = ~logFC, y = ~-log10(pvalue),
                           marker = list(size = 7, cmin = 0, cmax = 1, color = ~col, line = list(width=0.2, color = "grey89")),
                           opacity = 1,
                           text = ~paste0(gene, ", FDR=", signif(FDR, digits = 3)), hoverinfo = "text",
                           name = paste0("Found in user data (", nrow(data), ")")) #, name = "pull down"
        }
      }
      p
    }
    p <- p %>%
      layout(xaxis = list(range=c(min_x-0.5, max_x+0.5), showgrid = F), yaxis = list(range=c(min_y-0.5, max_y+0.5), showgrid = F),
             legend = list(orientation = 'h', y = -0.23))
    # title = paste0("p-value = ", vp_title), titlefont = list(size=15))
    if(input$c_colorscheme == "fdr" | input$c_colorscheme == "exac" | input$c_colorscheme == "user"){
      if(!is.null(c_upload_genes())){
        df <- ldply(c1_goi$d_g2s, data.frame)
        if(nrow(df) != 0){
          if(input$c_marker_text_goi == "yes_label"){
            vp_layer_genes <- vp_layer_for_uploaded_genes(p, df)
          } else if(input$c_marker_text_goi == "no_label"){
            vp_layer_genes <- vp_layer_for_uploaded_genes_no_text(p, df)
          }
          p <- vp_layer_genes
        } else{
          vp_layer_no_genes <- vp_layer_for_uploaded_genes_none(p, d)
          p <- vp_layer_no_genes
        }
      }
    } else if(input$c_colorscheme == "cbf"){
      if(!is.null(c_upload_genes())){
        df <- ldply(c1_goi$d_g2s, data.frame)
        if(nrow(df) != 0){
          if(input$c_marker_text_goi == "yes_label"){
            vp_layer_genes <- vp_layer_for_uploaded_genes_cbf(p, df)
          } else if(input$c_marker_text_goi == "no_label"){
            vp_layer_genes <- vp_layer_for_uploaded_genes_cbf_no_text(p, df)
          }
          p <- vp_layer_genes
        } #else{
        #   vp_layer_no_genes <- vp_layer_for_uploaded_genes_none_cbf(p, d)
        #   p <- vp_layer_no_genes
        # }
      }
    }
    p <- p %>% 
      layout(xaxis = list(range=c(min_x-0.5, max_x+0.5), showgrid = F), yaxis = list(range=c(min_y-0.5, max_y+0.5), showgrid = F)) %>%
      add_lines(x = c(min_x-0.5, max_x+0.5), y = -log10(input$c_pval_thresh), line = list(dash = "dash", width = 0.5, color = "#2b333e"), 
                name = '', hoverinfo = "text", text = paste0("pvalue = ", input$c_pval_thresh), showlegend = F) %>%
      add_lines(x = input$c_logfc_thresh_comb[1], y = c(min_y-0.5, max_y+0.5), line = list(dash = "dash", width = 0.5, color = "#252525"), 
                name = '', hoverinfo = "text", text = paste0("logFC = ", input$c_logfc_thresh_comb[1]), showlegend = F) %>%
      add_lines(x = input$c_logfc_thresh_comb[2], y = c(min_y-0.5, max_y+0.5), line = list(dash = "dash", width = 0.5, color = "#252525"), 
                name = '', hoverinfo = "text", text = paste0("logFC = ", input$c_logfc_thresh_comb[2]), showlegend = F)
  })
  
  c_vp1_goi_plus <- reactive({
    validate(
      need(!is.null(c_search_gene()), "")
    )
    p <- c_vp1_goi_layer()
    goi <- c_search_gene()
    orig_data <- c_pd1()
    searchgene <- orig_data[grepl(goi,orig_data$gene),]
    p1 <- search_volcano(p, searchgene)
    p1
  })
  
  # c_snp_vp1 <- eventReactive(input$c_make_plot_snp, {
  #   if(!is.null(c_snp())){
  #     d <- c_pd1()
  #     snp_interest <- c_SNP_to_gene(d)
  #     d_snp <- merge(d, snp_interest, by.x = 'gene', by.y = 'gene')
  #     list(d_snp=d_snp)
  #   }
  # })
  
  # c_vp1_snp_layer <- reactive({
  #   validate(
  #     need(!is.null(c_pd1()), "")
  #   )
  #   d <- c_pd1()
  #   c1_snp <- c_snp_vp1()
  #   min_x <- c_min_x()
  #   min_y <- c_min_y()
  #   max_x <- c_max_x()
  #   max_y <- c_max_y()
  #   # vp_title <- c_pd3_inweb_hypergeometric()
  #   if(input$c_colorscheme == "fdr"){
  #     req(input$c_color_snp_sig, input$c_color_snp_insig)
  #     data <- separate_to_groups_for_color_integrated(d, input$c_fdr_thresh, input$c_color_snp_sig, input$c_color_snp_insig)
  #     p <- plot_volcano_multiple_cond(data)
  #   } else if(input$c_colorscheme == "exac"){
  #     d$s <- exac$em_p_hi[match(d$gene, exac$GENE_NAME)]
  #     d$s[is.na(d$s)] <- 2
  #     below_thresh <- subset(d, s < 0.9)
  #     above_thresh <- subset(d, s >= 0.9)
  #     no_exist <- subset(d, s == 2)
  #     p <- plot_volcano_exac_multi(below_thresh, above_thresh, no_exist)
  #     p
  #   } else if(input$c_colorscheme == "cbf"){
  #     data <- separate_to_groups_for_cbf_integrated(d, input$c_fdr_thresh)
  #     p <- plot_volcano_multiple_cond(data)
  #   } else if(input$c_colorscheme == "user"){
  #     validate(
  #       need(!is.null(input$c_file_color), "Please upload file with gene and score")
  #     )
  #     d1 <- c_in_file_color()
  #     col_theme <- input$c_colorbrewer_theme_tab5
  #     if(input$c_colorscheme_style == "cont"){
  #       df <- separate_to_groups_for_color_continuous(d, d1, col_theme)
  #       data <- df$df1
  #       data1 <- df$no_exist
  #     } else if(input$c_colorscheme_style == "disc"){
  #       df <- separate_to_groups_for_color_discrete(d, d1, col_theme)
  #       data <- df$df1
  #       data1 <- df$no_exist
  #     }
  #     p <- plot_ly(showlegend = T, width = 320, height = 390)
  #     if(nrow(data1)>0){
  #       p <- add_markers(p, data = data1, x = ~logFC, y = ~-log10(pvalue),
  #                        marker = list(size = 7, line = list(width=0.1, color = "grey89"), cmin = 0, cmax = 1, color = "#f7f7f7"),
  #                        opacity = 0.8,
  #                        text = ~paste(gene), hoverinfo = "text", name = paste0("Not in user data (", nrow(df$no_exist), ")"))
  #     }
  #     if(nrow(data)>0){
  #       for(i in nrow(data)){
  #         p <- add_markers(p, data = data, x = ~logFC, y = ~-log10(pvalue),
  #                          marker = list(size = 7, cmin = 0, cmax = 1, color = ~col, line = list(width=0.2, color = "grey89")),
  #                          opacity = 1,
  #                          text = ~paste0(gene, ", FDR=", signif(FDR, digits = 3)), hoverinfo = "text",
  #                          name = paste0("Found in user data (", nrow(data), ")")) #, name = "pull down"
  #       }
  #     }
  #     p
  #   }
  #   p <- p%>%
  #     layout(xaxis = list(range=c(min_x-0.5, max_x+0.5), showgrid = F), yaxis = list(range=c(min_y-0.5, max_y+0.5), showgrid = F),
  #            legend = list(orientation = 'h', y = -0.23))
  #   #title = paste0("p-value = ", vp_title), titlefont = list(size=15))
  #   if(input$c_colorscheme == "fdr" | input$c_colorscheme == "exac" | input$c_colorscheme == "user"){
  #     if(!is.null(c_snp())){
  #       if(nrow(c1_snp$d_snp) != 0){
  #         snp_sgl <- subset(c1_snp$d_snp, Freq == 1)
  #         snp_mgl <- subset(c1_snp$d_snp, Freq != 1)
  #         col <- c(color = colorRampPalette(brewer.pal(3, input$c_colorbrewer_theme_snp))(2))
  #         if(nrow(snp_sgl) != 0){
  #           if(input$c_marker_text_snp == "yes_label"){
  #             vp_layer_snp2gene_sgl <- vp_layer_for_snp_to_gene_sgl(p, snp_sgl, col[1])
  #           } else if(input$c_marker_text_snp == "no_label"){
  #             vp_layer_snp2gene_sgl <- vp_layer_for_snp_to_gene_sgl_no_text(p, snp_sgl, col[1])
  #           }
  #           p <- vp_layer_snp2gene_sgl
  #         }
  #         if(nrow(snp_mgl) != 0){
  #           if(input$c_marker_text_snp == "yes_label"){
  #             vp_layer_snp2gene_mgl <- vp_layer_for_snp_to_gene_mgl(p, snp_mgl, col[2])
  #           } else if(input$c_marker_text_snp == "no_label"){
  #             vp_layer_snp2gene_mgl <- vp_layer_for_snp_to_gene_mgl_no_text(p, snp_mgl, col[2])
  #           }
  #           p <- vp_layer_snp2gene_mgl
  #         }
  #       } else{
  #         vp_layer_no_snp2gene <- vp_layer_for_snp_to_gene_none(p, d)
  #         p <- vp_layer_no_snp2gene
  #       }
  #     }
  #   } else if(input$c_colorscheme == "cbf"){
  #     if(!is.null(c_snp())){
  #       if(nrow(c1_snp$d_snp) != 0){
  #         snp_sgl <- subset(c1_snp$d_snp, Freq == 1)
  #         snp_mgl <- subset(c1_snp$d_snp, Freq != 1)
  #         if(nrow(snp_sgl) != 0){
  #           if(input$c_marker_text_snp == "yes_label"){
  #             vp_layer_snp2gene_sgl <- vp_layer_for_snp_to_gene_sgl_cbf(p, snp_sgl)
  #           } else if(input$c_marker_text_snp == "no_label"){
  #             vp_layer_snp2gene_sgl <- vp_layer_for_snp_to_gene_sgl_cbf_no_text(p, snp_sgl)
  #           }
  #           p <- vp_layer_snp2gene_sgl
  #         }
  #         if(nrow(snp_mgl) != 0){
  #           if(input$c_marker_text_snp == "yes_label"){
  #             vp_layer_snp2gene_mgl <- vp_layer_for_snp_to_gene_mgl_cbf(p, snp_mgl)
  #           } else if(input$c_marker_text_snp == "no_label"){
  #             vp_layer_snp2gene_mgl <- vp_layer_for_snp_to_gene_mgl_cbf_no_text(p, snp_mgl)
  #           }
  #           p <- vp_layer_snp2gene_mgl
  #         }
  #       } else{
  #         vp_layer_no_snp2gene <- vp_layer_for_snp_to_gene_none_cbf(p, d)
  #         p <- vp_layer_no_snp2gene
  #       }
  #     }
  #   }
  #   p <- p %>%
  #     layout(xaxis = list(range=c(min_x-0.5, max_x+0.5), showgrid = F), yaxis = list(range=c(min_y-0.5, max_y+0.5), showgrid = F)) %>%
  #     add_lines(x = c(min_x-0.5, max_x+0.5), y = -log10(input$c_pval_thresh), line = list(dash = "dash", width = 0.5, color = "#2b333e"),
  #               name = '', hoverinfo = "text", text = paste0("pvalue = ", input$c_pval_thresh), showlegend = F) %>%
  #     add_lines(x = input$c_logfc_thresh_comb[1], y = c(min_y-0.5, max_y+0.5), line = list(dash = "dash", width = 0.5, color = "#252525"),
  #               name = '', hoverinfo = "text", text = paste0("logFC = ", input$c_logfc_thresh_comb[1]), showlegend = F) %>%
  #     add_lines(x = input$c_logfc_thresh_comb[2], y = c(min_y-0.5, max_y+0.5), line = list(dash = "dash", width = 0.5, color = "#252525"),
  #               name = '', hoverinfo = "text", text = paste0("logFC = ", input$c_logfc_thresh_comb[2]), showlegend = F)
  # })
  # 
  # c_vp1_snp_plus <- reactive({
  #   validate(
  #     need(!is.null(c_search_gene()), "")
  #   )
  #   p <- c_vp1_snp_layer()
  #   goi <- c_search_gene()
  #   orig_data <- c_pd1()
  #   searchgene <- orig_data[grepl(goi,orig_data$gene),]
  #   p1 <- search_volcano(p, searchgene)
  #   p1
  # })
  
  c_vp2 <- reactive({
    d <- c_pd2()
    if(input$c_colorscheme == "fdr"){
      req(input$c_color_indv_sig, input$c_color_indv_insig)
      data <- separate_to_groups_for_color_integrated(d, input$c_fdr_thresh, input$c_color_indv_sig, input$c_color_indv_insig)
      p <- plot_volcano_multiple_cond(data)
    } else if(input$c_colorscheme == "exac"){
      d$s <- exac$em_p_hi[match(d$gene, exac$GENE_NAME)]
      d$s[is.na(d$s)] <- 2
      below_thresh <- subset(d, s < 0.9)
      above_thresh <- subset(d, s >= 0.9)
      no_exist <- subset(d, s == 2)
      p <- plot_volcano_exac_multi(below_thresh, above_thresh, no_exist)
    } else if(input$c_colorscheme == "cbf"){
      data <- separate_to_groups_for_cbf_integrated(d, input$c_fdr_thresh)
      p <- plot_volcano_multiple_cond(data)
    } else if(input$c_colorscheme == "user"){
      validate(
        need(!is.null(input$c_file_color), "Please upload file with gene and score")
      )
      d1 <- c_in_file_color()
      col_theme <- input$c_colorbrewer_theme
      if(input$c_colorscheme_style == "cont"){
        df <- separate_to_groups_for_color_continuous(d, d1, col_theme)
        data <- df$df1
        data1 <- df$no_exist
      } else if(input$c_colorscheme_style == "disc"){
        df <- separate_to_groups_for_color_discrete(d, d1, col_theme)
        data <- df$df1
        data1 <- df$no_exist
      }
      p <- plot_ly(showlegend = T, width = 320, height = 390)
      p <- add_markers(p, data = data1, x = ~logFC, y = ~-log10(pvalue),
                       marker = list(size = 7, line = list(width=0.1, color = "grey89"), cmin = 0, cmax = 1, color = "#f7f7f7"),
                       opacity = 0.8,
                       text = ~paste(gene), hoverinfo = "text", name = paste0("Not in user data (", nrow(df$no_exist), ")"))
      for(i in nrow(data)){
        p <- add_markers(p, data = data, x = ~logFC, y = ~-log10(pvalue),
                         marker = list(size = 7, cmin = 0, cmax = 1, color = ~col, line = list(width=0.2, color = "grey89")),
                         opacity = 1,
                         text = ~paste0(gene, ", FDR=", signif(FDR, digits = 3)), hoverinfo = "text",
                         name = paste0("Found in user data (", nrow(data), ")")) #, name = "pull down"
      }
      p
    }
    min_x <- c_min_x()
    min_y <- c_min_y()
    max_x <- c_max_x()
    max_y <- c_max_y()
    p <- p %>%
      layout(xaxis = list(range=c(min_x-0.5, max_x+0.5), showgrid = F), yaxis = list(range=c(min_y-0.5, max_y+0.5), showgrid = F), 
             legend = list(orientation = 'h', y = -0.23)) %>%
      add_lines(x = c(min_x-0.5, max_x+0.5), y = -log10(input$c_pval_thresh), line = list(dash = "dash", width = 0.5, color = "#2b333e"), 
                name = '', hoverinfo = "text", text = paste0("pvalue = ", input$c_pval_thresh), showlegend = F) %>%
      add_lines(x = input$c_logfc_thresh_comb[1], y = c(min_y-0.5, max_y+0.5), line = list(dash = "dash", width = 0.5, color = "#252525"), 
                name = '', hoverinfo = "text", text = paste0("logFC = ", input$c_logfc_thresh_comb[1]), showlegend = F) %>%
      add_lines(x = input$c_logfc_thresh_comb[2], y = c(min_y-0.5, max_y+0.5), line = list(dash = "dash", width = 0.5, color = "#252525"), 
                name = '', hoverinfo = "text", text = paste0("logFC = ", input$c_logfc_thresh_comb[2]), showlegend = F)
  })
  
  c_vp2_plus <- reactive({
    validate(
      need(!is.null(c_search_gene()), "")
    )
    p <- c_vp2()
    goi <- c_search_gene()
    orig_data <- c_pd2()
    searchgene <- orig_data[grepl(goi,orig_data$gene),]
    p1 <- search_volcano(p, searchgene)
    p1
  })
  
  c_vp2_inweb <- eventReactive(input$c_make_plot_inweb, {
    validate(
      need(!is.null(c_pd2()), "")
    )
    d <- c_pd2()
    if(!is.null(c_inweb_pd2())){
      inwebFile <- c_inweb_pd2()
      bait <- c_bait_in()
      d_in <- subset(d, gene %in% inwebFile$gene & gene != bait)
      list(d_in=d_in)
    }
  })
  
  c_vp2_inweb_layer <- reactive({
    validate(
      need(!is.null(c_pd2()), "")
    )
    data <- c_pd2()
    c2_multi_vp <- c_vp2_inweb()
    vp_title <- c_inweb_hypergeometric(data, c_inweb_pd2())
    p <- c_vp_inweb_layer(data, c2_multi_vp, vp_title)
  })
  
  c_vp2_inweb_plus <- reactive({
    validate(
      need(!is.null(c_search_gene()), "")
    )
    p <- c_vp2_inweb_layer()
    goi <- c_search_gene()
    orig_data <- c_pd2()
    searchgene <- orig_data[grepl(goi,orig_data$gene),]
    p1 <- search_volcano(p, searchgene)
    p1
  })
  
  c_goi_vp2 <- eventReactive(input$c_make_plot_goi, {
    if(!is.null(c_upload_genes())){
      d <- c_pd2()
      gene_interest <- c_genes_uploaded()
      d_g2s <- lapply(gene_interest, function(x) subset(d, gene %in% x) )
      list(d_g2s=d_g2s)
    }
  })
  
  c_vp2_goi_layer <- reactive({
    validate(
      need(!is.null(c_pd2()), "")
    )
    d <- c_pd2()
    c2_goi <- c_goi_vp2()
    min_x <- c_min_x()
    min_y <- c_min_y()
    max_x <- c_max_x()
    max_y <- c_max_y()
    col_goi <- input$c_colorbrewer_theme_goi
    # vp_title <- c_pd3_inweb_hypergeometric()
    if(input$c_colorscheme == "fdr"){
      req(input$c_color_goi_sig, input$c_color_goi_insig)
      data <- separate_to_groups_for_color_integrated(d, input$c_fdr_thresh, input$c_color_goi_sig, input$c_color_goi_insig)
      p <- plot_ly(colors = col_goi, showlegend = T, width = 300, height = 390)
      for(i in nrow(data)){
        p <- add_markers(p, data = data, x = ~logFC, y = ~-log10(pvalue),
                         marker = list(size = 6, cmin = 0, cmax = 1, color = ~col), 
                         opacity = 0.8, 
                         text = ~paste(gene), hoverinfo = "text", name = "pull down")
      }
      p 
    } else if(input$c_colorscheme == "exac"){
      d$s <- exac$em_p_hi[match(d$gene, exac$GENE_NAME)]
      d$s[is.na(d$s)] <- 2
      below_thresh <- subset(d, s < 0.9)
      above_thresh <- subset(d, s >= 0.9)
      no_exist <- subset(d, s == 2)
      p <- plot_ly(colors = col_goi, showlegend = T, width = 300, height = 390)
      p <- add_markers(p, data = below_thresh, x = ~logFC, y = ~-log10(pvalue),
                       marker = list(size = 8, line = list(width=0.1, color = 'black'), cmin = 0, cmax = 1, color = "#66c2a5"),
                       opacity = 0.8, 
                       text = ~paste(gene), hoverinfo = "text", name = paste0("pLI<0.9 (", nrow(below_thresh), ")"))
      p <- add_markers(p, data = above_thresh, x = ~logFC, y = ~-log10(pvalue),
                       marker = list(size = 8, line = list(width=0.1, color = "black"), cmin = 0, cmax = 1, color = "#fc8d62"),
                       opacity = 0.8, 
                       text = ~paste(gene), hoverinfo = "text", name = paste0("pLI>=0.9 (", nrow(above_thresh), ")"))
      p <- add_markers(p, data = no_exist, x = ~logFC, y = ~-log10(pvalue),
                       marker = list(size = 8, line = list(width=0.1, color = "black"), cmin = 0, cmax = 1, color = "#8da0cb"),
                       opacity = 0.8, 
                       text = ~paste(gene), hoverinfo = "text", name = paste0("not in ExAC (", nrow(no_exist), ")"))
      p
    } else if(input$c_colorscheme == "cbf"){
      data <- separate_to_groups_for_cbf_integrated(d, input$c_fdr_thresh)
      p <- plot_ly(colors = "Greys", showlegend = T, width = 300, height = 390)
      for(i in nrow(data)){
        p <- add_markers(p, data = data, x = ~logFC, y = ~-log10(pvalue),
                         marker = list(size = 6, cmin = 0, cmax = 1, color = ~col), 
                         opacity = 0.6, 
                         text = ~paste(gene), hoverinfo = "text", name = "pull down")
      }
      p 
    } else if(input$c_colorscheme == "user"){
      validate(
        need(!is.null(input$c_file_color), "Please upload file with gene and score")
      )
      d1 <- c_in_file_color()
      col_theme <- input$c_colorbrewer_theme_tab4
      if(input$c_colorscheme_style == "cont"){
        df <- separate_to_groups_for_color_continuous(d, d1, col_theme)
        data <- df$df1
        data1 <- df$no_exist
      } else if(input$c_colorscheme_style == "disc"){
        df <- separate_to_groups_for_color_discrete(d, d1, col_theme)
        data <- df$df1
        data1 <- df$no_exist
      }
      p <- plot_ly(showlegend = T, width = 320, height = 390)
      if(nrow(data1)>0){
        p <- add_markers(p, data = data1, x = ~logFC, y = ~-log10(pvalue),
                         marker = list(size = 7, line = list(width=0.1, color = "grey89"), cmin = 0, cmax = 1, color = "#f7f7f7"),
                         opacity = 0.8,
                         text = ~paste(gene), hoverinfo = "text", name = paste0("Not in user data (", nrow(df$no_exist), ")"))
      }
      if(nrow(data)>0){
        for(i in nrow(data)){
          p <- add_markers(p, data = data, x = ~logFC, y = ~-log10(pvalue),
                           marker = list(size = 7, cmin = 0, cmax = 1, color = ~col, line = list(width=0.2, color = "grey89")),
                           opacity = 1,
                           text = ~paste0(gene, ", FDR=", signif(FDR, digits = 3)), hoverinfo = "text",
                           name = paste0("Found in user data (", nrow(data), ")")) #, name = "pull down"
        }
      }
      p
    }
    p <- p%>%
      layout(xaxis = list(range=c(min_x-0.5, max_x+0.5), showgrid = F), yaxis = list(range=c(min_y-0.5, max_y+0.5), showgrid = F), 
             legend = list(orientation = 'h', y = -0.23))
    # title = paste0("p-value = ", vp_title), titlefont = list(size=15))
    if(input$c_colorscheme == "fdr" | input$c_colorscheme == "exac" | input$c_colorscheme == "user"){
      if(!is.null(c_upload_genes())){
        df <- ldply(c2_goi$d_g2s, data.frame)
        if(nrow(df) != 0){
          if(input$c_marker_text_goi == "yes_label"){
            vp_layer_genes <- vp_layer_for_uploaded_genes(p, df)
          } else if(input$c_marker_text_goi == "no_label"){
            vp_layer_genes <- vp_layer_for_uploaded_genes_no_text(p, df)
          }
          p <- vp_layer_genes
        } #else{
        #   vp_layer_no_genes <- vp_layer_for_uploaded_genes_none(p, d)
        #   p <- vp_layer_no_genes
        # }
      }
    } else if(input$c_colorscheme == "cbf"){
      if(!is.null(c_upload_genes())){
        df <- ldply(c2_goi$d_g2s, data.frame)
        if(nrow(df) != 0){
          if(input$c_marker_text_goi == "yes_label"){
            vp_layer_genes <- vp_layer_for_uploaded_genes_cbf(p, df)
          } else if(input$c_marker_text_goi == "no_label"){
            vp_layer_genes <- vp_layer_for_uploaded_genes_cbf_no_text(p, df)
          }
          p <- vp_layer_genes
        } #else{
        #   vp_layer_no_genes <- vp_layer_for_uploaded_genes_none_cbf(p, d)
        #   p <- vp_layer_no_genes
        # }
      }
    }
    p <- p %>% 
      layout(xaxis = list(range=c(min_x-0.5, max_x+0.5), showgrid = F), yaxis = list(range=c(min_y-0.5, max_y+0.5), showgrid = F)) %>%
      add_lines(x = c(min_x-0.5, max_x+0.5), y = -log10(input$c_pval_thresh), line = list(dash = "dash", width = 0.5, color = "#2b333e"), 
                name = '', hoverinfo = "text", text = paste0("pvalue = ", input$c_pval_thresh), showlegend = F) %>%
      add_lines(x = input$c_logfc_thresh_comb[1], y = c(min_y-0.5, max_y+0.5), line = list(dash = "dash", width = 0.5, color = "#252525"), 
                name = '', hoverinfo = "text", text = paste0("logFC = ", input$c_logfc_thresh_comb[1]), showlegend = F) %>%
      add_lines(x = input$c_logfc_thresh_comb[2], y = c(min_y-0.5, max_y+0.5), line = list(dash = "dash", width = 0.5, color = "#252525"), 
                name = '', hoverinfo = "text", text = paste0("logFC = ", input$c_logfc_thresh_comb[2]), showlegend = F)
  })
  
  c_vp2_goi_plus <- reactive({
    validate(
      need(!is.null(c_search_gene()), "")
    )
    p <- c_vp2_goi_layer()
    goi <- c_search_gene()
    orig_data <- c_pd2()
    searchgene <- orig_data[grepl(goi,orig_data$gene),]
    p1 <- search_volcano(p, searchgene)
    p1
  })
  
  # 
  # c_snp_vp2 <- eventReactive(input$c_make_plot_snp, {
  #   if(!is.null(c_snp())){
  #     d <- c_pd2()
  #     snp_interest <- c_SNP_to_gene(d)
  #     print(snp_interest)
  #     d_snp <- merge(d, snp_interest, by.x = 'gene', by.y = 'gene')
  #     list(d_snp=d_snp)
  #   }
  # })
  # 
  # c_vp2_snp_layer <- reactive({
  #   validate(
  #     need(!is.null(c_pd2()), "")
  #   )
  #   d <- c_pd2()
  #   c2_snp <- c_snp_vp2()
  #   min_x <- c_min_x()
  #   min_y <- c_min_y()
  #   max_x <- c_max_x()
  #   max_y <- c_max_y()
  #   # vp_title <- c_pd3_inweb_hypergeometric()
  #   if(input$c_colorscheme == "fdr"){
  #     req(input$c_color_snp_sig, input$c_color_snp_insig)
  #     data <- separate_to_groups_for_color_integrated(d, input$c_fdr_thresh, input$c_color_snp_sig, input$c_color_snp_insig)
  #     p <- plot_volcano_multiple_cond(data)
  #   } else if(input$c_colorscheme == "exac"){
  #     d$s <- exac$em_p_hi[match(d$gene, exac$GENE_NAME)]
  #     d$s[is.na(d$s)] <- 2
  #     below_thresh <- subset(d, s < 0.9)
  #     above_thresh <- subset(d, s >= 0.9)
  #     no_exist <- subset(d, s == 2)
  #     p <- plot_volcano_exac_multi(below_thresh, above_thresh, no_exist)
  #     p
  #   } else if(input$c_colorscheme == "cbf"){
  #     data <- separate_to_groups_for_cbf_integrated(d, input$c_fdr_thresh)
  #     p <- plot_volcano_multiple_cond(data)
  #   } else if(input$c_colorscheme == "user"){
  #     validate(
  #       need(!is.null(input$c_file_color), "Please upload file with gene and score")
  #     )
  #     d1 <- c_in_file_color()
  #     col_theme <- input$c_colorbrewer_theme_tab5
  #     if(input$c_colorscheme_style == "cont"){
  #       df <- separate_to_groups_for_color_continuous(d, d1, col_theme)
  #       data <- df$df1
  #       data1 <- df$no_exist
  #     } else if(input$c_colorscheme_style == "disc"){
  #       df <- separate_to_groups_for_color_discrete(d, d1, col_theme)
  #       data <- df$df1
  #       data1 <- df$no_exist
  #     }
  #     p <- plot_ly(showlegend = T, width = 320, height = 390)
  #     if(nrow(data1)>0){
  #       p <- add_markers(p, data = data1, x = ~logFC, y = ~-log10(pvalue),
  #                        marker = list(size = 7, line = list(width=0.1, color = "grey89"), cmin = 0, cmax = 1, color = "#f7f7f7"),
  #                        opacity = 0.8,
  #                        text = ~paste(gene), hoverinfo = "text", name = paste0("Not in user data (", nrow(df$no_exist), ")"))
  #     }
  #     if(nrow(data)>0){
  #       for(i in nrow(data)){
  #         p <- add_markers(p, data = data, x = ~logFC, y = ~-log10(pvalue),
  #                          marker = list(size = 7, cmin = 0, cmax = 1, color = ~col, line = list(width=0.2, color = "grey89")),
  #                          opacity = 1,
  #                          text = ~paste0(gene, ", FDR=", signif(FDR, digits = 3)), hoverinfo = "text",
  #                          name = paste0("Found in user data (", nrow(data), ")")) #, name = "pull down"
  #       }
  #     }
  #     p
  #   }
  #   p <- p%>%
  #     layout(xaxis = list(range=c(min_x-0.5, max_x+0.5), showgrid = F), yaxis = list(range=c(min_y-0.5, max_y+0.5), showgrid = F),
  #            legend = list(orientation = 'h', y = -0.23))
  #   #title = paste0("p-value = ", vp_title), titlefont = list(size=15))
  #   if(input$c_colorscheme == "fdr" | input$c_colorscheme == "exac" | input$c_colorscheme == "user"){
  #     if(!is.null(c_snp())){
  #       if(nrow(c2_snp$d_snp) != 0){
  #         snp_sgl <- subset(c2_snp$d_snp, Freq == 1)
  #         snp_mgl <- subset(c2_snp$d_snp, Freq != 1)
  #         col <- c(color = colorRampPalette(brewer.pal(3, input$c_colorbrewer_theme_snp))(2))
  #         if(nrow(snp_sgl) != 0){
  #           if(input$c_marker_text_snp == "yes_label"){
  #             vp_layer_snp2gene_sgl <- vp_layer_for_snp_to_gene_sgl(p, snp_sgl, col[1])
  #           } else if(input$c_marker_text_snp == "no_label"){
  #             vp_layer_snp2gene_sgl <- vp_layer_for_snp_to_gene_sgl_no_text(p, snp_sgl, col[1])
  #           }
  #           p <- vp_layer_snp2gene_sgl
  #         }
  #         if(nrow(snp_mgl) != 0){
  #           if(input$c_marker_text_snp == "yes_label"){
  #             vp_layer_snp2gene_mgl <- vp_layer_for_snp_to_gene_mgl(p, snp_mgl, col[2])
  #           } else if(input$c_marker_text_snp == "no_label"){
  #             vp_layer_snp2gene_mgl <- vp_layer_for_snp_to_gene_mgl_no_text(p, snp_mgl, col[2])
  #           }
  #           p <- vp_layer_snp2gene_mgl
  #         }
  #       } else{
  #         vp_layer_no_snp2gene <- vp_layer_for_snp_to_gene_none(p, d)
  #         p <- vp_layer_no_snp2gene
  #       }
  #     }
  #   } else if(input$c_colorscheme == "cbf"){
  #     if(!is.null(c_snp())){
  #       if(nrow(c2_snp$d_snp) != 0){
  #         snp_sgl <- subset(c2_snp$d_snp, Freq == 1)
  #         snp_mgl <- subset(c2_snp$d_snp, Freq != 1)
  #         if(nrow(snp_sgl) != 0){
  #           if(input$c_marker_text_snp == "yes_label"){
  #             vp_layer_snp2gene_sgl <- vp_layer_for_snp_to_gene_sgl_cbf(p, snp_sgl)
  #           } else if(input$c_marker_text_snp == "no_label"){
  #             vp_layer_snp2gene_sgl <- vp_layer_for_snp_to_gene_sgl_cbf_no_text(p, snp_sgl)
  #           }
  #           p <- vp_layer_snp2gene_sgl
  #         }
  #         if(nrow(snp_mgl) != 0){
  #           if(input$c_marker_text_snp == "yes_label"){
  #             vp_layer_snp2gene_mgl <- vp_layer_for_snp_to_gene_mgl_cbf(p, snp_mgl)
  #           } else if(input$c_marker_text_snp == "no_label"){
  #             vp_layer_snp2gene_mgl <- vp_layer_for_snp_to_gene_mgl_cbf_no_text(p, snp_mgl)
  #           }
  #           p <- vp_layer_snp2gene_mgl
  #         }
  #       } else{
  #         vp_layer_no_snp2gene <- vp_layer_for_snp_to_gene_none_cbf(p, d)
  #         p <- vp_layer_no_snp2gene
  #       }
  #     }
  #   }
  #   p <- p %>%
  #     layout(xaxis = list(range=c(min_x-0.5, max_x+0.5), showgrid = F), yaxis = list(range=c(min_y-0.5, max_y+0.5), showgrid = F)) %>%
  #     add_lines(x = c(min_x-0.5, max_x+0.5), y = -log10(input$c_pval_thresh), line = list(dash = "dash", width = 0.5, color = "#2b333e"),
  #               name = '', hoverinfo = "text", text = paste0("pvalue = ", input$c_pval_thresh), showlegend = F) %>%
  #     add_lines(x = input$c_logfc_thresh_comb[1], y = c(min_y-0.5, max_y+0.5), line = list(dash = "dash", width = 0.5, color = "#252525"),
  #               name = '', hoverinfo = "text", text = paste0("logFC = ", input$c_logfc_thresh_comb[1]), showlegend = F) %>%
  #     add_lines(x = input$c_logfc_thresh_comb[2], y = c(min_y-0.5, max_y+0.5), line = list(dash = "dash", width = 0.5, color = "#252525"),
  #               name = '', hoverinfo = "text", text = paste0("logFC = ", input$c_logfc_thresh_comb[2]), showlegend = F)
  # })
  # 
  # c_vp2_snp_plus <- reactive({
  #   validate(
  #     need(!is.null(c_search_gene()), "")
  #   )
  #   p <- c_vp2_snp_layer()
  #   goi <- c_search_gene()
  #   orig_data <- c_pd2()
  #   searchgene <- orig_data[grepl(goi,orig_data$gene),]
  #   p1 <- search_volcano(p, searchgene)
  #   p1
  # })
  
  c_vp3 <- reactive({
    d <- c_pd3()
    if(input$c_colorscheme == "fdr"){
      req(input$c_color_indv_sig, input$c_color_indv_insig)
      data <- separate_to_groups_for_color_integrated(d, input$c_fdr_thresh, input$c_color_indv_sig, input$c_color_indv_insig)
      p <- plot_volcano_multiple_cond(data)
    } else if(input$c_colorscheme == "exac"){
      d$s <- exac$em_p_hi[match(d$gene, exac$GENE_NAME)]
      d$s[is.na(d$s)] <- 2
      below_thresh <- subset(d, s < 0.9)
      above_thresh <- subset(d, s >= 0.9)
      no_exist <- subset(d, s == 2)
      p <- plot_volcano_exac_multi(below_thresh, above_thresh, no_exist)
    } else if(input$c_colorscheme == "cbf"){
      data <- separate_to_groups_for_cbf_integrated(d, input$c_fdr_thresh)
      p <- plot_volcano_multiple_cond(data)
    } else if(input$c_colorscheme == "user"){
      validate(
        need(!is.null(input$c_file_color), "Please upload file with gene and score")
      )
      d1 <- c_in_file_color()
      col_theme <- input$c_colorbrewer_theme
      if(input$c_colorscheme_style == "cont"){
        df <- separate_to_groups_for_color_continuous(d, d1, col_theme)
        data <- df$df1
        data1 <- df$no_exist
      } else if(input$c_colorscheme_style == "disc"){
        df <- separate_to_groups_for_color_discrete(d, d1, col_theme)
        data <- df$df1
        data1 <- df$no_exist
      }
      p <- plot_ly(showlegend = T, width = 320, height = 390)
      p <- add_markers(p, data = data1, x = ~logFC, y = ~-log10(pvalue),
                       marker = list(size = 7, line = list(width=0.1, color = "grey89"), cmin = 0, cmax = 1, color = "#f7f7f7"),
                       opacity = 0.8,
                       text = ~paste(gene), hoverinfo = "text", name = paste0("Not in user data (", nrow(df$no_exist), ")"))
      for(i in nrow(data)){
        p <- add_markers(p, data = data, x = ~logFC, y = ~-log10(pvalue),
                         marker = list(size = 7, cmin = 0, cmax = 1, color = ~col, line = list(width=0.2, color = "grey89")),
                         opacity = 1,
                         text = ~paste0(gene, ", FDR=", signif(FDR, digits = 3)), hoverinfo = "text",
                         name = paste0("Found in user data (", nrow(data), ")")) #, name = "pull down"
      }
      p
    }
    min_x <- c_min_x()
    min_y <- c_min_y()
    max_x <- c_max_x()
    max_y <- c_max_y()
    p <- p %>%
      layout(xaxis = list(range=c(min_x-0.5, max_x+0.5), showgrid = F), yaxis = list(range=c(min_y-0.5, max_y+0.5), showgrid = F), 
             legend = list(orientation = 'h', y = -0.23)) %>%
      add_lines(x = c(min_x-0.5, max_x+0.5), y = -log10(input$c_pval_thresh), line = list(dash = "dash", width = 0.5, color = "#2b333e"), 
                name = '', hoverinfo = "text", text = paste0("pvalue = ", input$c_pval_thresh), showlegend = F) %>%
      add_lines(x = input$c_logfc_thresh_comb[1], y = c(min_y-0.5, max_y+0.5), line = list(dash = "dash", width = 0.5, color = "#252525"), 
                name = '', hoverinfo = "text", text = paste0("logFC = ", input$c_logfc_thresh_comb[1]), showlegend = F) %>%
      add_lines(x = input$c_logfc_thresh_comb[2], y = c(min_y-0.5, max_y+0.5), line = list(dash = "dash", width = 0.5, color = "#252525"), 
                name = '', hoverinfo = "text", text = paste0("logFC = ", input$c_logfc_thresh_comb[2]), showlegend = F)
  })
  
  c_vp3_plus <- reactive({
    validate(
      need(!is.null(c_search_gene()), "")
    )
    p <- c_vp3()
    goi <- c_search_gene()
    orig_data <- c_pd3()
    searchgene <- orig_data[grepl(goi,orig_data$gene),]
    p1 <- search_volcano(p, searchgene)
    p1
  })
  
  c_vp3_inweb <- eventReactive(input$c_make_plot_inweb, {
    validate(
      need(!is.null(c_pd3()), "")
    )
    d <- c_pd3()
    if(!is.null(c_inweb_pd3())){
      inwebFile <- c_inweb_pd3()
      bait <- c_bait_in()
      d_in <- subset(d, gene %in% inwebFile$gene & gene != bait)
      list(d_in=d_in)
    }
  })
  
  c_vp3_inweb_layer <- reactive({
    validate(
      need(!is.null(c_pd3()), "")
    )
    data <- c_pd3()
    c3_multi_vp <- c_vp3_inweb()
    vp_title <- c_inweb_hypergeometric(data, c_inweb_pd3())
    p <- c_vp_inweb_layer(data, c3_multi_vp, vp_title)
  })
  
  c_vp3_inweb_plus <- reactive({
    validate(
      need(!is.null(c_search_gene()), "")
    )
    p <- c_vp3_inweb_layer()
    goi <- c_search_gene()
    orig_data <- c_pd3()
    searchgene <- orig_data[grepl(goi,orig_data$gene),]
    p1 <- search_volcano(p, searchgene)
    p1
  })
  
  c_goi_vp3 <- eventReactive(input$c_make_plot_goi, {
    if(!is.null(c_upload_genes())){
      d <- c_pd3()
      gene_interest <- c_genes_uploaded()
      d_g2s <- lapply(gene_interest, function(x) subset(d, gene %in% x) )
      list(d_g2s=d_g2s)
    }
  })
  
  c_vp3_goi_layer <- reactive({
    validate(
      need(!is.null(c_pd3()), "")
    )
    d <- c_pd3()
    c3_goi <- c_goi_vp3()
    min_x <- c_min_x()
    min_y <- c_min_y()
    max_x <- c_max_x()
    max_y <- c_max_y()
    col_goi <- input$c_colorbrewer_theme_goi
    # vp_title <- c_pd3_inweb_hypergeometric()
    if(input$c_colorscheme == "fdr"){
      req(input$c_color_goi_sig, input$c_color_goi_insig)
      data <- separate_to_groups_for_color_integrated(d, input$c_fdr_thresh, input$c_color_goi_sig, input$c_color_goi_insig)
      p <- plot_ly(colors = col_goi, showlegend = T, width = 300, height = 390)
      for(i in nrow(data)){
        p <- add_markers(p, data = data, x = ~logFC, y = ~-log10(pvalue),
                         marker = list(size = 6, cmin = 0, cmax = 1, color = ~col), 
                         opacity = 0.8, 
                         text = ~paste(gene), hoverinfo = "text", name = "pull down")
      }
      p 
    } else if(input$c_colorscheme == "exac"){
      d$s <- exac$em_p_hi[match(d$gene, exac$GENE_NAME)]
      d$s[is.na(d$s)] <- 2
      below_thresh <- subset(d, s < 0.9)
      above_thresh <- subset(d, s >= 0.9)
      no_exist <- subset(d, s == 2)
      p <- plot_ly(colors = col_goi, showlegend = T, width = 300, height = 390)
      p <- add_markers(p, data = below_thresh, x = ~logFC, y = ~-log10(pvalue),
                       marker = list(size = 8, line = list(width=0.1, color = 'black'), cmin = 0, cmax = 1, color = "#66c2a5"),
                       opacity = 0.8, 
                       text = ~paste(gene), hoverinfo = "text", name = paste0("pLI<0.9 (", nrow(below_thresh), ")"))
      p <- add_markers(p, data = above_thresh, x = ~logFC, y = ~-log10(pvalue),
                       marker = list(size = 8, line = list(width=0.1, color = "black"), cmin = 0, cmax = 1, color = "#fc8d62"),
                       opacity = 0.8, 
                       text = ~paste(gene), hoverinfo = "text", name = paste0("pLI>=0.9 (", nrow(above_thresh), ")"))
      p <- add_markers(p, data = no_exist, x = ~logFC, y = ~-log10(pvalue),
                       marker = list(size = 8, line = list(width=0.1, color = "black"), cmin = 0, cmax = 1, color = "#8da0cb"),
                       opacity = 0.8, 
                       text = ~paste(gene), hoverinfo = "text", name = paste0("not in ExAC (", nrow(no_exist), ")"))
      p
    } else if(input$c_colorscheme == "cbf"){
      data <- separate_to_groups_for_cbf_integrated(d, input$c_fdr_thresh)
      p <- plot_ly(colors = "Greys", showlegend = T, width = 300, height = 390)
      for(i in nrow(data)){
        p <- add_markers(p, data = data, x = ~logFC, y = ~-log10(pvalue),
                         marker = list(size = 6, cmin = 0, cmax = 1, color = ~col), 
                         opacity = 0.6, 
                         text = ~paste(gene), hoverinfo = "text", name = "pull down")
      }
      p 
    } else if(input$c_colorscheme == "user"){
      validate(
        need(!is.null(input$c_file_color), "Please upload file with gene and score")
      )
      d1 <- c_in_file_color()
      col_theme <- input$c_colorbrewer_theme_tab4
      if(input$c_colorscheme_style == "cont"){
        df <- separate_to_groups_for_color_continuous(d, d1, col_theme)
        data <- df$df1
        data1 <- df$no_exist
      } else if(input$c_colorscheme_style == "disc"){
        df <- separate_to_groups_for_color_discrete(d, d1, col_theme)
        data <- df$df1
        data1 <- df$no_exist
      }
      p <- plot_ly(showlegend = T, width = 320, height = 390)
      if(nrow(data1)>0){
        p <- add_markers(p, data = data1, x = ~logFC, y = ~-log10(pvalue),
                         marker = list(size = 7, line = list(width=0.1, color = "grey89"), cmin = 0, cmax = 1, color = "#f7f7f7"),
                         opacity = 0.8,
                         text = ~paste(gene), hoverinfo = "text", name = paste0("Not in user data (", nrow(df$no_exist), ")"))
      }
      if(nrow(data)>0){
        for(i in nrow(data)){
          p <- add_markers(p, data = data, x = ~logFC, y = ~-log10(pvalue),
                           marker = list(size = 7, cmin = 0, cmax = 1, color = ~col, line = list(width=0.2, color = "grey89")),
                           opacity = 1,
                           text = ~paste0(gene, ", FDR=", signif(FDR, digits = 3)), hoverinfo = "text",
                           name = paste0("Found in user data (", nrow(data), ")")) #, name = "pull down"
        }
      }
      p
    }
    p <- p%>%
      layout(xaxis = list(range=c(min_x-0.5, max_x+0.5), showgrid = F), yaxis = list(range=c(min_y-0.5, max_y+0.5), showgrid = F),
             legend = list(orientation = 'h', y = -0.23))
    # title = paste0("p-value = ", vp_title), titlefont = list(size=15))
    if(input$c_colorscheme == "fdr" | input$c_colorscheme == "exac" | input$c_colorscheme == "user"){
      if(!is.null(c_upload_genes())){
        df <- ldply(c3_goi$d_g2s, data.frame)
        if(nrow(df) != 0){
          if(input$c_marker_text_goi == "yes_label"){
            vp_layer_genes <- vp_layer_for_uploaded_genes(p, df)
          } else if(input$c_marker_text_goi == "no_label"){
            vp_layer_genes <- vp_layer_for_uploaded_genes_no_text(p, df)
          }
          p <- vp_layer_genes
        } #else{
        #   vp_layer_no_genes <- vp_layer_for_uploaded_genes_none(p, d)
        #   p <- vp_layer_no_genes
        # }
      }
    } else if(input$c_colorscheme == "cbf"){
      if(!is.null(c_upload_genes())){
        df <- ldply(c3_goi$d_g2s, data.frame)
        if(nrow(df) != 0){
          if(input$c_marker_text_goi == "yes_label"){
            vp_layer_genes <- vp_layer_for_uploaded_genes_cbf(p, df)
          } else if(input$c_marker_text_goi == "no_label"){
            vp_layer_genes <- vp_layer_for_uploaded_genes_cbf_no_text(p, df)
          }
          p <- vp_layer_genes
        } #else{
        #   vp_layer_no_genes <- vp_layer_for_uploaded_genes_none_cbf(p, d)
        #   p <- vp_layer_no_genes
        # }
      }
    }
    p <- p %>% 
      layout(xaxis = list(range=c(min_x-0.5, max_x+0.5), showgrid = F), yaxis = list(range=c(min_y-0.5, max_y+0.5), showgrid = F)) %>%
      add_lines(x = c(min_x-0.5, max_x+0.5), y = -log10(input$c_pval_thresh), line = list(dash = "dash", width = 0.5, color = "#2b333e"), 
                name = '', hoverinfo = "text", text = paste0("pvalue = ", input$c_pval_thresh), showlegend = F) %>%
      add_lines(x = input$c_logfc_thresh_comb[1], y = c(min_y-0.5, max_y+0.5), line = list(dash = "dash", width = 0.5, color = "#252525"), 
                name = '', hoverinfo = "text", text = paste0("logFC = ", input$c_logfc_thresh_comb[1]), showlegend = F) %>%
      add_lines(x = input$c_logfc_thresh_comb[2], y = c(min_y-0.5, max_y+0.5), line = list(dash = "dash", width = 0.5, color = "#252525"), 
                name = '', hoverinfo = "text", text = paste0("logFC = ", input$c_logfc_thresh_comb[2]), showlegend = F)
  })
  
  c_vp3_goi_plus <- reactive({
    validate(
      need(!is.null(c_search_gene()), "")
    )
    p <- c_vp3_goi_layer()
    goi <- c_search_gene()
    orig_data <- c_pd3()
    searchgene <- orig_data[grepl(goi,orig_data$gene),]
    p1 <- search_volcano(p, searchgene)
    p1
  })
  
  # c_snp_vp3 <- eventReactive(input$c_make_plot_snp, {
  #   if(!is.null(c_snp())){
  #     d <- c_pd3()
  #     snp_interest <- c_SNP_to_gene(d)
  #     d_snp <- merge(d, snp_interest, by.x = 'gene', by.y = 'gene')
  #     list(d_snp=d_snp)
  #   }
  # })
  # 
  # c_vp3_snp_layer <- reactive({
  #   validate(
  #     need(!is.null(c_pd3()), "")
  #   )
  #   d <- c_pd3()
  #   c3_snp <- c_snp_vp3()
  #   min_x <- c_min_x()
  #   min_y <- c_min_y()
  #   max_x <- c_max_x()
  #   max_y <- c_max_y()
  #   # vp_title <- c_pd3_inweb_hypergeometric()
  #   if(input$c_colorscheme == "fdr"){
  #     req(input$c_color_snp_sig, input$c_color_snp_insig)
  #     data <- separate_to_groups_for_color_integrated(d, input$c_fdr_thresh, input$c_color_snp_sig, input$c_color_snp_insig)
  #     p <- plot_volcano_multiple_cond(data)
  #   } else if(input$c_colorscheme == "exac"){
  #     d$s <- exac$em_p_hi[match(d$gene, exac$GENE_NAME)]
  #     d$s[is.na(d$s)] <- 2
  #     below_thresh <- subset(d, s < 0.9)
  #     above_thresh <- subset(d, s >= 0.9)
  #     no_exist <- subset(d, s == 2)
  #     p <- plot_volcano_exac_multi(below_thresh, above_thresh, no_exist)
  #     p
  #   } else if(input$c_colorscheme == "cbf"){
  #     data <- separate_to_groups_for_cbf_integrated(d, input$c_fdr_thresh)
  #     p <- plot_volcano_multiple_cond(data)
  #   } else if(input$c_colorscheme == "user"){
  #     validate(
  #       need(!is.null(input$c_file_color), "Please upload file with gene and score")
  #     )
  #     d1 <- c_in_file_color()
  #     col_theme <- input$c_colorbrewer_theme_tab5
  #     if(input$c_colorscheme_style == "cont"){
  #       df <- separate_to_groups_for_color_continuous(d, d1, col_theme)
  #       data <- df$df1
  #       data1 <- df$no_exist
  #     } else if(input$c_colorscheme_style == "disc"){
  #       df <- separate_to_groups_for_color_discrete(d, d1, col_theme)
  #       data <- df$df1
  #       data1 <- df$no_exist
  #     }
  #     p <- plot_ly(showlegend = T, width = 320, height = 390)
  #     if(nrow(data1)>0){
  #       p <- add_markers(p, data = data1, x = ~logFC, y = ~-log10(pvalue),
  #                        marker = list(size = 7, line = list(width=0.1, color = "grey89"), cmin = 0, cmax = 1, color = "#f7f7f7"),
  #                        opacity = 0.8,
  #                        text = ~paste(gene), hoverinfo = "text", name = paste0("Not in user data (", nrow(df$no_exist), ")"))
  #     }
  #     if(nrow(data)>0){
  #       for(i in nrow(data)){
  #         p <- add_markers(p, data = data, x = ~logFC, y = ~-log10(pvalue),
  #                          marker = list(size = 7, cmin = 0, cmax = 1, color = ~col, line = list(width=0.2, color = "grey89")),
  #                          opacity = 1,
  #                          text = ~paste0(gene, ", FDR=", signif(FDR, digits = 3)), hoverinfo = "text",
  #                          name = paste0("Found in user data (", nrow(data), ")")) #, name = "pull down"
  #       }
  #     }
  #     p
  #   }
  #   p <- p%>%
  #     layout(xaxis = list(range=c(min_x-0.5, max_x+0.5), showgrid = F), yaxis = list(range=c(min_y-0.5, max_y+0.5), showgrid = F),
  #            legend = list(orientation = 'h', y = -0.23))
  #   #title = paste0("p-value = ", vp_title), titlefont = list(size=15))
  #   if(input$c_colorscheme == "fdr" | input$c_colorscheme == "exac" | input$c_colorscheme == "user"){
  #     if(!is.null(c_snp())){
  #       if(nrow(c3_snp$d_snp) != 0){
  #         snp_sgl <- subset(c3_snp$d_snp, Freq == 1)
  #         snp_mgl <- subset(c3_snp$d_snp, Freq != 1)
  #         col <- c(color = colorRampPalette(brewer.pal(3, input$c_colorbrewer_theme_snp))(2))
  #         if(nrow(snp_sgl) != 0){
  #           if(input$c_marker_text_snp == "yes_label"){
  #             vp_layer_snp2gene_sgl <- vp_layer_for_snp_to_gene_sgl(p, snp_sgl, col[1])
  #           } else if(input$c_marker_text_snp == "no_label"){
  #             vp_layer_snp2gene_sgl <- vp_layer_for_snp_to_gene_sgl_no_text(p, snp_sgl, col[1])
  #           }
  #           p <- vp_layer_snp2gene_sgl
  #         }
  #         if(nrow(snp_mgl) != 0){
  #           if(input$c_marker_text_snp == "yes_label"){
  #             vp_layer_snp2gene_mgl <- vp_layer_for_snp_to_gene_mgl(p, snp_mgl, col[2])
  #           } else if(input$c_marker_text_snp == "no_label"){
  #             vp_layer_snp2gene_mgl <- vp_layer_for_snp_to_gene_mgl_no_text(p, snp_mgl, col[2])
  #           }
  #           p <- vp_layer_snp2gene_mgl
  #         }
  #       } else{
  #         vp_layer_no_snp2gene <- vp_layer_for_snp_to_gene_none(p, d)
  #         p <- vp_layer_no_snp2gene
  #       }
  #     }
  #   } else if(input$c_colorscheme == "cbf"){
  #     if(!is.null(c_snp())){
  #       if(nrow(c3_snp$d_snp) != 0){
  #         snp_sgl <- subset(c3_snp$d_snp, Freq == 1)
  #         snp_mgl <- subset(c3_snp$d_snp, Freq != 1)
  #         if(nrow(snp_sgl) != 0){
  #           if(input$c_marker_text_snp == "yes_label"){
  #             vp_layer_snp2gene_sgl <- vp_layer_for_snp_to_gene_sgl_cbf(p, snp_sgl)
  #           } else if(input$c_marker_text_snp == "no_label"){
  #             vp_layer_snp2gene_sgl <- vp_layer_for_snp_to_gene_sgl_cbf_no_text(p, snp_sgl)
  #           }
  #           p <- vp_layer_snp2gene_sgl
  #         }
  #         if(nrow(snp_mgl) != 0){
  #           if(input$c_marker_text_snp == "yes_label"){
  #             vp_layer_snp2gene_mgl <- vp_layer_for_snp_to_gene_mgl_cbf(p, snp_mgl)
  #           } else if(input$c_marker_text_snp == "no_label"){
  #             vp_layer_snp2gene_mgl <- vp_layer_for_snp_to_gene_mgl_cbf_no_text(p, snp_mgl)
  #           }
  #           p <- vp_layer_snp2gene_mgl
  #         }
  #       } else{
  #         vp_layer_no_snp2gene <- vp_layer_for_snp_to_gene_none_cbf(p, d)
  #         p <- vp_layer_no_snp2gene
  #       }
  #     }
  #   }
  #   p <- p %>%
  #     layout(xaxis = list(range=c(min_x-0.5, max_x+0.5), showgrid = F), yaxis = list(range=c(min_y-0.5, max_y+0.5), showgrid = F)) %>%
  #     add_lines(x = c(min_x-0.5, max_x+0.5), y = -log10(input$c_pval_thresh), line = list(dash = "dash", width = 0.5, color = "#2b333e"),
  #               name = '', hoverinfo = "text", text = paste0("pvalue = ", input$c_pval_thresh), showlegend = F) %>%
  #     add_lines(x = input$c_logfc_thresh_comb[1], y = c(min_y-0.5, max_y+0.5), line = list(dash = "dash", width = 0.5, color = "#252525"),
  #               name = '', hoverinfo = "text", text = paste0("logFC = ", input$c_logfc_thresh_comb[1]), showlegend = F) %>%
  #     add_lines(x = input$c_logfc_thresh_comb[2], y = c(min_y-0.5, max_y+0.5), line = list(dash = "dash", width = 0.5, color = "#252525"),
  #               name = '', hoverinfo = "text", text = paste0("logFC = ", input$c_logfc_thresh_comb[2]), showlegend = F)
  # })
  # 
  # c_vp3_snp_plus <- reactive({
  #   validate(
  #     need(!is.null(c_search_gene()), "")
  #   )
  #   p <- c_vp3_snp_layer()
  #   goi <- c_search_gene()
  #   orig_data <- c_pd3()
  #   searchgene <- orig_data[grepl(goi,orig_data$gene),]
  #   p1 <- search_volcano(p, searchgene)
  #   p1
  # })
  
  c1_vp_inweb_gg <- function(){
    validate(
      need(!is.null(c_pd1()), ""),
      need(!is.null(c_bait_in()), "")
    )
    data <- c_pd1()
    p <- c_vp_gg(data, input$c_color_inweb_sig, input$c_color_inweb_insig)
    inweb_d <- c_vp1_inweb()
    label <- input$c_marker_text_inweb
    if(input$c_colorscheme == "fdr" | input$c_colorscheme == "exac" | input$c_colorscheme == "user"){
      col <- input$c_color_inweb
      p <- vp_layer_for_inweb_sf_gg(p, inweb_d$d_in, col, label)
    } else if(input$c_colorscheme == "cbf"){
      p <- vp_layer_for_inweb_cbf_sf_gg(p, inweb_d$d_in, label)
    }
    p
  }
  
  c2_vp_inweb_gg <- function(){
    validate(
      need(!is.null(c_pd2()), ""),
      need(!is.null(c_bait_in()), "")
    )
    data <- c_pd2()
    p <- c_vp_gg(data, input$c_color_inweb_sig, input$c_color_inweb_insig)
    inweb_d <- c_vp2_inweb()
    label <- input$c_marker_text_inweb
    if(input$c_colorscheme == "fdr" | input$c_colorscheme == "exac" | input$c_colorscheme == "user"){
      col <- input$c_color_inweb
      p <- vp_layer_for_inweb_sf_gg(p, inweb_d$d_in, col, label)
    } else if(input$c_colorscheme == "cbf"){
      p <- vp_layer_for_inweb_cbf_sf_gg(p, inweb_d$d_in, label)
    }
    p
  }
  
  c3_vp_inweb_gg <- function(){
    validate(
      need(!is.null(c_pd3()), ""),
      need(!is.null(c_bait_in()), "")
    )
    data <- c_pd3()
    p <- c_vp_gg(data, input$c_color_inweb_sig, input$c_color_inweb_insig)
    inweb_d <- c_vp3_inweb()
    label <- input$c_marker_text_inweb
    if(input$c_colorscheme == "fdr" | input$c_colorscheme == "exac" | input$c_colorscheme == "user"){
      col <- input$c_color_inweb
      p <- vp_layer_for_inweb_sf_gg(p, inweb_d$d_in, col, label)
    } else if(input$c_colorscheme == "cbf"){
      p <- vp_layer_for_inweb_cbf_sf_gg(p, inweb_d$d_in, label)
    }
    p
  }
  
  c1_vp_inweb_plus_rep_gg <- function(){
    vp <- c1_vp_inweb_gg()
    d <- c_pd1()
    p <- c_vp_plus_rep_gg(vp, d)
    p
  }
  
  c2_vp_inweb_plus_rep_gg <- function(){
    vp <- c2_vp_inweb_gg()
    d <- c_pd2()
    p <- c_vp_plus_rep_gg(vp, d)
    p
  }
  
  c3_vp_inweb_plus_rep_gg <- function(){
    vp <- c3_vp_inweb_gg()
    d <- c_pd3()
    p <- c_vp_plus_rep_gg(vp, d)
    p
  }
  
  output$download_c_vp1_inweb_gg <- downloadHandler(
    filename = function() { paste("vp1_inweb_gg", '.png', sep='') },
    content = function(file) {
      if(is.null(c_search_gene())){
        ggsave(file, plot = c1_vp_inweb_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
      } else {
        ggsave(file, plot = c1_vp_inweb_plus_rep_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
      }
    }
  )
  
  output$download_c_vp2_inweb_gg <- downloadHandler(
    filename = function() { paste("vp2_inweb_gg", '.png', sep='') },
    content = function(file) {
      if(is.null(c_search_gene())){
        ggsave(file, plot = c2_vp_inweb_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
      } else {
        ggsave(file, plot = c2_vp_inweb_plus_rep_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
      }
    }
  )
  
  output$download_c_vp3_inweb_gg <- downloadHandler(
    filename = function() { paste("vp3_inweb_gg", '.png', sep='') },
    content = function(file) {
      if(is.null(c_search_gene())){
        ggsave(file, plot = c3_vp_inweb_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
      } else {
        ggsave(file, plot = c3_vp_inweb_plus_rep_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
      }
    }
  )
  
  c1_vp_goi_gg <- function(){
    validate(
      need(!is.null(c_pd1()), ""),
      need(!is.null(c_upload_genes()), "")
    )
    data <- c_pd1()
    c1_goi <- c_goi_vp1()
    p <- c_vp_gg(data, input$c_color_goi_sig, input$c_color_goi_insig)
    df <- ldply(c1_goi$d_g2s, data.frame)
    print(head(df))
    label <- input$c_marker_text_goi
    if(input$c_colorscheme == "fdr" | input$c_colorscheme == "exac" | input$c_colorscheme == "user"){
      col_goi <- input$c_colorbrewer_theme_goi
      if(nrow(df) != 0){
        p1 <- vp_layer_for_uploaded_genes_gg(p, df, col_goi, label)
      } else{
        p1 <- p
      }
    } else if(input$c_colorscheme == "cbf"){
      if(nrow(df) != 0){
        p1 <- vp_layer_for_uploaded_genes_cbf_gg(p, df, label)
      } else{
        p1 <- p
      }
    }
    p1
  }
  
  c2_vp_goi_gg <- function(){
    validate(
      need(!is.null(c_pd2()), ""),
      need(!is.null(c_upload_genes()), "")
    )
    data <- c_pd2()
    c2_goi <- c_goi_vp2()
    p <- c_vp_gg(data, input$c_color_goi_sig, input$c_color_goi_insig)
    df <- ldply(c2_goi$d_g2s, data.frame)
    label <- input$c_marker_text_goi
    if(input$c_colorscheme == "fdr" | input$c_colorscheme == "exac" | input$c_colorscheme == "user"){
      col_goi <- input$c_colorbrewer_theme_goi
      if(nrow(df) != 0){
        p1 <- vp_layer_for_uploaded_genes_gg(p, df, col_goi, label)
      } else{
        p1 <- p
      }
    } else if(input$c_colorscheme == "cbf"){
      if(nrow(df) != 0){
        p1 <- vp_layer_for_uploaded_genes_cbf_gg(p, df, label)
      } else{
        p1 <- p
      }
    }
    p1
  }
  
  c3_vp_goi_gg <- function(){
    validate(
      need(!is.null(c_pd3()), ""),
      need(!is.null(c_upload_genes()), "")
    )
    data <- c_pd3()
    c3_goi <- c_goi_vp3()
    p <- c_vp_gg(data, input$c_color_goi_sig, input$c_color_goi_insig)
    df <- ldply(c3_goi$d_g2s, data.frame)
    label <- input$c_marker_text_goi
    if(input$c_colorscheme == "fdr" | input$c_colorscheme == "exac" | input$c_colorscheme == "user"){
      col_goi <- input$c_colorbrewer_theme_goi
      if(nrow(df) != 0){
        p1 <- vp_layer_for_uploaded_genes_gg(p, df, col_goi, label)
      } else{
        p1 <- p
      }
    } else if(input$c_colorscheme == "cbf"){
      if(nrow(df) != 0){
        p1 <- vp_layer_for_uploaded_genes_cbf_gg(p, df, label)
      } else{
        p1 <- p
      }
    }
    p1
  }
  
  c1_vp_goi_plus_rep_gg <- function(){
    vp <- c1_vp_goi_gg()
    d <- c_pd1()
    p <- c_vp_plus_rep_gg(vp, d)
    p
  }
  
  c2_vp_goi_plus_rep_gg <- function(){
    vp <- c2_vp_goi_gg()
    d <- c_pd2()
    p <- c_vp_plus_rep_gg(vp, d)
    p
  }
  
  c3_vp_goi_plus_rep_gg <- function(){
    vp <- c3_vp_goi_gg()
    d <- c_pd3()
    p <- c_vp_plus_rep_gg(vp, d)
    p
  }
  
  output$download_c_vp1_goi_gg <- downloadHandler(
    filename = function() { paste("vp1_goi_gg", '.png', sep='') },
    content = function(file) {
      if(is.null(c_search_gene())){
        ggsave(file, plot = c1_vp_goi_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
      } else {
        ggsave(file, plot = c1_vp_goi_plus_rep_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
      }
    }
  )
  
  output$download_c_vp2_goi_gg <- downloadHandler(
    filename = function() { paste("vp2_goi_gg", '.png', sep='') },
    content = function(file) {
      if(is.null(c_search_gene())){
        ggsave(file, plot = c2_vp_goi_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
      } else {
        ggsave(file, plot = c2_vp_goi_plus_rep_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
      }
    }
  )
  
  # c1_vp_snp_gg <- function(){
  #   validate(
  #     need(!is.null(c_pd1()), ""),
  #     need(!is.null(c_snp()), "")
  #   )
  #   data <- c_pd1()
  #   p <- c_vp_gg(data, input$c_color_snp_sig, input$c_color_snp_insig)
  #   c_snp <- c_snp_vp1()
  #   label <- input$c_marker_text_snp
  #   if(nrow(c_snp$d_snp) != 0){
  #     snp_sgl <- subset(c_snp$d_snp, Freq == 1)
  #     snp_mgl <- subset(c_snp$d_snp, Freq != 1)
  #     col <- c(color = colorRampPalette(brewer.pal(3, input$c_colorbrewer_theme_snp))(2))
  #     if(nrow(snp_sgl) != 0){
  #       p <- vp_layer_for_snp_to_gene_sgl_gg(p, snp_sgl, col[1], label)
  #     }
  #     if(nrow(snp_mgl) != 0){
  #       p <- vp_layer_for_snp_to_gene_mgl_gg(p, snp_mgl, col[2], label)
  #     }
  #   } else{
  #     p
  #   }
  #   p
  # }
  # 
  # 
  # 
  # c2_vp_snp_gg <- function(){
  #   validate(
  #     need(!is.null(c_pd2()), ""),
  #     need(!is.null(c_snp()), "")
  #   )
  #   data <- c_pd2()
  #   p <- c_vp_gg(data, input$c_color_snp_sig, input$c_color_snp_insig)
  #   c_snp <- c_snp_vp2()
  #   label <- input$c_marker_text_snp
  #   if(nrow(c_snp$d_snp) != 0){
  #     snp_sgl <- subset(c_snp$d_snp, Freq == 1)
  #     snp_mgl <- subset(c_snp$d_snp, Freq != 1)
  #     col <- c(color = colorRampPalette(brewer.pal(3, input$c_colorbrewer_theme_snp))(2))
  #     if(nrow(snp_sgl) != 0){
  #       p <- vp_layer_for_snp_to_gene_sgl_gg(p, snp_sgl, col[1], label)
  #     }
  #     if(nrow(snp_mgl) != 0){
  #       p <- vp_layer_for_snp_to_gene_mgl_gg(p, snp_mgl, col[2], label)
  #     }
  #   } else{
  #     p
  #   }
  #   p}
  # 
  # c3_vp_snp_gg <- function(){
  #   validate(
  #     need(!is.null(c_pd3()), ""),
  #     need(!is.null(c_snp()), "")
  #   )
  #   data <- c_pd3()
  #   p <- c_vp_gg(data, input$c_color_snp_sig, input$c_color_snp_insig)
  #   c_snp <- c_snp_vp3()
  #   label <- input$c_marker_text_snp
  #   if(nrow(c_snp$d_snp) != 0){
  #     snp_sgl <- subset(c_snp$d_snp, Freq == 1)
  #     snp_mgl <- subset(c3_snp$d_snp, Freq != 1)
  #     col <- c(color = colorRampPalette(brewer.pal(3, input$c_colorbrewer_theme_snp))(2))
  #     if(nrow(snp_sgl) != 0){
  #       p <- vp_layer_for_snp_to_gene_sgl_gg(p, snp_sgl, col[1], label)
  #     }
  #     if(nrow(snp_mgl) != 0){
  #       p <- vp_layer_for_snp_to_gene_mgl_gg(p, snp_mgl, col[2], label)
  #     }
  #   } else{
  #     p
  #   }
  #   p
  # }
  # 
  # c1_vp_snp_plus_rep_gg <- function(){
  #   vp <- c1_vp_snp_gg()
  #   d <- c_pd1()
  #   p <- c_vp_plus_rep_gg(vp, d)
  #   p
  # }
  # 
  # c2_vp_snp_plus_rep_gg <- function(){
  #   vp <- c2_vp_snp_gg()
  #   d <- c_pd2()
  #   p <- c_vp_plus_rep_gg(vp, d)
  #   p
  # }
  # 
  # c3_vp_snp_plus_rep_gg <- function(){
  #   vp <- c3_vp_snp_gg()
  #   d <- c_pd3()
  #   p <- c_vp_plus_rep_gg(vp, d)
  #   p
  # }
  # 
  # output$download_c_vp1_snp_gg <- downloadHandler(
  #   filename = function() { paste("vp1_snp_gg", '.png', sep='') },
  #   content = function(file) {
  #     if(is.null(c_search_gene())){
  #       ggsave(file, plot = c1_vp_snp_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
  #     } else {
  #       ggsave(file, plot = c1_vp_snp_plus_rep_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
  #     }
  #   }
  # )
  # 
  # output$download_c_vp2_snp_gg <- downloadHandler(
  #   filename = function() { paste("vp2_snp_gg", '.png', sep='') },
  #   content = function(file) {
  #     if(is.null(c_search_gene())){
  #       ggsave(file, plot = c2_vp_snp_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
  #     } else {
  #       ggsave(file, plot = c2_vp_snp_plus_rep_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
  #     }
  #   }
  # )
  # 
  # 
  # output$download_c_vp3_snp_gg <- downloadHandler(
  #   filename = function() { paste("vp3_snp_gg", '.png', sep='') },
  #   content = function(file) {
  #     if(is.null(c_search_gene())){
  #       ggsave(file, plot = c3_vp_snp_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
  #     } else {
  #       ggsave(file, plot = c3_vp_snp_plus_rep_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
  #     }
  #   }
  # )
  
  output$download_c_vp3_goi_gg <- downloadHandler(
    filename = function() { paste("vp3_goi_gg", '.png', sep='') },
    content = function(file) {
      if(is.null(c_search_gene())){
        ggsave(file, plot = c3_vp_goi_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
      } else {
        ggsave(file, plot = c3_vp_goi_plus_rep_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
      }
    }
  )
  
  c_sp_gg <- function(d){
    if(input$c_colorscheme == "fdr"){
      req(input$c_color_indv_sig, input$c_color_indv_insig)
      data <- separate_to_groups_for_color_integrated(d, input$c_fdr_thresh, input$c_color_indv_sig, input$c_color_indv_insig)
      p <- plot_scatter_qc_gg(data)
    } else if(input$c_colorscheme == "exac"){
      d$s <- exac$em_p_hi[match(d$gene, exac$GENE_NAME)]
      d$s[is.na(d$s)] <- 2
      below_thresh <- subset(d, s < 0.9)
      above_thresh <- subset(d, s >= 0.9)
      no_exist <- subset(d, s == 2)
      p <- plot_scatter_exac_gg(below_thresh, above_thresh, no_exist)
    } else if(input$c_colorscheme == "cbf"){
      data <- separate_to_groups_for_cbf_integrated(d, input$c_fdr_thresh)
      p <- plot_scatter_qc_gg(data)
    } else if(input$c_colorscheme == "user"){
      validate(
        need(!is.null(input$c_file_color), "Please upload file with gene and score")
      )
      d1 <- c_in_file_color()
      req(input$c_colorbrewer_theme)
      col_theme <- input$c_colorbrewer_theme
      if(input$c_colorscheme_style == "cont"){
        df <- separate_to_groups_for_color_continuous(d, d1, col_theme)
        data <- df$df1
        data1 <- df$no_exist
      } else if(input$c_colorscheme_style == "disc"){
        df <- separate_to_groups_for_color_discrete(d, d1, col_theme)
        data <- df$df1
        data1 <- df$no_exist
        data2 <- rbind(data, data1)
        p <- ggplot(data = data2, aes(x = logFC, y = -log10(pvalue), text = gene)) +
          geom_point(data = data1, alpha = 0.5, size = 1.5, colour = "#f7f7f7") + 
          geom_point(data = data, alpha = 0.5, size = 1.5, colour = data$col) + 
          xlab("log2FC") + ylab("-log10(P)") +
          theme_minimal() +
          theme(axis.text.x = element_text(size=7),
                axis.title.x=element_text(size=8),
                axis.text.y=element_text(size=7),
                axis.title.y=element_text(size=8),
                panel.grid.minor = element_blank(),
                plot.margin = unit(c(1,1,1,1), "pt"))
      }
      p
    }
    p
  }
  
  c1_sp_gg <- function(){
    data <- c_pd1()
    print(head(data))
    p <- c_sp_gg(data)
    p
  }
  
  c2_sp_gg <- function(){
    data <- c_pd2()
    p <- c_sp_gg(data)
    p
  }
  
  c3_sp_gg <- function(){
    data <- c_pd3()
    p <- c_sp_gg(data)
    p
  }
  
  
  c_sp_plus_rep_gg <- function(p, orig_data){
    validate(
      need(!is.null(c_search_gene()), "")
    )
    goi <- c_search_gene()
    searchgene <- orig_data[grepl(goi,orig_data$gene),]
    p1 <- search_scatter_gg(p, searchgene)
    p1
  }
  
  c1_sp_plus_rep_gg <- function(){
    sp <- c1_sp_gg()
    d <- c_pd1()
    p <- c_sp_plus_rep_gg(sp, d)
    p
  }
  
  c2_sp_plus_rep_gg <- function(){
    sp <- c2_sp_gg()
    d <- c_pd2()
    p <- c_sp_plus_rep_gg(sp, d)
    p
  }
  
  c3_sp_plus_rep_gg <- function(){
    sp <- c3_sp_gg()
    d <- c_pd3()
    p <- c_sp_plus_rep_gg(sp, d)
    p
  }
  
  c_sp_pf_db_gg <- function(p, pf){
    validate(
      need(!is.null(c_pf_db()), "")
    )
    pf_col <- input$c_colorbrewer_theme_pf
    label <- input$c_marker_text_prot_fam_db
    
    if(input$c_colorscheme == "fdr" | input$c_colorscheme == "exac" | input$c_colorscheme == "user"){
      df <- ldply(pf$d_g2s, data.frame)
      if(nrow(df) != 0){
        sp_layer_genes <- sp_layer_for_uploaded_genes_gg(p, df, pf_col, label)
        p1 <- sp_layer_genes
      } else{
        p1 <- p
      }
    } else if(input$c_colorscheme == "cbf"){
      df <- ldply(pf$d_g2s, data.frame)
      if(nrow(df) != 0){
        sp_layer_genes <- sp_layer_for_uploaded_genes_cbf_gg(p, df, label)
        p1 <- sp_layer_genes
      } else{
        p1 <- p
      }
    }
    p1
  }
  
  c1_sp_pf_db_gg <- function(){
    validate(
      need(!is.null(c1_sp_gg()), "")
    )
    sp <- c1_sp_gg()
    pf_db <- c_pf_db_vp1()
    p <- c_sp_pf_db_gg(sp, pf_db)
    p
  }
  
  c2_sp_pf_db_gg <- function(){
    validate(
      need(!is.null(c2_sp_gg()), "")
    )
    sp <- c2_sp_gg()
    pf_db <- c_pf_db_vp2()
    p <- c_sp_pf_db_gg(sp, pf_db)
    p
  }
  
  c3_sp_pf_db_gg <- function(){
    validate(
      need(!is.null(c3_sp_gg()), "")
    )
    sp <- c3_sp_gg()
    pf_db <- c_pf_db_vp3()
    p <- c_sp_pf_db_gg(sp, pf_db)
    p
  }
  
  
  c_sp_pf_db_plus_gg <- function(p, orig_data){
    validate(
      need(!is.null(c_search_gene()), "")
    )
    goi <- c_search_gene()
    searchgene <- orig_data[grepl(goi,orig_data$gene),]
    p1 <- search_scatter_gg(p, searchgene)
    p1
  }
  
  c1_sp_pf_db_plus_gg <- function(){
    sp <- c1_sp_pf_db_gg()
    d <- c_pd1()
    p <- c_sp_pf_db_plus_gg(sp, d)
    p
  }
  
  c2_sp_pf_db_plus_gg <- function(){
    sp <- c2_sp_pf_db_gg()
    d <- c_pd2()
    p <- c_sp_pf_db_plus_gg(sp, d)
    p
  }
  
  c3_sp_pf_db_plus_gg <- function(){
    sp <- c3_sp_pf_db_gg()
    d <- c_pd3()
    p <- c_sp_pf_db_plus_gg(sp, d)
    p
  }
  
  output$download_c_sp1_gg <- downloadHandler(
    filename = function() { paste("sp1_gg", '.png', sep='') },
    content = function(file) {
      if(is.null(c_pf_db())){
        if(is.null(c_search_gene())){
          ggsave(file, plot = c1_sp_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
        } else {
          ggsave(file, plot = c1_sp_plus_rep_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
        }
      } else if(!is.null(c_pf_db())){
        if(is.null(c_search_gene())){
          ggsave(file, plot = c1_sp_pf_db_gg(), device = "png", width = 5, height = 4, units = "in", dpi = 300)
        } else{
          ggsave(file, plot = c1_sp_pf_db_plus_gg(), device = "png", width = 5, height = 4, units = "in", dpi = 300)
        }
      }
    }
  )
  
  output$download_c_sp2_gg <- downloadHandler(
    filename = function() { paste("sp2_gg", '.png', sep='') },
    content = function(file) {
      if(is.null(c_pf_db())){
        if(is.null(c_search_gene())){
          ggsave(file, plot = c2_sp_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
        } else {
          ggsave(file, plot = c2_sp_plus_rep_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
        }
      } else if(!is.null(c_pf_db())){
        if(is.null(c_search_gene())){
          ggsave(file, plot = c2_sp_pf_db_gg(), device = "png", width = 5, height = 4, units = "in", dpi = 300)
        } else{
          ggsave(file, plot = c2_sp_pf_db_plus_gg(), device = "png", width = 5, height = 4, units = "in", dpi = 300)
        }
      }
    }
  )
  
  output$download_c_sp3_gg <- downloadHandler(
    filename = function() { paste("sp3_gg", '.png', sep='') },
    content = function(file) {
      if(is.null(c_pf_db())){
        if(is.null(c_search_gene())){
          ggsave(file, plot = c3_sp_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
        } else {
          ggsave(file, plot = c3_sp_plus_rep_gg(), device = "png", width = 4, height = 4, units = "in", dpi = 300)
        }
      } else if(!is.null(c_pf_db())){
        if(is.null(c_search_gene())){
          ggsave(file, plot = c3_sp_pf_db_gg(), device = "png", width = 5, height = 4, units = "in", dpi = 300)
        } else{
          ggsave(file, plot = c3_sp_pf_db_plus_gg(), device = "png", width = 5, height = 4, units = "in", dpi = 300)
        }
      }
    }
  )
  
  c_sp_cor <- function(input_file){
    if("rep1" %in% colnames(input_file) & "rep2" %in% colnames(input_file)){
      cor <- signif(cor(input_file$rep1, input_file$rep2), 4)
      cor <- paste0("correlation coefficient: ", cor)
    } else if ("logFC" %in% colnames(input_file) & "FDR" %in% colnames(input_file) & "pvalue" %in% colnames(input_file)){
      return(NULL)
    }
  }
  
  c_sp <- function(d, cc){
    if(input$c_colorscheme == "fdr"){
      req(input$c_color_indv_sig, input$c_color_indv_insig)
      d1 <- separate_to_groups_for_color_integrated(d, input$c_fdr_thresh, input$c_color_indv_sig, input$c_color_indv_insig)
      p <- plot_scatter_multiple_cond(d, d1)
    } else if(input$c_colorscheme == "exac"){
      d$s <- exac$em_p_hi[match(d$gene, exac$GENE_NAME)]
      d$s[is.na(d$s)] <- 2
      below_thresh <- subset(d, s < 0.9)
      above_thresh <- subset(d, s >= 0.9)
      no_exist <- subset(d, s == 2)
      p <- plot_scatter_exac_multi(d, below_thresh, above_thresh, no_exist)
      p
    } else if(input$c_colorscheme == "cbf"){
      d1 <- separate_to_groups_for_cbf_integrated(d, input$c_fdr_thresh)
      p <- plot_scatter_multiple_cond(d, d1)
      p
    } else if(input$c_colorscheme == "user"){
      validate(
        need(!is.null(input$c_file_color), "Please upload file with gene and score")
      )
      d1 <- c_in_file_color()
      col_theme <- input$c_colorbrewer_theme
      if(input$c_colorscheme_style == "cont"){
        df <- separate_to_groups_for_color_continuous(d, d1, col_theme)
        data <- df$df1
        data1 <- df$no_exist
      } else if(input$c_colorscheme_style == "disc"){
        df <- separate_to_groups_for_color_discrete(d, d1, col_theme)
        data <- df$df1
        data1 <- df$no_exist
      }
      p <- plot_ly(showlegend = FALSE, width = 320, height = 320) 
      p <- add_lines(p, data = d, x = ~c((min(rep1, rep2)), (max(rep1, rep2))), y = ~c((min(rep1, rep2)), (max(rep1, rep2))),
                     text = "x=y", hoverinfo = "text",
                     line = list(dash = "dash", width = 1, color = "#252525"), showlegend = FALSE)
      p <- add_markers(p, data = data1, x = ~rep1, y = ~rep2,
                       marker = list(size = 7, line = list(width=0.1, color = "grey89"), cmin = 0, cmax = 1, color = "#f7f7f7"),
                       opacity = 0.8,
                       text = ~paste(gene), hoverinfo = "text", name = paste0("Not in user data (", nrow(df$no_exist), ")"))
      for(i in nrow(data)){
        p <- add_markers(p, data = data, x = ~rep1, y = ~rep2,
                         marker = list(size = 7, cmin = 0, cmax = 1, color = ~col, line = list(width=0.2, color = "grey89")),
                         opacity = 1,
                         text = ~paste0(gene, ", FDR=", signif(FDR, digits = 3)), hoverinfo = "text",
                         name = paste0("Found in user data (", nrow(data), ")")) #, name = "pull down"
      }
      p
    }
    p <- p %>%
      layout(xaxis = list(title = "rep1", range=~c((min(d$rep1, d$rep2))-1, (max(d$rep1, d$rep2))+1)), 
             yaxis = list(title = "rep2", range=~c((min(d$rep1, d$rep2))-1, (max(d$rep1, d$rep2))+1)), 
             title = cc, titlefont = list(size=12))
  }
  
  c_sp1 <- reactive({
    data <- c_pd1()
    corr <- c_sp_cor(data)
    p <- c_sp(data, corr)
  })
  
  c_sp1_plus <- reactive({
    validate(
      need(!is.null(c_search_gene()), "")
    )
    p <- c_sp1()
    goi <- c_search_gene()
    orig_data <- c_pd1()
    searchgene <- orig_data[grepl(goi,orig_data$gene),]
    p1 <- search_scatter(p, searchgene)
    p1
  })
  
  c_sp2 <- reactive({
    data <- c_pd2()
    corr <- c_sp_cor(data)
    p <- c_sp(data, corr)
  })
  
  c_sp2_plus <- reactive({
    validate(
      need(!is.null(c_search_gene()), "")
    )
    p <- c_sp2()
    goi <- c_search_gene()
    orig_data <- c_pd2()
    searchgene <- orig_data[grepl(goi,orig_data$gene),]
    p1 <- search_scatter(p, searchgene)
    p1
  })
  
  c_sp3 <- reactive({
    data <- c_pd3()
    corr <- c_sp_cor(data)
    p <- c_sp(data, corr)
  })
  
  c_sp3_plus <- reactive({
    validate(
      need(!is.null(c_search_gene()), "")
    )
    p <- c_sp3()
    goi <- c_search_gene()
    orig_data <- c_pd3()
    searchgene <- orig_data[grepl(goi,orig_data$gene),]
    p1 <- search_scatter(p, searchgene)
    p1
  })
  
  output$c_pc_plot_button <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_file_pulldown2 != '', "")
    )
    actionButton("c_pc_make_plot", "Generate plots")
  })
  
  # c_f1 <- reactive({
  c_f1 <- eventReactive(input$c_pc_make_plot, {
    f1 <- c_pd1()
    f1_subset <- subset(f1, FDR <= input$c_compare1_FDR_range[2] & FDR >= input$c_compare1_FDR_range[1]
                        & pvalue <= input$c_compare1_pvalue_range[2] & pvalue >= input$c_compare1_pvalue_range[1]
                        & logFC <= input$c_f1_logFC_range[2] & logFC >= input$c_f1_logFC_range[1])
  })
  
  # c_f2 <- reactive({
  c_f2 <- eventReactive(input$c_pc_make_plot, {
    f2 <- c_pd2()
    f2_subset <- subset(f2, FDR <= input$c_compare2_FDR_range[2] & FDR >= input$c_compare2_FDR_range[1]
                        & pvalue <= input$c_compare2_pvalue_range[2] & pvalue >= input$c_compare2_pvalue_range[1]
                        & logFC <= input$c_f2_logFC_range[2] & logFC >= input$c_f2_logFC_range[1])
  })
  
  # c_f3 <- reactive({
  c_f3 <- eventReactive(input$c_pc_make_plot, {
    f3 <- c_pd3()
    f3_subset <- subset(f3, FDR <= input$c_compare3_FDR_range[2] & FDR >= input$c_compare3_FDR_range[1]
                        & pvalue <= input$c_compare3_pvalue_range[2] & pvalue >= input$c_compare3_pvalue_range[1]
                        & logFC <= input$c_f3_logFC_range[2] & logFC >= input$c_f3_logFC_range[1])
  })
  
  c_f1_pf <- reactive({
    f1 <- c_pd1()
    f1_subset <- subset(f1, FDR <= input$c_f1_pf_FDR[2] & FDR >= input$c_f1_pf_FDR[1]
                        & pvalue <= input$c_f1_pf_pvalue[2] & pvalue >= input$c_f1_pf_pvalue[1]
                        & logFC <= input$c_f1_pf_logFC_range[2] & logFC >= input$c_f1_pf_logFC_range[1])
  })
  
  c_f2_pf <- reactive({
    f2 <- c_pd2()
    f2_subset <- subset(f2, FDR <= input$c_f2_pf_FDR[2] & FDR >= input$c_f2_pf_FDR[1]
                        & pvalue <= input$c_f2_pf_pvalue[2] & pvalue >= input$c_f2_pf_pvalue[1]
                        & logFC <= input$c_f2_pf_logFC_range[2] & logFC >= input$c_f2_pf_logFC_range[1])
  })
  
  c_f3_pf <- reactive({
    f3 <- c_pd3()
    f3_subset <- subset(f3, FDR <= input$c_f3_pf_FDR[2] & FDR >= input$c_f3_pf_FDR[1]
                        & pvalue <= input$c_f3_pf_pvalue[2] & pvalue >= input$c_f3_pf_pvalue[1]
                        & logFC <= input$c_f3_pf_logFC_range[2] & logFC >= input$c_f3_pf_logFC_range[1])
  })
  
  #comparison for file1
  c_compare1 <- reactive({
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
      f1 <- c_pd1()
      f1_subset <- c_f1()
      f2_subset <- c_f2()
      overlap1_2 <- subset(f1_subset, gene %in% f2_subset$gene)
      f1_subset <- subset(f1_subset, gene %!in% (overlap1_2)$gene)
      list(f1=f1, f1_subset=f1_subset, overlap1_2=overlap1_2)
    } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      f1 <- c_pd1()
      f1_subset <- c_f1()
      f2_subset <- c_f2()
      f3_subset <- c_f3()
      overlap1_2 <- subset(f1_subset, gene %in% f2_subset$gene)
      overlap1_3 <- subset(f1_subset, gene %in% f3_subset$gene)
      overlap1_2_3 <- subset(f1_subset, gene %in% f2_subset$gene & gene %in% f3_subset$gene)
      f1_subset <- subset(f1_subset, gene %!in% (overlap1_2)$gene & gene %!in% (overlap1_3)$gene & gene %!in% (overlap1_2_3)$gene)
      overlap1_2 <- subset(overlap1_2, gene %!in% (overlap1_2_3)$gene)
      overlap1_3 <- subset(overlap1_3, gene %!in% (overlap1_2_3)$gene)
      list(f1=f1, f1_subset=f1_subset, overlap1_2=overlap1_2, overlap1_3=overlap1_3, overlap1_2_3=overlap1_2_3)
    }
  })
  
  #comparison for file2
  c_compare2 <- reactive({
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
      f2 <- c_pd2()
      f1_subset <- c_f1()
      f2_subset <- c_f2()
      overlap2_1 <- subset(f2_subset, gene %in% f1_subset$gene)
      f2_subset <- subset(f2_subset, gene %!in% (overlap2_1)$gene)
      list(f2=f2, f2_subset=f2_subset, overlap2_1=overlap2_1)
    } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      f2 <- c_pd2()
      f1_subset <- c_f1()
      f2_subset <- c_f2()
      f3_subset <- c_f3()
      overlap2_1 <- subset(f2_subset, gene %in% f1_subset$gene)
      overlap2_3 <- subset(f2_subset, gene %in% f3_subset$gene)
      overlap2_1_3 <- subset(f2_subset, gene %in% f1_subset$gene & gene %in% f3_subset$gene)
      f2_subset <- subset(f2_subset, gene %!in% (overlap2_1)$gene & gene %!in% (overlap2_3)$gene & gene %!in% (overlap2_1_3)$gene)
      overlap2_1 <- subset(overlap2_1, gene %!in% (overlap2_1_3)$gene)
      overlap2_3 <- subset(overlap2_3, gene %!in% (overlap2_1_3)$gene)
      list(f2=f2, f2_subset=f2_subset, overlap2_1=overlap2_1, overlap2_3=overlap2_3, overlap2_1_3=overlap2_1_3)
    }
  })
  
  #comparison for file3
  c_compare3 <- reactive({
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      f3 <- c_pd3()
      f1_subset <- c_f1()
      f2_subset <- c_f2()
      f3_subset <- c_f3()
      overlap3_1 <- subset(f3_subset, gene %in% f1_subset$gene)
      overlap3_2 <- subset(f3_subset, gene %in% f2_subset$gene)
      overlap3_1_2 <- subset(f3_subset, gene %in% f1_subset$gene & gene %in% f2_subset$gene)
      f3_subset <- subset(f3_subset, gene %!in% (overlap3_1)$gene & gene %!in% (overlap3_2)$gene & gene %!in% (overlap3_1_2)$gene)
      overlap3_1 <- subset(overlap3_1, gene %!in% (overlap3_1_2)$gene)
      overlap3_2 <- subset(overlap3_2, gene %!in% (overlap3_1_2)$gene)
      list(f3=f3, f3_subset=f3_subset, overlap3_1=overlap3_1, overlap3_2=overlap3_2, overlap3_1_2=overlap3_1_2)
    }
  })
  
  #comparison for file1
  c_compare1_pf <- reactive({
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
      f1 <- c_pd1()
      f1_subset <- c_f1_pf()
      f2_subset <- c_f2_pf()
      overlap1_2 <- subset(f1_subset, gene %in% f2_subset$gene)
      f1_subset <- subset(f1_subset, gene %!in% (overlap1_2)$gene)
      list(f1=f1, f1_subset=f1_subset, overlap1_2=overlap1_2)
    } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      f1 <- c_pd1()
      f1_subset <- c_f1_pf()
      f2_subset <- c_f2_pf()
      f3_subset <- c_f3_pf()
      overlap1_2 <- subset(f1_subset, gene %in% f2_subset$gene)
      overlap1_3 <- subset(f1_subset, gene %in% f3_subset$gene)
      overlap1_2_3 <- subset(f1_subset, gene %in% f2_subset$gene & gene %in% f3_subset$gene)
      f1_subset <- subset(f1_subset, gene %!in% (overlap1_2)$gene & gene %!in% (overlap1_3)$gene & gene %!in% (overlap1_2_3)$gene)
      overlap1_2 <- subset(overlap1_2, gene %!in% (overlap1_2_3)$gene)
      overlap1_3 <- subset(overlap1_3, gene %!in% (overlap1_2_3)$gene)
      list(f1=f1, f1_subset=f1_subset, overlap1_2=overlap1_2, overlap1_3=overlap1_3, overlap1_2_3=overlap1_2_3)
    }
  })
  
  #comparison for file2
  c_compare2_pf <- reactive({
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
      f2 <- c_pd2()
      f1_subset <- c_f1_pf()
      f2_subset <- c_f2_pf()
      overlap2_1 <- subset(f2_subset, gene %in% f1_subset$gene)
      f2_subset <- subset(f2_subset, gene %!in% (overlap2_1)$gene)
      list(f2=f2, f2_subset=f2_subset, overlap2_1=overlap2_1)
    } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      f2 <- c_pd2()
      f1_subset <- c_f1_pf()
      f2_subset <- c_f2_pf()
      f3_subset <- c_f3_pf()
      overlap2_1 <- subset(f2_subset, gene %in% f1_subset$gene)
      overlap2_3 <- subset(f2_subset, gene %in% f3_subset$gene)
      overlap2_1_3 <- subset(f2_subset, gene %in% f1_subset$gene & gene %in% f3_subset$gene)
      f2_subset <- subset(f2_subset, gene %!in% (overlap2_1)$gene & gene %!in% (overlap2_3)$gene & gene %!in% (overlap2_1_3)$gene)
      overlap2_1 <- subset(overlap2_1, gene %!in% (overlap2_1_3)$gene)
      overlap2_3 <- subset(overlap2_3, gene %!in% (overlap2_1_3)$gene)
      list(f2=f2, f2_subset=f2_subset, overlap2_1=overlap2_1, overlap2_3=overlap2_3, overlap2_1_3=overlap2_1_3)
    }
  })
  
  #comparison for file3
  c_compare3_pf <- reactive({
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      f3 <- c_pd3()
      f1_subset <- c_f1_pf()
      f2_subset <- c_f2_pf()
      f3_subset <- c_f3_pf()
      overlap3_1 <- subset(f3_subset, gene %in% f1_subset$gene)
      overlap3_2 <- subset(f3_subset, gene %in% f2_subset$gene)
      overlap3_1_2 <- subset(f3_subset, gene %in% f1_subset$gene & gene %in% f2_subset$gene)
      f3_subset <- subset(f3_subset, gene %!in% (overlap3_1)$gene & gene %!in% (overlap3_2)$gene & gene %!in% (overlap3_1_2)$gene)
      overlap3_1 <- subset(overlap3_1, gene %!in% (overlap3_1_2)$gene)
      overlap3_2 <- subset(overlap3_2, gene %!in% (overlap3_1_2)$gene)
      list(f3=f3, f3_subset=f3_subset, overlap3_1=overlap3_1, overlap3_2=overlap3_2, overlap3_1_2=overlap3_1_2)
    }
  })
  
  c_compare1_plot <- eventReactive(input$c_pc_make_plot, { #reactive({
    d <- c_compare1()
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
      f1 <- d$f1
      f1_subset <- d$f1_subset
      overlap1_2 <- d$overlap1_2
      p1 <- compare_two_files_a(f1, f1_subset, overlap1_2)
    } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      f1 <- d$f1
      f1_subset <- d$f1_subset
      overlap1_2 <- d$overlap1_2
      overlap1_3 <- d$overlap1_3
      overlap1_2_3 <- d$overlap1_2_3
      p1 <- compare_two_files_aa(f1, f1_subset, overlap1_2, overlap1_3, overlap1_2_3)
    }
  })
  
  c_compare1_plot_scatter <- eventReactive(input$c_pc_make_plot, { #reactive({
    d <- c_compare1()
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
      f1 <- d$f1
      f1_subset <- d$f1_subset
      overlap1_2 <- d$overlap1_2
      p1 <- compare_two_files_a_scatter(f1, f1_subset, overlap1_2)
    } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      f1 <- d$f1
      f1_subset <- d$f1_subset
      overlap1_2 <- d$overlap1_2
      overlap1_3 <- d$overlap1_3
      overlap1_2_3 <- d$overlap1_2_3
      p1 <- compare_two_files_aa_scatter(f1, f1_subset, overlap1_2, overlap1_3, overlap1_2_3)
    }
  })
  
  c_compare1_plus <- reactive({
    validate(
      need(!is.null(c_search_gene()), "")
    )
    p <- c_compare1_plot()
    goi <- c_search_gene()
    orig_data <- c_pd1()
    searchgene <- orig_data[grepl(goi,orig_data$gene),]
    p1 <- search_volcano(p, searchgene)
    p1
  })
  
  c_compare1_plus_scatter <- reactive({
    validate(
      need(!is.null(c_search_gene()), "")
    )
    p <- c_compare1_plot_scatter()
    goi <- c_search_gene()
    orig_data <- c_pd1()
    searchgene <- orig_data[grepl(goi,orig_data$gene),]
    p1 <- search_scatter(p, searchgene)
    p1
  })
  
  c_compare2_plot <- eventReactive(input$c_pc_make_plot, { #reactive({
    d <- c_compare2()
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
      f2 <- d$f2
      f2_subset <- d$f2_subset
      overlap2_1 <- d$overlap2_1
      p2 <- compare_two_files_b(f2, f2_subset, overlap2_1)
    } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      f2 <- d$f2
      f2_subset <- d$f2_subset
      overlap2_1 <- d$overlap2_1
      overlap2_3 <- d$overlap2_3
      overlap2_1_3 <- d$overlap2_1_3
      p2 <- compare_two_files_bb(f2, f2_subset, overlap2_1, overlap2_3, overlap2_1_3)
    }
  })
  
  c_compare2_plot_scatter <- eventReactive(input$c_pc_make_plot, { #reactive({
    d <- c_compare2()
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
      f2 <- d$f2
      f2_subset <- d$f2_subset
      overlap2_1 <- d$overlap2_1
      p2 <- compare_two_files_b_scatter(f2, f2_subset, overlap2_1)
    } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      f2 <- d$f2
      f2_subset <- d$f2_subset
      overlap2_1 <- d$overlap2_1
      overlap2_3 <- d$overlap2_3
      overlap2_1_3 <- d$overlap2_1_3
      p2 <- compare_two_files_bb_scatter(f2, f2_subset, overlap2_1, overlap2_3, overlap2_1_3)
    }
  })
  
  c_compare2_plus <- reactive({
    validate(
      need(!is.null(c_search_gene()), "")
    )
    p <- c_compare2_plot()
    goi <- c_search_gene()
    orig_data <- c_pd2()
    searchgene <- orig_data[grepl(goi,orig_data$gene),]
    p2 <- search_volcano(p, searchgene)
    p2
  })
  
  c_compare2_plus_scatter <- reactive({
    validate(
      need(!is.null(c_search_gene()), "")
    )
    p <- c_compare2_plot_scatter()
    goi <- c_search_gene()
    orig_data <- c_pd2()
    searchgene <- orig_data[grepl(goi,orig_data$gene),]
    p2 <- search_scatter(p, searchgene)
    p2
  })
  
  c_compare3_plot <- eventReactive(input$c_pc_make_plot, { #reactive({
    d <- c_compare3()
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      f3 <- d$f3
      f3_subset <- d$f3_subset
      overlap3_1 <- d$overlap3_1
      overlap3_2 <- d$overlap3_2
      overlap3_1_2 <- d$overlap3_1_2
      p3 <- compare_two_files_cc(f3, f3_subset, overlap3_1, overlap3_2, overlap3_1_2)
    }
  })
  
  c_compare3_plot_scatter <- eventReactive(input$c_pc_make_plot, { #reactive({
    d <- c_compare3()
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      f3 <- d$f3
      f3_subset <- d$f3_subset
      overlap3_1 <- d$overlap3_1
      overlap3_2 <- d$overlap3_2
      overlap3_1_2 <- d$overlap3_1_2
      p3 <- compare_two_files_cc_scatter(f3, f3_subset, overlap3_1, overlap3_2, overlap3_1_2)
    }
  })
  
  c_compare3_plus <- reactive({
    validate(
      need(!is.null(c_search_gene()), "")
    )
    p <- c_compare3_plot()
    goi <- c_search_gene()
    orig_data <- c_pd3()
    searchgene <- orig_data[grepl(goi,orig_data$gene),]
    p3 <- search_volcano(p, searchgene)
    p3
  })
  
  c_compare3_plus_scatter <- reactive({
    validate(
      need(!is.null(c_search_gene()), "")
    )
    p <- c_compare3_plot_scatter()
    goi <- c_search_gene()
    orig_data <- c_pd3()
    searchgene <- orig_data[grepl(goi,orig_data$gene),]
    p3 <- search_scatter(p, searchgene)
    p3
  })
  
  c_compare1_pfe_plot <- eventReactive(input$c_make_bpf, {
    d <- c_compare1_pf()
    pfmsizing <- c_PF_marker()
    pfsort <- c_PF_sorting()
    increase_size <- input$c_PF_marker_freq
    data_selection <- c_pf_data_selection()
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
      withProgress(message = 'This may take a while', 
                   detail = 'Hold please', value = 0, {
                     pf_f1 <- d$f1
                     pf_f1s <- d$f1_subset
                     pf_overlap1_2 <- d$overlap1_2
                     incProgress(0.2)
                     #f1 subset
                     f1s_families <- assignFamily_inc_doubles(pf_f1s, data_selection)
                     f2s_families <- f1s_families
                     # f2s_families$logFC <- pf_f1$logFC[match(f2s_families$gene, pf_f1$gene)]
                     # f1s_gna <- addNames(pf_f1s)
                     incProgress(0.5)
                     #f1 and f2 overlap
                     o12_families <- assignFamily_inc_doubles(pf_overlap1_2, data_selection)
                     # o12_gna <- addNames(pf_overlap1_2)
                     incProgress(0.8)
                     f1_subset <- makePlotFamilies_1quadrant(f1s_families, pf_f1, pfsort) #f1s_gna, 
                     overlap1_2 <- makePlotFamilies_1quadrant(o12_families, pf_f1, pfsort) #o12_gna, 
                     incProgress(0.9)
                     if (pfmsizing == "change"){
                       p1 <- compare_two_files_pf_a_size(pf_f1, f1_subset, overlap1_2, increase_size) #, f1_subset[[2]] , overlap1_2[[2]]
                     } else if (pfmsizing == "no_change"){
                       p1 <- compare_two_files_pf_a(pf_f1, f1_subset, overlap1_2) #, f1_subset[[2]] , overlap1_2[[2]]
                     }
                   })
    } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      withProgress(message = 'This may take a while', 
                   detail = 'Hold please', value = 0, {
                     pf_f1 <- d$f1
                     pf_f1s <- d$f1_subset
                     pf_overlap1_2 <- d$overlap1_2
                     pf_overlap1_3 <- d$overlap1_3
                     pf_overlap1_2_3 <- d$overlap1_2_3
                     incProgress(0.1)
                     #f1 subset
                     f1s_families <- assignFamily_inc_doubles(pf_f1s, data_selection)
                     # f1s_gna <- addNames(pf_f1s)
                     incProgress(0.3)
                     #f1 and f2 overlap
                     o12_families <- assignFamily_inc_doubles(pf_overlap1_2, data_selection)
                     # o12_gna <- addNames(pf_overlap1_2)
                     incProgress(0.5)
                     #f1 and f3 overlap
                     o13_families <- assignFamily_inc_doubles(pf_overlap1_3, data_selection)
                     # o13_gna <- addNames(pf_overlap1_3)
                     incProgress(0.7)
                     #f1 and f2 overlap
                     o123_families <- assignFamily_inc_doubles(pf_overlap1_2_3, data_selection)
                     # o123_gna <- addNames(pf_overlap1_2_3)
                     incProgress(0.9)
                     f1_subset <- makePlotFamilies_1quadrant(f1s_families, pf_f1, pfsort) #f1s_gna, 
                     overlap1_2 <- makePlotFamilies_1quadrant(o12_families, pf_f1, pfsort) #o12_gna,
                     overlap1_3 <- makePlotFamilies_1quadrant(o13_families, pf_f1, pfsort) #o13_gna, 
                     overlap1_2_3 <- makePlotFamilies_1quadrant(o123_families, pf_f1, pfsort) #o123_gna, 
                     
                     if (pfmsizing == "change"){
                       p1 <- compare_two_files_pf_aa_size(pf_f1, f1_subset, overlap1_2, #f1_subset[[2]], overlap1_2[[2]], 
                                                          overlap1_3, overlap1_2_3, increase_size) #overlap1_3[[2]], overlap1_2_3[[2]], 
                     } else{
                       p1 <- compare_two_files_pf_aa(pf_f1, f1_subset, overlap1_2, #f1_subset[[2]], overlap1_2[[2]],
                                                     overlap1_3, overlap1_2_3) #overlap1_3[[2]], , overlap1_2_3[[2]]
                     }
                   })
    }
  })
  
  c_compare2_pfe_plot <- eventReactive(input$c_make_bpf, {
    d <- c_compare2_pf()
    pfmsizing <- c_PF_marker()
    pfsort <- c_PF_sorting()
    increase_size <- input$c_PF_marker_freq
    data_selection <- c_pf_data_selection()
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
      withProgress(message = 'This may take a while', 
                   detail = 'Hold please', value = 0, {
                     pf_f2 <- d$f2
                     pf_f2s <- d$f2_subset
                     pf_overlap2_1 <- d$overlap2_1
                     incProgress(0.2)
                     #f2 subset
                     f2s_families <- assignFamily_inc_doubles(pf_f2s, data_selection)
                     # f2s_gna <- addNames(pf_f2s)
                     incProgress(0.5)
                     #f1 and f2 overlap
                     o21_families <- assignFamily_inc_doubles(pf_overlap2_1, data_selection)
                     # o21_gna <- addNames(pf_overlap2_1)
                     incProgress(0.8)
                     f2_subset <- makePlotFamilies_1quadrant(f2s_families, pf_f2, pfsort) #, f2s_gna, 
                     overlap2_1 <- makePlotFamilies_1quadrant(o21_families, pf_f2, pfsort)#, o21_gna
                     incProgress(0.9)
                     if (pfmsizing == "change"){
                       p1 <- compare_two_files_pf_b_size(pf_f2, f2_subset, overlap2_1, increase_size) #, f2_subset[[2]] , overlap2_1[[2]]
                     } else if (pfmsizing == "no_change"){
                       p1 <- compare_two_files_pf_b(pf_f2, f2_subset, overlap2_1) #, f2_subset[[2]] , overlap2_1[[2]]
                     }
                   })
    } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      withProgress(message = 'This may take a while', 
                   detail = 'Hold please', value = 0, {
                     pf_f2 <- d$f2
                     pf_f2s <- d$f2_subset
                     pf_overlap2_1 <- d$overlap2_1
                     pf_overlap2_3 <- d$overlap2_3
                     pf_overlap2_1_3 <- d$overlap2_1_3
                     incProgress(0.1)
                     #f2 subset
                     f2s_families <- assignFamily_inc_doubles(pf_f2s, data_selection)
                     # f2s_gna <- addNames(pf_f2s)
                     incProgress(0.3)
                     #f1 and f2 overlap
                     o21_families <- assignFamily_inc_doubles(pf_overlap2_1, data_selection)
                     # o21_gna <- addNames(pf_overlap2_1)
                     incProgress(0.5)
                     #f1 and f3 overlap
                     o23_families <- assignFamily_inc_doubles(pf_overlap2_3, data_selection)
                     # o23_gna <- addNames(pf_overlap2_3)
                     incProgress(0.7)
                     #f1, f2 and f3 overlap
                     o213_families <- assignFamily_inc_doubles(pf_overlap2_1_3, data_selection)
                     # o213_gna <- addNames(pf_overlap2_1_3)
                     incProgress(0.9)
                     f2_subset <- makePlotFamilies_1quadrant(f2s_families, pf_f2, pfsort) #, f2s_gna
                     overlap2_1 <- makePlotFamilies_1quadrant(o21_families, pf_f2, pfsort) #, o21_gna
                     overlap2_3 <- makePlotFamilies_1quadrant(o23_families, pf_f2, pfsort) #, o23_gna
                     overlap2_1_3 <- makePlotFamilies_1quadrant(o213_families, pf_f2, pfsort) #, o213_gna
                     
                     if (pfmsizing == "change"){
                       p1 <- compare_two_files_pf_bb_size(pf_f2, f2_subset, overlap2_1, #, f2_subset[[2]] , overlap2_1[[2]]
                                                          overlap2_3, overlap2_1_3, increase_size) #, overlap2_3[[2]] , overlap2_1_3[[2]]
                     } else{
                       p1 <- compare_two_files_pf_bb(pf_f2, f2_subset, overlap2_1, #, f2_subset[[2]] , overlap2_1[[2]]
                                                     overlap2_3, overlap2_1_3) #, overlap2_3[[2]] , overlap2_1_3[[2]]
                     }
                   })
    }
  })
  
  c_compare3_pfe_plot <- eventReactive(input$c_make_bpf, {
    d <- c_compare3_pf()
    pfmsizing <- c_PF_marker()
    pfsort <- c_PF_sorting()
    increase_size <- input$c_PF_marker_freq
    data_selection <- c_pf_data_selection()
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      withProgress(message = 'This may take a while', 
                   detail = 'Hold please', value = 0, {
                     pf_f3 <- d$f3
                     pf_f3s <- d$f3_subset
                     pf_overlap3_1 <- d$overlap3_1
                     pf_overlap3_2 <- d$overlap3_2
                     pf_overlap3_1_2 <- d$overlap3_1_2
                     incProgress(0.1)
                     #f2 subset
                     f3s_families <- assignFamily_inc_doubles(pf_f3s, data_selection)
                     # f3s_gna <- addNames(pf_f3s)
                     incProgress(0.3)
                     #f1 and f2 overlap
                     o31_families <- assignFamily_inc_doubles(pf_overlap3_1, data_selection)
                     # o31_gna <- addNames(pf_overlap3_1)
                     incProgress(0.5)
                     #f1 and f3 overlap
                     o32_families <- assignFamily_inc_doubles(pf_overlap3_2, data_selection)
                     # o32_gna <- addNames(pf_overlap3_2)
                     incProgress(0.7)
                     #f1, f2 and f3 overlap
                     o312_families <- assignFamily_inc_doubles(pf_overlap3_1_2, data_selection)
                     # o312_gna <- addNames(pf_overlap3_1_2)
                     incProgress(0.9)
                     f3_subset <- makePlotFamilies_1quadrant(f3s_families, pf_f3, pfsort) #, f3s_gna
                     overlap3_1 <- makePlotFamilies_1quadrant(o31_families, pf_f3, pfsort) #, o31_gna
                     overlap3_2 <- makePlotFamilies_1quadrant(o32_families, pf_f3, pfsort) #, o32_gna
                     overlap3_1_2 <- makePlotFamilies_1quadrant(o312_families, pf_f3, pfsort) #, o312_gna
                     
                     if (pfmsizing == "change"){
                       p1 <- compare_two_files_pf_cc_size(pf_f3, f3_subset, overlap3_1, #, f3_subset[[2]] overlap3_1[[2]],
                                                          overlap3_2, overlap3_1_2, increase_size) #overlap3_2[[2]], , overlap3_1_2[[2]]
                     } else{
                       p1 <- compare_two_files_pf_cc(pf_f3, f3_subset, overlap3_1, #, f3_subset[[2]] overlap3_1[[2]],
                                                     overlap3_2, overlap3_1_2) #, overlap3_2[[2]] , overlap3_1_2[[2]]
                     }
                   })
    }
  })
  
  c_f1_unique <- reactive({
    d <- c_compare1()
    f1 <- d$f1_subset
    d1 <- data.frame(f1$gene)
    colnames(d1) <- c("f1")    
    d1 <- unique(sort(d1$f1))
  })
  
  c_f2_unique <- reactive({
    d <- c_compare2()
    f2 <- d$f2_subset
    d1 <- data.frame(f2$gene)
    colnames(d1) <- c("f2")    
    d1 <- unique(sort(d1$f2))
  })
  
  c_f1_and_f2 <- reactive({
    d <- c_compare2()
    f2 <- d$overlap2_1
    d1 <- data.frame(f2$gene)
    colnames(d1) <- c("f12")    
    d1 <- unique(sort(d1$f12))
  })
  
  c_f3_unique <- reactive({
    d <- c_compare3()
    f3 <- d$f3_subset
    d1 <- data.frame(f3$gene)
    colnames(d1) <- c("f3")    
    d1 <- unique(sort(d1$f3))
  })
  
  c_f1_and_f3 <- reactive({
    d <- c_compare3()
    f3 <- d$overlap3_1
    d1 <- data.frame(f3$gene)
    colnames(d1) <- c("f13")    
    d1 <- unique(sort(d1$f13))
  })
  
  c_f2_and_f3 <- reactive({
    d <- c_compare3()
    f3 <- d$overlap3_2
    d1 <- data.frame(f3$gene)
    colnames(d1) <- c("f23")    
    d1 <- unique(sort(d1$f23))
  })
  
  c_f1_f2_and_f3 <- reactive({
    d <- c_compare3()
    d1 <- data.frame((d$overlap3_1_2)$gene)
    colnames(d1) <- c("f123")    
    d1 <- unique(sort(d1$f123))
  })
  
  c_unique_dat <- eventReactive(input$c_pc_make_plot, { #reactive({
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
      f1 <- c_f1_unique()
      f2 <- c_f2_unique()
      f12 <- c_f1_and_f2()
      n <- max(length(f1), length(f2), length(f12))
      length(f1) <- n
      length(f2) <- n
      length(f12) <- n
      d <- cbind(f1, f2, f12)
      d[is.na(d)] <- " "
    } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      f1 <- c_f1_unique()
      f2 <- c_f2_unique()
      f12 <- c_f1_and_f2()
      f3 <- c_f3_unique()
      f13 <- c_f1_and_f3()
      f23 <- c_f2_and_f3()
      f123 <- c_f1_f2_and_f3()
      n <- max(length(f1), length(f2), length(f12), length(f3), length(f13), length(f23), length(f123))
      length(f1) <- n
      length(f2) <- n
      length(f12) <- n
      length(f3) <- n
      length(f13) <- n
      length(f23) <- n
      length(f123) <- n
      d <- cbind(f1, f2, f12, f3, f13, f23, f123)
      d[is.na(d)] <- " "
    }
    as.data.frame(d)
  })
  
  c_venndiagram <- eventReactive(input$c_pc_make_plot, { #reactive({
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
      f1_subset <- c_f1()
      f2_subset <- c_f2()
      x <- list()
      x[["File1"]] <- f1_subset$gene
      x[["File2"]] <- f2_subset$gene
      
      colors2 = c("#e41a1c", "#ffff33")
      
      v0 <- venn.diagram(x,
                         fill = colors2, margin=0.05, filename = NULL, resolution = 900, height = 400, force.unique = T,
                         sub = " ", sub.pos = c(0, 0), euler.d = F, scaled = F,
                         cat.cex = 1.1, cex = 2, cat.pos = c(180,180), cat.dist = c(0.05,0.05),
                         fontfamily = 'sans', cat.fontfamily = 'sans', main.fontfamily = 'sans')
    } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      f1_subset <- c_f1()
      f2_subset <- c_f2()
      f3_subset <- c_f3()
      x <- list()
      x[["File1"]] <- f1_subset$gene
      x[["File2"]] <- f2_subset$gene
      x[["File3"]] <- f3_subset$gene
      
      colors3 = c("#e41a1c", "#ffff33", "#1f78b4")
      
      v0 <- venn.diagram(x, 
                         fill = colors3, margin=0.05, filename = NULL, resolution = 900, height = 400, force.unique = T,
                         sub = " ", sub.pos = c(0, 0, 0), euler.d = F, scaled = F,
                         # cat.cex = 1.1, cex = 2, cat.pos = c(180,180,0), cat.dist = c(0.05,0.05,0.05),
                         fontfamily = 'sans', cat.fontfamily = 'sans', main.fontfamily = 'sans')
    }
    v0
  })
  
  #comparison for file1 inweb
  c_inweb_compare1 <- reactive({
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
      f1_inweb <- c_inweb_pd1()
      f1_inweb <- subset(f1_inweb, pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1])
      f2_inweb <- c_inweb_pd2()
      f2_inweb <- subset(f2_inweb, pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1])
      overlap1_2 <- subset(f1_inweb, gene %in% f2_inweb$gene)
      list(f1_inweb=f1_inweb, overlap1_2=overlap1_2)
    } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      f1_inweb <- c_inweb_pd1()
      f1_inweb <- subset(f1_inweb, pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1])
      f2_inweb <- c_inweb_pd2()
      f2_inweb <- subset(f2_inweb, pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1])
      f3_inweb <- c_inweb_pd3()
      f3_inweb <- subset(f3_inweb, pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1])
      overlap1_2 <- subset(f1_inweb, gene %in% f2_inweb$gene)
      overlap1_3 <- subset(f1_inweb, gene %in% f3_inweb$gene)
      overlap1_2_3 <- subset(f1_inweb, gene %in% f2_inweb$gene & gene %in% f3_inweb$gene)
      list(f1_inweb=f1_inweb, overlap1_2=overlap1_2, overlap1_3=overlap1_3, overlap1_2_3=overlap1_2_3)
    }
  })
  
  #comparison for file2 inweb
  c_inweb_compare2 <- reactive({
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
      f1_inweb <- c_inweb_pd1()
      f1_inweb <- subset(f1_inweb, pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1])
      f2_inweb <- c_inweb_pd2()
      f2_inweb <- subset(f2_inweb, pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1])
      overlap2_1 <- subset(f2_inweb, gene %in% f1_inweb$gene)
      list(f2_inweb=f2_inweb, overlap2_1=overlap2_1)
    } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      f1_inweb <- c_inweb_pd1()
      f1_inweb <- subset(f1_inweb, pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1])
      f2_inweb <- c_inweb_pd2()
      f2_inweb <- subset(f2_inweb, pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1])
      f3_inweb <- c_inweb_pd3()
      f3_inweb <- subset(f3_inweb, pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1])
      overlap2_1 <- subset(f2_inweb, gene %in% f1_inweb$gene)
      overlap2_3 <- subset(f2_inweb, gene %in% f3_inweb$gene)
      overlap2_1_3 <- subset(f2_inweb, gene %in% f1_inweb$gene & gene %in% f3_inweb$gene)
      list(f2_inweb=f2_inweb, overlap2_1=overlap2_1, overlap2_3=overlap2_3, overlap2_1_3=overlap2_1_3)
    }
  })
  
  #comparison for file3 inweb
  c_inweb_compare3 <- reactive({
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      f1_inweb <- c_inweb_pd1()
      f1_inweb <- subset(f1_inweb, pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1])
      f2_inweb <- c_inweb_pd2()
      f2_inweb <- subset(f2_inweb, pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1])
      f3_inweb <- c_inweb_pd3()
      f3_inweb <- subset(f3_inweb, pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1])
      overlap3_1 <- subset(f3_inweb, gene %in% f1_inweb$gene)
      overlap3_2 <- subset(f3_inweb, gene %in% f2_inweb$gene)
      overlap3_1_2 <- subset(f3_inweb, gene %in% f1_inweb$gene & gene %in% f2_inweb$gene)
      list(f3_inweb=f3_inweb, overlap3_1=overlap3_1, overlap3_2=overlap3_2, overlap3_1_2=overlap3_1_2)
    }
  })
  
  c_f1_unique_inweb <- reactive({
    d <- c_inweb_compare1()
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
      f1 <- subset(d$f1_inweb, gene %!in% (d$overlap1_2)$gene)
    } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      f1 <- subset(d$f1_inweb, gene %!in% (d$overlap1_2)$gene & gene %!in% (d$overlap1_3)$gene & gene %!in% (d$overlap1_2_3)$gene)
    }
    d1 <- data.frame(f1$gene)
    colnames(d1) <- c("f1")    
    d1 <- unique(sort(d1$f1))
  })
  
  c_f2_unique_inweb <- reactive({
    d <- c_inweb_compare2()
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
      f2 <- subset(d$f2_inweb, gene %!in% (d$overlap2_1)$gene)
    } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      f2 <- subset(d$f2_inweb, gene %!in% (d$overlap2_1)$gene & gene %!in% (d$overlap2_3)$gene & gene %!in% (d$overlap2_1_3)$gene)
    }
    d1 <- data.frame(f2$gene)
    colnames(d1) <- c("f2")    
    d1 <- unique(sort(d1$f2))
  })
  
  c_f1_and_f2_inweb <- reactive({
    d <- c_inweb_compare2()
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
      f2 <- d$overlap2_1
    } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      f2 <- subset(d$overlap2_1, gene %!in% (d$overlap2_1_3)$gene)
    }
    d1 <- data.frame(f2$gene)
    colnames(d1) <- c("f12")    
    d1 <- unique(sort(d1$f12))
  })
  
  c_f3_unique_inweb <- reactive({
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      d <- c_inweb_compare3()
      f3 <- subset(d$f3_inweb, gene %!in% (d$overlap3_1)$gene & gene %!in% (d$overlap3_2)$gene & gene %!in% (d$overlap3_1_2)$gene)
      d1 <- data.frame(f3$gene)
      colnames(d1) <- c("f3")    
      d1 <- unique(sort(d1$f3))
    }
  })
  
  c_f1_and_f3_inweb <- reactive({
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      d <- c_inweb_compare3()
      f3 <- subset(d$overlap3_1, gene %!in% (d$overlap3_1_2)$gene)
      d1 <- data.frame(f3$gene)
      colnames(d1) <- c("f13")    
      d1 <- unique(sort(d1$f13))
    }
  })
  
  c_f2_and_f3_inweb <- reactive({
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      d <- c_inweb_compare3()
      f3 <- subset(d$overlap3_2, gene %!in% (d$overlap3_1_2)$gene)
      d1 <- data.frame(f3$gene)
      colnames(d1) <- c("f23")    
      d1 <- unique(sort(d1$f23))
    }
  })
  
  c_f1_f2_and_f3_inweb <- reactive({
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      d <- c_inweb_compare3()
      d1 <- data.frame((d$overlap3_1_2)$gene)
      colnames(d1) <- c("f123")    
      d1 <- unique(sort(d1$f123))
    }
  })
  
  c_unique_inweb_dat <- reactive({
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
      f1 <- c_f1_unique_inweb()
      f2 <- c_f2_unique_inweb()
      f12 <- c_f1_and_f2_inweb()
      n <- max(length(f1), length(f2), length(f12))
      length(f1) <- n
      length(f2) <- n
      length(f12) <- n
      d <- cbind(f1, f2, f12)
      d[is.na(d)] <- " "
    } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      f1 <- c_f1_unique_inweb()
      f2 <- c_f2_unique_inweb()
      f12 <- c_f1_and_f2_inweb()
      f3 <- c_f3_unique_inweb()
      f13 <- c_f1_and_f3_inweb()
      f23 <- c_f2_and_f3_inweb()
      f123 <- c_f1_f2_and_f3_inweb()
      n <- max(length(f1), length(f2), length(f12), length(f3), length(f13), length(f23), length(f123))
      length(f1) <- n
      length(f2) <- n
      length(f12) <- n
      length(f3) <- n
      length(f13) <- n
      length(f23) <- n
      length(f123) <- n
      d <- cbind(f1, f2, f12, f3, f13, f23, f123)
      d[is.na(d)] <- " "
    }
    as.data.frame(d)
  })
  
  c_venndiagram_inweb <- reactive({
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
      if(!is.null(c_inweb_pd1()) & !is.null(c_in_pd2())){
        f1_inweb <- c_inweb_pd1()
        f1_inweb <- subset(f1_inweb, pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1])
        f2_inweb <- c_inweb_pd2()
        f2_inweb <- subset(f2_inweb, pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1])
        x <- list()
        x[["File1"]] <- f1_inweb$gene
        x[["File2"]] <- f2_inweb$gene
        
        colors2 = c("gray", "gray")
        
        v0 <- venn.diagram(x,
                           fill = colors2, margin=0.05, filename = NULL, resolution = 900, height = 400, force.unique = T,
                           sub = " ", sub.pos = c(0, 0), euler.d = F, scaled = F,
                           cat.cex = 1.1, cex = 2, cat.pos = c(180,180), cat.dist = c(0.05,0.05),
                           fontfamily = 'sans', cat.fontfamily = 'sans', main.fontfamily = 'sans')
      }
    } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      if(!is.null(c_inweb_pd1()) & !is.null(c_in_pd2()) & !is.null(c_in_pd3())){
        f1_inweb <- c_inweb_pd1()
        f1_inweb <- subset(f1_inweb, pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1])
        f2_inweb <- c_inweb_pd2()
        f2_inweb <- subset(f2_inweb, pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1])
        f3_inweb <- c_inweb_pd3()
        f3_inweb <- subset(f3_inweb, pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1])
        x <- list()
        x[["File1"]] <- f1_inweb$gene
        x[["File2"]] <- f2_inweb$gene
        x[["File3"]] <- f3_inweb$gene
        
        colors3 = c("gray", "gray", "gray")
        
        v0 <- venn.diagram(x, 
                           fill = colors3, margin=0.05, filename = NULL, resolution = 900, height = 400, force.unique = T,
                           sub = " ", sub.pos = c(0, 0, 0), euler.d = F, scaled = F,
                           # cat.cex = 1.1, cex = 2, cat.pos = c(180,180,0), cat.dist = c(0.05,0.05,0.05),
                           fontfamily = 'sans', cat.fontfamily = 'sans', main.fontfamily = 'sans')
      }
    }
  })
  
  c_inweb_hypergeometric <- function(d, inweb){
    pop <- subset(d, d$gene %in% inweb_combined$V1)
    pop1 <- unique(sort(pop$gene))
    sample <- inweb
    sample1 <- unique(sort(inweb$gene))
    # sample <- subset(pop, pop$gene %in% inweb$gene)
    success_pop <- subset(pop, pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1])
    success_pop1 <- unique(sort(success_pop$gene))
    success_samp <- subset(sample, sample$gene %in% success_pop$gene)
    success_samp1 <- unique(sort(success_samp$gene))
    rownames(success_samp) <- NULL
    samp_l <- length(sample1)
    pop_l <- length(pop1)
    success_pop_l <- length(success_pop1)
    success_samp_l <- length(success_samp1)
    pvalue <- phyper((success_samp_l-1), samp_l, (pop_l-samp_l), success_pop_l, lower.tail = F)
    pvalue <- signif(pvalue, 4)
  }
  
  c_pd1_inweb_hypergeometric <- reactive({
    d <- c_pd1()
    inweb <- c_inweb_pd1()
    pop <- subset(d, d$gene %in% inweb_combined$V1)
    pop1 <- unique(sort(pop$gene))
    sample <- inweb
    sample1 <- unique(sort(inweb$gene))
    # sample <- subset(pop, pop$gene %in% inweb$gene)
    success_pop <- subset(pop, pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1])
    success_pop1 <- unique(sort(success_pop$gene))
    success_samp <- subset(sample, sample$gene %in% success_pop$gene)
    success_samp1 <- unique(sort(success_samp$gene))
    rownames(success_samp) <- NULL
    samp_l <- length(sample1)
    pop_l <- length(pop1)
    success_pop_l <- length(success_pop1)
    success_samp_l <- length(success_samp1)
    pvalue <- phyper((success_samp_l-1), samp_l, (pop_l-samp_l), success_pop_l, lower.tail = F)
    pvalue <- signif(pvalue, 4)
  })
  
  c_pd2_inweb_hypergeometric <- reactive({
    d <- c_pd2()
    inweb <- c_inweb_pd2()
    pop <- subset(d, d$gene %in% inweb_combined$V1)
    pop1 <- unique(sort(pop$gene))
    sample <- inweb
    sample1 <- unique(sort(inweb$gene))
    # sample <- subset(pop, pop$gene %in% inweb$gene)
    success_pop <- subset(pop, pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1])
    success_pop1 <- unique(sort(success_pop$gene))
    success_samp <- subset(sample, sample$gene %in% success_pop$gene)
    success_samp1 <- unique(sort(success_samp$gene))
    rownames(success_samp) <- NULL
    samp_l <- length(sample1)
    pop_l <- length(pop1)
    success_pop_l <- length(success_pop1)
    success_samp_l <- length(success_samp1)
    pvalue <- phyper((success_samp_l-1), samp_l, (pop_l-samp_l), success_pop_l, lower.tail = F)
    pvalue <- signif(pvalue, 4)
  })
  
  c_pd3_inweb_hypergeometric <- reactive({
    d <- c_pd3()
    inweb <- c_inweb_pd3()
    pop <- subset(d, d$gene %in% inweb_combined$V1)
    pop1 <- unique(sort(pop$gene))
    sample <- inweb
    sample1 <- unique(sort(inweb$gene))
    # sample <- subset(pop, pop$gene %in% inweb$gene)
    success_pop <- subset(pop, pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1])
    success_pop1 <- unique(sort(success_pop$gene))
    success_samp <- subset(sample, sample$gene %in% success_pop$gene)
    success_samp1 <- unique(sort(success_samp$gene))
    rownames(success_samp) <- NULL
    samp_l <- length(sample1)
    pop_l <- length(pop1)
    success_pop_l <- length(success_pop1)
    success_samp_l <- length(success_samp1)
    pvalue <- phyper((success_samp_l-1), samp_l, (pop_l-samp_l), success_pop_l, lower.tail = F)
    pvalue <- signif(pvalue, 4)
  })
  
  c_venndiagram_goi <- reactive({
    validate(
      need(!is.null(c_genes_uploaded()), ""),
      need(!is.null(c_vp1_goi_layer()), ""),
      need(!is.null(c_vp2_goi_layer()), "")
    )
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
      if(!is.null(c_pd1()) & !is.null(c_pd2())){
        c1 <- c_pd1()
        c2 <- c_pd2()
        gene_interest <- c_genes_uploaded()
        df1_c1 <- lapply(gene_interest, function(x) subset(c1, gene %in% x & 
                                                             pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) )  
        df1_c2 <- lapply(gene_interest, function(x) subset(c2, gene %in% x & 
                                                             pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) ) 
        
        total1 <- as.data.frame(data.table::rbindlist(df1_c1))
        total2 <- as.data.frame(data.table::rbindlist(df1_c2))
        
        df1_c1 <- c(df1_c1, list(total = total1))
        df1_c2 <- c(df1_c2, list(total = total2))
        
        list_num <- input$c_goi_num_inputs
        
        x <- list()
        x[["File1"]] <-df1_c1[[list_num]]$gene
        x[["File2"]] <-df1_c2[[list_num]]$gene
        
        vd_title <- list_num
        colors2 = c("gray", "gray")
        
        v0 <- venn.diagram(x,
                           fill = colors2, margin=0.05, filename = NULL, resolution = 900, height = 400, force.unique = T,
                           sub = " ", sub.pos = c(0, 0), euler.d = F, scaled = F, main = vd_title,
                           cat.cex = 1.1, cex = 2, cat.pos = c(180,180), cat.dist = c(0.05,0.05),
                           fontfamily = 'sans', cat.fontfamily = 'sans', main.fontfamily = 'sans')
      }
    } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      if(!is.null(c_pd1()) & !is.null(c_pd2()) & !is.null(c_pd3())){
        c1 <- c_pd1()
        c2 <- c_pd2()
        c3 <- c_pd3()
        gene_interest <- c_genes_uploaded()
        df1_c1 <- lapply(gene_interest, function(x) subset(c1, gene %in% x & 
                                                             pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) ) 
        df1_c2 <- lapply(gene_interest, function(x) subset(c2, gene %in% x & 
                                                             pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) ) 
        df1_c3 <- lapply(gene_interest, function(x) subset(c3, gene %in% x & 
                                                             pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) ) 
        
        total1 <- as.data.frame(data.table::rbindlist(df1_c1))
        total2 <- as.data.frame(data.table::rbindlist(df1_c2))
        total3 <- as.data.frame(data.table::rbindlist(df1_c3))
        
        df1_c1 <- c(df1_c1, list(total = total1))
        df1_c2 <- c(df1_c2, list(total = total2))
        df1_c3 <- c(df1_c3, list(total = total3))
        
        list_num <- input$c_goi_num_inputs
        
        x <- list()
        x[["File1"]] <-df1_c1[[list_num]]$gene
        x[["File2"]] <-df1_c2[[list_num]]$gene
        x[["File3"]] <-df1_c3[[list_num]]$gene
        
        vd_title <- list_num
        colors3 = c("gray", "gray", "gray")
        
        v0 <- venn.diagram(x, 
                           fill = colors3, margin=0.05, filename = NULL, resolution = 900, height = 400, force.unique = T,
                           sub = " ", sub.pos = c(0, 0, 0), euler.d = F, scaled = F, main = vd_title,
                           # cat.cex = 1.1, cex = 2, cat.pos = c(180,180,0), cat.dist = c(0.05,0.05,0.05),
                           fontfamily = 'sans', cat.fontfamily = 'sans', main.fontfamily = 'sans')
      }
    }
  })
  
  #comparison for file1 goi
  c_goi_compare1 <- reactive({
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
      c1 <- c_pd1()
      c2 <- c_pd2()
      gene_interest <- c_genes_uploaded()
      df1_c1 <- lapply(gene_interest, function(x) subset(c1, gene %in% x & 
                                                           pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) )  
      df1_c2 <- lapply(gene_interest, function(x) subset(c2, gene %in% x & 
                                                           pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) )  
      
      total1 <- as.data.frame(data.table::rbindlist(df1_c1))
      total2 <- as.data.frame(data.table::rbindlist(df1_c2))
      
      df1_c1 <- c(df1_c1, list(total = total1))
      df1_c2 <- c(df1_c2, list(total = total2))
      
      list_num <- input$c_goi_num_inputs
      f1_goi <-df1_c1[[list_num]]
      f2_goi <-df1_c2[[list_num]]
      overlap1_2 <- subset(f1_goi, gene %in% f2_goi$gene)
      list(f1_goi=f1_goi, overlap1_2=overlap1_2)
    } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      c1 <- c_pd1()
      c2 <- c_pd2()
      c3 <- c_pd3()
      gene_interest <- c_genes_uploaded()
      df1_c1 <- lapply(gene_interest, function(x) subset(c1, gene %in% x & 
                                                           pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) )  
      df1_c2 <- lapply(gene_interest, function(x) subset(c2, gene %in% x & 
                                                           pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) )  
      df1_c3 <- lapply(gene_interest, function(x) subset(c3, gene %in% x & 
                                                           pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) )  
      
      total1 <- as.data.frame(data.table::rbindlist(df1_c1))
      total2 <- as.data.frame(data.table::rbindlist(df1_c2))
      total3 <- as.data.frame(data.table::rbindlist(df1_c3))
      
      df1_c1 <- c(df1_c1, list(total = total1))
      df1_c2 <- c(df1_c2, list(total = total2))
      df1_c3 <- c(df1_c3, list(total = total3))
      
      list_num <- input$c_goi_num_inputs
      f1_goi <-df1_c1[[list_num]]
      f2_goi <-df1_c2[[list_num]]
      f3_goi <-df1_c3[[list_num]]
      overlap1_2 <- subset(f1_goi, gene %in% f2_goi$gene)
      overlap1_3 <- subset(f1_goi, gene %in% f3_goi$gene)
      overlap1_2_3 <- subset(f1_goi, gene %in% f2_goi$gene & gene %in% f3_goi$gene)
      list(f1_goi=f1_goi, overlap1_2=overlap1_2, overlap1_3=overlap1_3, overlap1_2_3=overlap1_2_3)
    }
  })
  
  #comparison for file2 goi
  c_goi_compare2 <- reactive({
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
      c1 <- c_pd1()
      c2 <- c_pd2()
      gene_interest <- c_genes_uploaded()
      df1_c1 <- lapply(gene_interest, function(x) subset(c1, gene %in% x & 
                                                           pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) )  
      df1_c2 <- lapply(gene_interest, function(x) subset(c2, gene %in% x & 
                                                           pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) )  
      
      total1 <- as.data.frame(data.table::rbindlist(df1_c1))
      total2 <- as.data.frame(data.table::rbindlist(df1_c2))
      
      df1_c1 <- c(df1_c1, list(total = total1))
      df1_c2 <- c(df1_c2, list(total = total2))
      
      list_num <- input$c_goi_num_inputs
      f1_goi <-df1_c1[[list_num]]
      f2_goi <-df1_c2[[list_num]]
      overlap2_1 <- subset(f2_goi, gene %in% f1_goi$gene)
      list(f2_goi=f2_goi, overlap2_1=overlap2_1)
    } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      c1 <- c_pd1()
      c2 <- c_pd2()
      c3 <- c_pd3()
      gene_interest <- c_genes_uploaded()
      df1_c1 <- lapply(gene_interest, function(x) subset(c1, gene %in% x & 
                                                           pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) )  
      df1_c2 <- lapply(gene_interest, function(x) subset(c2, gene %in% x & 
                                                           pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) )  
      df1_c3 <- lapply(gene_interest, function(x) subset(c3, gene %in% x & 
                                                           pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) )  
      
      total1 <- as.data.frame(data.table::rbindlist(df1_c1))
      total2 <- as.data.frame(data.table::rbindlist(df1_c2))
      total3 <- as.data.frame(data.table::rbindlist(df1_c3))
      
      df1_c1 <- c(df1_c1, list(total = total1))
      df1_c2 <- c(df1_c2, list(total = total2))
      df1_c3 <- c(df1_c3, list(total = total3))
      
      list_num <- input$c_goi_num_inputs
      f1_goi <-df1_c1[[list_num]]
      f2_goi <-df1_c2[[list_num]]
      f3_goi <-df1_c3[[list_num]]
      overlap2_1 <- subset(f2_goi, gene %in% f1_goi$gene)
      overlap2_3 <- subset(f2_goi, gene %in% f3_goi$gene)
      overlap2_1_3 <- subset(f2_goi, gene %in% f1_goi$gene & gene %in% f3_goi$gene)
      list(f2_goi=f2_goi, overlap2_1=overlap2_1, overlap2_3=overlap2_3, overlap2_1_3=overlap2_1_3)
    }
  })
  
  #comparison for file3 goi
  c_goi_compare3 <- reactive({
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      c1 <- c_pd1()
      c2 <- c_pd2()
      c3 <- c_pd3()
      gene_interest <- c_genes_uploaded()
      df1_c1 <- lapply(gene_interest, function(x) subset(c1, gene %in% x & 
                                                           pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) )  
      df1_c2 <- lapply(gene_interest, function(x) subset(c2, gene %in% x & 
                                                           pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) )  
      df1_c3 <- lapply(gene_interest, function(x) subset(c3, gene %in% x & 
                                                           pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) )  
      
      total1 <- as.data.frame(data.table::rbindlist(df1_c1))
      total2 <- as.data.frame(data.table::rbindlist(df1_c2))
      total3 <- as.data.frame(data.table::rbindlist(df1_c3))
      
      df1_c1 <- c(df1_c1, list(total = total1))
      df1_c2 <- c(df1_c2, list(total = total2))
      df1_c3 <- c(df1_c3, list(total = total3))
      
      list_num <- input$c_goi_num_inputs
      f1_goi <-df1_c1[[list_num]]
      f2_goi <-df1_c2[[list_num]]
      f3_goi <-df1_c3[[list_num]]
      overlap3_1 <- subset(f3_goi, gene %in% f1_goi$gene)
      overlap3_2 <- subset(f3_goi, gene %in% f2_goi$gene)
      overlap3_1_2 <- subset(f3_goi, gene %in% f1_goi$gene & gene %in% f2_goi$gene)
      list(f3_goi=f3_goi, overlap3_1=overlap3_1, overlap3_2=overlap3_2, overlap3_1_2=overlap3_1_2)
    }
  })
  
  c_f1_unique_goi <- reactive({
    d <- c_goi_compare1()
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
      f1 <- subset(d$f1_goi, gene %!in% (d$overlap1_2)$gene)
    } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      f1 <- subset(d$f1_goi, gene %!in% (d$overlap1_2)$gene & gene %!in% (d$overlap1_3)$gene & gene %!in% (d$overlap1_2_3)$gene)
    }
    d1 <- data.frame(f1$gene)
    colnames(d1) <- c("f1")    
    d1 <- unique(sort(d1$f1))
  })
  
  c_f2_unique_goi <- reactive({
    d <- c_goi_compare2()
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
      f2 <- subset(d$f2_goi, gene %!in% (d$overlap2_1)$gene)
    } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      f2 <- subset(d$f2_goi, gene %!in% (d$overlap2_1)$gene & gene %!in% (d$overlap2_3)$gene & gene %!in% (d$overlap2_1_3)$gene)
    }
    d1 <- data.frame(f2$gene)
    colnames(d1) <- c("f2")    
    d1 <- unique(sort(d1$f2))
  })
  
  c_f1_and_f2_goi <- reactive({
    d <- c_goi_compare2()
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
      f2 <- d$overlap2_1
    } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      f2 <- subset(d$overlap2_1, gene %!in% (d$overlap2_1_3)$gene)
    }
    d1 <- data.frame(f2$gene)
    colnames(d1) <- c("f12")    
    d1 <- unique(sort(d1$f12))
  })
  
  c_f3_unique_goi <- reactive({
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      d <- c_goi_compare3()
      f3 <- subset(d$f3_goi, gene %!in% (d$overlap3_1)$gene & gene %!in% (d$overlap3_2)$gene & gene %!in% (d$overlap3_1_2)$gene)
      d1 <- data.frame(f3$gene)
      colnames(d1) <- c("f3")    
      d1 <- unique(sort(d1$f3))
    }
  })
  
  c_f1_and_f3_goi <- reactive({
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      d <- c_goi_compare3()
      f3 <- subset(d$overlap3_1, gene %!in% (d$overlap3_1_2)$gene)
      d1 <- data.frame(f3$gene)
      colnames(d1) <- c("f13")    
      d1 <- unique(sort(d1$f13))
    }
  })
  
  c_f2_and_f3_goi <- reactive({
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      d <- c_goi_compare3()
      f3 <- subset(d$overlap3_2, gene %!in% (d$overlap3_1_2)$gene)
      d1 <- data.frame(f3$gene)
      colnames(d1) <- c("f23")    
      d1 <- unique(sort(d1$f23))
    }
  })
  
  c_f1_f2_and_f3_goi <- reactive({
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      d <- c_goi_compare3()
      d1 <- data.frame((d$overlap3_1_2)$gene)
      colnames(d1) <- c("f123")    
      d1 <- unique(sort(d1$f123))
    }
  })
  
  c_unique_goi_dat <- reactive({
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
      f1 <- c_f1_unique_goi()
      f2 <- c_f2_unique_goi()
      f12 <- c_f1_and_f2_goi()
      n <- max(length(f1), length(f2), length(f12))
      length(f1) <- n
      length(f2) <- n
      length(f12) <- n
      d <- cbind(f1, f2, f12)
      d[is.na(d)] <- " "
    } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      f1 <- c_f1_unique_goi()
      f2 <- c_f2_unique_goi()
      f12 <- c_f1_and_f2_goi()
      f3 <- c_f3_unique_goi()
      f13 <- c_f1_and_f3_goi()
      f23 <- c_f2_and_f3_goi()
      f123 <- c_f1_f2_and_f3_goi()
      n <- max(length(f1), length(f2), length(f12), length(f3), length(f13), length(f23), length(f123))
      length(f1) <- n
      length(f2) <- n
      length(f12) <- n
      length(f3) <- n
      length(f13) <- n
      length(f23) <- n
      length(f123) <- n
      d <- cbind(f1, f2, f12, f3, f13, f23, f123)
      d[is.na(d)] <- " "
    }
    as.data.frame(d)
  })
  
  # c_venndiagram_snp <- reactive({
  #   validate(
  #     need(!is.null(c_snp()), ""),
  #     need(!is.null(c_vp1_snp_layer()), ""),
  #     need(!is.null(c_vp2_snp_layer()), "")
  #   )
  #   if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
  #     if(!is.null(c_pd1()) & !is.null(c_pd2())){
  #       c1 <- c_pd1()
  #       c2 <- c_pd2()
  #       c_total <- rbind(c1, c2)
  #       snp2gene <- c_SNP_to_gene(c_total)
  #       snp_interest <- split(snp2gene$gene, snp2gene$snpid)
  # 
  #       df1_c1 <- lapply(snp_interest, function(x) subset(c1, gene %in% x & 
  #                                                           pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) ) 
  #       df1_c2 <- lapply(snp_interest, function(x) subset(c2, gene %in% x & 
  #                                                           pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) ) 
  #       
  #       total1 <- as.data.frame(data.table::rbindlist(df1_c1))
  #       total2 <- as.data.frame(data.table::rbindlist(df1_c2))
  #       
  #       df1_c1 <- c(df1_c1, list(total = total1))
  #       df1_c2 <- c(df1_c2, list(total = total2))
  #       
  #       list_num <- input$c_snp_num_inputs
  # 
  #       x <- list()
  #       x[["File1"]] <-df1_c1[[list_num]]$gene
  #       x[["File2"]] <-df1_c2[[list_num]]$gene
  #       
  #       vd_title <- list_num
  #       colors2 = c("gray", "gray")
  #       
  #       v0 <- venn.diagram(x,
  #                          fill = colors2, margin=0.05, filename = NULL, resolution = 900, height = 400, force.unique = T,
  #                          sub = " ", sub.pos = c(0, 0), euler.d = F, scaled = F, main = vd_title,
  #                          cat.cex = 1.1, cex = 2, cat.pos = c(180,180), cat.dist = c(0.05,0.05),
  #                          fontfamily = 'sans', cat.fontfamily = 'sans', main.fontfamily = 'sans')
  #     }
  #   } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
  #     if(!is.null(c_pd1()) & !is.null(c_pd2()) & !is.null(c_pd3())){
  #       c1 <- c_pd1()
  #       c2 <- c_pd2()
  #       c3 <- c_pd3()
  #       c_total <- rbind(c1, c2, c3)
  #       snp2gene <- c_SNP_to_gene(c_total)
  #       snp_interest <- split(snp2gene$gene, snp2gene$snpid)
  #       df1_c1 <- lapply(snp_interest, function(x) subset(c1, gene %in% x & 
  #                                                           pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) ) 
  #       df1_c2 <- lapply(snp_interest, function(x) subset(c2, gene %in% x & 
  #                                                           pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) ) 
  #       df1_c3 <- lapply(snp_interest, function(x) subset(c3, gene %in% x & 
  #                                                           pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) ) 
  #       
  #       total1 <- as.data.frame(data.table::rbindlist(df1_c1))
  #       total2 <- as.data.frame(data.table::rbindlist(df1_c2))
  #       total3 <- as.data.frame(data.table::rbindlist(df1_c3))
  #       
  #       df1_c1 <- c(df1_c1, list(total = total1))
  #       df1_c2 <- c(df1_c2, list(total = total2))
  #       df1_c3 <- c(df1_c3, list(total = total3))
  #       
  #       
  #       list_num <- input$c_snp_num_inputs
  #       
  #       x <- list()
  #       x[["File1"]] <-df1_c1[[list_num]]$gene
  #       x[["File2"]] <-df1_c2[[list_num]]$gene
  #       x[["File3"]] <-df1_c3[[list_num]]$gene
  #       
  #       vd_title <- list_num
  #       colors3 = c("gray", "gray", "gray")
  #       
  #       v0 <- venn.diagram(x, 
  #                          fill = colors3, margin=0.05, filename = NULL, resolution = 900, height = 400, force.unique = T,
  #                          sub = " ", sub.pos = c(0, 0, 0), euler.d = F, scaled = F, main = vd_title,
  #                          # cat.cex = 1.1, cex = 2, cat.pos = c(180,180,0), cat.dist = c(0.05,0.05,0.05),
  #                          fontfamily = 'sans', cat.fontfamily = 'sans', main.fontfamily = 'sans')
  #     }
  #   }
  # })
  
  #comparison for file1 snp
  # c_snp_compare1 <- reactive({
  #   if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
  #     c1 <- c_pd1()
  #     c2 <- c_pd2()
  #     c_total <- rbind(c1, c2)
  #     snp2gene <- c_SNP_to_gene(c_total)
  # 
  #     snp_interest <- split(snp2gene$gene, snp2gene$snpid)
  #     df1_c1 <- lapply(snp_interest, function(x) subset(c1, gene %in% x & 
  #                                                         pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) )  
  #     df1_c2 <- lapply(snp_interest, function(x) subset(c2, gene %in% x & 
  #                                                         pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) )  
  #     
  #     total1 <- as.data.frame(data.table::rbindlist(df1_c1))
  #     total2 <- as.data.frame(data.table::rbindlist(df1_c2))
  #     
  #     df1_c1 <- c(df1_c1, list(total = total1))
  #     df1_c2 <- c(df1_c2, list(total = total2))
  #     
  #     list_num <- input$c_snp_num_inputs
  #     print("i am printing")
  #     print(list_num)
  #     f1_snp <-df1_c1[[list_num]]
  #     f2_snp <-df1_c2[[list_num]]
  #     overlap1_2 <- subset(f1_snp, gene %in% f2_snp$gene)
  #     list(f1_snp=f1_snp, overlap1_2=overlap1_2)
  #   } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
  #     c1 <- c_pd1()
  #     c2 <- c_pd2()
  #     c3 <- c_pd3()
  #     c_total <- rbind(c1, c2, c3)
  #     snp2gene <- c_SNP_to_gene(c_total)
  #     snp_interest <- split(snp2gene$gene, snp2gene$snpid)
  #     df1_c1 <- lapply(snp_interest, function(x) subset(c1, gene %in% x & 
  #                                                         pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) )  
  #     df1_c2 <- lapply(snp_interest, function(x) subset(c2, gene %in% x & 
  #                                                         pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) )  
  #     df1_c3 <- lapply(snp_interest, function(x) subset(c3, gene %in% x & 
  #                                                         pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) )  
  #     
  #     total1 <- as.data.frame(data.table::rbindlist(df1_c1))
  #     total2 <- as.data.frame(data.table::rbindlist(df1_c2))
  #     total3 <- as.data.frame(data.table::rbindlist(df1_c3))
  #     
  #     df1_c1 <- c(df1_c1, list(total = total1))
  #     df1_c2 <- c(df1_c2, list(total = total2))
  #     df1_c3 <- c(df1_c3, list(total = total3))
  #     
  #     list_num <- input$c_snp_num_inputs
  #     f1_snp <-df1_c1[[list_num]]
  #     f2_snp <-df1_c2[[list_num]]
  #     f3_snp <-df1_c3[[list_num]]
  #     overlap1_2 <- subset(f1_snp, gene %in% f2_snp$gene)
  #     overlap1_3 <- subset(f1_snp, gene %in% f3_snp$gene)
  #     overlap1_2_3 <- subset(f1_snp, gene %in% f2_snp$gene & gene %in% f3_snp$gene)
  #     list(f1_snp=f1_snp, overlap1_2=overlap1_2, overlap1_3=overlap1_3, overlap1_2_3=overlap1_2_3)
  #   }
  # })
  # 
  # #comparison for file2 snp
  # c_snp_compare2 <- reactive({
  #   if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
  #     c1 <- c_pd1()
  #     c2 <- c_pd2()
  #     c_total <- rbind(c1, c2)
  #     snp2gene <- c_SNP_to_gene(c_total)
  #     snp_interest <- split(snp2gene$gene, snp2gene$snpid)
  #     df1_c1 <- lapply(snp_interest, function(x) subset(c1, gene %in% x & 
  #                                                         pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) )  
  #     df1_c2 <- lapply(snp_interest, function(x) subset(c2, gene %in% x & 
  #                                                         pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) )  
  #     
  #     total1 <- as.data.frame(data.table::rbindlist(df1_c1))
  #     total2 <- as.data.frame(data.table::rbindlist(df1_c2))
  #     
  #     df1_c1 <- c(df1_c1, list(total = total1))
  #     df1_c2 <- c(df1_c2, list(total = total2))
  #     
  #     list_num <- input$c_snp_num_inputs
  #     f1_snp <-df1_c1[[list_num]]
  #     f2_snp <-df1_c2[[list_num]]
  #     overlap2_1 <- subset(f2_snp, gene %in% f1_snp$gene)
  #     list(f2_snp=f2_snp, overlap2_1=overlap2_1)
  #   } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
  #     c1 <- c_pd1()
  #     c2 <- c_pd2()
  #     c3 <- c_pd3()
  #     c_total <- rbind(c1, c2, c3)
  #     snp2gene <- c_SNP_to_gene(c_total)
  #     snp_interest <- split(snp2gene$gene, snp2gene$snpid)
  #     df1_c1 <- lapply(snp_interest, function(x) subset(c1, gene %in% x & 
  #                                                         pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) )  
  #     df1_c2 <- lapply(snp_interest, function(x) subset(c2, gene %in% x & 
  #                                                         pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) )  
  #     df1_c3 <- lapply(snp_interest, function(x) subset(c3, gene %in% x & 
  #                                                         pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) ) 
  #     
  #     total1 <- as.data.frame(data.table::rbindlist(df1_c1))
  #     total2 <- as.data.frame(data.table::rbindlist(df1_c2))
  #     total3 <- as.data.frame(data.table::rbindlist(df1_c3))
  #     
  #     df1_c1 <- c(df1_c1, list(total = total1))
  #     df1_c2 <- c(df1_c2, list(total = total2))
  #     df1_c3 <- c(df1_c3, list(total = total3))
  #     
  #     list_num <- input$c_snp_num_inputs
  #     f1_snp <-df1_c1[[list_num]]
  #     f2_snp <-df1_c2[[list_num]]
  #     f3_snp <-df1_c3[[list_num]]
  #     overlap2_1 <- subset(f2_snp, gene %in% f1_snp$gene)
  #     overlap2_3 <- subset(f2_snp, gene %in% f3_snp$gene)
  #     overlap2_1_3 <- subset(f2_snp, gene %in% f1_snp$gene & gene %in% f3_snp$gene)
  #     list(f2_snp=f2_snp, overlap2_1=overlap2_1, overlap2_3=overlap2_3, overlap2_1_3=overlap2_1_3)
  #   }
  # })
  # 
  # #comparison for file3 snp
  # c_snp_compare3 <- reactive({
  #   if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
  #     c1 <- c_pd1()
  #     c2 <- c_pd2()
  #     c3 <- c_pd3()
  #     c_total <- rbind(c1, c2, c3)
  #     snp2gene <- c_SNP_to_gene(c_total)
  #     snp_interest <- split(snp2gene$gene, snp2gene$snpid)
  #     df1_c1 <- lapply(snp_interest, function(x) subset(c1, gene %in% x & 
  #                                                         pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) )  
  #     df1_c2 <- lapply(snp_interest, function(x) subset(c2, gene %in% x & 
  #                                                         pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) )  
  #     df1_c3 <- lapply(snp_interest, function(x) subset(c3, gene %in% x & 
  #                                                         pvalue <= input$c_pval_thresh & FDR <= input$c_fdr_thresh & logFC <= input$c_logfc_thresh_comb[2] & logFC >= input$c_logfc_thresh_comb[1]) )  
  #     
  #     total1 <- as.data.frame(data.table::rbindlist(df1_c1))
  #     total2 <- as.data.frame(data.table::rbindlist(df1_c2))
  #     total3 <- as.data.frame(data.table::rbindlist(df1_c3))
  #     
  #     df1_c1 <- c(df1_c1, list(total = total1))
  #     df1_c2 <- c(df1_c2, list(total = total2))
  #     df1_c3 <- c(df1_c3, list(total = total3))
  #     
  #     list_num <- input$c_snp_num_inputs
  #     f1_snp <-df1_c1[[list_num]]
  #     f2_snp <-df1_c2[[list_num]]
  #     f3_snp <-df1_c3[[list_num]]
  #     overlap3_1 <- subset(f3_snp, gene %in% f1_snp$gene)
  #     overlap3_2 <- subset(f3_snp, gene %in% f2_snp$gene)
  #     overlap3_1_2 <- subset(f3_snp, gene %in% f1_snp$gene & gene %in% f2_snp$gene)
  #     list(f3_snp=f3_snp, overlap3_1=overlap3_1, overlap3_2=overlap3_2, overlap3_1_2=overlap3_1_2)
  #   }
  # })
  # 
  # c_f1_unique_snp <- reactive({
  #   d <- c_snp_compare1()
  #   if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
  #     f1 <- subset(d$f1_snp, gene %!in% (d$overlap1_2)$gene)
  #   } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
  #     f1 <- subset(d$f1_snp, gene %!in% (d$overlap1_2)$gene & gene %!in% (d$overlap1_3)$gene & gene %!in% (d$overlap1_2_3)$gene)
  #   }
  #   d1 <- data.frame(f1$gene)
  #   colnames(d1) <- c("f1")    
  #   d1 <- unique(sort(d1$f1))
  # })
  # 
  # c_f2_unique_snp <- reactive({
  #   d <- c_snp_compare2()
  #   if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
  #     f2 <- subset(d$f2_snp, gene %!in% (d$overlap2_1)$gene)
  #   } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
  #     f2 <- subset(d$f2_snp, gene %!in% (d$overlap2_1)$gene & gene %!in% (d$overlap2_3)$gene & gene %!in% (d$overlap2_1_3)$gene)
  #   }
  #   d1 <- data.frame(f2$gene)
  #   colnames(d1) <- c("f2")    
  #   d1 <- unique(sort(d1$f2))
  # })
  # 
  # c_f1_and_f2_snp <- reactive({
  #   d <- c_snp_compare2()
  #   if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
  #     f2 <- d$overlap2_1
  #   } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
  #     f2 <- subset(d$overlap2_1, gene %!in% (d$overlap2_1_3)$gene)
  #   }
  #   d1 <- data.frame(f2$gene)
  #   colnames(d1) <- c("f12")    
  #   d1 <- unique(sort(d1$f12))
  # })
  # 
  # c_f3_unique_snp <- reactive({
  #   if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
  #     d <- c_snp_compare3()
  #     f3 <- subset(d$f3_snp, gene %!in% (d$overlap3_1)$gene & gene %!in% (d$overlap3_2)$gene & gene %!in% (d$overlap3_1_2)$gene)
  #     d1 <- data.frame(f3$gene)
  #     colnames(d1) <- c("f3")    
  #     d1 <- unique(sort(d1$f3))
  #   }
  # })
  # 
  # c_f1_and_f3_snp <- reactive({
  #   if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
  #     d <- c_snp_compare3()
  #     f3 <- subset(d$overlap3_1, gene %!in% (d$overlap3_1_2)$gene)
  #     d1 <- data.frame(f3$gene)
  #     colnames(d1) <- c("f13")    
  #     d1 <- unique(sort(d1$f13))
  #   }
  # })
  # 
  # c_f2_and_f3_snp <- reactive({
  #   if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
  #     d <- c_snp_compare3()
  #     f3 <- subset(d$overlap3_2, gene %!in% (d$overlap3_1_2)$gene)
  #     d1 <- data.frame(f3$gene)
  #     colnames(d1) <- c("f23")    
  #     d1 <- unique(sort(d1$f23))
  #   }
  # })
  # 
  # c_f1_f2_and_f3_snp <- reactive({
  #   if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
  #     d <- c_snp_compare3()
  #     d1 <- data.frame((d$overlap3_1_2)$gene)
  #     colnames(d1) <- c("f123")    
  #     d1 <- unique(sort(d1$f123))
  #   }
  # })
  # 
  # c_unique_snp_dat <- reactive({
  #   if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
  #     f1 <- c_f1_unique_snp()
  #     f2 <- c_f2_unique_snp()
  #     f12 <- c_f1_and_f2_snp()
  #     n <- max(length(f1), length(f2), length(f12))
  #     length(f1) <- n
  #     length(f2) <- n
  #     length(f12) <- n
  #     d <- cbind(f1, f2, f12)
  #     d[is.na(d)] <- " "
  #   } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
  #     f1 <- c_f1_unique_snp()
  #     f2 <- c_f2_unique_snp()
  #     f12 <- c_f1_and_f2_snp()
  #     f3 <- c_f3_unique_snp()
  #     f13 <- c_f1_and_f3_snp()
  #     f23 <- c_f2_and_f3_snp()
  #     f123 <- c_f1_f2_and_f3_snp()
  #     n <- max(length(f1), length(f2), length(f12), length(f3), length(f13), length(f23), length(f123))
  #     length(f1) <- n
  #     length(f2) <- n
  #     length(f12) <- n
  #     length(f3) <- n
  #     length(f13) <- n
  #     length(f23) <- n
  #     length(f123) <- n
  #     d <- cbind(f1, f2, f12, f3, f13, f23, f123)
  #     d[is.na(d)] <- " "
  #   }
  #   as.data.frame(d)
  # })
  
  c_pf_db <- reactive({
    if(!is.null(input$c_pfam_db)){
      pf_db <- input$c_pfam_db
    }
  })
  
  c_pf_db_search <- reactive({
    pf_db <- c_pf_db()
    pf_db <- paste0("^", pf_db, "$")
    selected_pf <- prot_fam[grep(paste(pf_db,collapse='|'), names(prot_fam))]
    selected_pf <- lapply(selected_pf, function(x) x[!is.na(x)])
    selected_pf
  })
  
  c_pf_db_vp1 <- reactive({
    validate(
      need(input$c_file_pulldown1 != '', "")
    )
    if(!is.null(c_pf_db_search())){
      d <- c_pd1()
      gene_interest <- c_pf_db_search()
      d_g2s <- lapply(gene_interest, function(x) subset(d, gene %in% x) )
      list(d_g2s=d_g2s)
    }
  })
  
  c_pf_db_vp2 <- reactive({
    validate(
      need(input$c_file_pulldown2 != '', "")
    )
    if(!is.null(c_pf_db_search())){
      d <- c_pd2()
      gene_interest <- c_pf_db_search()
      d_g2s <- lapply(gene_interest, function(x) subset(d, gene %in% x) )
      list(d_g2s=d_g2s)
    }
  })
  
  c_pf_db_vp3 <- reactive({
    validate(
      need(input$c_file_pulldown3 != '', "")
    )
    if(!is.null(c_pf_db_search())){
      d <- c_pd3()
      gene_interest <- c_pf_db_search()
      d_g2s <- lapply(gene_interest, function(x) subset(d, gene %in% x) )
      list(d_g2s=d_g2s)
    }
  })
  
  c_pf1_cleanup <- reactive({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_pfam_db != '', "")
    )
    d <- c_orig_pd1()
    d_col <- colnames(d)
    if("rep1" %in% d_col & "rep2" %in% d_col){
      prot_remove <- unique(unlist(c_pf_db_search()))
      prot_cleaned <- subset(d, gene %!in% prot_remove)
      prot_cleaned
    }
  })
  
  c_pf2_cleanup <- reactive({
    validate(
      need(input$c_file_pulldown2 != '', ""),
      need(input$c_pfam_db != '', "")
    )
    d <- c_orig_pd2()
    d_col <- colnames(d)
    if("rep1" %in% d_col & "rep2" %in% d_col){
      prot_remove <- unique(unlist(c_pf_db_search()))
      prot_cleaned <- subset(d, gene %!in% prot_remove)
      prot_cleaned
    }
  })
  
  c_pf3_cleanup <- reactive({
    validate(
      need(input$c_file_pulldown3 != '', ""),
      need(input$c_pfam_db != '', "")
    )
    d <- c_orig_pd3()
    d_col <- colnames(d)
    if("rep1" %in% d_col & "rep2" %in% d_col){
      prot_remove <- unique(unlist(c_pf_db_search()))
      prot_cleaned <- subset(d, gene %!in% prot_remove)
      prot_cleaned
    }
  })
  
  c_vp1_pf_db_layer <- reactive({
    validate(
      need(!is.null(c_pd1()), ""),
      need(!is.null(c_pf_db()), "")
    )
    d <- c_pd1()
    c1_pf_db <- c_pf_db_vp1()
    min_x <- c_min_x()
    min_y <- c_min_y()
    max_x <- c_max_x()
    max_y <- c_max_y()
    pf_col <- input$c_colorbrewer_theme_pf
    if(input$c_colorscheme == "fdr"){
      req(input$c_color_indv_sig, input$c_color_indv_insig, input$c_colorbrewer_theme_pf)
      data <- separate_to_groups_for_color_integrated(d, input$c_fdr_thresh, input$c_color_indv_sig, input$c_color_indv_insig)
      p <- plot_ly(colors = pf_col, showlegend = T, width = 300, height = 390)
      for(i in nrow(data)){
        p <- add_markers(p, data = data, x = ~logFC, y = ~-log10(pvalue),
                         marker = list(size = 6, cmin = 0, cmax = 1, color = ~col), 
                         opacity = 0.8, 
                         text = ~paste(gene), hoverinfo = "text", name = "pull down")
      }
      p 
    } else if(input$c_colorscheme == "exac"){
      d$s <- exac$em_p_hi[match(d$gene, exac$GENE_NAME)]
      d$s[is.na(d$s)] <- 2
      below_thresh <- subset(d, s < 0.9)
      above_thresh <- subset(d, s >= 0.9)
      no_exist <- subset(d, s == 2)
      p <- plot_ly(colors = pf_col, showlegend = T, width = 300, height = 390)
      p <- add_markers(p, data = below_thresh, x = ~logFC, y = ~-log10(pvalue),
                       marker = list(size = 8, line = list(width=0.1, color = 'black'), cmin = 0, cmax = 1, color = "#66c2a5"),
                       opacity = 0.8, 
                       text = ~paste(gene), hoverinfo = "text", name = paste0("pLI<0.9 (", nrow(below_thresh), ")"))
      p <- add_markers(p, data = above_thresh, x = ~logFC, y = ~-log10(pvalue),
                       marker = list(size = 8, line = list(width=0.1, color = "black"), cmin = 0, cmax = 1, color = "#fc8d62"),
                       opacity = 0.8, 
                       text = ~paste(gene), hoverinfo = "text", name = paste0("pLI>=0.9 (", nrow(above_thresh), ")"))
      p <- add_markers(p, data = no_exist, x = ~logFC, y = ~-log10(pvalue),
                       marker = list(size = 8, line = list(width=0.1, color = "black"), cmin = 0, cmax = 1, color = "#8da0cb"),
                       opacity = 0.8, 
                       text = ~paste(gene), hoverinfo = "text", name = paste0("not in ExAC (", nrow(no_exist), ")"))
      p
    } else if(input$c_colorscheme == "cbf"){
      data <- separate_to_groups_for_cbf_integrated(d, input$c_fdr_thresh)
      p <- plot_ly(colors = "Greys", showlegend = T, width = 300, height = 390)
      for(i in nrow(data)){
        p <- add_markers(p, data = data, x = ~logFC, y = ~-log10(pvalue),
                         marker = list(size = 6, cmin = 0, cmax = 1, color = ~col), 
                         opacity = 0.6, 
                         text = ~paste(gene), hoverinfo = "text", name = "pull down")
      }
      p 
    } else if(input$c_colorscheme == "user"){
      validate(
        need(!is.null(input$c_file_color), "Please upload file with gene and score")
      )
      d1 <- c_in_file_color()
      col_theme <- input$c_colorbrewer_theme
      if(input$c_colorscheme_style == "cont"){
        df <- separate_to_groups_for_color_continuous(d, d1, col_theme)
        data <- df$df1
        data1 <- df$no_exist
      } else if(input$c_colorscheme_style == "disc"){
        df <- separate_to_groups_for_color_discrete(d, d1, col_theme)
        data <- df$df1
        data1 <- df$no_exist
      }
      p <- plot_ly(colors = pf_col, showlegend = T, width = 320, height = 390)
      p <- add_markers(p, data = data1, x = ~logFC, y = ~-log10(pvalue),
                       marker = list(size = 7, line = list(width=0.1, color = "grey89"), cmin = 0, cmax = 1, color = "#f7f7f7"),
                       opacity = 0.8,
                       text = ~paste(gene), hoverinfo = "text", name = paste0("Not in user data (", nrow(df$no_exist), ")"))
      for(i in nrow(data)){
        p <- add_markers(p, data = data, x = ~logFC, y = ~-log10(pvalue),
                         marker = list(size = 7, cmin = 0, cmax = 1, color = ~col, line = list(width=0.2, color = "grey89")),
                         opacity = 1,
                         text = ~paste0(gene, ", FDR=", signif(FDR, digits = 3)), hoverinfo = "text",
                         name = paste0("Found in user data (", nrow(data), ")")) #, name = "pull down"
      }
      p
    }
    p <- p%>%
      layout(xaxis = list(range=c(min_x-0.5, max_x+0.5), showgrid = F), yaxis = list(range=c(min_y-0.5, max_y+0.5), showgrid = F),
             legend = list(orientation = 'h', y = -0.23))
    if(input$c_colorscheme == "fdr" | input$c_colorscheme == "exac" | input$c_colorscheme == "user"){
      df <- ldply(c1_pf_db$d_g2s, data.frame)
      if(nrow(df) != 0){
        if(input$c_marker_text_prot_fam_db == "yes_label"){
          vp_layer_genes <- vp_layer_for_uploaded_genes(p, df)
        } else if(input$c_marker_text_prot_fam_db == "no_label"){
          vp_layer_genes <- vp_layer_for_uploaded_genes_no_text(p, df)
        }
        p <- vp_layer_genes
      } else{
        vp_layer_no_genes <- vp_layer_for_uploaded_genes_none(p, d)
        p <- vp_layer_no_genes
      }
      
    } else if(input$c_colorscheme == "cbf"){
      df <- ldply(c1_pf_db$d_g2s, data.frame)
      if(nrow(df) != 0){
        if(input$c_marker_text_prot_fam_db == "yes_label"){
          vp_layer_genes <- vp_layer_for_uploaded_genes_cbf(p, df)
        } else if(input$c_marker_text_prot_fam_db == "no_label"){
          vp_layer_genes <- vp_layer_for_uploaded_genes_cbf_no_text(p, df)
        }
        p <- vp_layer_genes
      } else{
        vp_layer_no_genes <- vp_layer_for_uploaded_genes_none_cbf(p, d)
        p <- vp_layer_no_genes
      }
      
    }
    p <- p %>% 
      layout(xaxis = list(range=c(min_x-0.5, max_x+0.5), showgrid = F), yaxis = list(range=c(min_y-0.5, max_y+0.5), showgrid = F)) %>%
      add_lines(x = c(min_x-0.5, max_x+0.5), y = -log10(input$c_pval_thresh), line = list(dash = "dash", width = 0.5, color = "#2b333e"), 
                name = '', hoverinfo = "text", text = paste0("pvalue = ", input$c_pval_thresh), showlegend = F) %>%
      add_lines(x = input$c_logfc_thresh_comb[1], y = c(min_y-0.5, max_y+0.5), line = list(dash = "dash", width = 0.5, color = "#252525"), 
                name = '', hoverinfo = "text", text = paste0("logFC = ", input$c_logfc_thresh_comb[1]), showlegend = F) %>%
      add_lines(x = input$c_logfc_thresh_comb[2], y = c(min_y-0.5, max_y+0.5), line = list(dash = "dash", width = 0.5, color = "#252525"), 
                name = '', hoverinfo = "text", text = paste0("logFC = ", input$c_logfc_thresh_comb[2]), showlegend = F)
  })
  
  c_vp2_pf_db_layer <- reactive({
    validate(
      need(!is.null(c_pd2()), ""),
      need(!is.null(c_pf_db()), "")
    )
    d <- c_pd2()
    c2_pf_db <- c_pf_db_vp2()
    min_x <- c_min_x()
    min_y <- c_min_y()
    max_x <- c_max_x()
    max_y <- c_max_y()
    pf_col <- input$c_colorbrewer_theme_pf
    if(input$c_colorscheme == "fdr"){
      req(input$c_color_indv_sig, input$c_color_indv_insig, input$c_colorbrewer_theme_pf)
      data <- separate_to_groups_for_color_integrated(d, input$c_fdr_thresh, input$c_color_indv_sig, input$c_color_indv_insig)
      p <- plot_ly(colors = pf_col, showlegend = T, width = 300, height = 390)
      for(i in nrow(data)){
        p <- add_markers(p, data = data, x = ~logFC, y = ~-log10(pvalue),
                         marker = list(size = 6, cmin = 0, cmax = 1, color = ~col), 
                         opacity = 0.8, 
                         text = ~paste(gene), hoverinfo = "text", name = "pull down")
      }
      p 
    } else if(input$c_colorscheme == "exac"){
      d$s <- exac$em_p_hi[match(d$gene, exac$GENE_NAME)]
      d$s[is.na(d$s)] <- 2
      below_thresh <- subset(d, s < 0.9)
      above_thresh <- subset(d, s >= 0.9)
      no_exist <- subset(d, s == 2)
      p <- plot_ly(colors = pf_col, showlegend = T, width = 300, height = 390)
      p <- add_markers(p, data = below_thresh, x = ~logFC, y = ~-log10(pvalue),
                       marker = list(size = 8, line = list(width=0.1, color = 'black'), cmin = 0, cmax = 1, color = "#66c2a5"),
                       opacity = 0.8, 
                       text = ~paste(gene), hoverinfo = "text", name = paste0("pLI<0.9 (", nrow(below_thresh), ")"))
      p <- add_markers(p, data = above_thresh, x = ~logFC, y = ~-log10(pvalue),
                       marker = list(size = 8, line = list(width=0.1, color = "black"), cmin = 0, cmax = 1, color = "#fc8d62"),
                       opacity = 0.8, 
                       text = ~paste(gene), hoverinfo = "text", name = paste0("pLI>=0.9 (", nrow(above_thresh), ")"))
      p <- add_markers(p, data = no_exist, x = ~logFC, y = ~-log10(pvalue),
                       marker = list(size = 8, line = list(width=0.1, color = "black"), cmin = 0, cmax = 1, color = "#8da0cb"),
                       opacity = 0.8, 
                       text = ~paste(gene), hoverinfo = "text", name = paste0("not in ExAC (", nrow(no_exist), ")"))
      p
    } else if(input$c_colorscheme == "cbf"){
      data <- separate_to_groups_for_cbf_integrated(d, input$c_fdr_thresh)
      p <- plot_ly(colors = "Greys", showlegend = T, width = 300, height = 390)
      for(i in nrow(data)){
        p <- add_markers(p, data = data, x = ~logFC, y = ~-log10(pvalue),
                         marker = list(size = 6, cmin = 0, cmax = 1, color = ~col), 
                         opacity = 0.6, 
                         text = ~paste(gene), hoverinfo = "text", name = "pull down")
      }
      p 
    } else if(input$c_colorscheme == "user"){
      validate(
        need(!is.null(input$c_file_color), "Please upload file with gene and score")
      )
      d1 <- c_in_file_color()
      col_theme <- input$c_colorbrewer_theme
      if(input$c_colorscheme_style == "cont"){
        df <- separate_to_groups_for_color_continuous(d, d1, col_theme)
        data <- df$df1
        data1 <- df$no_exist
      } else if(input$c_colorscheme_style == "disc"){
        df <- separate_to_groups_for_color_discrete(d, d1, col_theme)
        data <- df$df1
        data1 <- df$no_exist
      }
      p <- plot_ly(colors = pf_col, showlegend = T, width = 320, height = 390)
      p <- add_markers(p, data = data1, x = ~logFC, y = ~-log10(pvalue),
                       marker = list(size = 7, line = list(width=0.1, color = "grey89"), cmin = 0, cmax = 1, color = "#f7f7f7"),
                       opacity = 0.8,
                       text = ~paste(gene), hoverinfo = "text", name = paste0("Not in user data (", nrow(df$no_exist), ")"))
      for(i in nrow(data)){
        p <- add_markers(p, data = data, x = ~logFC, y = ~-log10(pvalue),
                         marker = list(size = 7, cmin = 0, cmax = 1, color = ~col, line = list(width=0.2, color = "grey89")),
                         opacity = 1,
                         text = ~paste0(gene, ", FDR=", signif(FDR, digits = 3)), hoverinfo = "text",
                         name = paste0("Found in user data (", nrow(data), ")")) #, name = "pull down"
      }
      p
    }
    p <- p%>%
      layout(xaxis = list(range=c(min_x-0.5, max_x+0.5), showgrid = F), yaxis = list(range=c(min_y-0.5, max_y+0.5), showgrid = F),
             legend = list(orientation = 'h', y = -0.23))
    if(input$c_colorscheme == "fdr" | input$c_colorscheme == "exac" | input$c_colorscheme == "user"){
      df <- ldply(c2_pf_db$d_g2s, data.frame)
      if(nrow(df) != 0){
        if(input$c_marker_text_prot_fam_db == "yes_label"){
          vp_layer_genes <- vp_layer_for_uploaded_genes(p, df)
        } else if(input$c_marker_text_prot_fam_db == "no_label"){
          vp_layer_genes <- vp_layer_for_uploaded_genes_no_text(p, df)
        }
        p <- vp_layer_genes
      } else{
        vp_layer_no_genes <- vp_layer_for_uploaded_genes_none(p, d)
        p <- vp_layer_no_genes
      }
      
    } else if(input$c_colorscheme == "cbf"){
      df <- ldply(c2_pf_db$d_g2s, data.frame)
      if(nrow(df) != 0){
        if(input$c_marker_text_prot_fam_db == "yes_label"){
          vp_layer_genes <- vp_layer_for_uploaded_genes_cbf(p, df)
        } else if(input$c_marker_text_prot_fam_db == "no_label"){
          vp_layer_genes <- vp_layer_for_uploaded_genes_cbf_no_text(p, df)
        }
        p <- vp_layer_genes
      } else{
        vp_layer_no_genes <- vp_layer_for_uploaded_genes_none_cbf(p, d)
        p <- vp_layer_no_genes
      }
      
    }
    p <- p %>% 
      layout(xaxis = list(range=c(min_x-0.5, max_x+0.5), showgrid = F), yaxis = list(range=c(min_y-0.5, max_y+0.5), showgrid = F)) %>%
      add_lines(x = c(min_x-0.5, max_x+0.5), y = -log10(input$c_pval_thresh), line = list(dash = "dash", width = 0.5, color = "#2b333e"), 
                name = '', hoverinfo = "text", text = paste0("pvalue = ", input$c_pval_thresh), showlegend = F) %>%
      add_lines(x = input$c_logfc_thresh_comb[1], y = c(min_y-0.5, max_y+0.5), line = list(dash = "dash", width = 0.5, color = "#252525"), 
                name = '', hoverinfo = "text", text = paste0("logFC = ", input$c_logfc_thresh_comb[1]), showlegend = F) %>%
      add_lines(x = input$c_logfc_thresh_comb[2], y = c(min_y-0.5, max_y+0.5), line = list(dash = "dash", width = 0.5, color = "#252525"), 
                name = '', hoverinfo = "text", text = paste0("logFC = ", input$c_logfc_thresh_comb[2]), showlegend = F)
  })
  
  c_vp3_pf_db_layer <- reactive({
    validate(
      need(!is.null(c_pd3()), ""),
      need(!is.null(c_pf_db()), "")
    )
    d <- c_pd3()
    c3_pf_db <- c_pf_db_vp3()
    min_x <- c_min_x()
    min_y <- c_min_y()
    max_x <- c_max_x()
    max_y <- c_max_y()
    pf_col <- input$c_colorbrewer_theme_pf
    if(input$c_colorscheme == "fdr"){
      req(input$c_color_indv_sig, input$c_color_indv_insig, input$c_colorbrewer_theme_pf)
      data <- separate_to_groups_for_color_integrated(d, input$c_fdr_thresh, input$c_color_indv_sig, input$c_color_indv_insig)
      p <- plot_ly(colors = pf_col, showlegend = T, width = 300, height = 390)
      for(i in nrow(data)){
        p <- add_markers(p, data = data, x = ~logFC, y = ~-log10(pvalue),
                         marker = list(size = 6, cmin = 0, cmax = 1, color = ~col), 
                         opacity = 0.8, 
                         text = ~paste(gene), hoverinfo = "text", name = "pull down")
      }
      p 
    } else if(input$c_colorscheme == "exac"){
      d$s <- exac$em_p_hi[match(d$gene, exac$GENE_NAME)]
      d$s[is.na(d$s)] <- 2
      below_thresh <- subset(d, s < 0.9)
      above_thresh <- subset(d, s >= 0.9)
      no_exist <- subset(d, s == 2)
      p <- plot_ly(colors = pf_col, showlegend = T, width = 300, height = 390)
      p <- add_markers(p, data = below_thresh, x = ~logFC, y = ~-log10(pvalue),
                       marker = list(size = 8, line = list(width=0.1, color = 'black'), cmin = 0, cmax = 1, color = "#66c2a5"),
                       opacity = 0.8, 
                       text = ~paste(gene), hoverinfo = "text", name = paste0("pLI<0.9 (", nrow(below_thresh), ")"))
      p <- add_markers(p, data = above_thresh, x = ~logFC, y = ~-log10(pvalue),
                       marker = list(size = 8, line = list(width=0.1, color = "black"), cmin = 0, cmax = 1, color = "#fc8d62"),
                       opacity = 0.8, 
                       text = ~paste(gene), hoverinfo = "text", name = paste0("pLI>=0.9 (", nrow(above_thresh), ")"))
      p <- add_markers(p, data = no_exist, x = ~logFC, y = ~-log10(pvalue),
                       marker = list(size = 8, line = list(width=0.1, color = "black"), cmin = 0, cmax = 1, color = "#8da0cb"),
                       opacity = 0.8, 
                       text = ~paste(gene), hoverinfo = "text", name = paste0("not in ExAC (", nrow(no_exist), ")"))
      p
    } else if(input$c_colorscheme == "cbf"){
      data <- separate_to_groups_for_cbf_integrated(d, input$c_fdr_thresh)
      p <- plot_ly(colors = "Greys", showlegend = T, width = 300, height = 390)
      for(i in nrow(data)){
        p <- add_markers(p, data = data, x = ~logFC, y = ~-log10(pvalue),
                         marker = list(size = 6, cmin = 0, cmax = 1, color = ~col), 
                         opacity = 0.6, 
                         text = ~paste(gene), hoverinfo = "text", name = "pull down")
      }
      p 
    } else if(input$c_colorscheme == "user"){
      validate(
        need(!is.null(input$c_file_color), "Please upload file with gene and score")
      )
      d1 <- c_in_file_color()
      col_theme <- input$c_colorbrewer_theme
      if(input$c_colorscheme_style == "cont"){
        df <- separate_to_groups_for_color_continuous(d, d1, col_theme)
        data <- df$df1
        data1 <- df$no_exist
      } else if(input$c_colorscheme_style == "disc"){
        df <- separate_to_groups_for_color_discrete(d, d1, col_theme)
        data <- df$df1
        data1 <- df$no_exist
      }
      p <- plot_ly(colors = pf_col, showlegend = T, width = 320, height = 390)
      p <- add_markers(p, data = data1, x = ~logFC, y = ~-log10(pvalue),
                       marker = list(size = 7, line = list(width=0.1, color = "grey89"), cmin = 0, cmax = 1, color = "#f7f7f7"),
                       opacity = 0.8,
                       text = ~paste(gene), hoverinfo = "text", name = paste0("Not in user data (", nrow(df$no_exist), ")"))
      for(i in nrow(data)){
        p <- add_markers(p, data = data, x = ~logFC, y = ~-log10(pvalue),
                         marker = list(size = 7, cmin = 0, cmax = 1, color = ~col, line = list(width=0.2, color = "grey89")),
                         opacity = 1,
                         text = ~paste0(gene, ", FDR=", signif(FDR, digits = 3)), hoverinfo = "text",
                         name = paste0("Found in user data (", nrow(data), ")")) #, name = "pull down"
      }
      p
    }
    p <- p%>%
      layout(xaxis = list(range=c(min_x-0.5, max_x+0.5), showgrid = F), yaxis = list(range=c(min_y-0.5, max_y+0.5), showgrid = F),
             legend = list(orientation = 'h', y = -0.23))
    if(input$c_colorscheme == "fdr" | input$c_colorscheme == "exac" | input$c_colorscheme == "user"){
      df <- ldply(c3_pf_db$d_g2s, data.frame)
      if(nrow(df) != 0){
        if(input$c_marker_text_prot_fam_db == "yes_label"){
          vp_layer_genes <- vp_layer_for_uploaded_genes(p, df)
        } else if(input$c_marker_text_prot_fam_db == "no_label"){
          vp_layer_genes <- vp_layer_for_uploaded_genes_no_text(p, df)
        }
        p <- vp_layer_genes
      } else{
        vp_layer_no_genes <- vp_layer_for_uploaded_genes_none(p, d)
        p <- vp_layer_no_genes
      }
      
    } else if(input$c_colorscheme == "cbf"){
      df <- ldply(c3_pf_db$d_g2s, data.frame)
      if(nrow(df) != 0){
        if(input$c_marker_text_prot_fam_db == "yes_label"){
          vp_layer_genes <- vp_layer_for_uploaded_genes_cbf(p, df)
        } else if(input$c_marker_text_prot_fam_db == "no_label"){
          vp_layer_genes <- vp_layer_for_uploaded_genes_cbf_no_text(p, df)
        }
        p <- vp_layer_genes
      } else{
        vp_layer_no_genes <- vp_layer_for_uploaded_genes_none_cbf(p, d)
        p <- vp_layer_no_genes
      }
      
    }
    p <- p %>% 
      layout(xaxis = list(range=c(min_x-0.5, max_x+0.5), showgrid = F), yaxis = list(range=c(min_y-0.5, max_y+0.5), showgrid = F)) %>%
      add_lines(x = c(min_x-0.5, max_x+0.5), y = -log10(input$c_pval_thresh), line = list(dash = "dash", width = 0.5, color = "#2b333e"), 
                name = '', hoverinfo = "text", text = paste0("pvalue = ", input$c_pval_thresh), showlegend = F) %>%
      add_lines(x = input$c_logfc_thresh_comb[1], y = c(min_y-0.5, max_y+0.5), line = list(dash = "dash", width = 0.5, color = "#252525"), 
                name = '', hoverinfo = "text", text = paste0("logFC = ", input$c_logfc_thresh_comb[1]), showlegend = F) %>%
      add_lines(x = input$c_logfc_thresh_comb[2], y = c(min_y-0.5, max_y+0.5), line = list(dash = "dash", width = 0.5, color = "#252525"), 
                name = '', hoverinfo = "text", text = paste0("logFC = ", input$c_logfc_thresh_comb[2]), showlegend = F)
  })
  
  c_sp1_pf_db_layer <- reactive({
    validate(
      need(!is.null(c_pd1()), ""),
      need(!is.null(c_pf_db()), "")
    )
    d <- c_pd1()
    c1_pf_db <- c_pf_db_vp1()
    cc <- c_sp_cor(d)
    pf_col <- input$c_colorbrewer_theme_pf
    if(input$c_colorscheme == "fdr"){
      req(input$c_color_indv_sig, input$c_color_indv_insig, input$c_colorbrewer_theme_pf)
      data <- separate_to_groups_for_color_integrated(d, input$a_fdr_thresh, input$c_color_indv_sig, input$c_color_indv_insig)
      p <- plot_ly(colors = pf_col, showlegend = F, width = 320, height = 320) 
      p <- add_lines(p, data = d, x = ~c(ceiling(min(rep1, rep2)), floor(max(rep1, rep2))), y = ~c(ceiling(min(rep1, rep2)), floor(max(rep1, rep2))),
                     line = list(dash = "dash", width = 1, color = "#252525"), showlegend = FALSE)
      for(i in nrow(data)){
        p <- add_markers(p, data = data, x = ~rep1, y = ~rep2, 
                         marker = list(size = 6, cmin = 0, cmax = 1, color = ~col), #line = list(width=0.1, color = "black"),
                         opacity = 0.8, 
                         text = ~paste(gene), hoverinfo = "text", name = "pull down")
      }
    } else if(input$c_colorscheme == "exac"){
      d$s <- exac$em_p_hi[match(d$gene, exac$GENE_NAME)]
      d$s[is.na(d$s)] <- 2
      below_thresh <- subset(d, s < 0.9)
      above_thresh <- subset(d, s >= 0.9)
      no_exist <- subset(d, s == 2)
      p <- plot_ly(colors = pf_col, showlegend = F, width = 320, height = 320) 
      p <- add_lines(p, data = d, x = ~c((min(rep1, rep2)), (max(rep1, rep2))), y = ~c((min(rep1, rep2)), (max(rep1, rep2))),
                     line = list(dash = "dash", width = 1, color = "#252525"), showlegend = FALSE)
      p <- add_markers(p, data = below_thresh, x = ~rep1, y = ~rep2, 
                       marker = list(size = 8, line = list(width=0.1, color = 'black'), cmin = 0, cmax = 1, color = "#66c2a5"),
                       opacity = 0.7, 
                       text = ~paste(gene), hoverinfo = "text") #, name = paste0("pLI<0.9 (", nrow(below_thresh), ")")
      p <- add_markers(p, data = above_thresh, x = ~rep1, y = ~rep2, 
                       marker = list(size = 8, line = list(width=0.1, color = "black"), cmin = 0, cmax = 1, color = "#fc8d62"),
                       opacity = 0.7, 
                       text = ~paste(gene), hoverinfo = "text") #, name = paste0("pLI>=0.9 (", nrow(above_thresh), ")")
      p <- add_markers(p, data = no_exist, x = ~rep1, y = ~rep2, 
                       marker = list(size = 8, line = list(width=0.1, color = "black"), cmin = 0, cmax = 1, color = "#8da0cb"),
                       opacity = 0.7, 
                       text = ~paste(gene), hoverinfo = "text") #, name = paste0("not in ExAC (", nrow(no_exist), ")")
      p
    } else if(input$c_colorscheme == "cbf"){
      data <- separate_to_groups_for_cbf_integrated(d, input$a_fdr_thresh)
      p <- plot_ly(colors = "Greys", showlegend = F, width = 320, height = 320) 
      p <- add_lines(p, data = d, x = ~c(ceiling(min(rep1, rep2)), floor(max(rep1, rep2))), y = ~c(ceiling(min(rep1, rep2)), floor(max(rep1, rep2))),
                     line = list(dash = "dash", width = 1, color = "#252525"), showlegend = FALSE)
      for(i in nrow(data)){
        p <- add_markers(p, data = data, x = ~rep1, y = ~rep2, 
                         marker = list(size = 6, cmin = 0, cmax = 1, color = ~col), #line = list(width=0.1, color = "black"), 
                         opacity = 0.6, 
                         text = ~paste(gene), hoverinfo = "text", name = "pull down")
      }
    } else if(input$c_colorscheme == "user"){
      validate(
        need(!is.null(input$c_file_color), "Please upload file with gene and score")
      )
      d1 <- c_in_file_color()
      col_theme <- input$c_colorbrewer_theme
      if(input$c_colorscheme_style == "cont"){
        df <- separate_to_groups_for_color_continuous(d, d1, col_theme)
        data <- df$df1
        data1 <- df$no_exist
      } else if(input$c_colorscheme_style == "disc"){
        df <- separate_to_groups_for_color_discrete(d, d1, col_theme)
        data <- df$df1
        data1 <- df$no_exist
      }
      p <- plot_ly(colors = pf_col, showlegend = FALSE, width = 320, height = 320) 
      p <- add_lines(p, data = d, x = ~c((min(rep1, rep2)), (max(rep1, rep2))), y = ~c((min(rep1, rep2)), (max(rep1, rep2))),
                     text = "x=y", hoverinfo = "text",
                     line = list(dash = "dash", width = 1, color = "#252525"), showlegend = FALSE)
      p <- add_markers(p, data = data1, x = ~rep1, y = ~rep2,
                       marker = list(size = 7, line = list(width=0.1, color = "grey89"), cmin = 0, cmax = 1, color = "#f7f7f7"),
                       opacity = 0.8,
                       text = ~paste(gene), hoverinfo = "text", name = paste0("Not in user data (", nrow(df$no_exist), ")"))
      for(i in nrow(data)){
        p <- add_markers(p, data = data, x = ~rep1, y = ~rep2,
                         marker = list(size = 7, cmin = 0, cmax = 1, color = ~col, line = list(width=0.2, color = "grey89")),
                         opacity = 1,
                         text = ~paste0(gene, ", FDR=", signif(FDR, digits = 3)), hoverinfo = "text",
                         name = paste0("Found in user data (", nrow(data), ")")) #, name = "pull down"
      }
      p
    }
    if(input$c_colorscheme == "fdr" | input$c_colorscheme == "exac" | input$c_colorscheme == "user"){
      df <- ldply(c1_pf_db$d_g2s, data.frame)
      if(nrow(df) != 0){
        if(input$c_marker_text_prot_fam_db == "yes_label"){
          sp_layer_genes <- sp_layer_for_uploaded_genes(p, df)
        } else if(input$c_marker_text_prot_fam_db == "no_label"){
          sp_layer_genes <- sp_layer_for_uploaded_genes_no_text(p, df)
        }
        p <- sp_layer_genes
      } else{
        sp_layer_no_genes <- sp_layer_for_uploaded_genes_none(p, d)
        p <- sp_layer_no_genes
      }
    } else if(input$c_colorscheme == "cbf"){
      df <- ldply(c1_pf_db$d_g2s, data.frame)
      if(nrow(df) != 0){
        if(input$c_marker_text_prot_fam_db == "yes_label"){
          sp_layer_genes <- sp_layer_for_uploaded_genes_cbf(p, df)
        } else if(input$c_marker_text_prot_fam_db == "no_label"){
          sp_layer_genes <- sp_layer_for_uploaded_genes_cbf_no_text(p, df)
        }
        p <- sp_layer_genes
      } else{
        sp_layer_no_genes <- sp_layer_for_uploaded_genes_none_cbf(p, d)
        p <- sp_layer_no_genes
      }
    }
    p <- p %>%
      layout(xaxis = list(title = "rep1", range=~c((min(d$rep1, d$rep2))-1, (max(d$rep1, d$rep2))+1)), 
             yaxis = list(title = "rep2", range=~c((min(d$rep1, d$rep2))-1, (max(d$rep1, d$rep2))+1)), 
             title = cc, titlefont = list(size=12))
  })
  
  c_sp2_pf_db_layer <- reactive({
    validate(
      need(!is.null(c_pd2()), ""),
      need(!is.null(c_pf_db()), "")
    )
    d <- c_pd2()
    c2_pf_db <- c_pf_db_vp2()
    cc <- c_sp_cor(d)
    pf_col <- input$c_colorbrewer_theme_pf
    if(input$c_colorscheme == "fdr"){
      req(input$c_color_indv_sig, input$c_color_indv_insig, input$c_colorbrewer_theme_pf)
      data <- separate_to_groups_for_color_integrated(d, input$a_fdr_thresh, input$c_color_indv_sig, input$c_color_indv_insig)
      p <- plot_ly(colors = pf_col, showlegend = F, width = 320, height = 320) 
      p <- add_lines(p, data = d, x = ~c(ceiling(min(rep1, rep2)), floor(max(rep1, rep2))), y = ~c(ceiling(min(rep1, rep2)), floor(max(rep1, rep2))),
                     line = list(dash = "dash", width = 1, color = "#252525"), showlegend = FALSE)
      for(i in nrow(data)){
        p <- add_markers(p, data = data, x = ~rep1, y = ~rep2, 
                         marker = list(size = 6, cmin = 0, cmax = 1, color = ~col), #line = list(width=0.1, color = "black"),
                         opacity = 0.8, 
                         text = ~paste(gene), hoverinfo = "text", name = "pull down")
      }
    } else if(input$c_colorscheme == "exac"){
      d$s <- exac$em_p_hi[match(d$gene, exac$GENE_NAME)]
      d$s[is.na(d$s)] <- 2
      below_thresh <- subset(d, s < 0.9)
      above_thresh <- subset(d, s >= 0.9)
      no_exist <- subset(d, s == 2)
      p <- plot_ly(colors = pf_col, showlegend = F, width = 320, height = 320) 
      p <- add_lines(p, data = d, x = ~c((min(rep1, rep2)), (max(rep1, rep2))), y = ~c((min(rep1, rep2)), (max(rep1, rep2))),
                     line = list(dash = "dash", width = 1, color = "#252525"), showlegend = FALSE)
      p <- add_markers(p, data = below_thresh, x = ~rep1, y = ~rep2, 
                       marker = list(size = 8, line = list(width=0.1, color = 'black'), cmin = 0, cmax = 1, color = "#66c2a5"),
                       opacity = 0.7, 
                       text = ~paste(gene), hoverinfo = "text", name = paste0("pLI<0.9 (", nrow(below_thresh), ")"))
      p <- add_markers(p, data = above_thresh, x = ~rep1, y = ~rep2, 
                       marker = list(size = 8, line = list(width=0.1, color = "black"), cmin = 0, cmax = 1, color = "#fc8d62"),
                       opacity = 0.7, 
                       text = ~paste(gene), hoverinfo = "text", name = paste0("pLI>=0.9 (", nrow(above_thresh), ")"))
      p <- add_markers(p, data = no_exist, x = ~rep1, y = ~rep2, 
                       marker = list(size = 8, line = list(width=0.1, color = "black"), cmin = 0, cmax = 1, color = "#8da0cb"),
                       opacity = 0.7, 
                       text = ~paste(gene), hoverinfo = "text", name = paste0("not in ExAC (", nrow(no_exist), ")"))
      p
    } else if(input$c_colorscheme == "cbf"){
      data <- separate_to_groups_for_cbf_integrated(d, input$a_fdr_thresh)
      p <- plot_ly(colors = "Greys", showlegend = F, width = 320, height = 320) 
      p <- add_lines(p, data = d, x = ~c(ceiling(min(rep1, rep2)), floor(max(rep1, rep2))), y = ~c(ceiling(min(rep1, rep2)), floor(max(rep1, rep2))),
                     line = list(dash = "dash", width = 1, color = "#252525"), showlegend = FALSE)
      for(i in nrow(data)){
        p <- add_markers(p, data = data, x = ~rep1, y = ~rep2, 
                         marker = list(size = 6, cmin = 0, cmax = 1, color = ~col), #line = list(width=0.1, color = "black"), 
                         opacity = 0.6, 
                         text = ~paste(gene), hoverinfo = "text", name = "pull down")
      }
    } else if(input$c_colorscheme == "user"){
      validate(
        need(!is.null(input$c_file_color), "Please upload file with gene and score")
      )
      d1 <- c_in_file_color()
      col_theme <- input$c_colorbrewer_theme
      if(input$c_colorscheme_style == "cont"){
        df <- separate_to_groups_for_color_continuous(d, d1, col_theme)
        data <- df$df1
        data1 <- df$no_exist
      } else if(input$c_colorscheme_style == "disc"){
        df <- separate_to_groups_for_color_discrete(d, d1, col_theme)
        data <- df$df1
        data1 <- df$no_exist
      }
      p <- plot_ly(colors = pf_col, showlegend = FALSE, width = 320, height = 320) 
      p <- add_lines(p, data = d, x = ~c((min(rep1, rep2)), (max(rep1, rep2))), y = ~c((min(rep1, rep2)), (max(rep1, rep2))),
                     text = "x=y", hoverinfo = "text",
                     line = list(dash = "dash", width = 1, color = "#252525"), showlegend = FALSE)
      p <- add_markers(p, data = data1, x = ~rep1, y = ~rep2,
                       marker = list(size = 7, line = list(width=0.1, color = "grey89"), cmin = 0, cmax = 1, color = "#f7f7f7"),
                       opacity = 0.8,
                       text = ~paste(gene), hoverinfo = "text", name = paste0("Not in user data (", nrow(df$no_exist), ")"))
      for(i in nrow(data)){
        p <- add_markers(p, data = data, x = ~rep1, y = ~rep2,
                         marker = list(size = 7, cmin = 0, cmax = 1, color = ~col, line = list(width=0.2, color = "grey89")),
                         opacity = 1,
                         text = ~paste0(gene, ", FDR=", signif(FDR, digits = 3)), hoverinfo = "text",
                         name = paste0("Found in user data (", nrow(data), ")")) #, name = "pull down"
      }
      p
    }
    if(input$c_colorscheme == "fdr" | input$c_colorscheme == "exac" | input$c_colorscheme == "user"){
      df <- ldply(c2_pf_db$d_g2s, data.frame)
      if(nrow(df) != 0){
        if(input$c_marker_text_prot_fam_db == "yes_label"){
          sp_layer_genes <- sp_layer_for_uploaded_genes(p, df)
        } else if(input$c_marker_text_prot_fam_db == "no_label"){
          sp_layer_genes <- sp_layer_for_uploaded_genes_no_text(p, df)
        }
        p <- sp_layer_genes
      } else{
        sp_layer_no_genes <- sp_layer_for_uploaded_genes_none(p, d)
        p <- sp_layer_no_genes
      }
    } else if(input$c_colorscheme == "cbf"){
      df <- ldply(c2_pf_db$d_g2s, data.frame)
      if(nrow(df) != 0){
        if(input$c_marker_text_prot_fam_db == "yes_label"){
          sp_layer_genes <- sp_layer_for_uploaded_genes_cbf(p, df)
        } else if(input$c_marker_text_prot_fam_db == "no_label"){
          sp_layer_genes <- sp_layer_for_uploaded_genes_cbf_no_text(p, df)
        }
        p <- sp_layer_genes
      } else{
        sp_layer_no_genes <- sp_layer_for_uploaded_genes_none_cbf(p, d)
        p <- sp_layer_no_genes
      }
    }
    p <- p %>%
      layout(xaxis = list(title = "rep1", range=~c((min(d$rep1, d$rep2))-1, (max(d$rep1, d$rep2))+1)), 
             yaxis = list(title = "rep2", range=~c((min(d$rep1, d$rep2))-1, (max(d$rep1, d$rep2))+1)), 
             title = cc, titlefont = list(size=12))
  })
  
  c_sp3_pf_db_layer <- reactive({
    validate(
      need(!is.null(c_pd3()), ""),
      need(!is.null(c_pf_db()), "")
    )
    d <- c_pd3()
    c3_pf_db <- c_pf_db_vp3()
    cc <- c_sp_cor(d)
    pf_col <- input$c_colorbrewer_theme_pf
    if(input$c_colorscheme == "fdr"){
      req(input$c_color_indv_sig, input$c_color_indv_insig, input$c_colorbrewer_theme_pf)
      data <- separate_to_groups_for_color_integrated(d, input$a_fdr_thresh, input$c_color_indv_sig, input$c_color_indv_insig)
      p <- plot_ly(colors = pf_col, showlegend = F, width = 320, height = 320) 
      p <- add_lines(p, data = d, x = ~c(ceiling(min(rep1, rep2)), floor(max(rep1, rep2))), y = ~c(ceiling(min(rep1, rep2)), floor(max(rep1, rep2))),
                     line = list(dash = "dash", width = 1, color = "#252525"), showlegend = FALSE)
      for(i in nrow(data)){
        p <- add_markers(p, data = data, x = ~rep1, y = ~rep2, 
                         marker = list(size = 6, cmin = 0, cmax = 1, color = ~col), #line = list(width=0.1, color = "black"),
                         opacity = 0.8, 
                         text = ~paste(gene), hoverinfo = "text", name = "pull down")
      }
    } else if(input$c_colorscheme == "exac"){
      d$s <- exac$em_p_hi[match(d$gene, exac$GENE_NAME)]
      d$s[is.na(d$s)] <- 2
      below_thresh <- subset(d, s < 0.9)
      above_thresh <- subset(d, s >= 0.9)
      no_exist <- subset(d, s == 2)
      p <- plot_ly(colors = pf_col, showlegend = F, width = 320, height = 320) 
      p <- add_lines(p, data = d, x = ~c((min(rep1, rep2)), (max(rep1, rep2))), y = ~c((min(rep1, rep2)), (max(rep1, rep2))),
                     line = list(dash = "dash", width = 1, color = "#252525"), showlegend = FALSE)
      p <- add_markers(p, data = below_thresh, x = ~rep1, y = ~rep2, 
                       marker = list(size = 8, line = list(width=0.1, color = 'black'), cmin = 0, cmax = 1, color = "#66c2a5"),
                       opacity = 0.7, 
                       text = ~paste(gene), hoverinfo = "text", name = paste0("pLI<0.9 (", nrow(below_thresh), ")"))
      p <- add_markers(p, data = above_thresh, x = ~rep1, y = ~rep2, 
                       marker = list(size = 8, line = list(width=0.1, color = "black"), cmin = 0, cmax = 1, color = "#fc8d62"),
                       opacity = 0.7, 
                       text = ~paste(gene), hoverinfo = "text", name = paste0("pLI>=0.9 (", nrow(above_thresh), ")"))
      p <- add_markers(p, data = no_exist, x = ~rep1, y = ~rep2, 
                       marker = list(size = 8, line = list(width=0.1, color = "black"), cmin = 0, cmax = 1, color = "#8da0cb"),
                       opacity = 0.7, 
                       text = ~paste(gene), hoverinfo = "text", name = paste0("not in ExAC (", nrow(no_exist), ")"))
      p
    } else if(input$c_colorscheme == "cbf"){
      data <- separate_to_groups_for_cbf_integrated(d, input$a_fdr_thresh)
      p <- plot_ly(colors = "Greys", showlegend = F, width = 320, height = 320) 
      p <- add_lines(p, data = d, x = ~c(ceiling(min(rep1, rep2)), floor(max(rep1, rep2))), y = ~c(ceiling(min(rep1, rep2)), floor(max(rep1, rep2))),
                     line = list(dash = "dash", width = 1, color = "#252525"), showlegend = FALSE)
      for(i in nrow(data)){
        p <- add_markers(p, data = data, x = ~rep1, y = ~rep2, 
                         marker = list(size = 6, cmin = 0, cmax = 1, color = ~col), #line = list(width=0.1, color = "black"), 
                         opacity = 0.6, 
                         text = ~paste(gene), hoverinfo = "text", name = "pull down")
      }
    } else if(input$c_colorscheme == "user"){
      validate(
        need(!is.null(input$c_file_color), "Please upload file with gene and score")
      )
      d1 <- c_in_file_color()
      col_theme <- input$c_colorbrewer_theme
      if(input$c_colorscheme_style == "cont"){
        df <- separate_to_groups_for_color_continuous(d, d1, col_theme)
        data <- df$df1
        data1 <- df$no_exist
      } else if(input$c_colorscheme_style == "disc"){
        df <- separate_to_groups_for_color_discrete(d, d1, col_theme)
        data <- df$df1
        data1 <- df$no_exist
      }
      p <- plot_ly(colors = pf_col, showlegend = FALSE, width = 550, height = 550) 
      p <- add_lines(p, data = d, x = ~c((min(rep1, rep2)), (max(rep1, rep2))), y = ~c((min(rep1, rep2)), (max(rep1, rep2))),
                     text = "x=y", hoverinfo = "text",
                     line = list(dash = "dash", width = 1, color = "#252525"), showlegend = FALSE)
      p <- add_markers(p, data = data1, x = ~rep1, y = ~rep2,
                       marker = list(size = 7, line = list(width=0.1, color = "grey89"), cmin = 0, cmax = 1, color = "#f7f7f7"),
                       opacity = 0.8,
                       text = ~paste(gene), hoverinfo = "text", name = paste0("Not in user data (", nrow(df$no_exist), ")"))
      for(i in nrow(data)){
        p <- add_markers(p, data = data, x = ~rep1, y = ~rep2,
                         marker = list(size = 7, cmin = 0, cmax = 1, color = ~col, line = list(width=0.2, color = "grey89")),
                         opacity = 1,
                         text = ~paste0(gene, ", FDR=", signif(FDR, digits = 3)), hoverinfo = "text",
                         name = paste0("Found in user data (", nrow(data), ")")) #, name = "pull down"
      }
      p
    }
    if(input$c_colorscheme == "fdr" | input$c_colorscheme == "exac" | input$c_colorscheme == "user"){
      df <- ldply(c3_pf_db$d_g2s, data.frame)
      if(nrow(df) != 0){
        if(input$c_marker_text_prot_fam_db == "yes_label"){
          sp_layer_genes <- sp_layer_for_uploaded_genes(p, df)
        } else if(input$c_marker_text_prot_fam_db == "no_label"){
          sp_layer_genes <- sp_layer_for_uploaded_genes_no_text(p, df)
        }
        p <- sp_layer_genes
      } else{
        sp_layer_no_genes <- sp_layer_for_uploaded_genes_none(p, d)
        p <- sp_layer_no_genes
      }
    } else if(input$c_colorscheme == "cbf"){
      df <- ldply(c3_pf_db$d_g2s, data.frame)
      if(nrow(df) != 0){
        if(input$c_marker_text_prot_fam_db == "yes_label"){
          sp_layer_genes <- sp_layer_for_uploaded_genes_cbf(p, df)
        } else if(input$c_marker_text_prot_fam_db == "no_label"){
          sp_layer_genes <- sp_layer_for_uploaded_genes_cbf_no_text(p, df)
        }
        p <- sp_layer_genes
      } else{
        sp_layer_no_genes <- sp_layer_for_uploaded_genes_none_cbf(p, d)
        p <- sp_layer_no_genes
      }
    }
    p <- p %>%
      layout(xaxis = list(title = "rep1", range=~c((min(d$rep1, d$rep2))-1, (max(d$rep1, d$rep2))+1)), 
             yaxis = list(title = "rep2", range=~c((min(d$rep1, d$rep2))-1, (max(d$rep1, d$rep2))+1)), 
             title = cc, titlefont = list(size=12))
  })
  
  output$c_FDR_colorbar <- renderPlot({
    validate(
      need(input$c_file_pulldown1 != '', "")
    )
    c_vp_colorbar_dl()
  })
  
  output$c_inweb_colorbar <- renderPlot({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_make_plot_inweb != 0, "")
    )
    c_vp_colorbar_inweb()
  })
  
  output$c_goi_colorbar <- renderPlot({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_make_plot_goi != 0, "")
    )
    c_vp_colorbar_goi()
  })
  
  # output$c_snp_colorbar <- renderPlot({
  #   validate(
  #     need(input$c_file_pulldown1 != '', ""),
  #     need(input$c_make_plot_snp != 0, "")
  #   )
  #   c_vp_colorbar_snp()
  # })
  
  output$VolcanoPlot_c1 <- renderPlotly({
    validate(
      need(input$c_file_pulldown1 != '', "Upload file")
    )
    if(is.null(c_pf_db())){
      if(is.null(c_search_gene())){
        c_vp1()
      } else{
        c_vp1_plus()
      }
    } else if(!is.null(c_pf_db())){
      c_vp1_pf_db_layer()
    }
  })
  
  
  output$VolcanoPlot_c1_inweb <- renderPlotly({
    validate(
      need(input$c_file_pulldown1 != '', "Upload file")
    )
    if(is.null(c_search_gene())){
      c_vp1_inweb_layer()
    } else{
      c_vp1_inweb_plus()
    }
  })
  
  output$VolcanoPlot_c1_goi <- renderPlotly({
    validate(
      need(input$c_file_pulldown1 != '', "Upload file")
    )
    if(is.null(c_search_gene())){
      c_vp1_goi_layer()
    } else{
      c_vp1_goi_plus()
    }
  })
  
  # output$VolcanoPlot_c1_snp <- renderPlotly({
  #   validate(
  #     need(input$c_file_pulldown1 != '', "Upload file")
  #   )
  #   if(is.null(c_search_gene())){
  #     c_vp1_snp_layer()
  #   } else{
  #     c_vp1_snp_plus()
  #   }
  # })
  
  output$VolcanoPlot_c2 <- renderPlotly({
    validate(
      need(input$c_file_pulldown2 != '', "Upload file")
    )
    if(is.null(c_pf_db())){
      if(is.null(c_search_gene())){
        c_vp2()
      } else{
        c_vp2_plus()
      }
    } else if(!is.null(c_pf_db())){
      c_vp2_pf_db_layer()
    }
  })
  
  output$VolcanoPlot_c2_inweb <- renderPlotly({
    validate(
      need(input$c_file_pulldown2 != '', "Upload file")
    )
    if(is.null(c_search_gene())){
      c_vp2_inweb_layer()
    } else{
      c_vp2_inweb_plus()
    }
  })
  
  output$VolcanoPlot_c2_goi <- renderPlotly({
    validate(
      need(input$c_file_pulldown2 != '', "Upload file")
    )
    if(is.null(c_search_gene())){
      c_vp2_goi_layer()
    } else{
      c_vp2_goi_plus()
    }
  })
  
  # output$VolcanoPlot_c2_snp <- renderPlotly({
  #   validate(
  #     need(input$c_file_pulldown2 != '', "Upload file")
  #   )
  #   if(is.null(c_search_gene())){
  #     c_vp2_snp_layer()
  #   } else{
  #     c_vp2_snp_plus()
  #   }
  # })
  # 
  output$VolcanoPlot_c3 <- renderPlotly({
    validate(
      need(input$c_file_pulldown3 != '', "Upload file")
    )
    if(is.null(c_pf_db())){
      if(is.null(c_search_gene())){
        c_vp3()
      } else{
        c_vp3_plus()
      }
    } else if(!is.null(c_pf_db())){
      c_vp3_pf_db_layer()
    }
  })
  
  output$VolcanoPlot_c3_inweb <- renderPlotly({
    validate(
      need(input$c_file_pulldown3 != '', "Upload file")
    )
    if(is.null(c_search_gene())){
      c_vp3_inweb_layer()
    } else{
      c_vp3_inweb_plus()
    }
  })
  
  output$VolcanoPlot_c3_goi <- renderPlotly({
    validate(
      need(input$c_file_pulldown3 != '', "Upload file")
    )
    if(is.null(c_search_gene())){
      c_vp3_goi_layer()
    } else{
      c_vp3_goi_plus()
    }
  })
  # 
  # output$VolcanoPlot_c3_snp <- renderPlotly({
  #   validate(
  #     need(input$c_file_pulldown3 != '', "Upload file")
  #   )
  #   if(is.null(c_search_gene())){
  #     c_vp3_snp_layer()
  #   } else{
  #     c_vp3_snp_plus()
  #   }
  # })
  
  output$c_VennDiagram_inweb <- renderPlot({
    validate(
      need(input$c_file_pulldown1 != '', "Upload file"),
      need(input$c_file_pulldown2 != '', "")
    )
    v0 <- c_venndiagram_inweb()
    grid.newpage()
    grid.draw(v0)
  })
  
  output$c_VennDiagram_goi <- renderPlot({
    validate(
      need(input$c_file_pulldown1 != '', "Upload file"),
      need(input$c_file_pulldown2 != '', "")
    )
    v0 <- c_venndiagram_goi()
    grid.newpage()
    grid.draw(v0)
  })
  
  # output$c_VennDiagram_snp <- renderPlot({
  #   validate(
  #     need(input$c_file_pulldown1 != '', "Upload file"),
  #     need(input$c_file_pulldown2 != '', "")
  #   )
  #   v0 <- c_venndiagram_snp()
  #   grid.newpage()
  #   grid.draw(v0)
  # })
  
  output$ScatterPlot_c1 <- renderPlotly({
    validate(
      need(input$c_file_pulldown1 != '', "Upload file")
    )
    input_file <- c_in_pd1()
    d_col <- colnames(input_file)
    if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col &
       "rep1" %in% d_col & "rep2" %in% d_col){
      if(is.null(c_pf_db())){
        if(is.null(c_search_gene())){
          c_sp1()
        } else{
          c_sp1_plus()
        }
      }else if(!is.null(c_pf_db())){
        c_sp1_pf_db_layer()
      }
    } else if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col){
      validate(
        need("rep1" %in% d_col & "rep2" %in% d_col, "Must have rep1 and rep2 values.")
      )
    } else if("rep1" %in% d_col & "rep2" %in% d_col){
      if(is.null(c_pf_db())){
        if(is.null(c_search_gene())){
          c_sp1()
        } else{
          c_sp1_plus()
        }
      }else if(!is.null(c_pf_db())){
        c_sp1_pf_db_layer()
      }
    }
  })
  
  output$ScatterPlot_c2 <- renderPlotly({
    validate(
      need(input$c_file_pulldown2 != '', "Upload file")
    )
    input_file <- c_in_pd2()
    d_col <- colnames(input_file)
    if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col &
       "rep1" %in% d_col & "rep2" %in% d_col){
      if(is.null(c_pf_db())){
        if(is.null(c_search_gene())){
          c_sp2()
        } else{
          c_sp2_plus()
        }
      }else if(!is.null(c_pf_db())){
        c_sp2_pf_db_layer()
      }
    } else if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col){
      validate(
        need("rep1" %in% d_col & "rep2" %in% d_col, "Must have rep1 and rep2 values.")
      )
    } else if("rep1" %in% d_col & "rep2" %in% d_col){
      if(is.null(c_pf_db())){
        if(is.null(c_search_gene())){
          c_sp2()
        } else{
          c_sp2_plus()
        }
      }else if(!is.null(c_pf_db())){
        c_sp2_pf_db_layer()
      }
    }
  })
  
  output$ScatterPlot_c3 <- renderPlotly({
    validate(
      need(input$c_file_pulldown3 != '', "Upload file")
    )
    input_file <- c_in_pd3()
    d_col <- colnames(input_file)
    if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col &
       "rep1" %in% d_col & "rep2" %in% d_col){
      if(is.null(c_pf_db())){
        if(is.null(c_search_gene())){
          c_sp3()
        } else{
          c_sp3_plus()
        }
      }else if(!is.null(c_pf_db())){
        c_sp3_pf_db_layer()
      }
    } else if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col){
      validate(
        need("rep1" %in% d_col & "rep2" %in% d_col, "Must have rep1 and rep2 values.")
      )
    } else if("rep1" %in% d_col & "rep2" %in% d_col){
      if(is.null(c_pf_db())){
        if(is.null(c_search_gene())){
          c_sp3()
        } else{
          c_sp3_plus()
        }
      }else if(!is.null(c_pf_db())){
        c_sp3_pf_db_layer()
      }
    }
  })
  
  output$c_comparison_text <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_file_pulldown2 != '', "")
    )
    HTML("<b>User-defined significance thresholds:</b>")
  })
  
  output$c_comparison_text_f1 <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_file_pulldown2 != '', "")
    )
    HTML("<b>Controllers for file1:</b>")
  })
  
  output$c_comparison_text_f2 <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_file_pulldown2 != '', "")
    )
    HTML("<b>Controllers for file2:</b>")
  })
  
  output$c_comparison_text_f3 <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_file_pulldown2 != '', ""),
      need(input$c_file_pulldown3 != '', "")
    )
    HTML("<b>Controllers for file3:</b>")
  })
  
  output$comparison1 <- renderPlotly({
    validate(
      need(input$c_file_pulldown1 != '', "Upload file"),
      need(input$c_file_pulldown2 != '', "")
    )
    if(is.null(c_search_gene())){
      c_compare1_plot()
    } else{
      c_compare1_plus()
    }
  })
  
  output$comparison1_scatter <- renderPlotly({
    validate(
      need(input$c_file_pulldown1 != '', "Upload file"),
      need(input$c_file_pulldown2 != '', "")
    )
    if(is.null(c_search_gene())){
      c_compare1_plot_scatter()
    } else{
      c_compare1_plus_scatter()
    }
  })
  
  output$comparison1_pf <- renderPlotly({
    validate(
      need(input$c_file_pulldown1 != '', "Upload file"),
      need(input$c_file_pulldown2 != '', "")
    )
    c_compare1_pfe_plot()
  })
  
  output$comparison2 <- renderPlotly({
    validate(
      need(input$c_file_pulldown1 != '', "Upload file"),
      need(input$c_file_pulldown2 != '', "")
    )
    if(is.null(c_search_gene())){
      c_compare2_plot()
    } else{
      c_compare2_plus()
    }
  })
  
  output$comparison2_scatter <- renderPlotly({
    validate(
      need(input$c_file_pulldown1 != '', "Upload file"),
      need(input$c_file_pulldown2 != '', "")
    )
    if(is.null(c_search_gene())){
      c_compare2_plot_scatter()
    } else{
      c_compare2_plus_scatter()
    }
  })
  
  output$comparison2_pf <- renderPlotly({
    validate(
      need(input$c_file_pulldown1 != '', "Upload file"),
      need(input$c_file_pulldown2 != '', "")
    )
    c_compare2_pfe_plot()
  })
  
  output$comparison3 <- renderPlotly({
    validate(
      need(input$c_file_pulldown1 != '', "Upload file"),
      need(input$c_file_pulldown2 != '', ""),
      need(input$c_file_pulldown3 != '', "")
    )
    if(is.null(c_search_gene())){
      c_compare3_plot()
    } else{
      c_compare3_plus()
    }
  })
  
  output$comparison3_scatter <- renderPlotly({
    validate(
      need(input$c_file_pulldown1 != '', "Upload file"),
      need(input$c_file_pulldown2 != '', ""),
      need(input$c_file_pulldown3 != '', "")
    )
    if(is.null(c_search_gene())){
      c_compare3_plot_scatter()
    } else{
      c_compare3_plus_scatter()
    }
  })
  
  output$comparison3_pf <- renderPlotly({
    validate(
      need(input$c_file_pulldown1 != '', "Upload file"),
      need(input$c_file_pulldown2 != '', ""),
      need(input$c_file_pulldown3 != '', "")
    )
    c_compare3_pfe_plot()
  })
  
  output$c_VennDiagram_legend <- renderUI({
    validate(
      need(input$c_file_pulldown1 != '', "Upload file"),
      need(input$c_file_pulldown2 != '', "")
    )
    if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
      c_compare2()
      img(src = 'pc_2.png')
    }
    else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
      c_compare3()
      img(src = 'pc_3.png')
    }
  })
  
  output$c_VennDiagram <- renderPlot({
    validate(
      need(input$c_file_pulldown1 != '', "Upload file"),
      need(input$c_file_pulldown2 != '', "")
    )
    v0 <- c_venndiagram()
    grid.newpage()
    grid.draw(v0)
  })
  
  output$c_unique <- renderTable({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_file_pulldown2 != '', "")
    )
    c_unique_dat()
  },
  caption.placement = getOption("xtable.caption.placement", "top"), 
  caption.width = getOption("xtable.caption.width", NULL))
  
  output$c_unique_inweb <- renderTable({
    validate(
      need(input$c_file_pulldown1 != '', ""),
      need(input$c_file_pulldown2 != '', "")
    )
    c_unique_inweb_dat()
  },
  caption.placement = getOption("xtable.caption.placement", "top"), 
  caption.width = getOption("xtable.caption.width", NULL))
  
  output$c_unique_goi <- renderTable({
    validate(
      need(!is.null(c_vp1_goi_layer()), ""),
      need(!is.null(c_vp2_goi_layer()), "")
    )
    c_unique_goi_dat()
  },
  caption.placement = getOption("xtable.caption.placement", "top"), 
  caption.width = getOption("xtable.caption.width", NULL))
  
  # output$c_unique_snp <- renderTable({
  #   validate(
  #     need(!is.null(c_vp1_snp_layer()), ""),
  #     need(!is.null(c_vp2_snp_layer()), "")
  #   )
  #   c_unique_snp_dat()
  # },
  # caption.placement = getOption("xtable.caption.placement", "top"), 
  # caption.width = getOption("xtable.caption.width", NULL))
  
  output$c1_download_mapped_uniprot <- downloadHandler(
    filename = function() {
      paste("protein-to-gene-names1", ".txt", sep = "")
    },
    content = function(file) {
      write.table(c_pd1_converted(), file, sep = "\t", col.names = T, row.names = F, quote = F)
    }
  )
  
  output$c1_download_replications_calculated <- downloadHandler(
    filename = function() {
      paste("input_calculated1", ".txt", sep = "")
    },
    content = function(file) {
      write.table(c_pd1(), file, sep = "\t", col.names = T, row.names = F, quote = F)
    }
  )
  
  output$c2_download_mapped_uniprot <- downloadHandler(
    filename = function() {
      paste("protein-to-gene-names2", ".txt", sep = "")
    },
    content = function(file) {
      write.table(c_pd2_converted(), file, sep = "\t", col.names = T, row.names = F, quote = F)
    }
  )
  
  output$c2_download_replications_calculated <- downloadHandler(
    filename = function() {
      paste("input_calculated2", ".txt", sep = "")
    },
    content = function(file) {
      write.table(c_pd2(), file, sep = "\t", col.names = T, row.names = F, quote = F)
    }
  )
  
  output$c3_download_mapped_uniprot <- downloadHandler(
    filename = function() {
      paste("protein-to-gene-names3", ".txt", sep = "")
    },
    content = function(file) {
      write.table(c_pd3_converted(), file, sep = "\t", col.names = T, row.names = F, quote = F)
    }
  )
  
  output$c3_download_replications_calculated <- downloadHandler(
    filename = function() {
      paste("input_calculated3", ".txt", sep = "")
    },
    content = function(file) {
      write.table(c_pd3(), file, sep = "\t", col.names = T, row.names = F, quote = F)
    }
  )
  
  # output$c_download_snp_to_genes <- downloadHandler(
  #   filename = function() {
  #     paste("snp-to-gene_mfc", ".txt", sep = "")
  #   },
  #   content = function(file) {
  #     write.table(c_SNP_to_gene(), file, sep = "\t", col.names = T, row.names = F, quote = F)
  #   }
  # )
  
  output$c_download_basic_plots <- downloadHandler(
    filename = "basic-plots-mfc.html",
    content = function(file) {
      if(!is.null(input$c_file_pulldown1) & is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
        df <- c_orig_pd1()
        
        d_all5 <- c("logFC" %in% colnames(df) & "FDR" %in% colnames(df) & "pvalue" %in% colnames(df) &
                      "rep1" %in% colnames(df) & "rep2" %in% colnames(df))
        d_ttest <- c("logFC" %in% colnames(df) & "FDR" %in% colnames(df) & "pvalue" %in% colnames(df))
        d_reps <- c("rep1" %in% colnames(df) & "rep2" %in% colnames(df))
        
        if(d_all5){
          if(is.null(c_search_gene())){
            params <- list(a = c_vp1(), b = c_sp1(), g = c_vp_colorbar_dl())
          } else{
            params <- list(a = c_vp1_plus(), b = c_sp1_plus(), g = c_vp_colorbar_dl())
          }
        } else if(d_ttest){
          if(is.null(c_search_gene())){
            params <- list(a = c_vp1(), b = c_sp_blank(), g = c_vp_colorbar_dl())
          } else{
            params <- list(a = c_vp1_plus(), g = c_vp_colorbar_dl())
          }
        } else if(d_reps){
          if(is.null(c_search_gene())){
            params <- list(a = c_vp1(), b = c_sp1(), g = c_vp_colorbar_dl())
          } else{
            params <- list(a = c_vp1_plus(), b = c_sp1_plus(), g = c_vp_colorbar_dl())
          }
        }
      } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
        df <- c_orig_pd1()
        df1 <- c_orig_pd2()
        
        d_all5 <- c("logFC" %in% colnames(df) & "FDR" %in% colnames(df) & "pvalue" %in% colnames(df) &
                      "rep1" %in% colnames(df) & "rep2" %in% colnames(df))
        d_ttest <- c("logFC" %in% colnames(df) & "FDR" %in% colnames(df) & "pvalue" %in% colnames(df))
        d_reps <- c("rep1" %in% colnames(df) & "rep2" %in% colnames(df))
        
        d1_all5 <- c("logFC" %in% colnames(df1) & "FDR" %in% colnames(df1) & "pvalue" %in% colnames(df1) &
                       "rep1" %in% colnames(df1) & "rep2" %in% colnames(df1))
        d1_ttest <- c("logFC" %in% colnames(df1) & "FDR" %in% colnames(df1) & "pvalue" %in% colnames(df1))
        d1_reps <- c("rep1" %in% colnames(df1) & "rep2" %in% colnames(df1))
        
        if((d_all5 | d_reps) & (d1_all5 | d1_reps)){
          if(is.null(c_search_gene())){
            p <- subplot(c_vp1(), c_vp2(), shareY = T) %>% layout(width = 640)
            p1 <- subplot(c_sp1(), c_sp2(), shareY = T) %>% layout(width = 640, title = "")
            params <- list(a = p, b = p1, g = c_vp_colorbar_dl())
          } else{
            p_plus <- subplot(c_vp1_plus(), c_vp2_plus(), shareY = T) %>% layout(width = 640)
            p1_plus <- subplot(c_sp1_plus(), c_sp1_plus(), shareY = T) %>% layout(width = 640, title = "")
            params <- list(a = p_plus, b = p1_plus, g = c_vp_colorbar_dl())
          }
        } else if((d_all5 | d_reps) & d1_ttest){
          if(is.null(c_search_gene())){
            p <- subplot(c_vp1(), c_vp2(), shareY = T) %>% layout(width = 640)
            p1 <- subplot(c_sp1(), plot_scatter_white(c_pd1()), shareY = T) %>% layout(width = 640, title = "")
            params <- list(a = p, b = p1, g = c_vp_colorbar_dl())
          } else{
            p_plus <- subplot(c_vp1_plus(), c_vp2_plus(), shareY = T) %>% layout(width = 640)
            p1_plus <- subplot(c_sp1_plus(), plot_scatter_white(c_pd1()), shareY = T) %>% layout(width = 640, title = "")
            params <- list(a = p_plus, b = p1_plus, g = c_vp_colorbar_dl())
          }
        } else if(d_ttest & (d1_all5 | d1_reps)){
          if(is.null(c_search_gene())){
            p <- subplot(c_vp1(), c_vp2(), shareY = T) %>% layout(width = 640)
            p1 <- subplot(plot_scatter_white(c_pd2()), c_sp2(), shareY = T) %>% layout(width = 640, title = "")
            params <- list(a = p, b = p1, g = c_vp_colorbar_dl())
          } else{
            p_plus <- subplot(c_vp1_plus(), c_vp2_plus(), shareY = T) %>% layout(width = 640)
            p1_plus <- subplot(plot_scatter_white(c_pd2()), c_sp2_plus(), shareY = T) %>% layout(width = 640, title = "")
            params <- list(a = p_plus, b = p1_plus, g = c_vp_colorbar_dl())
          }
        } else if(d_ttest & d1_ttest){
          if(is.null(c_search_gene())){
            p <- subplot(c_vp1(), c_vp2(), shareY = T) %>% layout(width = 640)
            params <- list(a = p, g = c_vp_colorbar_dl())
          } else{
            p_plus <- subplot(c_vp1_plus(), c_vp2_plus(), shareY = T) %>% layout(width = 640)
            params <- list(a = p_plus, g = c_vp_colorbar_dl())
          }
        } 
      } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
        df <- c_orig_pd1()
        df1 <- c_orig_pd2()
        df2 <- c_orig_pd3()
        
        d_all5 <- c("logFC" %in% colnames(df) & "FDR" %in% colnames(df) & "pvalue" %in% colnames(df) &
                      "rep1" %in% colnames(df) & "rep2" %in% colnames(df))
        d_ttest <- c("logFC" %in% colnames(df) & "FDR" %in% colnames(df) & "pvalue" %in% colnames(df))
        d_reps <- c("rep1" %in% colnames(df) & "rep2" %in% colnames(df))
        
        d1_all5 <- c("logFC" %in% colnames(df1) & "FDR" %in% colnames(df1) & "pvalue" %in% colnames(df1) &
                       "rep1" %in% colnames(df1) & "rep2" %in% colnames(df1))
        d1_ttest <- c("logFC" %in% colnames(df1) & "FDR" %in% colnames(df1) & "pvalue" %in% colnames(df1))
        d1_reps <- c("rep1" %in% colnames(df1) & "rep2" %in% colnames(df1))
        
        d2_all5 <- c("logFC" %in% colnames(df2) & "FDR" %in% colnames(df2) & "pvalue" %in% colnames(df2) &
                       "rep1" %in% colnames(df2) & "rep2" %in% colnames(df2))
        d2_ttest <- c("logFC" %in% colnames(df2) & "FDR" %in% colnames(df2) & "pvalue" %in% colnames(df2))
        d2_reps <- c("rep1" %in% colnames(df2) & "rep2" %in% colnames(df2))
        
        if((d_all5 | d_reps) & (d1_all5 | d1_reps) & (d2_all5 | d2_reps)){
          if(is.null(c_search_gene())){
            p <- subplot(c_vp1(), c_vp2(), c_vp3(), shareY = T) %>% layout(width = 960)
            p1 <- subplot(c_sp1(), c_sp2(), c_sp3(), shareY = T) %>% layout(width = 960, title = "")
            params <- list(a = p, b = p1, g = c_vp_colorbar_dl())
          } else{
            p_plus <- subplot(c_vp1_plus(), c_vp2_plus(), c_vp3_plus(), shareY = T) %>% layout(width = 960)
            p1_plus <- subplot(c_sp1_plus(), c_sp2_plus(), c_sp3_plus(), shareY = T) %>% layout(width = 960, title = "")
            params <- list(a = p_plus, b = p1_plus, g = c_vp_colorbar_dl())
          }
        } else if((d_all5 | d_reps) & (d1_all5 | d1_reps) & d2_ttest){
          if(is.null(c_search_gene())){
            p <- subplot(c_vp1(), c_vp2(), c_vp3(), shareY = T) %>% layout(width = 960)
            p1 <- subplot(c_sp1(), c_sp2(), plot_scatter_white(c_pd2()), shareY = T) %>% layout(width = 960, title = "")
            params <- list(a = p, b = p1, g = c_vp_colorbar_dl())
          } else{
            p_plus <- subplot(c_vp1_plus(), c_vp2_plus(), c_vp3_plus(), shareY = T) %>% layout(width = 960)
            p1_plus <- subplot(c_sp1_plus(), c_sp2_plus(), plot_scatter_white(c_pd2()), shareY = T) %>% layout(width = 960, title = "")
            params <- list(a = p_plus, b = p1_plus, g = c_vp_colorbar_dl())
          }
        } else if((d_all5 | d_reps) & d1_ttest & (d2_all5 | d2_reps)){
          if(is.null(c_search_gene())){
            p <- subplot(c_vp1(), c_vp2(), c_vp3(), shareY = T) %>% layout(width = 960)
            p1 <- subplot(c_sp1(), plot_scatter_white(c_pd1()), c_sp3(), shareY = T) %>% layout(width = 960, title = "")
            params <- list(a = p, b = p1, g = c_vp_colorbar_dl())
          } else{
            p_plus <- subplot(c_vp1_plus(), c_vp2_plus(), c_vp3_plus(), shareY = T) %>% layout(width = 960)
            p1_plus <- subplot(c_sp1_plus(), plot_scatter_white(c_pd1()), c_sp3_plus(), shareY = T) %>% layout(width = 960, title = "")
            params <- list(a = p_plus, b = p1_plus, g = c_vp_colorbar_dl())
          }
        } else if((d_all5 | d_reps) & d1_ttest & d2_ttest){
          if(is.null(c_search_gene())){
            p <- subplot(c_vp1(), c_vp2(), c_vp3(), shareY = T) %>% layout(width = 960)
            p1 <- subplot(c_sp1(), plot_scatter_white(c_pd1()), plot_scatter_white(c_pd1()), shareY = T) %>% layout(width = 960, title = "")
            params <- list(a = p, b = p1, g = c_vp_colorbar_dl())
          } else{
            p_plus <- subplot(c_vp1_plus(), c_vp2_plus(), c_vp3_plus(), shareY = T) %>% layout(width = 960)
            p1_plus <- subplot(c_sp1_plus(), plot_scatter_white(c_pd1()), plot_scatter_white(c_pd1()), shareY = T) %>% layout(width = 960, title = "")
            params <- list(a = p_plus, b = p1_plus, g = c_vp_colorbar_dl())
          }
        } else if(d_ttest & (d1_all5 | d1_reps) & (d2_all5 | d2_reps)){
          if(is.null(c_search_gene())){
            p <- subplot(c_vp1(), c_vp2(), c_vp3(), shareY = T) %>% layout(width = 960)
            p1 <- subplot(plot_scatter_white(c_pd2()), c_sp2(), c_sp3(), shareY = T) %>% layout(width = 960, title = "")
            params <- list(a = p, b = p1, g = c_vp_colorbar_dl())
          } else{
            p_plus <- subplot(c_vp1_plus(), c_vp2_plus(), c_vp3_plus(), shareY = T) %>% layout(width = 960)
            p1_plus <- subplot(plot_scatter_white(c_pd2()), c_sp2_plus(), c_sp3_plus(), shareY = T) %>% layout(width = 960, title = "")
            params <- list(a = p_plus, b = p1_plus, g = c_vp_colorbar_dl())
          }
        } else if(d_ttest & (d1_all5 | d1_reps) & d2_ttest){
          if(is.null(c_search_gene())){
            p <- subplot(c_vp1(), c_vp2(), c_vp3(), shareY = T) %>% layout(width = 960)
            p1 <- subplot(plot_scatter_white(c_pd2()), c_sp2(), plot_scatter_white(c_pd2()), shareY = T) %>% layout(width = 960, title = "")
            params <- list(a = p, b = p1, g = c_vp_colorbar_dl())
          } else{
            p_plus <- subplot(c_vp1_plus(), c_vp2_plus(), c_vp3_plus(), shareY = T) %>% layout(width = 960)
            p1_plus <- subplot(plot_scatter_white(c_pd2()), c_sp2_plus(), plot_scatter_white(c_pd2()), shareY = T) %>% layout(width = 960, title = "")
            params <- list(a = p_plus, b = p1_plus, g = c_vp_colorbar_dl())
          }
        } else if(d_ttest & d1_ttest & (d2_all5 | d2_reps)){
          if(is.null(c_search_gene())){
            p <- subplot(c_vp1(), c_vp2(), c_vp3(), shareY = T) %>% layout(width = 960)
            p1 <- subplot(plot_scatter_white(c_pd3()), plot_scatter_white(c_pd3()), c_sp3(), shareY = T) %>% layout(width = 960, title = "")
            params <- list(a = p, b = p1, g = c_vp_colorbar_dl())
          } else{
            p_plus <- subplot(c_vp1_plus(), c_vp2_plus(), c_vp3_plus(), shareY = T) %>% layout(width = 960)
            p1_plus <- subplot(plot_scatter_white(c_pd3()), plot_scatter_white(c_pd3()), c_sp3_plus(), shareY = T) %>% layout(width = 960, title = "")
            params <- list(a = p_plus, b = p1_plus, g = c_vp_colorbar_dl())
          }
        } else if(d_ttest & d1_ttest & d2_ttest){
          if(is.null(c_search_gene())){
            p <- subplot(c_vp1(), c_vp2(), c_vp3(), shareY = T) %>% layout(width = 960)
            params <- list(a = p, g = c_vp_colorbar_dl())
          } else{
            p_plus <- subplot(c_vp1_plus(), c_vp2_plus(), c_vp3_plus(), shareY = T) %>% layout(width = 960)
            params <- list(a = p_plus, g = c_vp_colorbar_dl())
          }
        }
      }
      rmarkdown::render("scripts/basic-mfc.Rmd", 
                        output_file = file,
                        params = params
      )
    }
  )
  
  output$c_download_protein_comparisons <- downloadHandler(
    filename = "comparisons.html",
    content = function(file) {
      if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
        if(is.null(c_search_gene())){
          p <- subplot(c_compare1_plot(), c_compare2_plot(), shareY = T) %>% layout(width = 640)
          params <- list(a = p, d = c_venndiagram(), e = c_unique_dat())
        } else{
          p <- subplot(c_compare1_plus(), c_compare2_plus(), shareY = T) %>% layout(width = 640)
          params <- list(a = p, d = c_venndiagram(), e = c_unique_dat())
        }
      } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
        if(is.null(c_search_gene())){
          p <- subplot(c_compare1_plot(), c_compare2_plot(), c_compare3_plot(), shareY = T) %>% layout(width = 960)
          params <- list(a = p, d = c_venndiagram(), e = c_unique_dat())
        } else{
          p <- subplot(c_compare1_plus(), c_compare2_plus(), c_compare3_plus(), shareY = T) %>% layout(width = 960)
          params <- list(a = p, d = c_venndiagram(), e = c_unique_dat())
        }
      }
      rmarkdown::render("scripts/pc-mfc.Rmd", 
                        output_file = file,
                        params = params
      )
    }
  )
  
  output$c_download_inweb <- downloadHandler(
    filename = "MFC-inweb.html",
    content = function(file) {
      if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
        if(is.null(c_search_gene())){
          p <- subplot(c_vp1_inweb_layer(), c_vp2_inweb_layer(), shareY = T) %>% layout(width = 640, title = "")
          params <- list(a = p, d = c_venndiagram_inweb(), e = c_unique_inweb_dat())
        } else{
          p <- subplot(c_vp1_inweb_plus(), c_vp2_inweb_plus(), shareY = T) %>% layout(width = 640, title = "")
          params <- list(a = p, d = c_venndiagram_inweb(), e = c_unique_inweb_dat())
        }
      } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
        if(is.null(c_search_gene())){
          p <- subplot(c_vp1_inweb_layer(), c_vp2_inweb_layer(), c_vp3_inweb_layer(), shareY = T) %>% layout(width = 960, title = "")
          params <- list(a = p, d = c_venndiagram_inweb(), e = c_unique_inweb_dat())
        } else{
          p <- subplot(c_vp1_inweb_plus(), c_vp2_inweb_plus(), c_vp3_inweb_plus(), shareY = T) %>% layout(width = 960, title = "")
          params <- list(a = p, d = c_venndiagram_inweb(), e = c_unique_inweb_dat())
        }
      }
      rmarkdown::render("scripts/pc-mfc.Rmd", 
                        output_file = file,
                        params = params
      )
    }
  )
  
  output$c_download_goi <- downloadHandler(
    filename = "MFC-goi.html",
    content = function(file) {
      if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
        if(is.null(c_search_gene())){
          p <- subplot(c_vp1_goi_layer(), c_vp2_goi_layer(), shareY = T) %>% layout(width = 640, title = "")
          params <- list(a = p, d = c_venndiagram_goi(), e = c_unique_goi_dat())
        } else{
          p <- subplot(c_vp1_goi_plus(), c_vp2_goi_plus(), shareY = T) %>% layout(width = 640, title = "")
          params <- list(a = p, d = c_venndiagram_goi(), e = c_unique_goi_dat())
        }
      } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
        if(is.null(c_search_gene())){
          p <- subplot(c_vp1_goi_layer(), c_vp2_goi_layer(), c_vp3_goi_layer(), shareY = T) %>% layout(width = 960, title = "")
          params <- list(a = p, d = c_venndiagram_goi(), e = c_unique_goi_dat())
        } else{
          p <- subplot(c_vp1_goi_plus(), c_vp2_goi_plus(), c_vp3_goi_plus(), shareY = T) %>% layout(width = 960, title = "")
          params <- list(a = p, d = c_venndiagram_goi(), e = c_unique_goi_dat())
        }
      }
      rmarkdown::render("scripts/pc-mfc.Rmd", 
                        output_file = file,
                        params = params
      )
    }
  )
  
  # output$c_download_snp <- downloadHandler(
  #   filename = "MFC-snp.html",
  #   content = function(file) {
  #     if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
  #       if(is.null(c_search_gene())){
  #         p <- subplot(c_vp1_snp_layer(), c_vp2_snp_layer(), shareY = T) %>% layout(width = 640, title = "")
  #         params <- list(a = p, d = c_venndiagram_snp(), e = c_unique_snp_dat())
  #       } else{
  #         p <- subplot(c_vp1_snp_plus(), c_vp2_snp_plus(), shareY = T) %>% layout(width = 640, title = "")
  #         params <- list(a = p, d = c_venndiagram_snp(), e = c_unique_snp_dat())
  #       }
  #     } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
  #       if(is.null(c_search_gene())){
  #         p <- subplot(c_vp1_snp_layer(), c_vp2_snp_layer(), c_vp3_snp_layer(), shareY = T) %>% layout(width = 960, title = "")
  #         params <- list(a = p, d = c_venndiagram_snp(), e = c_unique_snp_dat())
  #       } else{
  #         p <- subplot(c_vp1_snp_plus(), c_vp2_snp_plus(), c_vp3_snp_plus(), shareY = T) %>% layout(width = 960, title = "")
  #         params <- list(a = p, d = c_venndiagram_snp(), e = c_unique_snp_dat())
  #       }
  #     }
  #     rmarkdown::render("scripts/pc-mfc.Rmd", 
  #                       output_file = file,
  #                       params = params
  #     )
  #   }
  # )
  
  output$c_download_protein_fams <- downloadHandler(
    filename = "MFC-prot-fam.html",
    content = function(file) {
      if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & is.null(input$c_file_pulldown3)){
        params <- list(a = c_compare1_pfe_plot(), b = c_compare2_pfe_plot())
      } else if(!is.null(input$c_file_pulldown1) & !is.null(input$c_file_pulldown2) & !is.null(input$c_file_pulldown3)){
        params <- list(a = c_compare1_pfe_plot(), b = c_compare2_pfe_plot(), c = c_compare3_pfe_plot())
      }
      rmarkdown::render("scripts/pf-mfc.Rmd", 
                        output_file = file,
                        params = params
      )
    }
  )
  
  output$download_c1_pf_cleaned_input <- downloadHandler(
    filename = function() {
      paste("input1-pf-removed", ".txt", sep = "")
    },
    content = function(file) {
      write.table(c_pf1_cleanup(), file, sep = "\t", col.names = T, row.names = F, quote = F)
    }
  )
  
  output$download_c2_pf_cleaned_input <- downloadHandler(
    filename = function() {
      paste("input2-pf-removed", ".txt", sep = "")
    },
    content = function(file) {
      write.table(c_pf2_cleanup(), file, sep = "\t", col.names = T, row.names = F, quote = F)
    }
  )
  
  output$download_c3_pf_cleaned_input <- downloadHandler(
    filename = function() {
      paste("input3-pf-removed", ".txt", sep = "")
    },
    content = function(file) {
      write.table(c_pf3_cleanup(), file, sep = "\t", col.names = T, row.names = F, quote = F)
    }
  )
  
  observe({
    if(!is.null(c_pd1())){
      d <- c_orig_pd1()
      d_col <- colnames(d)
      if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col &
         "rep1" %in% d_col & "rep2" %in% d_col){
        shinyjs::show("download_c_vp1_gg")
        shinyjs::show("download_c_sp1_gg")
      } else if("rep1" %in% d_col & "rep2" %in% d_col){
        shinyjs::show("download_c_vp1_gg")
        shinyjs::show("download_c_sp1_gg")
      } else if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col) {
        shinyjs::hide("download_c_sp1_gg")
      }
    } else if(is.null(c_pd1())){
      shinyjs::hide("download_c_vp1_gg")
      shinyjs::hide("download_c_sp1_gg")
    }
  })
  
  observe({
    if(!is.null(c_pd2())){
      d <- c_orig_pd2()
      d_col <- colnames(d)
      if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col &
         "rep1" %in% d_col & "rep2" %in% d_col){
        shinyjs::show("download_c_vp2_gg")
        shinyjs::show("download_c_sp2_gg")
      } else if("rep1" %in% d_col & "rep2" %in% d_col){
        shinyjs::show("download_c_vp2_gg")
        shinyjs::show("download_c_sp2_gg")
      } else if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col) {
        shinyjs::hide("download_c_sp2_gg")
      }
    } else if(is.null(c_pd2())){
      shinyjs::hide("download_c_vp2_gg")
      shinyjs::hide("download_c_sp2_gg")
    }
  })
  
  observe({
    if(!is.null(c_pd3())){
      d <- c_orig_pd3()
      d_col <- colnames(d)
      if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col &
         "rep1" %in% d_col & "rep2" %in% d_col){
        shinyjs::show("download_c_vp3_gg")
        shinyjs::show("download_c_sp3_gg")
      } else if("rep1" %in% d_col & "rep2" %in% d_col){
        shinyjs::show("download_c_vp3_gg")
        shinyjs::show("download_c_sp3_gg")
      } else if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col) {
        shinyjs::hide("download_c_sp3_gg")
      }
    } else if(is.null(c_pd3())){
      shinyjs::hide("download_c_vp3_gg")
      shinyjs::hide("download_c_sp3_gg")
    }
  })
  
  observe({
    if(!is.null(c_pf_db())){
      if(!is.null(c_pd1())){
        d <- c_orig_pd1()
        d_col <- colnames(d)
        if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col &
           "rep1" %in% d_col & "rep2" %in% d_col){
          shinyjs::hide("download_c1_pf_cleaned_input")
        } else if("rep1" %in% d_col & "rep2" %in% d_col){
          shinyjs::show("download_c1_pf_cleaned_input")
        } else if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col) {
          shinyjs::hide("download_c1_pf_cleaned_input")
        }
      }
    } else if(is.null(c_pf_db())){
      shinyjs::hide("download_c1_pf_cleaned_input")
    }
  })
  
  
  observe({
    if(!is.null(c_pf_db())){
      if(!is.null(c_pd2())){
        d <- c_orig_pd2()
        d_col <- colnames(d)
        if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col &
           "rep1" %in% d_col & "rep2" %in% d_col){
          shinyjs::hide("download_c2_pf_cleaned_input")
        } else if("rep1" %in% d_col & "rep2" %in% d_col){
          shinyjs::show("download_c2_pf_cleaned_input")
        } else if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col) {
          shinyjs::hide("download_c2_pf_cleaned_input")
        }
      }
    } else if(is.null(c_pf_db())){
      shinyjs::hide("download_c2_pf_cleaned_input")
    }
  })
  
  observe({
    if(!is.null(c_pf_db())){
      if(!is.null(c_pd3())){
        d <- c_orig_pd3()
        d_col <- colnames(d)
        if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col &
           "rep1" %in% d_col & "rep2" %in% d_col){
          shinyjs::hide("download_c3_pf_cleaned_input")
        } else if("rep1" %in% d_col & "rep2" %in% d_col){
          shinyjs::show("download_c3_pf_cleaned_input")
        } else if("logFC" %in% d_col & "FDR" %in% d_col & "pvalue" %in% d_col) {
          shinyjs::hide("download_c3_pf_cleaned_input")
        }
      }
    } else if(is.null(c_pf_db())){
      shinyjs::hide("download_c3_pf_cleaned_input")
    }
  })
  
  observe({
    if(!is.null(input$c_file_pulldown1)){
      df <- c_in_pd1()
      if("accession_number" %in% colnames(df)){
        shinyjs::enable("c1_download_mapped_uniprot")
      }
    }
  })
  
  observe({
    if(!is.null(input$c_file_pulldown1)){
      df <- c_in_pd1()
      if("gene" %in% colnames(df)){
        shinyjs::disable("c1_download_mapped_uniprot")
      }
    }
  })
  
  observe({
    if(!is.null(input$c_file_pulldown1)){
      df <- c_in_pd1()
      if("rep1" %in% colnames(df) & "rep2" %in% colnames(df)){
        shinyjs::enable("c1_download_replications_calculated")
      }
    }
  })
  
  observe({
    if(!is.null(input$c_file_pulldown1)){
      df <- c_in_pd1()
      if("logFC" %in% colnames(df) & "FDR" %in% colnames(df) & "pvalue" %in% colnames(df)){
        shinyjs::disable("c1_download_replications_calculated")
      }
    }
  })
  
  observe({
    if(!is.null(input$c_file_pulldown2)){
      df <- c_in_pd2()
      if("accession_number" %in% colnames(df)){
        shinyjs::enable("c2_download_mapped_uniprot")
      }
    }
  })
  
  observe({
    if(!is.null(input$c_file_pulldown2)){
      df <- c_in_pd2()
      if("gene" %in% colnames(df)){
        shinyjs::disable("c2_download_mapped_uniprot")
      }
    }
  })
  
  observe({
    if(!is.null(input$c_file_pulldown2)){
      df <- c_in_pd2()
      if("rep1" %in% colnames(df) & "rep2" %in% colnames(df)){
        shinyjs::enable("c2_download_replications_calculated")
      }
    }
  })
  
  observe({
    if(!is.null(input$c_file_pulldown2)){
      df <- c_in_pd2()
      if("logFC" %in% colnames(df) & "FDR" %in% colnames(df) & "pvalue" %in% colnames(df)){
        shinyjs::disable("c2_download_replications_calculated")
      }
    }
  })
  
  observe({
    if(!is.null(input$c_file_pulldown3)){
      df <- c_in_pd3()
      if("accession_number" %in% colnames(df)){
        shinyjs::enable("c2_download_mapped_uniprot")
      }
    }
  })
  
  observe({
    if(!is.null(input$c_file_pulldown3)){
      df <- c_in_pd3()
      if("gene" %in% colnames(df)){
        shinyjs::disable("c2_download_mapped_uniprot")
      }
    }
  })
  
  observe({
    if(!is.null(input$c_file_pulldown3)){
      df <- c_in_pd3()
      if("rep1" %in% colnames(df) & "rep2" %in% colnames(df)){
        shinyjs::enable("c3_download_replications_calculated")
      }
    }
  })
  
  observe({
    if(!is.null(input$c_file_pulldown3)){
      df <- c_in_pd3()
      if("logFC" %in% colnames(df) & "FDR" %in% colnames(df) & "pvalue" %in% colnames(df)){
        shinyjs::disable("c3_download_replications_calculated")
      }
    }
  })
  
  observe({
    if (is.null(input$c_file_pulldown1)){
      shinyjs::disable("c1_download_mapped_uniprot")
      shinyjs::disable("c1_download_replications_calculated")
      # shinyjs::disable("c_download_snp_to_genes")
      shinyjs::disable("c_download_basic_plots")
      shinyjs::disable("c_download_protein_comparisons")
      # shinyjs::disable("c_download_snp")
      shinyjs::disable("c_download_goi")
      shinyjs::disable("c_download_inweb")
      shinyjs::disable("c_download_protein_fams")
    } else {
      shinyjs::enable("c_download_basic_plots")
    }
  })
  
  observe({
    if (is.null(input$c_file_pulldown2)){
      shinyjs::disable("c2_download_mapped_uniprot")
      shinyjs::disable("c2_download_replications_calculated")
      # shinyjs::disable("c_download_snp_to_genes")
      shinyjs::disable("c_download_protein_comparisons")
      # shinyjs::disable("c_download_snp")
      shinyjs::disable("c_download_goi")
      shinyjs::disable("c_download_inweb")
      shinyjs::disable("c_download_protein_fams")
    } else {
      shinyjs::enable("c_download_protein_comparisons")
    }
  })
  
  observe({
    if (is.null(input$c_file_pulldown3)){
      shinyjs::disable("c3_download_mapped_uniprot")
      shinyjs::disable("c3_download_replications_calculated")
      # shinyjs::disable("c_download_snp_to_genes")
      # shinyjs::disable("c_download_snp")
      shinyjs::disable("c_download_goi")
      shinyjs::disable("c_download_inweb")
      shinyjs::disable("c_download_protein_fams")
    }
  })
  
  # observe({
  #   if (is.null(input$c_file_SNP)){
  #     shinyjs::disable("c_download_snp_to_genes")
  #     shinyjs::disable("c_download_snp")
  #   } else {
  #     shinyjs::enable("c_download_snp_to_genes")
  #     shinyjs::enable("c_download_snp")
  #   }
  # })
  
  observe({
    if (is.null(input$c_file_genes)){
      shinyjs::disable("c_download_goi")
    } else {
      shinyjs::enable("c_download_goi")
    }
  })
  
  observe({
    if (input$c_make_plot_inweb == 0 || is.null(input$c_make_plot_inweb)){
      shinyjs::disable("c_download_inweb")
    } else {
      shinyjs::enable("c_download_inweb")
    }
  })
  
  # observe({
  #   if (is.null(input$c_file_pulldown2)){
  #     shinyjs::disable("c_download_protein_fams")
  #   } else {
  #     if(!is.null(c_compare2_pfe_plot())){
  #       shinyjs::enable("c_download_protein_fams")
  #     }
  #   }
  # })
  
  ##### GENERAL #####
  #documentation
  
  getPage_guide<-function() {
    return(tags$iframe(src = "how-to.html"
                       , style="width:100%;",  frameborder="0"
                       , height = "3100px"))
  }
  
  output$how_to_guide <- renderUI({
    getPage_guide()
  })
  
  
  # output$documentation <- renderUI({
  #   return(includeHTML("documentation/doc_0316.html")
  #   )
  # })
  
})



